{
  "cells": [
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {
        "id": "8jDjsj690T5O"
      },
      "source": [
        "# Imports"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "2Mcb2bjS0QP7",
        "outputId": "fe293228-8115-48ac-a980-788ef5e08802"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Requirement already satisfied: langchain in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.0.219)\n",
            "Requirement already satisfied: PyYAML>=5.4.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (6.0)\n",
            "Requirement already satisfied: SQLAlchemy<3,>=1.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.4.39)\n",
            "Requirement already satisfied: aiohttp<4.0.0,>=3.8.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (3.8.3)\n",
            "Requirement already satisfied: async-timeout<5.0.0,>=4.0.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (4.0.2)\n",
            "Requirement already satisfied: dataclasses-json<0.6.0,>=0.5.7 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (0.5.8)\n",
            "Requirement already satisfied: langchainplus-sdk>=0.0.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (0.0.17)\n",
            "Requirement already satisfied: numexpr<3.0.0,>=2.8.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (2.8.4)\n",
            "Requirement already satisfied: numpy<2,>=1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.23.5)\n",
            "Requirement already satisfied: openapi-schema-pydantic<2.0,>=1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.2.4)\n",
            "Requirement already satisfied: pydantic<2,>=1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.10.9)\n",
            "Requirement already satisfied: requests<3,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (2.29.0)\n",
            "Requirement already satisfied: tenacity<9.0.0,>=8.1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (8.2.2)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (22.1.0)\n",
            "Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (2.0.4)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (6.0.2)\n",
            "Requirement already satisfied: yarl<2.0,>=1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.8.1)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.3.3)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.2.0)\n",
            "Requirement already satisfied: marshmallow<4.0.0,>=3.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (3.19.0)\n",
            "Requirement already satisfied: marshmallow-enum<2.0.0,>=1.5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (1.5.1)\n",
            "Requirement already satisfied: typing-inspect>=0.4.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (0.9.0)\n",
            "Requirement already satisfied: typing-extensions>=4.2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pydantic<2,>=1->langchain) (4.6.3)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (3.4)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (1.26.16)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (2023.5.7)\n",
            "Requirement already satisfied: greenlet!=0.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from SQLAlchemy<3,>=1.4->langchain) (2.0.1)\n",
            "Requirement already satisfied: packaging>=17.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from marshmallow<4.0.0,>=3.3.0->dataclasses-json<0.6.0,>=0.5.7->langchain) (23.1)\n",
            "Requirement already satisfied: mypy-extensions>=0.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from typing-inspect>=0.4.0->dataclasses-json<0.6.0,>=0.5.7->langchain) (1.0.0)\n",
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Requirement already satisfied: openai in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.27.4)\n",
            "Requirement already satisfied: requests>=2.20 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from openai) (2.29.0)\n",
            "Requirement already satisfied: tqdm in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from openai) (4.65.0)\n",
            "Requirement already satisfied: aiohttp in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from openai) (3.8.3)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.20->openai) (2.0.4)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.20->openai) (3.4)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.20->openai) (1.26.16)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.20->openai) (2023.5.7)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (22.1.0)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (6.0.2)\n",
            "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (4.0.2)\n",
            "Requirement already satisfied: yarl<2.0,>=1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (1.8.1)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (1.3.3)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (1.2.0)\n",
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Requirement already satisfied: langchain in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.0.219)\n",
            "Requirement already satisfied: boto3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (1.26.163)\n",
            "Requirement already satisfied: PyYAML>=5.4.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (6.0)\n",
            "Requirement already satisfied: SQLAlchemy<3,>=1.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.4.39)\n",
            "Requirement already satisfied: aiohttp<4.0.0,>=3.8.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (3.8.3)\n",
            "Requirement already satisfied: async-timeout<5.0.0,>=4.0.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (4.0.2)\n",
            "Requirement already satisfied: dataclasses-json<0.6.0,>=0.5.7 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (0.5.8)\n",
            "Requirement already satisfied: langchainplus-sdk>=0.0.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (0.0.17)\n",
            "Requirement already satisfied: numexpr<3.0.0,>=2.8.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (2.8.4)\n",
            "Requirement already satisfied: numpy<2,>=1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.23.5)\n",
            "Requirement already satisfied: openapi-schema-pydantic<2.0,>=1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.2.4)\n",
            "Requirement already satisfied: pydantic<2,>=1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.10.9)\n",
            "Requirement already satisfied: requests<3,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (2.29.0)\n",
            "Requirement already satisfied: tenacity<9.0.0,>=8.1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (8.2.2)\n",
            "Requirement already satisfied: botocore<1.30.0,>=1.29.163 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from boto3) (1.29.163)\n",
            "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from boto3) (1.0.1)\n",
            "Requirement already satisfied: s3transfer<0.7.0,>=0.6.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from boto3) (0.6.1)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (22.1.0)\n",
            "Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (2.0.4)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (6.0.2)\n",
            "Requirement already satisfied: yarl<2.0,>=1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.8.1)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.3.3)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.2.0)\n",
            "Requirement already satisfied: python-dateutil<3.0.0,>=2.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from botocore<1.30.0,>=1.29.163->boto3) (2.8.2)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.25.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from botocore<1.30.0,>=1.29.163->boto3) (1.26.16)\n",
            "Requirement already satisfied: marshmallow<4.0.0,>=3.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (3.19.0)\n",
            "Requirement already satisfied: marshmallow-enum<2.0.0,>=1.5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (1.5.1)\n",
            "Requirement already satisfied: typing-inspect>=0.4.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (0.9.0)\n",
            "Requirement already satisfied: typing-extensions>=4.2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pydantic<2,>=1->langchain) (4.6.3)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (3.4)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (2023.5.7)\n",
            "Requirement already satisfied: greenlet!=0.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from SQLAlchemy<3,>=1.4->langchain) (2.0.1)\n",
            "Requirement already satisfied: packaging>=17.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from marshmallow<4.0.0,>=3.3.0->dataclasses-json<0.6.0,>=0.5.7->langchain) (23.1)\n",
            "Requirement already satisfied: six>=1.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from python-dateutil<3.0.0,>=2.1->botocore<1.30.0,>=1.29.163->boto3) (1.16.0)\n",
            "Requirement already satisfied: mypy-extensions>=0.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from typing-inspect>=0.4.0->dataclasses-json<0.6.0,>=0.5.7->langchain) (1.0.0)\n",
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Requirement already satisfied: pypdf in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (3.11.1)\n",
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Requirement already satisfied: semanticscholar in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.4.1)\n",
            "Requirement already satisfied: requests in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from semanticscholar) (2.29.0)\n",
            "Requirement already satisfied: tenacity in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from semanticscholar) (8.2.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->semanticscholar) (2.0.4)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->semanticscholar) (3.4)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->semanticscholar) (1.26.16)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->semanticscholar) (2023.5.7)\n",
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Requirement already satisfied: metapub in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.5.5)\n",
            "Requirement already satisfied: setuptools in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (67.8.0)\n",
            "Requirement already satisfied: lxml in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (4.9.2)\n",
            "Requirement already satisfied: requests in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (2.29.0)\n",
            "Requirement already satisfied: eutils in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (0.6.0)\n",
            "Requirement already satisfied: habanero in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (1.2.3)\n",
            "Requirement already satisfied: tabulate in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (0.9.0)\n",
            "Requirement already satisfied: cssselect in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (1.2.0)\n",
            "Requirement already satisfied: unidecode in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (1.3.6)\n",
            "Requirement already satisfied: docopt in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (0.6.2)\n",
            "Requirement already satisfied: six in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (1.16.0)\n",
            "Requirement already satisfied: tox in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (4.6.3)\n",
            "Requirement already satisfied: pytest in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (7.4.0)\n",
            "Requirement already satisfied: coloredlogs in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (15.0.1)\n",
            "Requirement already satisfied: python-Levenshtein in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (0.21.1)\n",
            "Requirement already satisfied: humanfriendly>=9.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from coloredlogs->metapub) (10.0)\n",
            "Requirement already satisfied: pytz in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from eutils->metapub) (2022.7)\n",
            "Requirement already satisfied: tqdm in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from habanero->metapub) (4.65.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->metapub) (2.0.4)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->metapub) (3.4)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->metapub) (1.26.16)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->metapub) (2023.5.7)\n",
            "Requirement already satisfied: iniconfig in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (2.0.0)\n",
            "Requirement already satisfied: packaging in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (23.1)\n",
            "Requirement already satisfied: pluggy<2.0,>=0.12 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (1.2.0)\n",
            "Requirement already satisfied: exceptiongroup>=1.0.0rc8 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (1.1.1)\n",
            "Requirement already satisfied: tomli>=1.0.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (2.0.1)\n",
            "Requirement already satisfied: Levenshtein==0.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from python-Levenshtein->metapub) (0.21.1)\n",
            "Requirement already satisfied: rapidfuzz<4.0.0,>=2.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from Levenshtein==0.21.1->python-Levenshtein->metapub) (3.1.1)\n",
            "Requirement already satisfied: cachetools>=5.3.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (5.3.1)\n",
            "Requirement already satisfied: chardet>=5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (5.1.0)\n",
            "Requirement already satisfied: colorama>=0.4.6 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (0.4.6)\n",
            "Requirement already satisfied: filelock>=3.12.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (3.12.2)\n",
            "Requirement already satisfied: platformdirs>=3.5.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (3.8.0)\n",
            "Requirement already satisfied: pyproject-api>=1.5.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (1.5.2)\n",
            "Requirement already satisfied: virtualenv>=20.23.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (20.23.1)\n",
            "Requirement already satisfied: distlib<1,>=0.3.6 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from virtualenv>=20.23.1->tox->metapub) (0.3.6)\n",
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Requirement already satisfied: tiktoken in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.4.0)\n",
            "Requirement already satisfied: regex>=2022.1.18 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tiktoken) (2022.7.9)\n",
            "Requirement already satisfied: requests>=2.26.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tiktoken) (2.29.0)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken) (2.0.4)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken) (3.4)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken) (1.26.16)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken) (2023.5.7)\n",
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Requirement already satisfied: chromadb in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.3.26)\n",
            "Requirement already satisfied: pandas>=1.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (1.5.3)\n",
            "Requirement already satisfied: requests>=2.28 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (2.29.0)\n",
            "Requirement already satisfied: pydantic>=1.9 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (1.10.9)\n",
            "Requirement already satisfied: hnswlib>=0.7 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (0.7.0)\n",
            "Requirement already satisfied: clickhouse-connect>=0.5.7 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (0.6.5)\n",
            "Requirement already satisfied: duckdb>=0.7.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (0.8.1)\n",
            "Requirement already satisfied: fastapi>=0.85.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (0.99.1)\n",
            "Requirement already satisfied: uvicorn[standard]>=0.18.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (0.22.0)\n",
            "Requirement already satisfied: numpy>=1.21.6 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (1.23.5)\n",
            "Requirement already satisfied: posthog>=2.4.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (3.0.1)\n",
            "Requirement already satisfied: typing-extensions>=4.5.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (4.6.3)\n",
            "Requirement already satisfied: pulsar-client>=3.1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (3.2.0)\n",
            "Requirement already satisfied: onnxruntime>=1.14.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (1.15.1)\n",
            "Requirement already satisfied: tokenizers>=0.13.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (0.13.2)\n",
            "Requirement already satisfied: tqdm>=4.65.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (4.65.0)\n",
            "Requirement already satisfied: overrides>=7.3.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from chromadb) (7.3.1)\n",
            "Requirement already satisfied: certifi in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from clickhouse-connect>=0.5.7->chromadb) (2023.5.7)\n",
            "Requirement already satisfied: importlib-metadata in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from clickhouse-connect>=0.5.7->chromadb) (6.0.0)\n",
            "Requirement already satisfied: urllib3>=1.26 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from clickhouse-connect>=0.5.7->chromadb) (1.26.16)\n",
            "Requirement already satisfied: pytz in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from clickhouse-connect>=0.5.7->chromadb) (2022.7)\n",
            "Requirement already satisfied: zstandard in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from clickhouse-connect>=0.5.7->chromadb) (0.21.0)\n",
            "Requirement already satisfied: lz4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from clickhouse-connect>=0.5.7->chromadb) (4.3.2)\n",
            "Requirement already satisfied: starlette<0.28.0,>=0.27.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from fastapi>=0.85.1->chromadb) (0.27.0)\n",
            "Requirement already satisfied: coloredlogs in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from onnxruntime>=1.14.1->chromadb) (15.0.1)\n",
            "Requirement already satisfied: flatbuffers in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from onnxruntime>=1.14.1->chromadb) (23.5.26)\n",
            "Requirement already satisfied: packaging in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from onnxruntime>=1.14.1->chromadb) (23.1)\n",
            "Requirement already satisfied: protobuf in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from onnxruntime>=1.14.1->chromadb) (4.23.4)\n",
            "Requirement already satisfied: sympy in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from onnxruntime>=1.14.1->chromadb) (1.11.1)\n",
            "Requirement already satisfied: python-dateutil>=2.8.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pandas>=1.3->chromadb) (2.8.2)\n",
            "Requirement already satisfied: six>=1.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from posthog>=2.4.0->chromadb) (1.16.0)\n",
            "Requirement already satisfied: monotonic>=1.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from posthog>=2.4.0->chromadb) (1.6)\n",
            "Requirement already satisfied: backoff>=1.10.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from posthog>=2.4.0->chromadb) (2.2.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.28->chromadb) (2.0.4)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.28->chromadb) (3.4)\n",
            "Requirement already satisfied: click>=7.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (8.0.4)\n",
            "Requirement already satisfied: h11>=0.8 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (0.14.0)\n",
            "Requirement already satisfied: httptools>=0.5.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (0.5.0)\n",
            "Requirement already satisfied: python-dotenv>=0.13 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (1.0.0)\n",
            "Requirement already satisfied: pyyaml>=5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (6.0)\n",
            "Requirement already satisfied: uvloop!=0.15.0,!=0.15.1,>=0.14.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (0.17.0)\n",
            "Requirement already satisfied: watchfiles>=0.13 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (0.19.0)\n",
            "Requirement already satisfied: websockets>=10.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (11.0.3)\n",
            "Requirement already satisfied: anyio<5,>=3.4.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from starlette<0.28.0,>=0.27.0->fastapi>=0.85.1->chromadb) (3.7.1)\n",
            "Requirement already satisfied: humanfriendly>=9.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from coloredlogs->onnxruntime>=1.14.1->chromadb) (10.0)\n",
            "Requirement already satisfied: zipp>=0.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from importlib-metadata->clickhouse-connect>=0.5.7->chromadb) (3.11.0)\n",
            "Requirement already satisfied: mpmath>=0.19 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages/mpmath-1.2.1-py3.10.egg (from sympy->onnxruntime>=1.14.1->chromadb) (1.2.1)\n",
            "Requirement already satisfied: sniffio>=1.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from anyio<5,>=3.4.0->starlette<0.28.0,>=0.27.0->fastapi>=0.85.1->chromadb) (1.3.0)\n",
            "Requirement already satisfied: exceptiongroup in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from anyio<5,>=3.4.0->starlette<0.28.0,>=0.27.0->fastapi>=0.85.1->chromadb) (1.1.1)\n",
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Requirement already satisfied: rouge-score in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.1.2)\n",
            "Requirement already satisfied: absl-py in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from rouge-score) (1.4.0)\n",
            "Requirement already satisfied: nltk in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from rouge-score) (3.8.1)\n",
            "Requirement already satisfied: numpy in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from rouge-score) (1.23.5)\n",
            "Requirement already satisfied: six>=1.14.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from rouge-score) (1.16.0)\n",
            "Requirement already satisfied: click in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from nltk->rouge-score) (8.0.4)\n",
            "Requirement already satisfied: joblib in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from nltk->rouge-score) (1.2.0)\n",
            "Requirement already satisfied: regex>=2021.8.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from nltk->rouge-score) (2022.7.9)\n",
            "Requirement already satisfied: tqdm in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from nltk->rouge-score) (4.65.0)\n",
            "Note: you may need to restart the kernel to use updated packages.\n",
            "Requirement already satisfied: datasets in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (2.12.0)\n",
            "Requirement already satisfied: transformers[sentencepiece] in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (4.29.2)\n",
            "Requirement already satisfied: numpy>=1.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (1.23.5)\n",
            "Requirement already satisfied: pyarrow>=8.0.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (11.0.0)\n",
            "Requirement already satisfied: dill<0.3.7,>=0.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (0.3.6)\n",
            "Requirement already satisfied: pandas in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (1.5.3)\n",
            "Requirement already satisfied: requests>=2.19.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (2.29.0)\n",
            "Requirement already satisfied: tqdm>=4.62.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (4.65.0)\n",
            "Requirement already satisfied: xxhash in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (2.0.2)\n",
            "Requirement already satisfied: multiprocess in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (0.70.14)\n",
            "Requirement already satisfied: fsspec[http]>=2021.11.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (2023.4.0)\n",
            "Requirement already satisfied: aiohttp in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (3.8.3)\n",
            "Requirement already satisfied: huggingface-hub<1.0.0,>=0.11.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (0.15.1)\n",
            "Requirement already satisfied: packaging in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (23.1)\n",
            "Requirement already satisfied: responses<0.19 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (0.13.3)\n",
            "Requirement already satisfied: pyyaml>=5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from datasets) (6.0)\n",
            "Requirement already satisfied: filelock in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers[sentencepiece]) (3.12.2)\n",
            "Requirement already satisfied: regex!=2019.12.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers[sentencepiece]) (2022.7.9)\n",
            "Requirement already satisfied: tokenizers!=0.11.3,<0.14,>=0.11.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers[sentencepiece]) (0.13.2)\n",
            "Requirement already satisfied: sentencepiece!=0.1.92,>=0.1.91 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers[sentencepiece]) (0.1.99)\n",
            "Collecting protobuf<=3.20.2 (from transformers[sentencepiece])\n",
            "  Using cached protobuf-3.20.2-cp310-cp310-manylinux_2_12_x86_64.manylinux2010_x86_64.whl (1.1 MB)\n",
            "Requirement already satisfied: attrs>=17.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->datasets) (22.1.0)\n",
            "Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->datasets) (2.0.4)\n",
            "Requirement already satisfied: multidict<7.0,>=4.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->datasets) (6.0.2)\n",
            "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->datasets) (4.0.2)\n",
            "Requirement already satisfied: yarl<2.0,>=1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->datasets) (1.8.1)\n",
            "Requirement already satisfied: frozenlist>=1.1.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->datasets) (1.3.3)\n",
            "Requirement already satisfied: aiosignal>=1.1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->datasets) (1.2.0)\n",
            "Requirement already satisfied: typing-extensions>=3.7.4.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface-hub<1.0.0,>=0.11.0->datasets) (4.6.3)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.19.0->datasets) (3.4)\n",
            "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.19.0->datasets) (1.26.16)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.19.0->datasets) (2023.5.7)\n",
            "Requirement already satisfied: six in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from responses<0.19->datasets) (1.16.0)\n",
            "Requirement already satisfied: python-dateutil>=2.8.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pandas->datasets) (2.8.2)\n",
            "Requirement already satisfied: pytz>=2020.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pandas->datasets) (2022.7)\n",
            "Installing collected packages: protobuf\n",
            "  Attempting uninstall: protobuf\n",
            "    Found existing installation: protobuf 4.23.4\n",
            "    Uninstalling protobuf-4.23.4:\n",
            "      Successfully uninstalled protobuf-4.23.4\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "tensorflow 2.12.0 requires protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3, but you have protobuf 3.20.2 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed protobuf-3.20.2\n",
            "Note: you may need to restart the kernel to use updated packages.\n"
          ]
        }
      ],
      "source": [
        "%pip install langchain\n",
        "%pip install openai\n",
        "%pip install langchain boto3\n",
        "%pip install pypdf\n",
        "%pip install semanticscholar\n",
        "%pip install metapub\n",
        "%pip install tiktoken\n",
        "%pip install chromadb\n",
        "%pip install rouge-score\n",
        "%pip install datasets transformers[sentencepiece]\n",
        "from nltk.translate.bleu_score import sentence_bleu\n",
        "from langchain.llms import OpenAI\n",
        "from langchain import PromptTemplate, SagemakerEndpoint\n",
        "from langchain.llms.sagemaker_endpoint import LLMContentHandler\n",
        "from langchain.chains.question_answering import load_qa_chain\n",
        "from langchain.chains.summarize import load_summarize_chain\n",
        "import json\n",
        "from langchain.document_loaders import TextLoader\n",
        "from langchain.document_loaders.csv_loader import CSVLoader\n",
        "from langchain.document_loaders import JSONLoader\n",
        "from langchain.document_loaders import PyPDFLoader\n",
        "from langchain.docstore.document import Document\n",
        "from langchain.text_splitter import CharacterTextSplitter\n",
        "from pathlib import Path\n",
        "from pprint import pprint\n",
        "from langchain import OpenAI, PromptTemplate, LLMChain\n",
        "from langchain.chains.mapreduce import MapReduceChain\n",
        "from langchain.prompts import PromptTemplate\n",
        "from langchain.embeddings.sentence_transformer import SentenceTransformerEmbeddings\n",
        "from langchain.vectorstores import Chroma\n",
        "\n",
        "import os\n",
        "from glob import glob\n",
        "import requests\n",
        "import urllib"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {
        "id": "UHCR3R9V5CGk"
      },
      "outputs": [],
      "source": [
        "from langchain.chains import RetrievalQA\n",
        "from langchain.embeddings.openai import OpenAIEmbeddings\n",
        "from langchain.vectorstores import Chroma\n",
        "from langchain.text_splitter import RecursiveCharacterTextSplitter\n",
        "from langchain.chains import VectorDBQA\n",
        "from rouge_score import rouge_scorer"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {
        "id": "mKq1OISGCENq"
      },
      "source": [
        "# Load Data"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "YeZFxU0TGSOV",
        "outputId": "5d0f04e5-891d-4279-cc75-d98cbdd2c122"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "2023-07-20 21:14:44 ip-172-31-90-198 metapub.config[1829] WARNING NCBI_API_KEY was not set.\n"
          ]
        }
      ],
      "source": [
        "# from google.colab import drive\n",
        "# from google.colab import files\n",
        "import pandas as pd\n",
        "import numpy as np\n",
        "import csv\n",
        "from metapub import PubMedFetcher\n",
        "import chromadb"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 4,
      "metadata": {},
      "outputs": [],
      "source": [
        "import os\n",
        "import yaml\n",
        "from sqlalchemy import create_engine\n",
        "from sqlalchemy import text\n",
        "\n",
        "# Create engine to connect to database\n",
        "engine = create_engine(f'postgresql://postgres:qRd71PqwOsbv62WAvboR@database-1.cuaho2dof33c.us-east-1.rds.amazonaws.com:5432/postgres')"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 5,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>empty</th>\n",
              "      <th>corpusid</th>\n",
              "      <th>text</th>\n",
              "      <th>id</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1410</td>\n",
              "      <td>250929149</td>\n",
              "      <td>\\nOPEN ACCESS EDITED BY\\n\\n\\nAlexander Nikolae...</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1411</td>\n",
              "      <td>215793097</td>\n",
              "      <td>\\n\\n\\n\\nCi Song \\nState Key Laboratory of Repr...</td>\n",
              "      <td>2</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1412</td>\n",
              "      <td>246020601</td>\n",
              "      <td>\\nCOVID-19 and the Otolaryngology Residency Ma...</td>\n",
              "      <td>3</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1413</td>\n",
              "      <td>18163935</td>\n",
              "      <td>\\nAnalysis of biophysical and functional conse...</td>\n",
              "      <td>4</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1414</td>\n",
              "      <td>237935836</td>\n",
              "      <td>\\nImpact of Plant-Based Meat Alternatives on t...</td>\n",
              "      <td>5</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "  empty   corpusid                                               text  id\n",
              "0  1410  250929149  \\nOPEN ACCESS EDITED BY\\n\\n\\nAlexander Nikolae...   1\n",
              "1  1411  215793097  \\n\\n\\n\\nCi Song \\nState Key Laboratory of Repr...   2\n",
              "2  1412  246020601  \\nCOVID-19 and the Otolaryngology Residency Ma...   3\n",
              "3  1413   18163935  \\nAnalysis of biophysical and functional conse...   4\n",
              "4  1414  237935836  \\nImpact of Plant-Based Meat Alternatives on t...   5"
            ]
          },
          "execution_count": 5,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "### Retreive Full Text from Table\n",
        "\n",
        "table_name = 'fulltext'\n",
        "\n",
        "sql = text(f''' \n",
        "    SELECT empty, corpusid, text, id FROM {table_name} LIMIT 10;\n",
        "''')\n",
        "\n",
        "with engine.connect() as conn: \n",
        "    query = conn.execute(sql)\n",
        "\n",
        "full_text = pd.DataFrame(query.fetchall())\n",
        "full_text.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "metadata": {},
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "['abstract_encodings_test4', 'abstracts', 'abstracts_encodings', 'abstracts_encodings_test3', 'assayTerms', 'assayterms', 'attributes', 'authors', 'copy_test', 'fulltext']\n"
          ]
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "/tmp/ipykernel_1829/1043810782.py:1: SADeprecationWarning: The Engine.table_names() method is deprecated and will be removed in a future release.  Please refer to Inspector.get_table_names(). (deprecated since: 1.4)\n",
            "  print(engine.table_names())\n"
          ]
        }
      ],
      "source": [
        "print(engine.table_names())"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 7,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>paperId</th>\n",
              "      <th>corpusId</th>\n",
              "      <th>abstract</th>\n",
              "      <th>input_ids</th>\n",
              "      <th>attention_mask</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>859c91de1ab22aeff85558dcf676ee5ffc4981a5</td>\n",
              "      <td>33235381</td>\n",
              "      <td>Summary This work aims at applying concepts of...</td>\n",
              "      <td>[20698, 100, 161, 3, 8345, 44, 6247, 6085, 13,...</td>\n",
              "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>2460a37a3305b3bb072770e5bb57ed95496ecf80</td>\n",
              "      <td>72970529</td>\n",
              "      <td>Summary Objectives: To diagnose the hospital i...</td>\n",
              "      <td>[20698, 27919, 7, 10, 304, 18730, 8, 2833, 251...</td>\n",
              "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>e92f481f3be6f0956e0cd6c160a2a384c4eacd76</td>\n",
              "      <td>26375252</td>\n",
              "      <td>Holdoff et al.1 described a retrospective, mon...</td>\n",
              "      <td>[8470, 1647, 3, 15, 17, 491, 5, 536, 3028, 3, ...</td>\n",
              "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>d51f04cfdc8fe907e4a66e948028bf09f3a6af7a</td>\n",
              "      <td>11026954</td>\n",
              "      <td>1. In forty‐one out of forty‐seven dogs under ...</td>\n",
              "      <td>[1300, 86, 19662, 2, 782, 91, 13, 19662, 2, 7,...</td>\n",
              "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>111d6c65ad374861a3c9c70b210996b76b0a7080</td>\n",
              "      <td>95308098</td>\n",
              "      <td>Calculated and observed excited singlet state ...</td>\n",
              "      <td>[18555, 920, 11, 6970, 2787, 712, 17, 538, 703...</td>\n",
              "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                    paperId  corpusId  \\\n",
              "0  859c91de1ab22aeff85558dcf676ee5ffc4981a5  33235381   \n",
              "1  2460a37a3305b3bb072770e5bb57ed95496ecf80  72970529   \n",
              "2  e92f481f3be6f0956e0cd6c160a2a384c4eacd76  26375252   \n",
              "3  d51f04cfdc8fe907e4a66e948028bf09f3a6af7a  11026954   \n",
              "4  111d6c65ad374861a3c9c70b210996b76b0a7080  95308098   \n",
              "\n",
              "                                            abstract  \\\n",
              "0  Summary This work aims at applying concepts of...   \n",
              "1  Summary Objectives: To diagnose the hospital i...   \n",
              "2  Holdoff et al.1 described a retrospective, mon...   \n",
              "3  1. In forty‐one out of forty‐seven dogs under ...   \n",
              "4  Calculated and observed excited singlet state ...   \n",
              "\n",
              "                                           input_ids  \\\n",
              "0  [20698, 100, 161, 3, 8345, 44, 6247, 6085, 13,...   \n",
              "1  [20698, 27919, 7, 10, 304, 18730, 8, 2833, 251...   \n",
              "2  [8470, 1647, 3, 15, 17, 491, 5, 536, 3028, 3, ...   \n",
              "3  [1300, 86, 19662, 2, 782, 91, 13, 19662, 2, 7,...   \n",
              "4  [18555, 920, 11, 6970, 2787, 712, 17, 538, 703...   \n",
              "\n",
              "                                      attention_mask  \n",
              "0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  \n",
              "1  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  \n",
              "2  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  \n",
              "3  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  \n",
              "4  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  "
            ]
          },
          "execution_count": 7,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "### Retreive Abstract Text from Table\n",
        "from sqlalchemy import text\n",
        "\n",
        "table_name = 'abstracts_encodings'\n",
        "\n",
        "sql = text(f''' \n",
        "    SELECT * FROM {table_name} as ae\n",
        "     Limit 10;\n",
        "''')\n",
        "\n",
        "with engine.connect() as conn: \n",
        "    query = conn.execute(sql)\n",
        "\n",
        "abstracts = pd.DataFrame(query.fetchall())\n",
        "abstracts.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 8,
      "metadata": {},
      "outputs": [],
      "source": [
        "# ### Retreive Abstract Text from Table\n",
        "# from sqlalchemy import text\n",
        "\n",
        "# table_name = 'abstracts_encodings'\n",
        "\n",
        "# sql = text(f''' \n",
        "#     SELECT ae.corpusId, abstract, input_ids , cast(ae.corpusId as int) FROM {table_name} as ae\n",
        "#     where \"ae\".\"corpusId\" == 72970529;\n",
        "# ''')\n",
        "\n",
        "# with engine.connect() as conn: \n",
        "#     query = conn.execute(sql)\n",
        "\n",
        "# abstracts = pd.DataFrame(query.fetchall())\n",
        "# abstracts.head()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 11,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "paperId           object\n",
              "corpusId          object\n",
              "abstract          object\n",
              "input_ids         object\n",
              "attention_mask    object\n",
              "dtype: object"
            ]
          },
          "execution_count": 11,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "abstracts.dtypes"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 9,
      "metadata": {},
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "2023-07-20 21:15:06 ip-172-31-90-198 chromadb.telemetry.posthog[1829] INFO Anonymized telemetry enabled. See https://docs.trychroma.com/telemetry for more information.\n"
          ]
        },
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "Nanosecond heartbeat on server 1689887706805504034000\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "[Collection(name=langchain_store),\n",
              " Collection(name=abstracts),\n",
              " Collection(name=fulltext),\n",
              " Collection(name=specter_abstracts)]"
            ]
          },
          "execution_count": 9,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "# Create chroma client\n",
        "from chromadb.config import Settings\n",
        "import dask.dataframe as dd\n",
        "from dask.distributed import Client, LocalCluster, progress\n",
        "chroma = chromadb.Client(Settings(chroma_api_impl=\"rest\",\n",
        "                                  chroma_server_host=\"52.23.195.129\", # EC2 instance public IPv4\n",
        "                                  chroma_server_http_port=8000))\n",
        "\n",
        "print(\"Nanosecond heartbeat on server\", chroma.heartbeat()) # returns a nanosecond heartbeat. Useful for making sure the client remains connected.\n",
        "\n",
        "# Check Existing connections\n",
        "chroma.list_collections()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "# vector_em <=> (select vector_em from abstract_encodings_test4 where vector_em is not null limit 1) as sim\n",
        "# ALTER TABLE abstract_encodings_test4 ADD COLUMN vector_em vector(768);\n",
        "# psql --host=database-1.cuaho2dof33c.us-east-1.rds.amazonaws.com --port=5432 --username=postgres --password --dbname=postgres"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 15,
      "metadata": {},
      "outputs": [
        {
          "ename": "NotSupportedError",
          "evalue": "(psycopg2.errors.FeatureNotSupported) extension \"pgml\" is not available\nDETAIL:  Could not open extension control file \"/rdsdbbin/postgres-15.2.R2/share/extension/pgml.control\": No such file or directory.\nHINT:  The extension must first be installed on the system where PostgreSQL is running.\n\n[SQL: \n                               create extension pgml;]\n(Background on this error at: https://sqlalche.me/e/14/tw8g)",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mFeatureNotSupported\u001b[0m                       Traceback (most recent call last)",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1819\u001b[0m, in \u001b[0;36mConnection._execute_context\u001b[0;34m(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)\u001b[0m\n\u001b[1;32m   1818\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m evt_handled:\n\u001b[0;32m-> 1819\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mdialect\u001b[39m.\u001b[39;49mdo_execute(\n\u001b[1;32m   1820\u001b[0m             cursor, statement, parameters, context\n\u001b[1;32m   1821\u001b[0m         )\n\u001b[1;32m   1823\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_has_events \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mengine\u001b[39m.\u001b[39m_has_events:\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/engine/default.py:732\u001b[0m, in \u001b[0;36mDefaultDialect.do_execute\u001b[0;34m(self, cursor, statement, parameters, context)\u001b[0m\n\u001b[1;32m    731\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mdo_execute\u001b[39m(\u001b[39mself\u001b[39m, cursor, statement, parameters, context\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m--> 732\u001b[0m     cursor\u001b[39m.\u001b[39;49mexecute(statement, parameters)\n",
            "\u001b[0;31mFeatureNotSupported\u001b[0m: extension \"pgml\" is not available\nDETAIL:  Could not open extension control file \"/rdsdbbin/postgres-15.2.R2/share/extension/pgml.control\": No such file or directory.\nHINT:  The extension must first be installed on the system where PostgreSQL is running.\n",
            "\nThe above exception was the direct cause of the following exception:\n",
            "\u001b[0;31mNotSupportedError\u001b[0m                         Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[15], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m sql_create_postgres_env \u001b[39m=\u001b[39m text(\u001b[39mf\u001b[39m\u001b[39m'''\u001b[39m\n\u001b[1;32m      2\u001b[0m \u001b[39m                               create extension pgml;\u001b[39m\u001b[39m'''\u001b[39m)\n\u001b[1;32m      3\u001b[0m \u001b[39mwith\u001b[39;00m engine\u001b[39m.\u001b[39mconnect() \u001b[39mas\u001b[39;00m conn: \n\u001b[0;32m----> 4\u001b[0m     query \u001b[39m=\u001b[39m conn\u001b[39m.\u001b[39;49mexecute(sql_create_postgres_env)\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1306\u001b[0m, in \u001b[0;36mConnection.execute\u001b[0;34m(self, statement, *multiparams, **params)\u001b[0m\n\u001b[1;32m   1302\u001b[0m     util\u001b[39m.\u001b[39mraise_(\n\u001b[1;32m   1303\u001b[0m         exc\u001b[39m.\u001b[39mObjectNotExecutableError(statement), replace_context\u001b[39m=\u001b[39merr\n\u001b[1;32m   1304\u001b[0m     )\n\u001b[1;32m   1305\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m-> 1306\u001b[0m     \u001b[39mreturn\u001b[39;00m meth(\u001b[39mself\u001b[39;49m, multiparams, params, _EMPTY_EXECUTION_OPTS)\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/sql/elements.py:332\u001b[0m, in \u001b[0;36mClauseElement._execute_on_connection\u001b[0;34m(self, connection, multiparams, params, execution_options, _force)\u001b[0m\n\u001b[1;32m    328\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m_execute_on_connection\u001b[39m(\n\u001b[1;32m    329\u001b[0m     \u001b[39mself\u001b[39m, connection, multiparams, params, execution_options, _force\u001b[39m=\u001b[39m\u001b[39mFalse\u001b[39;00m\n\u001b[1;32m    330\u001b[0m ):\n\u001b[1;32m    331\u001b[0m     \u001b[39mif\u001b[39;00m _force \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39msupports_execution:\n\u001b[0;32m--> 332\u001b[0m         \u001b[39mreturn\u001b[39;00m connection\u001b[39m.\u001b[39;49m_execute_clauseelement(\n\u001b[1;32m    333\u001b[0m             \u001b[39mself\u001b[39;49m, multiparams, params, execution_options\n\u001b[1;32m    334\u001b[0m         )\n\u001b[1;32m    335\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    336\u001b[0m         \u001b[39mraise\u001b[39;00m exc\u001b[39m.\u001b[39mObjectNotExecutableError(\u001b[39mself\u001b[39m)\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1498\u001b[0m, in \u001b[0;36mConnection._execute_clauseelement\u001b[0;34m(self, elem, multiparams, params, execution_options)\u001b[0m\n\u001b[1;32m   1486\u001b[0m compiled_cache \u001b[39m=\u001b[39m execution_options\u001b[39m.\u001b[39mget(\n\u001b[1;32m   1487\u001b[0m     \u001b[39m\"\u001b[39m\u001b[39mcompiled_cache\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mengine\u001b[39m.\u001b[39m_compiled_cache\n\u001b[1;32m   1488\u001b[0m )\n\u001b[1;32m   1490\u001b[0m compiled_sql, extracted_params, cache_hit \u001b[39m=\u001b[39m elem\u001b[39m.\u001b[39m_compile_w_cache(\n\u001b[1;32m   1491\u001b[0m     dialect\u001b[39m=\u001b[39mdialect,\n\u001b[1;32m   1492\u001b[0m     compiled_cache\u001b[39m=\u001b[39mcompiled_cache,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1496\u001b[0m     linting\u001b[39m=\u001b[39m\u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdialect\u001b[39m.\u001b[39mcompiler_linting \u001b[39m|\u001b[39m compiler\u001b[39m.\u001b[39mWARN_LINTING,\n\u001b[1;32m   1497\u001b[0m )\n\u001b[0;32m-> 1498\u001b[0m ret \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_execute_context(\n\u001b[1;32m   1499\u001b[0m     dialect,\n\u001b[1;32m   1500\u001b[0m     dialect\u001b[39m.\u001b[39;49mexecution_ctx_cls\u001b[39m.\u001b[39;49m_init_compiled,\n\u001b[1;32m   1501\u001b[0m     compiled_sql,\n\u001b[1;32m   1502\u001b[0m     distilled_params,\n\u001b[1;32m   1503\u001b[0m     execution_options,\n\u001b[1;32m   1504\u001b[0m     compiled_sql,\n\u001b[1;32m   1505\u001b[0m     distilled_params,\n\u001b[1;32m   1506\u001b[0m     elem,\n\u001b[1;32m   1507\u001b[0m     extracted_params,\n\u001b[1;32m   1508\u001b[0m     cache_hit\u001b[39m=\u001b[39;49mcache_hit,\n\u001b[1;32m   1509\u001b[0m )\n\u001b[1;32m   1510\u001b[0m \u001b[39mif\u001b[39;00m has_events:\n\u001b[1;32m   1511\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdispatch\u001b[39m.\u001b[39mafter_execute(\n\u001b[1;32m   1512\u001b[0m         \u001b[39mself\u001b[39m,\n\u001b[1;32m   1513\u001b[0m         elem,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1517\u001b[0m         ret,\n\u001b[1;32m   1518\u001b[0m     )\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1862\u001b[0m, in \u001b[0;36mConnection._execute_context\u001b[0;34m(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)\u001b[0m\n\u001b[1;32m   1859\u001b[0m             branched\u001b[39m.\u001b[39mclose()\n\u001b[1;32m   1861\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mBaseException\u001b[39;00m \u001b[39mas\u001b[39;00m e:\n\u001b[0;32m-> 1862\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_handle_dbapi_exception(\n\u001b[1;32m   1863\u001b[0m         e, statement, parameters, cursor, context\n\u001b[1;32m   1864\u001b[0m     )\n\u001b[1;32m   1866\u001b[0m \u001b[39mreturn\u001b[39;00m result\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/engine/base.py:2043\u001b[0m, in \u001b[0;36mConnection._handle_dbapi_exception\u001b[0;34m(self, e, statement, parameters, cursor, context)\u001b[0m\n\u001b[1;32m   2041\u001b[0m     util\u001b[39m.\u001b[39mraise_(newraise, with_traceback\u001b[39m=\u001b[39mexc_info[\u001b[39m2\u001b[39m], from_\u001b[39m=\u001b[39me)\n\u001b[1;32m   2042\u001b[0m \u001b[39melif\u001b[39;00m should_wrap:\n\u001b[0;32m-> 2043\u001b[0m     util\u001b[39m.\u001b[39;49mraise_(\n\u001b[1;32m   2044\u001b[0m         sqlalchemy_exception, with_traceback\u001b[39m=\u001b[39;49mexc_info[\u001b[39m2\u001b[39;49m], from_\u001b[39m=\u001b[39;49me\n\u001b[1;32m   2045\u001b[0m     )\n\u001b[1;32m   2046\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   2047\u001b[0m     util\u001b[39m.\u001b[39mraise_(exc_info[\u001b[39m1\u001b[39m], with_traceback\u001b[39m=\u001b[39mexc_info[\u001b[39m2\u001b[39m])\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/util/compat.py:208\u001b[0m, in \u001b[0;36mraise_\u001b[0;34m(***failed resolving arguments***)\u001b[0m\n\u001b[1;32m    205\u001b[0m     exception\u001b[39m.\u001b[39m__cause__ \u001b[39m=\u001b[39m replace_context\n\u001b[1;32m    207\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 208\u001b[0m     \u001b[39mraise\u001b[39;00m exception\n\u001b[1;32m    209\u001b[0m \u001b[39mfinally\u001b[39;00m:\n\u001b[1;32m    210\u001b[0m     \u001b[39m# credit to\u001b[39;00m\n\u001b[1;32m    211\u001b[0m     \u001b[39m# https://cosmicpercolator.com/2016/01/13/exception-leaks-in-python-2-and-3/\u001b[39;00m\n\u001b[1;32m    212\u001b[0m     \u001b[39m# as the __traceback__ object creates a cycle\u001b[39;00m\n\u001b[1;32m    213\u001b[0m     \u001b[39mdel\u001b[39;00m exception, replace_context, from_, with_traceback\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/engine/base.py:1819\u001b[0m, in \u001b[0;36mConnection._execute_context\u001b[0;34m(self, dialect, constructor, statement, parameters, execution_options, *args, **kw)\u001b[0m\n\u001b[1;32m   1817\u001b[0m                 \u001b[39mbreak\u001b[39;00m\n\u001b[1;32m   1818\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m evt_handled:\n\u001b[0;32m-> 1819\u001b[0m         \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mdialect\u001b[39m.\u001b[39;49mdo_execute(\n\u001b[1;32m   1820\u001b[0m             cursor, statement, parameters, context\n\u001b[1;32m   1821\u001b[0m         )\n\u001b[1;32m   1823\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_has_events \u001b[39mor\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mengine\u001b[39m.\u001b[39m_has_events:\n\u001b[1;32m   1824\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mdispatch\u001b[39m.\u001b[39mafter_cursor_execute(\n\u001b[1;32m   1825\u001b[0m         \u001b[39mself\u001b[39m,\n\u001b[1;32m   1826\u001b[0m         cursor,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1830\u001b[0m         context\u001b[39m.\u001b[39mexecutemany,\n\u001b[1;32m   1831\u001b[0m     )\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sqlalchemy/engine/default.py:732\u001b[0m, in \u001b[0;36mDefaultDialect.do_execute\u001b[0;34m(self, cursor, statement, parameters, context)\u001b[0m\n\u001b[1;32m    731\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39mdo_execute\u001b[39m(\u001b[39mself\u001b[39m, cursor, statement, parameters, context\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m--> 732\u001b[0m     cursor\u001b[39m.\u001b[39;49mexecute(statement, parameters)\n",
            "\u001b[0;31mNotSupportedError\u001b[0m: (psycopg2.errors.FeatureNotSupported) extension \"pgml\" is not available\nDETAIL:  Could not open extension control file \"/rdsdbbin/postgres-15.2.R2/share/extension/pgml.control\": No such file or directory.\nHINT:  The extension must first be installed on the system where PostgreSQL is running.\n\n[SQL: \n                               create extension pgml;]\n(Background on this error at: https://sqlalche.me/e/14/tw8g)"
          ]
        }
      ],
      "source": [
        "sql_create_postgres_env = text(f'''\n",
        "                               create extension pgml;''')\n",
        "with engine.connect() as conn: \n",
        "    query = conn.execute(sql_create_postgres_env)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "{'ids': ['220647965-0',\n",
              "  '220647965-1',\n",
              "  '220647965-2',\n",
              "  '220647965-3',\n",
              "  '220647965-4',\n",
              "  '220647965-5',\n",
              "  '20044380-0',\n",
              "  '20044380-1',\n",
              "  '20044380-2',\n",
              "  '20044380-3'],\n",
              " 'embeddings': [[-0.005119905341416597,\n",
              "   -0.07151228189468384,\n",
              "   -0.11324868351221085,\n",
              "   -0.09201928228139877,\n",
              "   0.00045634983689524233,\n",
              "   -0.024387119337916374,\n",
              "   0.028379959985613823,\n",
              "   0.04509858414530754,\n",
              "   0.05559356138110161,\n",
              "   0.012055506929755211,\n",
              "   -0.009709623642265797,\n",
              "   0.02123657427728176,\n",
              "   -0.0021696484182029963,\n",
              "   0.08374569565057755,\n",
              "   -0.05272151529788971,\n",
              "   0.11446069180965424,\n",
              "   0.011344498954713345,\n",
              "   0.009295452386140823,\n",
              "   -0.04702450707554817,\n",
              "   0.0028786389157176018,\n",
              "   -0.03945611044764519,\n",
              "   0.03999094292521477,\n",
              "   0.11128133535385132,\n",
              "   0.043151360005140305,\n",
              "   -0.07247650623321533,\n",
              "   0.007482609245926142,\n",
              "   0.009368007071316242,\n",
              "   0.054855529218912125,\n",
              "   -0.04732478782534599,\n",
              "   -0.06813782453536987,\n",
              "   0.08444996178150177,\n",
              "   0.04073921591043472,\n",
              "   -0.03744032606482506,\n",
              "   -0.05311126261949539,\n",
              "   -0.020215507596731186,\n",
              "   -0.0006251035956665874,\n",
              "   -0.09467754513025284,\n",
              "   0.027652250602841377,\n",
              "   0.03385273739695549,\n",
              "   -0.05787116289138794,\n",
              "   -0.0011094070505350828,\n",
              "   -0.03742683306336403,\n",
              "   -0.022017724812030792,\n",
              "   0.016742000356316566,\n",
              "   0.02952498383820057,\n",
              "   -0.12797077000141144,\n",
              "   -0.08032695949077606,\n",
              "   -0.05143599584698677,\n",
              "   0.07123086601495743,\n",
              "   0.06003155559301376,\n",
              "   -0.07456756383180618,\n",
              "   0.027782410383224487,\n",
              "   -0.01605100743472576,\n",
              "   0.02990124560892582,\n",
              "   -0.05412716791033745,\n",
              "   -0.035381536930799484,\n",
              "   -0.06878578662872314,\n",
              "   -0.028757641091942787,\n",
              "   -0.014204716309905052,\n",
              "   -0.06728637218475342,\n",
              "   0.02684232033789158,\n",
              "   -0.040794823318719864,\n",
              "   0.017295753583312035,\n",
              "   -0.018061229959130287,\n",
              "   -0.030021455138921738,\n",
              "   -0.019281182438135147,\n",
              "   -0.02884584851562977,\n",
              "   -0.01780518889427185,\n",
              "   0.0013925704406574368,\n",
              "   0.024884071201086044,\n",
              "   -0.055800169706344604,\n",
              "   -0.062318842858076096,\n",
              "   0.02584764175117016,\n",
              "   0.01533286739140749,\n",
              "   -0.046591866761446,\n",
              "   0.013576552271842957,\n",
              "   0.11364924162626266,\n",
              "   0.00576383713632822,\n",
              "   -0.014953678473830223,\n",
              "   0.0061080073937773705,\n",
              "   0.03629374876618385,\n",
              "   0.005544818006455898,\n",
              "   0.0836009606719017,\n",
              "   0.08921158313751221,\n",
              "   0.009148911572992802,\n",
              "   0.08508393168449402,\n",
              "   0.04495849087834358,\n",
              "   0.03634541109204292,\n",
              "   0.03495343402028084,\n",
              "   -0.008176715113222599,\n",
              "   0.04589567705988884,\n",
              "   0.17109236121177673,\n",
              "   0.04085356369614601,\n",
              "   0.024069976061582565,\n",
              "   -0.040606580674648285,\n",
              "   0.03141756355762482,\n",
              "   0.003983448725193739,\n",
              "   0.03276918828487396,\n",
              "   -0.04044102504849434,\n",
              "   0.039461150765419006,\n",
              "   0.04055006057024002,\n",
              "   -0.02630348689854145,\n",
              "   -0.07194703072309494,\n",
              "   -0.048160258680582047,\n",
              "   -0.06355971097946167,\n",
              "   -0.034395527094602585,\n",
              "   0.010917656123638153,\n",
              "   0.007130216807126999,\n",
              "   0.0059699397534132,\n",
              "   -0.026198433712124825,\n",
              "   -0.05512506887316704,\n",
              "   -0.04510047659277916,\n",
              "   -0.022665243595838547,\n",
              "   0.037295520305633545,\n",
              "   -0.0271278228610754,\n",
              "   0.049319252371788025,\n",
              "   0.0059521677903831005,\n",
              "   -0.009002920240163803,\n",
              "   0.016252409666776657,\n",
              "   0.047067079693078995,\n",
              "   0.04762159660458565,\n",
              "   -0.018225345760583878,\n",
              "   0.004621571861207485,\n",
              "   -0.0789421945810318,\n",
              "   0.006487048231065273,\n",
              "   0.04680205136537552,\n",
              "   -0.034601885825395584,\n",
              "   1.1631167626501135e-32,\n",
              "   -0.019548239186406136,\n",
              "   0.07630410045385361,\n",
              "   0.04814128205180168,\n",
              "   0.020194096490740776,\n",
              "   0.08964180946350098,\n",
              "   0.0514264740049839,\n",
              "   -0.04263899475336075,\n",
              "   -0.027049794793128967,\n",
              "   -0.10513424873352051,\n",
              "   -0.056634411215782166,\n",
              "   -0.1339838206768036,\n",
              "   -0.09844935685396194,\n",
              "   0.0553702637553215,\n",
              "   0.0303665678948164,\n",
              "   -0.057359784841537476,\n",
              "   -0.04482853412628174,\n",
              "   0.007602156605571508,\n",
              "   0.05849496275186539,\n",
              "   -0.02500443533062935,\n",
              "   -0.0831679031252861,\n",
              "   0.0620962530374527,\n",
              "   0.04151422902941704,\n",
              "   -0.02811223641037941,\n",
              "   -0.032999832183122635,\n",
              "   -0.039081886410713196,\n",
              "   0.03795275464653969,\n",
              "   0.001799903460778296,\n",
              "   0.09782125055789948,\n",
              "   -0.015848448500037193,\n",
              "   0.004430828616023064,\n",
              "   -0.02467116340994835,\n",
              "   0.01627720147371292,\n",
              "   -0.07830329239368439,\n",
              "   -0.08740530908107758,\n",
              "   0.0013250014744699001,\n",
              "   -0.05934200808405876,\n",
              "   -0.07312677055597305,\n",
              "   -0.028593609109520912,\n",
              "   0.05484054610133171,\n",
              "   0.017013663426041603,\n",
              "   -0.036652572453022,\n",
              "   -0.020379377529025078,\n",
              "   -0.05990813672542572,\n",
              "   0.05172919109463692,\n",
              "   0.02035595290362835,\n",
              "   -0.03389650955796242,\n",
              "   -0.003111664205789566,\n",
              "   0.0690518245100975,\n",
              "   -0.03227884694933891,\n",
              "   0.013788189738988876,\n",
              "   0.02476329728960991,\n",
              "   -0.07302097976207733,\n",
              "   0.004393118899315596,\n",
              "   0.06894262135028839,\n",
              "   -0.02210146002471447,\n",
              "   0.026108527556061745,\n",
              "   -0.0995739996433258,\n",
              "   0.04361740127205849,\n",
              "   0.06726258993148804,\n",
              "   -0.026258528232574463,\n",
              "   -0.08698582649230957,\n",
              "   -0.028778154402971268,\n",
              "   -0.03855820372700691,\n",
              "   0.01689876988530159,\n",
              "   -0.008590102195739746,\n",
              "   -0.08192413300275803,\n",
              "   0.004668745212256908,\n",
              "   -0.04599344730377197,\n",
              "   0.01686832122504711,\n",
              "   0.04104551672935486,\n",
              "   0.06147473305463791,\n",
              "   -0.02566719986498356,\n",
              "   0.06916435807943344,\n",
              "   0.06073461472988129,\n",
              "   0.09907449036836624,\n",
              "   0.022538213059306145,\n",
              "   0.06964915245771408,\n",
              "   0.02408255822956562,\n",
              "   0.013375760987401009,\n",
              "   0.07022766023874283,\n",
              "   -0.03462446480989456,\n",
              "   0.005080505274236202,\n",
              "   -0.06774523109197617,\n",
              "   -0.0003278667281847447,\n",
              "   -0.08148021250963211,\n",
              "   -0.10703947395086288,\n",
              "   -0.03164586424827576,\n",
              "   -0.011165131814777851,\n",
              "   0.01786208339035511,\n",
              "   -0.00710050156340003,\n",
              "   0.02633124589920044,\n",
              "   -0.08655443042516708,\n",
              "   -0.05937293544411659,\n",
              "   -0.0007504976820200682,\n",
              "   -0.009011083282530308,\n",
              "   -1.2094040567656488e-32,\n",
              "   0.09958890080451965,\n",
              "   0.04268789663910866,\n",
              "   0.03468776494264603,\n",
              "   -0.07310187071561813,\n",
              "   0.051267433911561966,\n",
              "   -0.015804564580321312,\n",
              "   -0.030436517670750618,\n",
              "   -0.09577895700931549,\n",
              "   0.06486444920301437,\n",
              "   -0.014715990982949734,\n",
              "   0.1285076141357422,\n",
              "   0.012807497754693031,\n",
              "   0.04072052612900734,\n",
              "   0.007929117418825626,\n",
              "   -0.013541744090616703,\n",
              "   0.0698506310582161,\n",
              "   0.07938341796398163,\n",
              "   -0.014638402499258518,\n",
              "   -0.02948484942317009,\n",
              "   0.0178113654255867,\n",
              "   -0.012324081733822823,\n",
              "   0.11676028370857239,\n",
              "   -0.12352794408798218,\n",
              "   0.01935862936079502,\n",
              "   -0.01658787950873375,\n",
              "   0.013984955847263336,\n",
              "   0.059635527431964874,\n",
              "   -0.01932441256940365,\n",
              "   0.03391094133257866,\n",
              "   0.0916295051574707,\n",
              "   0.0027666790410876274,\n",
              "   0.015913188457489014,\n",
              "   -0.0532645583152771,\n",
              "   0.00767865777015686,\n",
              "   0.05161130428314209,\n",
              "   0.008888592012226582,\n",
              "   0.07349412888288498,\n",
              "   -0.08208175748586655,\n",
              "   -0.018297873437404633,\n",
              "   0.01630297116935253,\n",
              "   0.006101557053625584,\n",
              "   0.042103253304958344,\n",
              "   -0.0364467017352581,\n",
              "   -0.05424446985125542,\n",
              "   0.051235634833574295,\n",
              "   -0.009894569404423237,\n",
              "   -0.009885326027870178,\n",
              "   0.05941745266318321,\n",
              "   -0.009575922042131424,\n",
              "   -0.03654801845550537,\n",
              "   -0.0050314306281507015,\n",
              "   -0.04146185889840126,\n",
              "   0.02507575787603855,\n",
              "   0.00393027113750577,\n",
              "   0.11595418304204941,\n",
              "   0.032028064131736755,\n",
              "   0.015193917788565159,\n",
              "   -0.08922804892063141,\n",
              "   0.04010188207030296,\n",
              "   -0.012138963676989079,\n",
              "   0.0033400743268430233,\n",
              "   0.006384639535099268,\n",
              "   0.0489225834608078,\n",
              "   0.059150487184524536,\n",
              "   0.029915504157543182,\n",
              "   0.04230307415127754,\n",
              "   -0.019809197634458542,\n",
              "   0.023347605019807816,\n",
              "   0.043769240379333496,\n",
              "   -0.050795089453458786,\n",
              "   -0.013771635480225086,\n",
              "   0.01744106411933899,\n",
              "   -0.008307875134050846,\n",
              "   0.06968345493078232,\n",
              "   -0.011915896087884903,\n",
              "   0.047568779438734055,\n",
              "   -0.057258427143096924,\n",
              "   -0.004690965171903372,\n",
              "   -0.035875409841537476,\n",
              "   0.014761083759367466,\n",
              "   0.02923683635890484,\n",
              "   -0.04007578641176224,\n",
              "   0.025890300050377846,\n",
              "   -0.07591235637664795,\n",
              "   -0.047286488115787506,\n",
              "   -0.040010642260313034,\n",
              "   0.009800011292099953,\n",
              "   0.041199956089258194,\n",
              "   -0.004457308445125818,\n",
              "   -0.06256028264760971,\n",
              "   0.03500619903206825,\n",
              "   -0.03164411708712578,\n",
              "   -0.008729718625545502,\n",
              "   0.1448160856962204,\n",
              "   0.05098456144332886,\n",
              "   -5.667649816132325e-08,\n",
              "   0.04552558436989784,\n",
              "   -0.14494885504245758,\n",
              "   -0.01714271865785122,\n",
              "   -0.11624489724636078,\n",
              "   -0.05519966036081314,\n",
              "   0.019103841856122017,\n",
              "   -0.018483661115169525,\n",
              "   -0.010680902749300003,\n",
              "   0.10378718376159668,\n",
              "   0.029039600864052773,\n",
              "   -0.0006207487313076854,\n",
              "   0.051209114491939545,\n",
              "   -0.00207632128149271,\n",
              "   0.061567965894937515,\n",
              "   -0.026897646486759186,\n",
              "   0.021153651177883148,\n",
              "   0.016178082674741745,\n",
              "   -0.02501019462943077,\n",
              "   -0.048041604459285736,\n",
              "   -0.019774576649069786,\n",
              "   -0.00681261857971549,\n",
              "   -0.00655484339222312,\n",
              "   -0.036808691918849945,\n",
              "   -0.04307912290096283,\n",
              "   0.0727047547698021,\n",
              "   -0.033699240535497665,\n",
              "   -0.020984407514333725,\n",
              "   -0.00042101586586795747,\n",
              "   0.019858133047819138,\n",
              "   -0.04467127099633217,\n",
              "   -0.022004634141921997,\n",
              "   -0.004555344115942717,\n",
              "   0.10267955809831619,\n",
              "   0.005125486757606268,\n",
              "   0.010916639119386673,\n",
              "   0.009279392659664154,\n",
              "   0.018223803490400314,\n",
              "   -0.06698024272918701,\n",
              "   0.008098178543150425,\n",
              "   0.11122651398181915,\n",
              "   0.05366033688187599,\n",
              "   -0.08099259436130524,\n",
              "   -0.054397668689489365,\n",
              "   -0.040087372064590454,\n",
              "   -0.011957738548517227,\n",
              "   -0.09655839204788208,\n",
              "   0.026221277192234993,\n",
              "   -0.013105328194797039,\n",
              "   -0.07770681381225586,\n",
              "   -0.08727739751338959,\n",
              "   -0.020705267786979675,\n",
              "   0.03227957710623741,\n",
              "   0.0013156547211110592,\n",
              "   -0.033838801085948944,\n",
              "   -0.023670010268688202,\n",
              "   0.07184009253978729,\n",
              "   -0.06651438027620316,\n",
              "   0.05404360219836235,\n",
              "   0.03070203959941864,\n",
              "   -0.09840656816959381,\n",
              "   0.0018445114837959409,\n",
              "   -0.05030288174748421,\n",
              "   0.029676176607608795,\n",
              "   -0.023130908608436584],\n",
              "  [0.04754897207021713,\n",
              "   -0.09599931538105011,\n",
              "   -0.12249145656824112,\n",
              "   -0.03506482020020485,\n",
              "   0.03120911866426468,\n",
              "   -0.058353062719106674,\n",
              "   0.03759680315852165,\n",
              "   0.12144768983125687,\n",
              "   0.03653908893465996,\n",
              "   0.002077482407912612,\n",
              "   -0.05691555142402649,\n",
              "   0.1020406112074852,\n",
              "   0.03591756150126457,\n",
              "   0.09282480180263519,\n",
              "   -0.04883643984794617,\n",
              "   0.049470290541648865,\n",
              "   0.02363290823996067,\n",
              "   0.05159487947821617,\n",
              "   -0.029133696109056473,\n",
              "   0.028843004256486893,\n",
              "   0.04180411621928215,\n",
              "   -0.049803514033555984,\n",
              "   0.08560922741889954,\n",
              "   0.038932912051677704,\n",
              "   -0.004261158872395754,\n",
              "   0.03136803209781647,\n",
              "   0.015728341415524483,\n",
              "   0.026381636038422585,\n",
              "   -0.08761588484048843,\n",
              "   -0.023588305339217186,\n",
              "   0.030127665027976036,\n",
              "   0.024885188788175583,\n",
              "   0.06436506658792496,\n",
              "   -0.11429942399263382,\n",
              "   -0.011076908558607101,\n",
              "   0.008002480491995811,\n",
              "   -0.04761311784386635,\n",
              "   -0.05445028468966484,\n",
              "   0.07678429037332535,\n",
              "   0.058800917118787766,\n",
              "   0.035475149750709534,\n",
              "   0.029401864856481552,\n",
              "   0.002136436989530921,\n",
              "   0.04510597884654999,\n",
              "   -0.03010631538927555,\n",
              "   -0.07361655682325363,\n",
              "   -0.01254186499863863,\n",
              "   0.007711658254265785,\n",
              "   0.007012758869677782,\n",
              "   0.04614781588315964,\n",
              "   -0.022273816168308258,\n",
              "   0.0033507707994431257,\n",
              "   -0.08582726866006851,\n",
              "   0.022026779130101204,\n",
              "   -0.020616047084331512,\n",
              "   -0.04837225377559662,\n",
              "   -0.060514312237501144,\n",
              "   0.018661439418792725,\n",
              "   -0.04165787249803543,\n",
              "   -0.02546721324324608,\n",
              "   -0.05072234943509102,\n",
              "   0.028349237516522408,\n",
              "   -0.0067651160061359406,\n",
              "   -0.021297426894307137,\n",
              "   -0.042600229382514954,\n",
              "   0.022207967936992645,\n",
              "   -0.015927987173199654,\n",
              "   -0.0661701187491417,\n",
              "   0.05289275571703911,\n",
              "   0.027193808928132057,\n",
              "   -0.08615998178720474,\n",
              "   -0.009290697053074837,\n",
              "   0.07807563245296478,\n",
              "   0.047383662313222885,\n",
              "   -0.028689615428447723,\n",
              "   0.013470714911818504,\n",
              "   0.14666812121868134,\n",
              "   -0.030292369425296783,\n",
              "   -0.09789205342531204,\n",
              "   -0.03805076703429222,\n",
              "   0.03427945449948311,\n",
              "   -0.03056107833981514,\n",
              "   -0.013117102906107903,\n",
              "   0.04098721593618393,\n",
              "   0.05052072927355766,\n",
              "   0.02532719448208809,\n",
              "   0.017721815034747124,\n",
              "   0.049221061170101166,\n",
              "   0.06046115979552269,\n",
              "   -0.011978153139352798,\n",
              "   0.06355240941047668,\n",
              "   0.1192157194018364,\n",
              "   -0.06401997804641724,\n",
              "   -0.01774844154715538,\n",
              "   -0.00219049584120512,\n",
              "   0.011837578378617764,\n",
              "   -0.03237229958176613,\n",
              "   0.011629258282482624,\n",
              "   0.01865018904209137,\n",
              "   -0.011450367979705334,\n",
              "   0.04600301384925842,\n",
              "   -0.06102687492966652,\n",
              "   -0.010289940983057022,\n",
              "   0.044211748987436295,\n",
              "   0.03459462150931358,\n",
              "   -0.010272328741848469,\n",
              "   -0.029550453647971153,\n",
              "   0.05249602720141411,\n",
              "   -0.07888416200876236,\n",
              "   0.046373941004276276,\n",
              "   0.004709053318947554,\n",
              "   0.011831622570753098,\n",
              "   0.029506182298064232,\n",
              "   0.04462137818336487,\n",
              "   -0.012500576674938202,\n",
              "   0.016723722219467163,\n",
              "   0.04171103239059448,\n",
              "   -0.061377473175525665,\n",
              "   -0.0678846463561058,\n",
              "   0.01159389317035675,\n",
              "   0.040679436177015305,\n",
              "   0.05079971253871918,\n",
              "   0.10708418488502502,\n",
              "   -0.09292297810316086,\n",
              "   0.05471237748861313,\n",
              "   0.0480477549135685,\n",
              "   0.035550083965063095,\n",
              "   4.8208257596290374e-33,\n",
              "   -0.015341594815254211,\n",
              "   0.022604618221521378,\n",
              "   0.02159273810684681,\n",
              "   0.05291718617081642,\n",
              "   0.05212090536952019,\n",
              "   0.14011360704898834,\n",
              "   -0.028377383947372437,\n",
              "   -0.0009408874902874231,\n",
              "   0.01907791756093502,\n",
              "   -0.054290737956762314,\n",
              "   -0.052072349935770035,\n",
              "   -0.01890314742922783,\n",
              "   0.038489799946546555,\n",
              "   -0.021395228803157806,\n",
              "   -0.022542214021086693,\n",
              "   0.07216061651706696,\n",
              "   0.06104939430952072,\n",
              "   0.030358286574482918,\n",
              "   0.05943337455391884,\n",
              "   0.05423574522137642,\n",
              "   0.011575485579669476,\n",
              "   -0.04850291833281517,\n",
              "   -0.035140324383974075,\n",
              "   0.041355375200510025,\n",
              "   -0.003909564111381769,\n",
              "   0.016126582399010658,\n",
              "   -0.014214370399713516,\n",
              "   0.05328750237822533,\n",
              "   -0.004567726515233517,\n",
              "   -0.025327635928988457,\n",
              "   -0.030662985518574715,\n",
              "   -0.047847647219896317,\n",
              "   -0.0239198487251997,\n",
              "   -0.04038700461387634,\n",
              "   0.020052233710885048,\n",
              "   -0.006951584946364164,\n",
              "   -0.12550963461399078,\n",
              "   -0.00746981892734766,\n",
              "   0.05635659396648407,\n",
              "   0.037748783826828,\n",
              "   -0.005650071892887354,\n",
              "   -0.0401458777487278,\n",
              "   0.03502785041928291,\n",
              "   0.03190479055047035,\n",
              "   -0.007140517700463533,\n",
              "   0.046578921377658844,\n",
              "   -0.09677569568157196,\n",
              "   0.0973796397447586,\n",
              "   0.03228207305073738,\n",
              "   -0.04275044426321983,\n",
              "   0.030430354177951813,\n",
              "   -0.07830113172531128,\n",
              "   0.036858197301626205,\n",
              "   0.013516340404748917,\n",
              "   -0.03301665186882019,\n",
              "   0.039747025817632675,\n",
              "   -0.06225854903459549,\n",
              "   -0.002769975457340479,\n",
              "   0.06764226406812668,\n",
              "   -0.039492085576057434,\n",
              "   -0.044161275029182434,\n",
              "   -0.015444722026586533,\n",
              "   -0.014758514240384102,\n",
              "   0.07797279208898544,\n",
              "   0.03499316796660423,\n",
              "   -0.07369222491979599,\n",
              "   -0.06385480612516403,\n",
              "   0.005820122547447681,\n",
              "   0.0009148523095063865,\n",
              "   -0.0014461346436291933,\n",
              "   0.05260365456342697,\n",
              "   -0.09510993957519531,\n",
              "   0.013614585623145103,\n",
              "   0.032214947044849396,\n",
              "   0.025627728551626205,\n",
              "   0.03863541781902313,\n",
              "   0.06983011960983276,\n",
              "   0.034000806510448456,\n",
              "   0.042467232793569565,\n",
              "   0.03277308866381645,\n",
              "   0.026128029450774193,\n",
              "   0.02257632091641426,\n",
              "   -0.02367941476404667,\n",
              "   -0.010921219363808632,\n",
              "   -0.06426136940717697,\n",
              "   -0.028443412855267525,\n",
              "   -0.049905095249414444,\n",
              "   0.03983194753527641,\n",
              "   0.06330952793359756,\n",
              "   0.001284802216105163,\n",
              "   0.005681908223778009,\n",
              "   -0.09784693270921707,\n",
              "   -0.08247257769107819,\n",
              "   -0.04586615785956383,\n",
              "   -0.027006227523088455,\n",
              "   -4.2052362658346985e-33,\n",
              "   0.07591462135314941,\n",
              "   0.06667027622461319,\n",
              "   0.013572216965258121,\n",
              "   -0.0329526886343956,\n",
              "   -0.0014490581816062331,\n",
              "   0.007555150426924229,\n",
              "   0.015372338704764843,\n",
              "   -0.01814132183790207,\n",
              "   0.04981497675180435,\n",
              "   -0.05605309084057808,\n",
              "   0.09484498202800751,\n",
              "   0.020784132182598114,\n",
              "   -0.1048479750752449,\n",
              "   -0.06686093658208847,\n",
              "   -0.007957644760608673,\n",
              "   0.00184071552939713,\n",
              "   -0.007741341833025217,\n",
              "   0.09998281300067902,\n",
              "   0.01787465438246727,\n",
              "   0.013896933756768703,\n",
              "   -0.07081195712089539,\n",
              "   0.07593423873186111,\n",
              "   -0.04230687394738197,\n",
              "   0.07922469079494476,\n",
              "   0.055383797734975815,\n",
              "   -0.051691994071006775,\n",
              "   0.01842000149190426,\n",
              "   -0.054792262613773346,\n",
              "   0.055097971111536026,\n",
              "   0.0587284155189991,\n",
              "   0.0981043353676796,\n",
              "   -0.004112342372536659,\n",
              "   -0.011754763312637806,\n",
              "   0.019278766587376595,\n",
              "   0.05704247206449509,\n",
              "   0.051865462213754654,\n",
              "   0.011841927655041218,\n",
              "   -0.05514282360672951,\n",
              "   -0.016873735934495926,\n",
              "   0.06228704750537872,\n",
              "   0.01636938564479351,\n",
              "   0.005548679269850254,\n",
              "   -0.04670711234211922,\n",
              "   -0.050734568387269974,\n",
              "   0.0029721956234425306,\n",
              "   0.021706508472561836,\n",
              "   -0.03788985311985016,\n",
              "   -0.02470095269382,\n",
              "   0.009648068808019161,\n",
              "   -0.07722797244787216,\n",
              "   0.009504587389528751,\n",
              "   -0.10378143191337585,\n",
              "   -0.022319231182336807,\n",
              "   -0.05419734865427017,\n",
              "   0.03196612000465393,\n",
              "   0.020694022998213768,\n",
              "   0.06263825297355652,\n",
              "   -0.017445949837565422,\n",
              "   0.09675604850053787,\n",
              "   -0.0717785581946373,\n",
              "   0.02176824025809765,\n",
              "   -0.0037372312508523464,\n",
              "   0.013767766766250134,\n",
              "   -0.037754811346530914,\n",
              "   -0.08451786637306213,\n",
              "   0.023272963240742683,\n",
              "   0.01247422955930233,\n",
              "   0.08438902348279953,\n",
              "   0.03773995861411095,\n",
              "   0.015696905553340912,\n",
              "   -0.0014700808096677065,\n",
              "   0.04137663543224335,\n",
              "   0.01948942244052887,\n",
              "   0.03569202125072479,\n",
              "   -0.02163391187787056,\n",
              "   0.08608226478099823,\n",
              "   -0.10272490233182907,\n",
              "   -0.01761697232723236,\n",
              "   -0.023262515664100647,\n",
              "   0.0068892245180904865,\n",
              "   0.019358377903699875,\n",
              "   -0.038749706000089645,\n",
              "   -0.01810838282108307,\n",
              "   0.003360619070008397,\n",
              "   -0.09227333217859268,\n",
              "   -0.08846118301153183,\n",
              "   -0.0024244787637144327,\n",
              "   0.05970274284482002,\n",
              "   -0.04324223846197128,\n",
              "   -0.038793034851551056,\n",
              "   0.008663279935717583,\n",
              "   -0.09933847934007645,\n",
              "   -0.031031712889671326,\n",
              "   0.09803178906440735,\n",
              "   0.07035988569259644,\n",
              "   -5.033028216416824e-08,\n",
              "   0.02616390958428383,\n",
              "   -0.028603767976164818,\n",
              "   -0.01575288362801075,\n",
              "   0.0032925670966506004,\n",
              "   -0.10347512364387512,\n",
              "   -0.021368063986301422,\n",
              "   0.009896285831928253,\n",
              "   -0.011925403960049152,\n",
              "   0.10655615478754044,\n",
              "   0.039198294281959534,\n",
              "   -0.12536610662937164,\n",
              "   0.05760817974805832,\n",
              "   -0.008687611669301987,\n",
              "   0.056026168167591095,\n",
              "   -0.09648285061120987,\n",
              "   0.029314449056982994,\n",
              "   -0.049683742225170135,\n",
              "   -0.053971052169799805,\n",
              "   -0.019483722746372223,\n",
              "   0.03456566110253334,\n",
              "   -0.05515388399362564,\n",
              "   0.03005577065050602,\n",
              "   -0.08844467252492905,\n",
              "   0.0013486731331795454,\n",
              "   0.09236694872379303,\n",
              "   -0.061275236308574677,\n",
              "   0.01694602146744728,\n",
              "   0.010253600776195526,\n",
              "   0.017964381724596024,\n",
              "   -0.009386492893099785,\n",
              "   0.08920087665319443,\n",
              "   -0.033671241253614426,\n",
              "   0.05729757249355316,\n",
              "   0.00883860606700182,\n",
              "   -0.034399088472127914,\n",
              "   -0.058227553963661194,\n",
              "   2.4008761556615354e-06,\n",
              "   0.00026894506299868226,\n",
              "   0.03231215476989746,\n",
              "   0.0966804102063179,\n",
              "   -0.0006483122706413269,\n",
              "   -0.13197526335716248,\n",
              "   -0.02092840149998665,\n",
              "   0.005620595999062061,\n",
              "   -0.0580979660153389,\n",
              "   -0.044831257313489914,\n",
              "   -0.07905002683401108,\n",
              "   0.023498041555285454,\n",
              "   -0.07024542987346649,\n",
              "   -0.05900215357542038,\n",
              "   0.05686967447400093,\n",
              "   0.034116845577955246,\n",
              "   -0.05662116780877113,\n",
              "   -0.05700559541583061,\n",
              "   -0.03738594055175781,\n",
              "   0.004343108739703894,\n",
              "   -0.010400013998150826,\n",
              "   -0.0012106066569685936,\n",
              "   -0.011448941193521023,\n",
              "   -0.10172943025827408,\n",
              "   -0.05888138711452484,\n",
              "   -0.022057009860873222,\n",
              "   0.09900611639022827,\n",
              "   -0.047255273908376694],\n",
              "  [0.04430558159947395,\n",
              "   -0.04939784109592438,\n",
              "   -0.10671574622392654,\n",
              "   -0.07076289504766464,\n",
              "   0.02284681797027588,\n",
              "   -0.003881315467879176,\n",
              "   0.003441686276346445,\n",
              "   -0.0067611937411129475,\n",
              "   0.03427709639072418,\n",
              "   0.028871923685073853,\n",
              "   -0.06177197024226189,\n",
              "   0.07321511954069138,\n",
              "   -0.018481049686670303,\n",
              "   0.0754205584526062,\n",
              "   -0.02378339320421219,\n",
              "   0.08134376257658005,\n",
              "   -0.024160318076610565,\n",
              "   0.008200409822165966,\n",
              "   0.002350153401494026,\n",
              "   0.09102530777454376,\n",
              "   0.005506679881364107,\n",
              "   -0.03940657526254654,\n",
              "   0.0884193480014801,\n",
              "   0.11464128643274307,\n",
              "   -0.054324690252542496,\n",
              "   0.01651771180331707,\n",
              "   -0.01939319260418415,\n",
              "   0.05421070754528046,\n",
              "   -0.11129456758499146,\n",
              "   -0.01677924394607544,\n",
              "   0.02322273515164852,\n",
              "   -0.012605133466422558,\n",
              "   -0.007189351599663496,\n",
              "   -0.01913660205900669,\n",
              "   -0.029641639441251755,\n",
              "   0.016909370198845863,\n",
              "   -0.06819482147693634,\n",
              "   0.010263640433549881,\n",
              "   -0.01556757278740406,\n",
              "   0.010489658452570438,\n",
              "   -0.018440699204802513,\n",
              "   -0.03900550678372383,\n",
              "   -0.022432737052440643,\n",
              "   0.03372691199183464,\n",
              "   -0.0016659221146255732,\n",
              "   -0.12373381108045578,\n",
              "   -0.013393537141382694,\n",
              "   -0.035567719489336014,\n",
              "   0.03621762618422508,\n",
              "   0.07960447669029236,\n",
              "   -0.028209950774908066,\n",
              "   0.017184386029839516,\n",
              "   -0.02261047437787056,\n",
              "   0.06223399564623833,\n",
              "   -0.02301080897450447,\n",
              "   -0.040364522486925125,\n",
              "   -0.06262093037366867,\n",
              "   0.060131072998046875,\n",
              "   0.002218645764514804,\n",
              "   0.0033661257475614548,\n",
              "   0.02201487310230732,\n",
              "   -0.049613550305366516,\n",
              "   0.006738787516951561,\n",
              "   0.03135613352060318,\n",
              "   -0.08018023520708084,\n",
              "   0.020679611712694168,\n",
              "   -0.05215210095047951,\n",
              "   -0.04313893988728523,\n",
              "   -0.007338490337133408,\n",
              "   0.06200836971402168,\n",
              "   -0.06332732737064362,\n",
              "   -0.06420469284057617,\n",
              "   0.004921340383589268,\n",
              "   -0.05738041177392006,\n",
              "   -0.031210074201226234,\n",
              "   0.024489646777510643,\n",
              "   0.1368878334760666,\n",
              "   -0.0006052659591659904,\n",
              "   -0.08073415607213974,\n",
              "   -0.040211766958236694,\n",
              "   0.0008659328450448811,\n",
              "   -0.015070569701492786,\n",
              "   0.07852237671613693,\n",
              "   0.05294131487607956,\n",
              "   0.030073044821619987,\n",
              "   0.008667532354593277,\n",
              "   0.03795597329735756,\n",
              "   0.055878616869449615,\n",
              "   0.0972471833229065,\n",
              "   -0.0011922618141397834,\n",
              "   0.06241978704929352,\n",
              "   0.12166573107242584,\n",
              "   0.05177908390760422,\n",
              "   0.024768806993961334,\n",
              "   0.010039793327450752,\n",
              "   0.05288737267255783,\n",
              "   -0.018374262377619743,\n",
              "   0.07322068512439728,\n",
              "   0.016171075403690338,\n",
              "   -0.01958109438419342,\n",
              "   0.09443827718496323,\n",
              "   -0.03451709449291229,\n",
              "   -0.07516231387853622,\n",
              "   -0.025600198656320572,\n",
              "   0.01907218247652054,\n",
              "   0.024653464555740356,\n",
              "   0.047786030918359756,\n",
              "   0.02360055409371853,\n",
              "   -0.059054676443338394,\n",
              "   0.007712078280746937,\n",
              "   -0.032877445220947266,\n",
              "   -0.0004211578634567559,\n",
              "   -0.04269711300730705,\n",
              "   0.043044861406087875,\n",
              "   -0.04482262581586838,\n",
              "   0.04907730594277382,\n",
              "   0.06605110317468643,\n",
              "   -0.046055328100919724,\n",
              "   -0.01795269548892975,\n",
              "   0.05198762193322182,\n",
              "   0.02069828473031521,\n",
              "   -0.05918119475245476,\n",
              "   0.10929329693317413,\n",
              "   -0.06544575095176697,\n",
              "   0.027278201654553413,\n",
              "   0.04240702837705612,\n",
              "   -0.017203209921717644,\n",
              "   2.9198530296776972e-33,\n",
              "   0.01842149905860424,\n",
              "   0.051367029547691345,\n",
              "   0.07997197657823563,\n",
              "   0.05227328836917877,\n",
              "   0.0928584411740303,\n",
              "   0.05125187709927559,\n",
              "   -0.05770467594265938,\n",
              "   -0.05160534754395485,\n",
              "   -0.051093511283397675,\n",
              "   -0.030276155099272728,\n",
              "   -0.035652536898851395,\n",
              "   0.038674090057611465,\n",
              "   0.026724141091108322,\n",
              "   0.0012944587506353855,\n",
              "   -0.0054378085769712925,\n",
              "   -0.0033921825233846903,\n",
              "   0.003277544630691409,\n",
              "   0.022457143291831017,\n",
              "   0.07833579182624817,\n",
              "   -0.069615937769413,\n",
              "   0.02479606680572033,\n",
              "   -0.007210674230009317,\n",
              "   -0.06638972461223602,\n",
              "   0.004885429050773382,\n",
              "   -0.08269781619310379,\n",
              "   0.1032952070236206,\n",
              "   -0.015567154623568058,\n",
              "   0.13142281770706177,\n",
              "   -0.03982086852192879,\n",
              "   -0.002983539365231991,\n",
              "   -0.06804335862398148,\n",
              "   -0.08420955389738083,\n",
              "   -0.0004245394666213542,\n",
              "   -0.04081076756119728,\n",
              "   0.05155022442340851,\n",
              "   0.015934502705931664,\n",
              "   -0.10053496062755585,\n",
              "   0.025752075016498566,\n",
              "   0.04359005391597748,\n",
              "   0.04582374915480614,\n",
              "   0.0013357562711462379,\n",
              "   -0.07022296637296677,\n",
              "   0.04341771453619003,\n",
              "   0.003732943907380104,\n",
              "   -0.016562845557928085,\n",
              "   0.021124612540006638,\n",
              "   -0.09151643514633179,\n",
              "   0.05903155729174614,\n",
              "   -0.04218932241201401,\n",
              "   0.0022119542118161917,\n",
              "   0.0017747117672115564,\n",
              "   -0.10560678690671921,\n",
              "   -0.010392316617071629,\n",
              "   0.10978240519762039,\n",
              "   -0.0405663400888443,\n",
              "   -0.00077438959851861,\n",
              "   -0.10214361548423767,\n",
              "   0.004093633033335209,\n",
              "   0.10419374704360962,\n",
              "   -0.025548851117491722,\n",
              "   -0.035608816891908646,\n",
              "   -0.04727397486567497,\n",
              "   0.008865220472216606,\n",
              "   0.03344832360744476,\n",
              "   -0.008452929556369781,\n",
              "   -0.03211777284741402,\n",
              "   -0.005538241937756538,\n",
              "   -0.05066912993788719,\n",
              "   -0.046657562255859375,\n",
              "   0.012726940214633942,\n",
              "   0.02219613827764988,\n",
              "   -0.029618460685014725,\n",
              "   0.06744260340929031,\n",
              "   0.040066491812467575,\n",
              "   0.11793194711208344,\n",
              "   0.0962710902094841,\n",
              "   0.03258958086371422,\n",
              "   0.010976981371641159,\n",
              "   0.021405939012765884,\n",
              "   0.11904305964708328,\n",
              "   0.002764233620837331,\n",
              "   -0.04750685393810272,\n",
              "   -0.06570931524038315,\n",
              "   0.01955220103263855,\n",
              "   -0.1054409071803093,\n",
              "   -0.06385575234889984,\n",
              "   -0.066709965467453,\n",
              "   -0.03666678071022034,\n",
              "   0.05938633531332016,\n",
              "   -0.012139959260821342,\n",
              "   -0.006411433219909668,\n",
              "   -0.09048042446374893,\n",
              "   -0.002574553247541189,\n",
              "   -0.07409331202507019,\n",
              "   -0.008316499181091785,\n",
              "   -3.8428379696319106e-33,\n",
              "   0.07843245565891266,\n",
              "   0.036823198199272156,\n",
              "   0.09106526523828506,\n",
              "   -0.07915870100259781,\n",
              "   0.011752921156585217,\n",
              "   -0.019181422889232635,\n",
              "   0.038751501590013504,\n",
              "   -0.10276996344327927,\n",
              "   0.045111216604709625,\n",
              "   -0.016605453565716743,\n",
              "   0.05670111998915672,\n",
              "   -0.05034114792943001,\n",
              "   -0.0506882406771183,\n",
              "   -0.04143417999148369,\n",
              "   -0.06425902247428894,\n",
              "   -0.013989181257784367,\n",
              "   0.09482020139694214,\n",
              "   -0.041479840874671936,\n",
              "   -0.005340139847248793,\n",
              "   -0.0006981782498769462,\n",
              "   -0.01538971345871687,\n",
              "   0.042639702558517456,\n",
              "   -0.07399160414934158,\n",
              "   0.019892320036888123,\n",
              "   -0.022369051352143288,\n",
              "   0.04785332828760147,\n",
              "   0.06649685651063919,\n",
              "   -0.06433720141649246,\n",
              "   0.04925636574625969,\n",
              "   0.07275956869125366,\n",
              "   -0.011652534827589989,\n",
              "   0.01739323139190674,\n",
              "   -0.03703956678509712,\n",
              "   -0.021970950067043304,\n",
              "   0.06521044671535492,\n",
              "   0.0017824972746893764,\n",
              "   -0.049179673194885254,\n",
              "   -0.05293317139148712,\n",
              "   -0.0034938689786940813,\n",
              "   -0.030940880998969078,\n",
              "   -0.003972321283072233,\n",
              "   0.018803073093295097,\n",
              "   -0.019572928547859192,\n",
              "   -0.013947251252830029,\n",
              "   0.03799665719270706,\n",
              "   0.016401143744587898,\n",
              "   -0.0450119823217392,\n",
              "   0.04783952608704567,\n",
              "   -0.02725907787680626,\n",
              "   -0.0322113111615181,\n",
              "   0.0034689605236053467,\n",
              "   0.016739197075366974,\n",
              "   -0.038093019276857376,\n",
              "   0.01512171234935522,\n",
              "   0.11755391210317612,\n",
              "   0.023060401901602745,\n",
              "   0.020496046170592308,\n",
              "   -0.04529250040650368,\n",
              "   0.10796350240707397,\n",
              "   -0.01792421005666256,\n",
              "   -0.01922920159995556,\n",
              "   0.021042313426733017,\n",
              "   -0.044100284576416016,\n",
              "   -0.0403483621776104,\n",
              "   -0.04509023204445839,\n",
              "   0.05719512328505516,\n",
              "   -0.05655814707279205,\n",
              "   0.018428973853588104,\n",
              "   0.05059010162949562,\n",
              "   -0.004776509944349527,\n",
              "   0.01848025433719158,\n",
              "   -0.0031320354901254177,\n",
              "   0.0042704809457063675,\n",
              "   0.09976226836442947,\n",
              "   -0.12310265004634857,\n",
              "   0.08242256939411163,\n",
              "   0.009545455686748028,\n",
              "   -0.0977431982755661,\n",
              "   -0.029454058036208153,\n",
              "   0.013185277581214905,\n",
              "   -0.009366988204419613,\n",
              "   -0.05478347837924957,\n",
              "   -0.00045852866605855525,\n",
              "   -0.08731849491596222,\n",
              "   -0.0858197957277298,\n",
              "   -0.026266176253557205,\n",
              "   0.02253800630569458,\n",
              "   0.02236064337193966,\n",
              "   -0.06251342594623566,\n",
              "   -0.021152207627892494,\n",
              "   0.03788017854094505,\n",
              "   -0.044384490698575974,\n",
              "   -0.017811601981520653,\n",
              "   0.15392737090587616,\n",
              "   -0.012926789931952953,\n",
              "   -4.290220445568593e-08,\n",
              "   0.02515467256307602,\n",
              "   -0.05946589261293411,\n",
              "   0.016534943133592606,\n",
              "   -0.0897609069943428,\n",
              "   -0.04577358439564705,\n",
              "   0.03190572187304497,\n",
              "   -0.005513042211532593,\n",
              "   -0.001626893994398415,\n",
              "   0.10503950715065002,\n",
              "   -0.006992722395807505,\n",
              "   0.0042443000711500645,\n",
              "   0.062495406717061996,\n",
              "   0.05396689474582672,\n",
              "   0.030683625489473343,\n",
              "   -0.08943285048007965,\n",
              "   0.022043775767087936,\n",
              "   0.022710168734192848,\n",
              "   -0.05263720825314522,\n",
              "   -0.02085109055042267,\n",
              "   0.004992135800421238,\n",
              "   -0.04174676910042763,\n",
              "   0.029724115505814552,\n",
              "   -0.1121496632695198,\n",
              "   -0.047826915979385376,\n",
              "   0.04410449042916298,\n",
              "   -0.01272748876363039,\n",
              "   -0.014187154360115528,\n",
              "   0.054478514939546585,\n",
              "   -0.016191400587558746,\n",
              "   -0.020166317000985146,\n",
              "   0.0191185362637043,\n",
              "   -0.012027298100292683,\n",
              "   0.04640508443117142,\n",
              "   0.017011381685733795,\n",
              "   -0.02194802649319172,\n",
              "   0.007754157762974501,\n",
              "   0.01923062838613987,\n",
              "   -0.045264750719070435,\n",
              "   -0.0045274789445102215,\n",
              "   0.03447504714131355,\n",
              "   -0.008958771824836731,\n",
              "   -0.10202329605817795,\n",
              "   0.004175727255642414,\n",
              "   0.015680180862545967,\n",
              "   -0.07296290993690491,\n",
              "   -0.04309242218732834,\n",
              "   -0.06648224592208862,\n",
              "   -7.386628340100287e-07,\n",
              "   -0.05392366275191307,\n",
              "   0.0066924975253641605,\n",
              "   0.006456849630922079,\n",
              "   0.04881441965699196,\n",
              "   0.008313767611980438,\n",
              "   -0.032032713294029236,\n",
              "   -0.025678817182779312,\n",
              "   0.012827564962208271,\n",
              "   0.02017514780163765,\n",
              "   -0.0273521039634943,\n",
              "   -0.04871321842074394,\n",
              "   -0.09318479150533676,\n",
              "   0.009189230389893055,\n",
              "   -0.07157769799232483,\n",
              "   0.07180710881948471,\n",
              "   -0.012334059923887253],\n",
              "  [0.015300806611776352,\n",
              "   -0.05033532530069351,\n",
              "   -0.02606043964624405,\n",
              "   -0.06612824648618698,\n",
              "   -0.027208875864744186,\n",
              "   -0.053510747849941254,\n",
              "   -0.006247733719646931,\n",
              "   0.11434321105480194,\n",
              "   -0.058660220354795456,\n",
              "   0.06983420997858047,\n",
              "   -0.03699962794780731,\n",
              "   0.10548923909664154,\n",
              "   0.036978986114263535,\n",
              "   0.1133391335606575,\n",
              "   -0.04932854697108269,\n",
              "   0.07119440287351608,\n",
              "   0.04157157242298126,\n",
              "   -0.0005585629842244089,\n",
              "   -0.04492654651403427,\n",
              "   0.10031778365373611,\n",
              "   0.08576197177171707,\n",
              "   -0.03507520630955696,\n",
              "   1.2997871635889169e-05,\n",
              "   -0.0031630643643438816,\n",
              "   -0.0036674838047474623,\n",
              "   0.08459857851266861,\n",
              "   -0.06346458941698074,\n",
              "   0.08482882380485535,\n",
              "   0.02551846206188202,\n",
              "   -0.0056559559889137745,\n",
              "   0.004635779652744532,\n",
              "   0.03542393445968628,\n",
              "   0.02218570187687874,\n",
              "   -0.06432101130485535,\n",
              "   -0.05229254066944122,\n",
              "   0.045781828463077545,\n",
              "   -0.09598082304000854,\n",
              "   -0.01799815706908703,\n",
              "   -0.02418646030128002,\n",
              "   -0.020531512796878815,\n",
              "   0.043668147176504135,\n",
              "   0.00023747028899379075,\n",
              "   0.008538501337170601,\n",
              "   0.012785462662577629,\n",
              "   0.056338269263505936,\n",
              "   -0.10863268375396729,\n",
              "   0.0012000497663393617,\n",
              "   -0.0033508329652249813,\n",
              "   0.005519385915249586,\n",
              "   0.014820164069533348,\n",
              "   -0.06640491634607315,\n",
              "   0.02811010740697384,\n",
              "   0.028615161776542664,\n",
              "   0.019096914678812027,\n",
              "   -0.07859443128108978,\n",
              "   0.04281962290406227,\n",
              "   -0.04812699556350708,\n",
              "   0.000545613351278007,\n",
              "   -0.021235391497612,\n",
              "   -0.05493992194533348,\n",
              "   -0.06941404193639755,\n",
              "   0.019864507019519806,\n",
              "   -0.025939835235476494,\n",
              "   -0.006869212724268436,\n",
              "   -0.02519308216869831,\n",
              "   0.02981368824839592,\n",
              "   -0.08568748086690903,\n",
              "   -0.04712914302945137,\n",
              "   0.04971260577440262,\n",
              "   0.031024156138300896,\n",
              "   -0.08474816381931305,\n",
              "   -0.030385510995984077,\n",
              "   0.029563259333372116,\n",
              "   -0.01307621318846941,\n",
              "   -0.08550500124692917,\n",
              "   0.023418370634317398,\n",
              "   0.1316150575876236,\n",
              "   -0.03508102148771286,\n",
              "   -0.12769542634487152,\n",
              "   0.03337292745709419,\n",
              "   0.06100654602050781,\n",
              "   0.015611632727086544,\n",
              "   0.02466270513832569,\n",
              "   -0.003244138788431883,\n",
              "   -0.012821070849895477,\n",
              "   0.07502785325050354,\n",
              "   0.051691215485334396,\n",
              "   0.09698909521102905,\n",
              "   0.01353833544999361,\n",
              "   -0.0476524643599987,\n",
              "   0.08066227287054062,\n",
              "   0.12122957408428192,\n",
              "   -0.04233802855014801,\n",
              "   0.06749195605516434,\n",
              "   -0.03675338625907898,\n",
              "   -0.019452789798378944,\n",
              "   0.04395657032728195,\n",
              "   0.08385539799928665,\n",
              "   0.016657082363963127,\n",
              "   0.010411909781396389,\n",
              "   0.046458132565021515,\n",
              "   -0.058609552681446075,\n",
              "   -0.06084072217345238,\n",
              "   -0.027430947870016098,\n",
              "   -0.04064089432358742,\n",
              "   -0.06483811140060425,\n",
              "   0.1007215827703476,\n",
              "   0.029445059597492218,\n",
              "   -0.0012769625755026937,\n",
              "   0.056097857654094696,\n",
              "   -0.0444808155298233,\n",
              "   -0.01144553441554308,\n",
              "   -0.02682020515203476,\n",
              "   0.13299284875392914,\n",
              "   0.030029892921447754,\n",
              "   0.03904673084616661,\n",
              "   0.03400139883160591,\n",
              "   -0.008695908822119236,\n",
              "   -0.008650442585349083,\n",
              "   0.017154347151517868,\n",
              "   0.0442286916077137,\n",
              "   -0.025733275339007378,\n",
              "   0.1062588319182396,\n",
              "   -0.11652091145515442,\n",
              "   0.035248592495918274,\n",
              "   0.0395047552883625,\n",
              "   0.03202306479215622,\n",
              "   4.600577056535312e-33,\n",
              "   -0.02825024724006653,\n",
              "   0.06470081955194473,\n",
              "   0.02095739357173443,\n",
              "   0.029755478724837303,\n",
              "   0.026114782318472862,\n",
              "   0.029976829886436462,\n",
              "   -0.05476893484592438,\n",
              "   3.6873923818347976e-05,\n",
              "   -0.007426613476127386,\n",
              "   -0.04148268327116966,\n",
              "   -0.07311934977769852,\n",
              "   0.0008742264471948147,\n",
              "   -0.018918292596936226,\n",
              "   0.06357485055923462,\n",
              "   -0.08280778676271439,\n",
              "   0.009482775814831257,\n",
              "   0.02647768147289753,\n",
              "   0.017834501340985298,\n",
              "   0.05745552107691765,\n",
              "   -0.0054448796436190605,\n",
              "   0.016206135973334312,\n",
              "   -0.00068799959262833,\n",
              "   -0.09475845843553543,\n",
              "   0.0428958497941494,\n",
              "   -0.06352747976779938,\n",
              "   0.07146415114402771,\n",
              "   -0.04313735291361809,\n",
              "   0.1347581148147583,\n",
              "   -0.02157747372984886,\n",
              "   -0.01835036650300026,\n",
              "   0.03763151168823242,\n",
              "   -0.026899293065071106,\n",
              "   -0.02091858722269535,\n",
              "   -0.09787055104970932,\n",
              "   -0.002980474615469575,\n",
              "   0.02933424711227417,\n",
              "   -0.11251500248908997,\n",
              "   0.026499414816498756,\n",
              "   0.05522492527961731,\n",
              "   -0.057807620614767075,\n",
              "   0.01777433604001999,\n",
              "   -0.012997880578041077,\n",
              "   0.06553748995065689,\n",
              "   0.04086989909410477,\n",
              "   0.01333136297762394,\n",
              "   0.015919268131256104,\n",
              "   -0.12795938551425934,\n",
              "   0.07723043113946915,\n",
              "   0.04746080935001373,\n",
              "   -0.060516975820064545,\n",
              "   -0.003888467326760292,\n",
              "   -0.03830735385417938,\n",
              "   -0.015433058142662048,\n",
              "   0.03565208986401558,\n",
              "   -0.08121555298566818,\n",
              "   0.06230736896395683,\n",
              "   -0.11394697427749634,\n",
              "   -0.019907886162400246,\n",
              "   0.07773712277412415,\n",
              "   0.0015556379221379757,\n",
              "   -0.01050668302923441,\n",
              "   -0.0072922431863844395,\n",
              "   0.0021932905074208975,\n",
              "   0.07278294861316681,\n",
              "   0.031017040833830833,\n",
              "   -0.02376851625740528,\n",
              "   -0.024031944572925568,\n",
              "   -0.04014386609196663,\n",
              "   -0.024760734289884567,\n",
              "   -0.016328223049640656,\n",
              "   0.019334523007273674,\n",
              "   -0.029523655772209167,\n",
              "   -0.03400893881917,\n",
              "   0.030843472108244896,\n",
              "   0.06401965767145157,\n",
              "   0.025157969444990158,\n",
              "   0.12191402912139893,\n",
              "   0.04750640317797661,\n",
              "   -0.0028456139843910933,\n",
              "   0.024103669449687004,\n",
              "   0.02088390849530697,\n",
              "   -0.03249940648674965,\n",
              "   -0.09712483733892441,\n",
              "   -0.061656612902879715,\n",
              "   -0.06442782282829285,\n",
              "   -0.029010465368628502,\n",
              "   -0.008882076479494572,\n",
              "   0.019052723422646523,\n",
              "   0.005874416325241327,\n",
              "   -0.04186178371310234,\n",
              "   0.030491266399621964,\n",
              "   -0.10265567898750305,\n",
              "   -0.06932510435581207,\n",
              "   -0.0652630627155304,\n",
              "   0.010299017652869225,\n",
              "   -4.7931707856580046e-33,\n",
              "   0.08384299278259277,\n",
              "   0.054645147174596786,\n",
              "   0.027863003313541412,\n",
              "   0.03163491562008858,\n",
              "   0.04947919398546219,\n",
              "   0.007717473432421684,\n",
              "   0.08163201063871384,\n",
              "   -0.007087323348969221,\n",
              "   0.0491538904607296,\n",
              "   -0.019860921427607536,\n",
              "   0.048092931509017944,\n",
              "   0.04757761210203171,\n",
              "   -0.04100252315402031,\n",
              "   -0.05304224416613579,\n",
              "   -0.05299782380461693,\n",
              "   0.03798389434814453,\n",
              "   0.027896009385585785,\n",
              "   -0.04665188491344452,\n",
              "   0.01914728619158268,\n",
              "   -0.01593714952468872,\n",
              "   -0.03613561391830444,\n",
              "   0.04978889599442482,\n",
              "   -0.05776796117424965,\n",
              "   0.03566349670290947,\n",
              "   0.009186452254652977,\n",
              "   0.03642295300960541,\n",
              "   0.051256511360406876,\n",
              "   -0.08227387815713882,\n",
              "   -0.003470823634415865,\n",
              "   0.046560827642679214,\n",
              "   0.031616270542144775,\n",
              "   -0.013356056064367294,\n",
              "   -0.06251294910907745,\n",
              "   -0.028174826875329018,\n",
              "   0.021358586847782135,\n",
              "   0.03401495888829231,\n",
              "   -0.058824386447668076,\n",
              "   -0.10409049689769745,\n",
              "   -0.004267408978193998,\n",
              "   -0.01306991372257471,\n",
              "   0.014474552124738693,\n",
              "   0.04910227283835411,\n",
              "   -0.035939641296863556,\n",
              "   -0.01906707137823105,\n",
              "   0.03573057800531387,\n",
              "   0.037180542945861816,\n",
              "   -0.015942176803946495,\n",
              "   -0.0519355908036232,\n",
              "   0.013192940503358841,\n",
              "   0.0025907603558152914,\n",
              "   -0.017473863437771797,\n",
              "   0.011061190627515316,\n",
              "   -0.010125790722668171,\n",
              "   -0.004061148036271334,\n",
              "   0.07471423596143723,\n",
              "   0.005690866149961948,\n",
              "   0.04754598066210747,\n",
              "   0.03528459370136261,\n",
              "   -0.009841728955507278,\n",
              "   -0.03683149814605713,\n",
              "   0.028199439868330956,\n",
              "   0.01016178261488676,\n",
              "   -0.06584703177213669,\n",
              "   -0.06300108879804611,\n",
              "   -0.02720361016690731,\n",
              "   0.0009217330371029675,\n",
              "   0.027044396847486496,\n",
              "   0.07899289578199387,\n",
              "   0.02037009224295616,\n",
              "   -6.907564966240898e-05,\n",
              "   0.02980802394449711,\n",
              "   0.013630518689751625,\n",
              "   0.04704480245709419,\n",
              "   0.03714956343173981,\n",
              "   -0.048256054520606995,\n",
              "   0.12391503900289536,\n",
              "   -0.12289538979530334,\n",
              "   -0.04041560739278793,\n",
              "   -0.06975235044956207,\n",
              "   -0.029981128871440887,\n",
              "   0.01768212951719761,\n",
              "   -0.005270324181765318,\n",
              "   -0.009452668018639088,\n",
              "   -0.01937820203602314,\n",
              "   -0.02480100654065609,\n",
              "   -0.07910358160734177,\n",
              "   0.018407445400953293,\n",
              "   -0.015883831307291985,\n",
              "   -0.07374367117881775,\n",
              "   0.004353758879005909,\n",
              "   0.02243201993405819,\n",
              "   -0.0822196751832962,\n",
              "   0.016616439446806908,\n",
              "   0.0915377214550972,\n",
              "   0.0433531329035759,\n",
              "   -5.7722996160691764e-08,\n",
              "   0.03834333270788193,\n",
              "   -0.05493075028061867,\n",
              "   0.003399936482310295,\n",
              "   -0.028756247833371162,\n",
              "   -0.09754471480846405,\n",
              "   0.038534924387931824,\n",
              "   -0.014082049019634724,\n",
              "   0.0011220538290217519,\n",
              "   0.11715172976255417,\n",
              "   0.030921494588255882,\n",
              "   -0.0454137809574604,\n",
              "   0.023832155391573906,\n",
              "   -0.052530694752931595,\n",
              "   0.06452032178640366,\n",
              "   -0.06694744527339935,\n",
              "   0.03935542702674866,\n",
              "   -0.017528096213936806,\n",
              "   -0.006117761135101318,\n",
              "   -0.022780293598771095,\n",
              "   -0.038591813296079636,\n",
              "   -0.019803708419203758,\n",
              "   -0.029424909502267838,\n",
              "   -0.07258126884698868,\n",
              "   -0.013730701059103012,\n",
              "   0.06365437060594559,\n",
              "   0.019647270441055298,\n",
              "   -0.007449701428413391,\n",
              "   0.03794712573289871,\n",
              "   -0.059649981558322906,\n",
              "   -0.043239776045084,\n",
              "   0.04816657677292824,\n",
              "   -0.02630792185664177,\n",
              "   0.11804257333278656,\n",
              "   0.004668090958148241,\n",
              "   -0.002586547052487731,\n",
              "   0.03270614892244339,\n",
              "   -0.013869973830878735,\n",
              "   -0.011346927843987942,\n",
              "   0.060158856213092804,\n",
              "   0.045883312821388245,\n",
              "   0.020776772871613503,\n",
              "   -0.11779853701591492,\n",
              "   0.03384646400809288,\n",
              "   -0.040863037109375,\n",
              "   -0.05138816311955452,\n",
              "   -0.03895723074674606,\n",
              "   -0.059107568114995956,\n",
              "   0.018962714821100235,\n",
              "   -0.027449196204543114,\n",
              "   -0.047768186777830124,\n",
              "   0.058953914791345596,\n",
              "   0.03505602478981018,\n",
              "   -0.05232943594455719,\n",
              "   -0.11876652389764786,\n",
              "   0.00241226889193058,\n",
              "   0.024868721142411232,\n",
              "   0.004137774929404259,\n",
              "   0.04554281756281853,\n",
              "   0.003208477981388569,\n",
              "   -0.09530226141214371,\n",
              "   -0.027236558496952057,\n",
              "   -0.011465154588222504,\n",
              "   0.009285690262913704,\n",
              "   -0.05414365977048874],\n",
              "  [-0.03384682536125183,\n",
              "   -0.046434078365564346,\n",
              "   -0.11194693297147751,\n",
              "   -0.044590290635824203,\n",
              "   0.018052542582154274,\n",
              "   -0.013743879273533821,\n",
              "   0.0007921134820207953,\n",
              "   0.04952997341752052,\n",
              "   0.056835707277059555,\n",
              "   0.043675970286130905,\n",
              "   -0.00010025903611676767,\n",
              "   0.05730408430099487,\n",
              "   -0.045669734477996826,\n",
              "   0.12031290680170059,\n",
              "   0.01470320113003254,\n",
              "   0.13906316459178925,\n",
              "   -0.024183928966522217,\n",
              "   0.02708592638373375,\n",
              "   0.021744370460510254,\n",
              "   0.06383905559778214,\n",
              "   -0.04184212163090706,\n",
              "   -0.023920537903904915,\n",
              "   0.0314127542078495,\n",
              "   -0.010499967262148857,\n",
              "   -0.07305944710969925,\n",
              "   0.05862334370613098,\n",
              "   0.033213794231414795,\n",
              "   0.03566982224583626,\n",
              "   -0.09061732143163681,\n",
              "   -0.04581175744533539,\n",
              "   0.011309230700135231,\n",
              "   0.017296992242336273,\n",
              "   0.009705259464681149,\n",
              "   -0.10317390412092209,\n",
              "   -0.013771049678325653,\n",
              "   -0.026366306468844414,\n",
              "   -0.15107017755508423,\n",
              "   0.006763246841728687,\n",
              "   0.031008459627628326,\n",
              "   -0.05080081522464752,\n",
              "   -0.00962933897972107,\n",
              "   -0.029634129256010056,\n",
              "   -0.01620166003704071,\n",
              "   -0.012563338503241539,\n",
              "   -0.04689954221248627,\n",
              "   -0.10424111038446426,\n",
              "   -0.011503939516842365,\n",
              "   0.012853802181780338,\n",
              "   7.278261909959838e-05,\n",
              "   0.02054615132510662,\n",
              "   -0.018325606361031532,\n",
              "   -0.02439482882618904,\n",
              "   -0.023536164313554764,\n",
              "   0.11230651289224625,\n",
              "   0.0024075363762676716,\n",
              "   -0.0033409472089260817,\n",
              "   -0.08158592134714127,\n",
              "   0.006119279656559229,\n",
              "   0.05539510026574135,\n",
              "   -0.0019512248691171408,\n",
              "   0.018662450835108757,\n",
              "   0.036692604422569275,\n",
              "   -0.002721191616728902,\n",
              "   -0.05309600383043289,\n",
              "   -0.13208892941474915,\n",
              "   0.002483714371919632,\n",
              "   -0.00850831437855959,\n",
              "   -0.045265521854162216,\n",
              "   0.02399696595966816,\n",
              "   0.05093597620725632,\n",
              "   -0.06885187327861786,\n",
              "   -0.03919825330376625,\n",
              "   -0.00894866045564413,\n",
              "   0.02235427126288414,\n",
              "   0.03153650835156441,\n",
              "   0.09185383468866348,\n",
              "   0.04888517037034035,\n",
              "   0.06970402598381042,\n",
              "   -0.03967821225523949,\n",
              "   -0.051267486065626144,\n",
              "   0.03882226720452309,\n",
              "   0.0037223349791020155,\n",
              "   0.0594562403857708,\n",
              "   0.045045044273138046,\n",
              "   -0.005935220513492823,\n",
              "   0.05146004259586334,\n",
              "   0.006250666454434395,\n",
              "   0.0821956917643547,\n",
              "   0.05352373048663139,\n",
              "   -0.057172469794750214,\n",
              "   0.13950397074222565,\n",
              "   0.10301381349563599,\n",
              "   -0.01637295074760914,\n",
              "   -0.03294647857546806,\n",
              "   0.017661932855844498,\n",
              "   0.006273060571402311,\n",
              "   0.04684040695428848,\n",
              "   0.13268014788627625,\n",
              "   0.010931987315416336,\n",
              "   -0.0010388519149273634,\n",
              "   0.029938019812107086,\n",
              "   -0.023714587092399597,\n",
              "   -0.11568386107683182,\n",
              "   -0.059704434126615524,\n",
              "   -0.06154708191752434,\n",
              "   0.05163013935089111,\n",
              "   0.04559953883290291,\n",
              "   0.027879273518919945,\n",
              "   0.01697939820587635,\n",
              "   -0.005869424901902676,\n",
              "   -0.0932701975107193,\n",
              "   -0.013221919536590576,\n",
              "   0.011668967083096504,\n",
              "   0.005064832977950573,\n",
              "   -0.045393265783786774,\n",
              "   0.04393000900745392,\n",
              "   0.04324520006775856,\n",
              "   0.020273035392165184,\n",
              "   0.020178262144327164,\n",
              "   0.05480322614312172,\n",
              "   0.04561406001448631,\n",
              "   0.02133021503686905,\n",
              "   0.04489261656999588,\n",
              "   0.0038777703884989023,\n",
              "   0.04940216988325119,\n",
              "   0.04769263416528702,\n",
              "   -0.05265219137072563,\n",
              "   3.2452945181746017e-33,\n",
              "   0.027861997485160828,\n",
              "   0.07232290506362915,\n",
              "   0.08259421586990356,\n",
              "   0.024801138788461685,\n",
              "   0.050467804074287415,\n",
              "   -0.016747858375310898,\n",
              "   -0.061031125485897064,\n",
              "   -0.08794402331113815,\n",
              "   -0.06430806964635849,\n",
              "   -0.058254316449165344,\n",
              "   -0.09483949840068817,\n",
              "   -0.023821264505386353,\n",
              "   -0.007288612425327301,\n",
              "   0.006783533375710249,\n",
              "   -0.04870389774441719,\n",
              "   0.05808968096971512,\n",
              "   0.03335981070995331,\n",
              "   0.009366365149617195,\n",
              "   0.0767613872885704,\n",
              "   -0.05851490795612335,\n",
              "   0.05252682417631149,\n",
              "   0.04801051691174507,\n",
              "   -0.04130867123603821,\n",
              "   0.0313536673784256,\n",
              "   -0.048738691955804825,\n",
              "   0.06030692160129547,\n",
              "   -0.0019279682310298085,\n",
              "   0.05835690349340439,\n",
              "   0.017616143450140953,\n",
              "   0.003143765963613987,\n",
              "   -0.06459914147853851,\n",
              "   -0.06881866604089737,\n",
              "   -0.034002695232629776,\n",
              "   -0.04189924895763397,\n",
              "   0.03533893823623657,\n",
              "   0.049557581543922424,\n",
              "   -0.10491184145212173,\n",
              "   -0.024159733206033707,\n",
              "   0.0016285073943436146,\n",
              "   -0.04144606739282608,\n",
              "   0.025852041319012642,\n",
              "   -0.013324057683348656,\n",
              "   -0.04500512033700943,\n",
              "   -0.00022422939946409315,\n",
              "   -0.008049865253269672,\n",
              "   0.032576099038124084,\n",
              "   -0.04925881326198578,\n",
              "   0.11202028393745422,\n",
              "   0.012430927716195583,\n",
              "   -0.018977807834744453,\n",
              "   0.001501243794336915,\n",
              "   -0.025740060955286026,\n",
              "   -0.026269538328051567,\n",
              "   0.07436473667621613,\n",
              "   -0.02415236458182335,\n",
              "   0.056971944868564606,\n",
              "   -0.041595760732889175,\n",
              "   -0.02850436419248581,\n",
              "   0.13010816276073456,\n",
              "   -0.04470004513859749,\n",
              "   -0.011819494888186455,\n",
              "   -0.04884433001279831,\n",
              "   -0.0102206626906991,\n",
              "   0.009706391021609306,\n",
              "   0.026526572182774544,\n",
              "   0.041618142277002335,\n",
              "   -0.07214371860027313,\n",
              "   0.03329960256814957,\n",
              "   -0.03465348854660988,\n",
              "   0.008276804350316525,\n",
              "   0.026162104681134224,\n",
              "   0.017371779307723045,\n",
              "   0.020484399050474167,\n",
              "   0.004323875997215509,\n",
              "   0.09737979620695114,\n",
              "   0.04805285483598709,\n",
              "   -0.000849218457005918,\n",
              "   -0.01472888607531786,\n",
              "   0.05867625027894974,\n",
              "   0.06286070495843887,\n",
              "   0.047089606523513794,\n",
              "   -0.0273162629455328,\n",
              "   -0.05177586153149605,\n",
              "   -0.027692921459674835,\n",
              "   -0.08737253397703171,\n",
              "   -0.08483754098415375,\n",
              "   -0.056081317365169525,\n",
              "   0.009455982595682144,\n",
              "   0.006226274650543928,\n",
              "   0.003944357391446829,\n",
              "   -0.010766880586743355,\n",
              "   -0.08671600371599197,\n",
              "   -0.07898686081171036,\n",
              "   -0.005782697815448046,\n",
              "   -0.006391216069459915,\n",
              "   -4.393931394474492e-33,\n",
              "   0.06394848972558975,\n",
              "   -0.008199726231396198,\n",
              "   0.06851865351200104,\n",
              "   -0.007963220588862896,\n",
              "   0.06666988134384155,\n",
              "   -0.005245376378297806,\n",
              "   0.008513348177075386,\n",
              "   -0.14174558222293854,\n",
              "   0.01648031733930111,\n",
              "   -0.03596743941307068,\n",
              "   0.07044246792793274,\n",
              "   -0.019315630197525024,\n",
              "   -0.046423956751823425,\n",
              "   -0.03090604394674301,\n",
              "   -0.025636013597249985,\n",
              "   -0.003471335396170616,\n",
              "   0.028445370495319366,\n",
              "   0.041938669979572296,\n",
              "   0.0064414492808282375,\n",
              "   0.03388826549053192,\n",
              "   -0.052158039063215256,\n",
              "   -0.03329373151063919,\n",
              "   -0.07783928513526917,\n",
              "   -0.018307331949472427,\n",
              "   0.012684715911746025,\n",
              "   0.013003087602555752,\n",
              "   0.09521661698818207,\n",
              "   -0.026989232748746872,\n",
              "   0.045864153653383255,\n",
              "   0.1283973902463913,\n",
              "   0.010556082241237164,\n",
              "   0.03405970335006714,\n",
              "   -0.06744331121444702,\n",
              "   -0.04606332629919052,\n",
              "   0.07575666159391403,\n",
              "   0.04680733010172844,\n",
              "   -0.0011331485584378242,\n",
              "   -0.03937479853630066,\n",
              "   0.020836398005485535,\n",
              "   -0.016114328056573868,\n",
              "   0.009949072264134884,\n",
              "   0.008312813937664032,\n",
              "   -0.05357819050550461,\n",
              "   -0.057175952941179276,\n",
              "   0.09436352550983429,\n",
              "   0.0883435383439064,\n",
              "   -0.01820462942123413,\n",
              "   0.06111488491296768,\n",
              "   -0.023619383573532104,\n",
              "   -0.08056185394525528,\n",
              "   0.00212118704803288,\n",
              "   0.049100905656814575,\n",
              "   -0.03315277770161629,\n",
              "   -0.019726281985640526,\n",
              "   0.06000806391239166,\n",
              "   0.0782412439584732,\n",
              "   0.04754316434264183,\n",
              "   -0.0011486824369058013,\n",
              "   0.06866683810949326,\n",
              "   -0.05276508629322052,\n",
              "   -0.06340164691209793,\n",
              "   -0.04438497871160507,\n",
              "   -0.014593232423067093,\n",
              "   -0.04016584903001785,\n",
              "   -0.04944553226232529,\n",
              "   0.06104544550180435,\n",
              "   0.00710559356957674,\n",
              "   0.10303036123514175,\n",
              "   -0.015417439863085747,\n",
              "   -0.02603851445019245,\n",
              "   -0.01013453584164381,\n",
              "   -0.030720479786396027,\n",
              "   0.017141452059149742,\n",
              "   0.07406685501337051,\n",
              "   -0.01697433926165104,\n",
              "   0.023515312001109123,\n",
              "   0.0020161366555839777,\n",
              "   -0.06523876637220383,\n",
              "   -0.03801950812339783,\n",
              "   0.040586065500974655,\n",
              "   0.07784769684076309,\n",
              "   -0.048165950924158096,\n",
              "   0.014295601285994053,\n",
              "   -0.0670376792550087,\n",
              "   -0.003660918213427067,\n",
              "   0.027370579540729523,\n",
              "   0.05005485564470291,\n",
              "   0.03766689822077751,\n",
              "   -0.041694581508636475,\n",
              "   -0.03962719812989235,\n",
              "   0.026466721668839455,\n",
              "   -0.007305516395717859,\n",
              "   -0.014774305745959282,\n",
              "   0.16303087770938873,\n",
              "   0.09775920957326889,\n",
              "   -5.238158706788454e-08,\n",
              "   0.02882448211312294,\n",
              "   -0.032983265817165375,\n",
              "   -0.03551928326487541,\n",
              "   -0.03937181457877159,\n",
              "   -0.019237956032156944,\n",
              "   0.023704463616013527,\n",
              "   -0.06388872861862183,\n",
              "   0.023000136017799377,\n",
              "   -0.0022177291102707386,\n",
              "   -0.029448404908180237,\n",
              "   -0.04365456476807594,\n",
              "   0.02456916682422161,\n",
              "   0.052971553057432175,\n",
              "   0.050252560526132584,\n",
              "   -0.014083025977015495,\n",
              "   0.021167941391468048,\n",
              "   0.03031516633927822,\n",
              "   -0.08512283116579056,\n",
              "   -0.009532848373055458,\n",
              "   0.014824870973825455,\n",
              "   -0.02070593275129795,\n",
              "   0.04595709964632988,\n",
              "   -0.06493514031171799,\n",
              "   -0.053963880985975266,\n",
              "   0.0032438805792480707,\n",
              "   -0.08753059804439545,\n",
              "   -0.048591408878564835,\n",
              "   -0.04510284587740898,\n",
              "   -0.03119264543056488,\n",
              "   -0.013889207504689693,\n",
              "   -0.0021651065908372402,\n",
              "   0.042408768087625504,\n",
              "   0.006023063790053129,\n",
              "   -0.0018810722976922989,\n",
              "   0.0596277080476284,\n",
              "   -0.017537107691168785,\n",
              "   -0.07062455266714096,\n",
              "   -0.05370854213833809,\n",
              "   0.039576053619384766,\n",
              "   -9.955449786502868e-05,\n",
              "   0.028976712375879288,\n",
              "   -0.060101158916950226,\n",
              "   -0.05614648014307022,\n",
              "   -0.019500868394970894,\n",
              "   -0.050075188279151917,\n",
              "   -0.01861896552145481,\n",
              "   -0.018114468082785606,\n",
              "   0.043191246688365936,\n",
              "   -0.12335304170846939,\n",
              "   -0.015748469159007072,\n",
              "   0.033449139446020126,\n",
              "   0.0324353463947773,\n",
              "   0.005081276409327984,\n",
              "   -0.06822095066308975,\n",
              "   0.010957499034702778,\n",
              "   0.06690119951963425,\n",
              "   -0.017143510282039642,\n",
              "   -0.006274893879890442,\n",
              "   -0.04897230863571167,\n",
              "   -0.15481780469417572,\n",
              "   -0.03810912370681763,\n",
              "   -0.09833988547325134,\n",
              "   0.04126523807644844,\n",
              "   -0.00038758880691602826],\n",
              "  [-0.07344125956296921,\n",
              "   -0.03951113298535347,\n",
              "   -0.05670949071645737,\n",
              "   -0.06290439516305923,\n",
              "   0.0653381422162056,\n",
              "   -0.04494765028357506,\n",
              "   0.02565426006913185,\n",
              "   0.07183143496513367,\n",
              "   0.03520757704973221,\n",
              "   0.0222232174128294,\n",
              "   -0.015454672276973724,\n",
              "   -0.043145809322595596,\n",
              "   0.01805424503982067,\n",
              "   0.02606745809316635,\n",
              "   -0.06518726795911789,\n",
              "   0.05800975114107132,\n",
              "   0.01639765128493309,\n",
              "   0.0014543519355356693,\n",
              "   -0.0033000365365296602,\n",
              "   0.029375210404396057,\n",
              "   -0.11150682717561722,\n",
              "   -0.014013808220624924,\n",
              "   0.011486141011118889,\n",
              "   -0.014665608294308186,\n",
              "   -0.04919281229376793,\n",
              "   -0.020680800080299377,\n",
              "   0.048402972519397736,\n",
              "   -0.006356858182698488,\n",
              "   -0.04567641019821167,\n",
              "   -0.05466002598404884,\n",
              "   0.021715743467211723,\n",
              "   0.038800742477178574,\n",
              "   -0.010570801794528961,\n",
              "   -0.0451790913939476,\n",
              "   0.05577639117836952,\n",
              "   -0.043326813727617264,\n",
              "   -0.06343500316143036,\n",
              "   0.01409864891320467,\n",
              "   -0.04973018914461136,\n",
              "   -0.025775739923119545,\n",
              "   0.04133350029587746,\n",
              "   -0.01933341473340988,\n",
              "   0.006805329583585262,\n",
              "   -0.031396906822919846,\n",
              "   0.05083554983139038,\n",
              "   -0.036450307816267014,\n",
              "   -0.06301011890172958,\n",
              "   -0.01967771165072918,\n",
              "   0.009978072717785835,\n",
              "   0.08478686213493347,\n",
              "   0.014751208014786243,\n",
              "   0.058224182575941086,\n",
              "   -0.03786243125796318,\n",
              "   0.031086646020412445,\n",
              "   -0.036075737327337265,\n",
              "   0.05780674144625664,\n",
              "   -0.05588824301958084,\n",
              "   0.0331900529563427,\n",
              "   0.016480926424264908,\n",
              "   -0.030711760744452477,\n",
              "   0.03988686203956604,\n",
              "   -0.09261190891265869,\n",
              "   -0.009009705856442451,\n",
              "   0.000608569651376456,\n",
              "   0.02927234396338463,\n",
              "   0.043126776814460754,\n",
              "   0.04613226652145386,\n",
              "   -0.03952356055378914,\n",
              "   0.08225112408399582,\n",
              "   0.023815728724002838,\n",
              "   -0.10787832736968994,\n",
              "   0.03279080241918564,\n",
              "   -0.01674807257950306,\n",
              "   -0.04014987125992775,\n",
              "   0.12277621030807495,\n",
              "   -0.01531509030610323,\n",
              "   -0.011574733071029186,\n",
              "   -0.04631628096103668,\n",
              "   0.04845732823014259,\n",
              "   0.0021038593258708715,\n",
              "   -0.0491606704890728,\n",
              "   0.11191745102405548,\n",
              "   0.0641474649310112,\n",
              "   0.033274006098508835,\n",
              "   -0.01636728085577488,\n",
              "   0.0649443119764328,\n",
              "   -0.03321629390120506,\n",
              "   0.027247074991464615,\n",
              "   0.0213128924369812,\n",
              "   -0.05573389679193497,\n",
              "   0.06637932360172272,\n",
              "   0.0012523940531536937,\n",
              "   -0.07449902594089508,\n",
              "   -0.018288612365722656,\n",
              "   -0.049107860773801804,\n",
              "   -0.05106345936655998,\n",
              "   0.1275004893541336,\n",
              "   0.03428817167878151,\n",
              "   -0.012868063524365425,\n",
              "   0.008416389115154743,\n",
              "   0.03714946657419205,\n",
              "   0.051841553300619125,\n",
              "   -0.08808247745037079,\n",
              "   -0.08069698512554169,\n",
              "   0.0021538978908210993,\n",
              "   0.05336211994290352,\n",
              "   0.06035717949271202,\n",
              "   0.0002958379627671093,\n",
              "   -0.03696843609213829,\n",
              "   -0.011883750557899475,\n",
              "   -0.03609202057123184,\n",
              "   -0.04178794473409653,\n",
              "   0.07995539158582687,\n",
              "   -0.008181636221706867,\n",
              "   -0.051803749054670334,\n",
              "   0.057960592210292816,\n",
              "   -0.024161865934729576,\n",
              "   -0.04454784095287323,\n",
              "   0.049895718693733215,\n",
              "   -0.03769709914922714,\n",
              "   -0.00642816349864006,\n",
              "   -0.019590601325035095,\n",
              "   0.0099482461810112,\n",
              "   0.03351851552724838,\n",
              "   0.018750714138150215,\n",
              "   0.04771622642874718,\n",
              "   -0.09467949718236923,\n",
              "   5.006031506756812e-33,\n",
              "   -0.02767982892692089,\n",
              "   -0.024612456560134888,\n",
              "   0.035551704466342926,\n",
              "   3.827998443739489e-05,\n",
              "   0.03931132331490517,\n",
              "   -0.03424378111958504,\n",
              "   -0.0623149499297142,\n",
              "   0.009520858526229858,\n",
              "   -0.09492705762386322,\n",
              "   0.008689365349709988,\n",
              "   -0.08799479901790619,\n",
              "   -0.0730772390961647,\n",
              "   -0.02030389942228794,\n",
              "   -0.024134129285812378,\n",
              "   -0.024335308000445366,\n",
              "   -0.0630473867058754,\n",
              "   -0.03413032367825508,\n",
              "   0.08552061766386032,\n",
              "   -0.05578377842903137,\n",
              "   -0.10190416872501373,\n",
              "   0.04162723198533058,\n",
              "   0.10578396916389465,\n",
              "   -0.03944644704461098,\n",
              "   -0.012513224966824055,\n",
              "   0.008938648737967014,\n",
              "   0.0890456810593605,\n",
              "   -0.030724363401532173,\n",
              "   0.0487724170088768,\n",
              "   0.0008863897528499365,\n",
              "   0.008269795216619968,\n",
              "   -0.02982879802584648,\n",
              "   -0.02079922892153263,\n",
              "   -0.05465138331055641,\n",
              "   0.030314456671476364,\n",
              "   0.03078761324286461,\n",
              "   0.00660003861412406,\n",
              "   -0.05399608239531517,\n",
              "   0.0020011260639876127,\n",
              "   0.03441103175282478,\n",
              "   0.024449583142995834,\n",
              "   -0.030196523293852806,\n",
              "   0.02912638522684574,\n",
              "   0.07890783995389938,\n",
              "   -0.0029773013666272163,\n",
              "   -0.007126358803361654,\n",
              "   -0.10914356261491776,\n",
              "   -0.037140779197216034,\n",
              "   0.09594319015741348,\n",
              "   -0.018213611096143723,\n",
              "   0.0494011789560318,\n",
              "   -0.007012459449470043,\n",
              "   -0.02192230522632599,\n",
              "   0.004212110303342342,\n",
              "   0.05800753831863403,\n",
              "   -0.06499926745891571,\n",
              "   -0.006020214408636093,\n",
              "   -0.05102815851569176,\n",
              "   -0.020683519542217255,\n",
              "   0.02496558055281639,\n",
              "   0.06714262813329697,\n",
              "   -0.09901636093854904,\n",
              "   0.043395739048719406,\n",
              "   -0.010049564763903618,\n",
              "   -0.005175880156457424,\n",
              "   0.0033030277118086815,\n",
              "   0.10502327233552933,\n",
              "   -0.0691889077425003,\n",
              "   0.0033020395785570145,\n",
              "   0.0210532508790493,\n",
              "   0.07315415889024734,\n",
              "   0.01781795173883438,\n",
              "   0.06005064770579338,\n",
              "   0.06778709590435028,\n",
              "   0.019637659192085266,\n",
              "   0.05923471227288246,\n",
              "   -0.017114756628870964,\n",
              "   0.05719313025474548,\n",
              "   -0.034326449036598206,\n",
              "   0.06731786578893661,\n",
              "   -0.045010801404714584,\n",
              "   -0.0087785879150033,\n",
              "   0.02019348368048668,\n",
              "   -0.03727559372782707,\n",
              "   -0.010456888005137444,\n",
              "   -0.13606974482536316,\n",
              "   -0.16608978807926178,\n",
              "   -0.011424255557358265,\n",
              "   -0.12037096917629242,\n",
              "   -0.027727318927645683,\n",
              "   -0.005698586348444223,\n",
              "   0.03421090915799141,\n",
              "   -0.04811132699251175,\n",
              "   -0.0206235833466053,\n",
              "   -0.04179053753614426,\n",
              "   -0.044682033360004425,\n",
              "   -5.900149244190707e-33,\n",
              "   0.04982445761561394,\n",
              "   -0.025158271193504333,\n",
              "   0.027548898011446,\n",
              "   -0.0020404160022735596,\n",
              "   -0.06607021391391754,\n",
              "   0.0821015015244484,\n",
              "   0.05706958472728729,\n",
              "   -0.08983028680086136,\n",
              "   0.05152098834514618,\n",
              "   -0.09960824996232986,\n",
              "   0.030701907351613045,\n",
              "   0.10562457889318466,\n",
              "   -0.010049344040453434,\n",
              "   0.035377297550439835,\n",
              "   -0.002527472795918584,\n",
              "   0.08585604280233383,\n",
              "   -0.05733009800314903,\n",
              "   -0.054479122161865234,\n",
              "   0.028597112745046616,\n",
              "   0.008429571986198425,\n",
              "   0.03385701775550842,\n",
              "   -0.0354146771132946,\n",
              "   -0.0002904644061345607,\n",
              "   0.10729018598794937,\n",
              "   -0.014453490264713764,\n",
              "   -0.0033232797868549824,\n",
              "   0.10586864501237869,\n",
              "   -0.018468305468559265,\n",
              "   0.0898473709821701,\n",
              "   0.08308905363082886,\n",
              "   -0.024894611909985542,\n",
              "   0.07416185736656189,\n",
              "   -0.073221355676651,\n",
              "   0.06816442310810089,\n",
              "   0.024681178852915764,\n",
              "   0.004605820868164301,\n",
              "   0.04458968713879585,\n",
              "   0.06949180364608765,\n",
              "   -0.09470117092132568,\n",
              "   -0.03863633796572685,\n",
              "   0.01933283917605877,\n",
              "   -0.04208847135305405,\n",
              "   -0.03859807178378105,\n",
              "   0.04309050738811493,\n",
              "   0.09059486538171768,\n",
              "   0.062199268490076065,\n",
              "   -0.03695636987686157,\n",
              "   0.06343310326337814,\n",
              "   -0.043196093291044235,\n",
              "   -0.020744923502206802,\n",
              "   -0.06497923284769058,\n",
              "   -0.021890059113502502,\n",
              "   0.02561626024544239,\n",
              "   0.02861826866865158,\n",
              "   0.01668761670589447,\n",
              "   0.09675616025924683,\n",
              "   0.03261512145400047,\n",
              "   -0.014959724619984627,\n",
              "   -0.03196023777127266,\n",
              "   -0.04952366650104523,\n",
              "   -0.006917055696249008,\n",
              "   -0.03988396376371384,\n",
              "   0.021022427827119827,\n",
              "   -0.05483800545334816,\n",
              "   0.03454023227095604,\n",
              "   0.018747694790363312,\n",
              "   0.009009426459670067,\n",
              "   0.0186674352735281,\n",
              "   0.02413845993578434,\n",
              "   -0.020931575447320938,\n",
              "   0.08368521928787231,\n",
              "   -0.0014120028354227543,\n",
              "   0.06917941570281982,\n",
              "   -0.019850246608257294,\n",
              "   0.011925635859370232,\n",
              "   -0.04341932758688927,\n",
              "   0.04833506792783737,\n",
              "   -0.13493074476718903,\n",
              "   0.009594342671334743,\n",
              "   -0.012860861606895924,\n",
              "   -0.05763867869973183,\n",
              "   0.010592828504741192,\n",
              "   0.11231637001037598,\n",
              "   -0.068573497235775,\n",
              "   -0.04226931557059288,\n",
              "   0.0397782139480114,\n",
              "   0.000434533430961892,\n",
              "   0.0634017214179039,\n",
              "   0.015245284885168076,\n",
              "   0.030788281932473183,\n",
              "   0.03153841197490692,\n",
              "   0.053258996456861496,\n",
              "   -0.0038680648431181908,\n",
              "   0.10337628424167633,\n",
              "   0.06015908718109131,\n",
              "   -5.613646791857718e-08,\n",
              "   0.07540380209684372,\n",
              "   0.003678932087495923,\n",
              "   -0.0017827774863690138,\n",
              "   -0.04098271578550339,\n",
              "   -0.06861995160579681,\n",
              "   0.03266307711601257,\n",
              "   -0.009104199707508087,\n",
              "   0.010300107300281525,\n",
              "   -0.05382181704044342,\n",
              "   -0.08396041393280029,\n",
              "   -0.006292405072599649,\n",
              "   0.002157615963369608,\n",
              "   -0.03695129603147507,\n",
              "   -0.013937323354184628,\n",
              "   0.05003126710653305,\n",
              "   0.09713509678840637,\n",
              "   -0.08744002133607864,\n",
              "   -0.01189634669572115,\n",
              "   -0.07756398618221283,\n",
              "   -0.03169942647218704,\n",
              "   -0.1107034832239151,\n",
              "   0.041830990463495255,\n",
              "   -0.02489359676837921,\n",
              "   -0.05358879640698433,\n",
              "   0.01413396093994379,\n",
              "   -0.04430892691016197,\n",
              "   -0.03645027428865433,\n",
              "   0.030870236456394196,\n",
              "   0.04097191244363785,\n",
              "   -0.007725152652710676,\n",
              "   -0.015746479853987694,\n",
              "   -0.02086399495601654,\n",
              "   -0.012622544541954994,\n",
              "   0.027568595483899117,\n",
              "   0.08604206144809723,\n",
              "   0.053847264498472214,\n",
              "   -0.07997340708971024,\n",
              "   -0.011522980406880379,\n",
              "   0.002367734909057617,\n",
              "   0.011289053596556187,\n",
              "   -0.05974890664219856,\n",
              "   0.0034975125454366207,\n",
              "   -0.0961129441857338,\n",
              "   -0.03129688650369644,\n",
              "   -0.035481005907058716,\n",
              "   -0.010340562090277672,\n",
              "   0.020022807642817497,\n",
              "   0.046924833208322525,\n",
              "   -0.005553742405027151,\n",
              "   0.02930901013314724,\n",
              "   0.002250964054837823,\n",
              "   0.004224800039082766,\n",
              "   0.05189524590969086,\n",
              "   -0.04526336491107941,\n",
              "   -0.046410974115133286,\n",
              "   0.09890253096818924,\n",
              "   -0.07636462152004242,\n",
              "   -0.014728537760674953,\n",
              "   0.09904889762401581,\n",
              "   -0.03327077627182007,\n",
              "   -0.06363773345947266,\n",
              "   0.011561092920601368,\n",
              "   -0.0057481424883008,\n",
              "   0.010478324256837368],\n",
              "  [-0.15599121153354645,\n",
              "   -0.017828764393925667,\n",
              "   -0.03289751708507538,\n",
              "   0.013790144585072994,\n",
              "   0.04164232686161995,\n",
              "   -0.07639668136835098,\n",
              "   -0.06565874069929123,\n",
              "   0.021636657416820526,\n",
              "   -0.004342050291597843,\n",
              "   -0.0987490639090538,\n",
              "   0.006738640367984772,\n",
              "   -0.06740302592515945,\n",
              "   -0.012253758497536182,\n",
              "   0.027786854654550552,\n",
              "   -0.05270654335618019,\n",
              "   0.0007518502534367144,\n",
              "   -0.03683526813983917,\n",
              "   0.002571715274825692,\n",
              "   -0.007683835458010435,\n",
              "   0.06563149392604828,\n",
              "   -0.017273811623454094,\n",
              "   -0.04520547762513161,\n",
              "   -0.0111648915335536,\n",
              "   0.0027703046798706055,\n",
              "   -0.05454453453421593,\n",
              "   0.11596576869487762,\n",
              "   0.04010789468884468,\n",
              "   -0.008627166040241718,\n",
              "   -0.05900363624095917,\n",
              "   -0.08620648086071014,\n",
              "   0.09873872250318527,\n",
              "   0.03963106498122215,\n",
              "   -0.003592441324144602,\n",
              "   -0.0505002923309803,\n",
              "   -0.017430391162633896,\n",
              "   0.048118218779563904,\n",
              "   -0.0597684271633625,\n",
              "   0.021200843155384064,\n",
              "   -0.03261758014559746,\n",
              "   -0.08217208832502365,\n",
              "   0.02585502713918686,\n",
              "   0.03584004566073418,\n",
              "   0.0733409896492958,\n",
              "   0.024469802156090736,\n",
              "   0.018718261271715164,\n",
              "   -0.08342058211565018,\n",
              "   0.0003380624984856695,\n",
              "   -0.029917098581790924,\n",
              "   0.030049260705709457,\n",
              "   -0.0398460254073143,\n",
              "   0.0007268677582032979,\n",
              "   0.047084372490644455,\n",
              "   -0.00029067264404147863,\n",
              "   0.0948949009180069,\n",
              "   0.030542461201548576,\n",
              "   0.008768490515649319,\n",
              "   -0.027495242655277252,\n",
              "   -0.0923766940832138,\n",
              "   -0.04846620932221413,\n",
              "   -0.021609177812933922,\n",
              "   0.09510432183742523,\n",
              "   0.009402402676641941,\n",
              "   -0.046620555222034454,\n",
              "   -0.022235548123717308,\n",
              "   0.05994461849331856,\n",
              "   0.05647854879498482,\n",
              "   0.017336033284664154,\n",
              "   -0.07991591840982437,\n",
              "   0.0026722000911831856,\n",
              "   -0.07018083333969116,\n",
              "   0.00199131085537374,\n",
              "   -0.0071725244633853436,\n",
              "   -0.03914652764797211,\n",
              "   0.0632510706782341,\n",
              "   0.09386290609836578,\n",
              "   0.16562508046627045,\n",
              "   0.05602819845080376,\n",
              "   -0.026165898889303207,\n",
              "   0.014079928398132324,\n",
              "   -0.0651375874876976,\n",
              "   0.043617039918899536,\n",
              "   0.06470620632171631,\n",
              "   0.0651920959353447,\n",
              "   -0.015679705888032913,\n",
              "   0.04674239829182625,\n",
              "   0.08515764772891998,\n",
              "   -0.03610684722661972,\n",
              "   0.017314352095127106,\n",
              "   -0.10770324617624283,\n",
              "   -0.025787051767110825,\n",
              "   0.09693893045186996,\n",
              "   0.05915530025959015,\n",
              "   -0.03249015659093857,\n",
              "   -0.001205497421324253,\n",
              "   -0.010420762933790684,\n",
              "   0.03972010314464569,\n",
              "   0.060240477323532104,\n",
              "   0.11389265209436417,\n",
              "   -0.00524791469797492,\n",
              "   0.019406326115131378,\n",
              "   0.02524520643055439,\n",
              "   0.07286601513624191,\n",
              "   -0.10020752251148224,\n",
              "   -0.02361133322119713,\n",
              "   -0.06945937126874924,\n",
              "   0.00930691510438919,\n",
              "   0.03909323364496231,\n",
              "   0.08647729456424713,\n",
              "   0.01259023230522871,\n",
              "   -0.02917812392115593,\n",
              "   0.0043888273648917675,\n",
              "   0.01892603188753128,\n",
              "   -0.031475916504859924,\n",
              "   0.01422277744859457,\n",
              "   0.0009909081272780895,\n",
              "   -0.02519908919930458,\n",
              "   -0.02120649442076683,\n",
              "   0.004107894375920296,\n",
              "   -0.0412076897919178,\n",
              "   0.004582997411489487,\n",
              "   0.005329868756234646,\n",
              "   -0.02378825657069683,\n",
              "   -0.04168723523616791,\n",
              "   -0.005604663398116827,\n",
              "   -0.009753415361046791,\n",
              "   0.0003028219216503203,\n",
              "   -0.15553542971611023,\n",
              "   1.9151799016121722e-32,\n",
              "   1.0291477337887045e-05,\n",
              "   0.041943226009607315,\n",
              "   0.03723989427089691,\n",
              "   -0.011874537914991379,\n",
              "   -0.0410168394446373,\n",
              "   -0.053238801658153534,\n",
              "   -0.07461314648389816,\n",
              "   -0.031856101006269455,\n",
              "   -0.06122202426195145,\n",
              "   -0.0045726606622338295,\n",
              "   -0.10851902514696121,\n",
              "   -0.021802198141813278,\n",
              "   -0.011687912046909332,\n",
              "   -0.008205716498196125,\n",
              "   -0.08326350152492523,\n",
              "   -0.06511089205741882,\n",
              "   0.03577977418899536,\n",
              "   0.010941846296191216,\n",
              "   -0.0006353099015541375,\n",
              "   -0.13925839960575104,\n",
              "   0.0475321039557457,\n",
              "   0.1070038378238678,\n",
              "   -0.028680667281150818,\n",
              "   0.007534502074122429,\n",
              "   0.06997915357351303,\n",
              "   0.002218842739239335,\n",
              "   -0.0970190018415451,\n",
              "   -0.02556091733276844,\n",
              "   -0.01269677747040987,\n",
              "   -0.004120286088436842,\n",
              "   -0.016439346596598625,\n",
              "   -0.044027283787727356,\n",
              "   -0.04612261801958084,\n",
              "   -0.01123539824038744,\n",
              "   0.029230285435914993,\n",
              "   -0.01844751462340355,\n",
              "   -0.007933974266052246,\n",
              "   -0.025706125423312187,\n",
              "   0.006421374622732401,\n",
              "   0.001312385662458837,\n",
              "   0.03098042868077755,\n",
              "   0.019726354628801346,\n",
              "   -0.04266539216041565,\n",
              "   0.07758059352636337,\n",
              "   0.014067232608795166,\n",
              "   -0.07580499351024628,\n",
              "   -0.013674188405275345,\n",
              "   0.048875268548727036,\n",
              "   0.03695593401789665,\n",
              "   -0.03708096221089363,\n",
              "   0.016582386568188667,\n",
              "   -0.014546478167176247,\n",
              "   0.006368817761540413,\n",
              "   -0.002080911537632346,\n",
              "   -0.0779535248875618,\n",
              "   0.0011416421039029956,\n",
              "   0.004768436774611473,\n",
              "   -0.033425651490688324,\n",
              "   0.019564883783459663,\n",
              "   -0.003404716495424509,\n",
              "   0.07033448666334152,\n",
              "   0.03678061440587044,\n",
              "   -0.021470267325639725,\n",
              "   0.024191169068217278,\n",
              "   0.028692489489912987,\n",
              "   0.012951898388564587,\n",
              "   -0.047266822308301926,\n",
              "   -0.007668190635740757,\n",
              "   0.014420018531382084,\n",
              "   0.13992749154567719,\n",
              "   -0.014518102630972862,\n",
              "   -0.006475939881056547,\n",
              "   0.04599352926015854,\n",
              "   0.023005183786153793,\n",
              "   0.09843076020479202,\n",
              "   0.0006603957153856754,\n",
              "   0.019187942147254944,\n",
              "   -0.042366933077573776,\n",
              "   0.08913294970989227,\n",
              "   0.04209429770708084,\n",
              "   -0.019451238214969635,\n",
              "   0.01078027579933405,\n",
              "   -0.00632232241332531,\n",
              "   -0.02421281486749649,\n",
              "   -0.038054969161748886,\n",
              "   0.01142249908298254,\n",
              "   0.02440648153424263,\n",
              "   -0.12238789349794388,\n",
              "   -0.029925113543868065,\n",
              "   -0.02444758638739586,\n",
              "   0.013146919198334217,\n",
              "   -0.03812021017074585,\n",
              "   -0.023388132452964783,\n",
              "   -0.06115012988448143,\n",
              "   -0.07682834565639496,\n",
              "   -1.9591674879671176e-32,\n",
              "   0.03634266182780266,\n",
              "   -0.03857668489217758,\n",
              "   0.08208579570055008,\n",
              "   -0.09483206272125244,\n",
              "   -0.025487219914793968,\n",
              "   -0.02977714128792286,\n",
              "   -0.015469351783394814,\n",
              "   -0.01632772758603096,\n",
              "   0.06801830977201462,\n",
              "   -0.047674525529146194,\n",
              "   0.1657445728778839,\n",
              "   -0.05517817661166191,\n",
              "   0.041392505168914795,\n",
              "   -0.06923437118530273,\n",
              "   0.05443684011697769,\n",
              "   0.06280560046434402,\n",
              "   -0.0536186546087265,\n",
              "   0.019441723823547363,\n",
              "   0.01738905906677246,\n",
              "   0.030458739027380943,\n",
              "   -0.07033364474773407,\n",
              "   0.01040327176451683,\n",
              "   -0.0915820524096489,\n",
              "   -0.028479384258389473,\n",
              "   0.005977049469947815,\n",
              "   0.009791024029254913,\n",
              "   0.02306944690644741,\n",
              "   -0.04282185435295105,\n",
              "   0.13924680650234222,\n",
              "   0.06009618937969208,\n",
              "   -0.05046583339571953,\n",
              "   0.07420317083597183,\n",
              "   -0.08952342718839645,\n",
              "   0.046817611902952194,\n",
              "   -0.059034716337919235,\n",
              "   0.10248982161283493,\n",
              "   -0.017207663506269455,\n",
              "   -0.026617735624313354,\n",
              "   0.007180023938417435,\n",
              "   0.013921275734901428,\n",
              "   -0.03042466752231121,\n",
              "   0.06127174198627472,\n",
              "   -0.0426027774810791,\n",
              "   0.05509069189429283,\n",
              "   0.09189946204423904,\n",
              "   -0.0003071880782954395,\n",
              "   -0.01939995214343071,\n",
              "   0.02060120180249214,\n",
              "   0.01943814568221569,\n",
              "   0.030761046335101128,\n",
              "   0.08819066733121872,\n",
              "   0.038095198571681976,\n",
              "   -0.022944271564483643,\n",
              "   0.017243947833776474,\n",
              "   0.02058529295027256,\n",
              "   0.006442823447287083,\n",
              "   0.08900119364261627,\n",
              "   -0.06653596460819244,\n",
              "   -0.028902171179652214,\n",
              "   -0.041261691600084305,\n",
              "   -0.028748100623488426,\n",
              "   -0.012855516746640205,\n",
              "   0.008256436325609684,\n",
              "   -0.02957835979759693,\n",
              "   -0.01631796360015869,\n",
              "   0.005851649679243565,\n",
              "   0.00437278114259243,\n",
              "   0.08438915014266968,\n",
              "   0.012927032075822353,\n",
              "   -0.04942731931805611,\n",
              "   0.02273545227944851,\n",
              "   0.01255032978951931,\n",
              "   -0.007603548467159271,\n",
              "   -0.05099232867360115,\n",
              "   0.03945634141564369,\n",
              "   -0.0006005493341945112,\n",
              "   -0.08665180206298828,\n",
              "   0.033554915338754654,\n",
              "   -0.022782936692237854,\n",
              "   0.01948009431362152,\n",
              "   0.0275324247777462,\n",
              "   0.0422142893075943,\n",
              "   -0.015948228538036346,\n",
              "   -0.02834119275212288,\n",
              "   0.06738927960395813,\n",
              "   -0.028347279876470566,\n",
              "   -0.003128207288682461,\n",
              "   0.04887156933546066,\n",
              "   -0.010116452351212502,\n",
              "   -0.013310092501342297,\n",
              "   0.01729426346719265,\n",
              "   -0.04741394892334938,\n",
              "   0.02495388127863407,\n",
              "   0.09900463372468948,\n",
              "   0.025950923562049866,\n",
              "   -6.991210454998509e-08,\n",
              "   0.0240623876452446,\n",
              "   -0.03060942143201828,\n",
              "   0.014296864159405231,\n",
              "   -0.061511293053627014,\n",
              "   0.005456590559333563,\n",
              "   0.07356054335832596,\n",
              "   0.029499327763915062,\n",
              "   -0.07189377397298813,\n",
              "   0.024569716304540634,\n",
              "   -0.006864326540380716,\n",
              "   -0.031030138954520226,\n",
              "   -0.027342062443494797,\n",
              "   -0.025462167337536812,\n",
              "   -0.12573033571243286,\n",
              "   -0.00912666879594326,\n",
              "   0.038907721638679504,\n",
              "   -0.0551227368414402,\n",
              "   -0.06499834358692169,\n",
              "   -0.032714877277612686,\n",
              "   0.007022416219115257,\n",
              "   0.05200384557247162,\n",
              "   0.005665014963597059,\n",
              "   -0.06891735643148422,\n",
              "   -0.010781751945614815,\n",
              "   0.01684555597603321,\n",
              "   -0.057023026049137115,\n",
              "   -0.02776891179382801,\n",
              "   -0.024998119100928307,\n",
              "   -0.09751434624195099,\n",
              "   -0.0153344152495265,\n",
              "   -0.03216630220413208,\n",
              "   -0.013331156224012375,\n",
              "   0.02278798818588257,\n",
              "   -0.0034657625947147608,\n",
              "   0.11475620418787003,\n",
              "   -0.016360685229301453,\n",
              "   -0.002167198108509183,\n",
              "   -0.009005521424114704,\n",
              "   -0.035810891538858414,\n",
              "   0.021557169035077095,\n",
              "   -0.06691092252731323,\n",
              "   -0.04974118992686272,\n",
              "   -0.0920242890715599,\n",
              "   0.023854246363043785,\n",
              "   0.057527218014001846,\n",
              "   -0.01370405312627554,\n",
              "   -0.04312523081898689,\n",
              "   0.1220654621720314,\n",
              "   0.04898874834179878,\n",
              "   -0.048492658883333206,\n",
              "   -0.022905420511960983,\n",
              "   -0.03199787810444832,\n",
              "   -3.399925117264502e-05,\n",
              "   -0.056849874556064606,\n",
              "   -0.04649583622813225,\n",
              "   0.08518394827842712,\n",
              "   -0.04213585704565048,\n",
              "   0.0682656541466713,\n",
              "   0.04296359047293663,\n",
              "   -0.08361482620239258,\n",
              "   0.0683722123503685,\n",
              "   -0.06031876429915428,\n",
              "   0.02224717102944851,\n",
              "   -0.036381226032972336],\n",
              "  [-0.13378705084323883,\n",
              "   -0.00020904232223983854,\n",
              "   0.021255938336253166,\n",
              "   0.006092768162488937,\n",
              "   0.06174902245402336,\n",
              "   -0.04425463825464249,\n",
              "   0.004724225029349327,\n",
              "   0.10375066101551056,\n",
              "   0.01454250980168581,\n",
              "   -0.021283719688653946,\n",
              "   -0.003874776419252157,\n",
              "   -0.045549213886260986,\n",
              "   0.009764819405972958,\n",
              "   0.026340162381529808,\n",
              "   -0.08093556016683578,\n",
              "   -0.05628526955842972,\n",
              "   0.05527655407786369,\n",
              "   0.10000326484441757,\n",
              "   0.01204758696258068,\n",
              "   -0.016866348683834076,\n",
              "   0.008439265191555023,\n",
              "   0.019390864297747612,\n",
              "   -0.01647820882499218,\n",
              "   -0.03095291182398796,\n",
              "   -0.0034922990016639233,\n",
              "   0.02537483721971512,\n",
              "   0.03290877863764763,\n",
              "   0.035556141287088394,\n",
              "   -0.03937261179089546,\n",
              "   -0.04282468557357788,\n",
              "   -0.022373702377080917,\n",
              "   -0.10731305181980133,\n",
              "   0.0780852809548378,\n",
              "   -0.020693272352218628,\n",
              "   0.03272348642349243,\n",
              "   0.09880125522613525,\n",
              "   0.002325813751667738,\n",
              "   -0.0367012694478035,\n",
              "   6.938100705156103e-05,\n",
              "   -0.09009543806314468,\n",
              "   0.02685239538550377,\n",
              "   0.07540017366409302,\n",
              "   0.0029341126792132854,\n",
              "   0.0010991219896823168,\n",
              "   0.005978124216198921,\n",
              "   0.0008349725976586342,\n",
              "   0.02583148144185543,\n",
              "   0.056724242866039276,\n",
              "   0.04257264360785484,\n",
              "   -0.006876200903207064,\n",
              "   0.004100016318261623,\n",
              "   0.05093569681048393,\n",
              "   0.06179390847682953,\n",
              "   0.05237903445959091,\n",
              "   0.018909379839897156,\n",
              "   -0.014108269475400448,\n",
              "   -0.1004192903637886,\n",
              "   -0.02549107000231743,\n",
              "   -0.056913841515779495,\n",
              "   -0.008386572822928429,\n",
              "   0.015006177127361298,\n",
              "   -0.03309747204184532,\n",
              "   0.021740484982728958,\n",
              "   -0.0722595751285553,\n",
              "   0.003858489915728569,\n",
              "   -0.047168027609586716,\n",
              "   0.03208151459693909,\n",
              "   0.017703866586089134,\n",
              "   0.04954499006271362,\n",
              "   -0.014368418604135513,\n",
              "   -0.03788745030760765,\n",
              "   0.04686778038740158,\n",
              "   0.05945776775479317,\n",
              "   0.011775286868214607,\n",
              "   -0.01932477578520775,\n",
              "   0.08772866427898407,\n",
              "   0.0940905511379242,\n",
              "   -0.04315733164548874,\n",
              "   0.01767677627503872,\n",
              "   -0.026118528097867966,\n",
              "   0.04696308821439743,\n",
              "   0.03980596736073494,\n",
              "   0.06007546931505203,\n",
              "   0.054541122168302536,\n",
              "   0.02032739296555519,\n",
              "   0.05744168907403946,\n",
              "   -0.04343205317854881,\n",
              "   0.0684308409690857,\n",
              "   -0.09668438136577606,\n",
              "   0.012297107838094234,\n",
              "   0.030061649158596992,\n",
              "   0.06732950359582901,\n",
              "   0.03473277762532234,\n",
              "   -0.04118957743048668,\n",
              "   -0.021588914096355438,\n",
              "   0.03696761652827263,\n",
              "   -0.061698492616415024,\n",
              "   0.08840988576412201,\n",
              "   0.018266426399350166,\n",
              "   -0.0020639286376535892,\n",
              "   0.02101675420999527,\n",
              "   0.084483802318573,\n",
              "   -0.01958450675010681,\n",
              "   -0.0893668532371521,\n",
              "   -0.029144128784537315,\n",
              "   -0.043325938284397125,\n",
              "   0.03468981012701988,\n",
              "   0.1074497401714325,\n",
              "   -0.013789028860628605,\n",
              "   -0.013040578924119473,\n",
              "   0.011293202638626099,\n",
              "   -0.003872802946716547,\n",
              "   0.07079554349184036,\n",
              "   0.030566466972231865,\n",
              "   -0.028842831030488014,\n",
              "   0.04856693744659424,\n",
              "   0.018502771854400635,\n",
              "   -0.04787253215909004,\n",
              "   -0.08645288646221161,\n",
              "   -0.04017230495810509,\n",
              "   -0.012310116551816463,\n",
              "   -0.05876871570944786,\n",
              "   0.048159752041101456,\n",
              "   0.0048305015079677105,\n",
              "   -0.010747101157903671,\n",
              "   -0.04356355220079422,\n",
              "   -0.01468747854232788,\n",
              "   5.8499027381648086e-33,\n",
              "   0.006851104088127613,\n",
              "   0.01689487136900425,\n",
              "   -0.027858613058924675,\n",
              "   0.06736142188310623,\n",
              "   0.03830292448401451,\n",
              "   0.012483657337725163,\n",
              "   -0.03516365587711334,\n",
              "   0.02444990538060665,\n",
              "   -0.029747065156698227,\n",
              "   0.020571613684296608,\n",
              "   -0.050338614732027054,\n",
              "   -0.13310359418392181,\n",
              "   0.018693024292588234,\n",
              "   -0.07795161753892899,\n",
              "   -0.05788653343915939,\n",
              "   -0.017367076128721237,\n",
              "   0.001422171131707728,\n",
              "   0.0277074184268713,\n",
              "   0.05024829879403114,\n",
              "   -0.0014992643846198916,\n",
              "   -0.02575492113828659,\n",
              "   0.013974837027490139,\n",
              "   -0.0039244042709469795,\n",
              "   -0.048675429075956345,\n",
              "   0.07292091101408005,\n",
              "   -0.04478670656681061,\n",
              "   -0.042953912168741226,\n",
              "   -0.05789024382829666,\n",
              "   0.039144136011600494,\n",
              "   -0.034371569752693176,\n",
              "   0.01616348698735237,\n",
              "   -0.06915563344955444,\n",
              "   -0.07985413819551468,\n",
              "   -0.01245156954973936,\n",
              "   0.049681250005960464,\n",
              "   -0.025830121710896492,\n",
              "   0.022379927337169647,\n",
              "   0.053834110498428345,\n",
              "   -0.02257762849330902,\n",
              "   -0.007148267235606909,\n",
              "   0.009498516097664833,\n",
              "   0.07817158848047256,\n",
              "   -0.029721591621637344,\n",
              "   -0.004905322100967169,\n",
              "   0.07512064278125763,\n",
              "   -0.05167799070477486,\n",
              "   -0.05921897664666176,\n",
              "   0.07924772799015045,\n",
              "   -0.005391351412981749,\n",
              "   0.013526463881134987,\n",
              "   0.08794397115707397,\n",
              "   -0.00017886490968521684,\n",
              "   0.13092342019081116,\n",
              "   -0.1274404525756836,\n",
              "   -0.05892479792237282,\n",
              "   -0.005825945641845465,\n",
              "   -0.005263681523501873,\n",
              "   0.02408376708626747,\n",
              "   0.020568814128637314,\n",
              "   0.06040176749229431,\n",
              "   -0.011390259489417076,\n",
              "   0.07583601027727127,\n",
              "   -0.07817091792821884,\n",
              "   0.042683497071266174,\n",
              "   0.00964423269033432,\n",
              "   -0.009584533050656319,\n",
              "   -0.04416210576891899,\n",
              "   -0.002139682648703456,\n",
              "   -0.014682849869132042,\n",
              "   0.09757671505212784,\n",
              "   -0.07736610621213913,\n",
              "   -0.03926219418644905,\n",
              "   0.010514338500797749,\n",
              "   0.016338612884283066,\n",
              "   -0.04094240069389343,\n",
              "   -0.0569651760160923,\n",
              "   -0.039296623319387436,\n",
              "   0.05413582921028137,\n",
              "   0.10256274044513702,\n",
              "   0.049101993441581726,\n",
              "   0.019008632749319077,\n",
              "   0.003530129324644804,\n",
              "   -0.016708722338080406,\n",
              "   -0.03156580403447151,\n",
              "   -0.047814223915338516,\n",
              "   0.005357971880584955,\n",
              "   0.0003161749627906829,\n",
              "   -0.007929271087050438,\n",
              "   0.12250718474388123,\n",
              "   0.01715567149221897,\n",
              "   0.11549869924783707,\n",
              "   -0.11331024020910263,\n",
              "   -0.09082376211881638,\n",
              "   -0.008013132028281689,\n",
              "   -0.020019590854644775,\n",
              "   -4.763650081747026e-33,\n",
              "   0.019374927505850792,\n",
              "   0.040321819484233856,\n",
              "   -0.010549443773925304,\n",
              "   -0.05190705507993698,\n",
              "   -0.06651037186384201,\n",
              "   0.02478213608264923,\n",
              "   0.021961495280265808,\n",
              "   0.047166842967271805,\n",
              "   0.08027593791484833,\n",
              "   -0.05257406085729599,\n",
              "   0.1711561381816864,\n",
              "   -0.003146778093650937,\n",
              "   0.019370131194591522,\n",
              "   -0.01686842180788517,\n",
              "   0.009295282885432243,\n",
              "   0.08755060285329819,\n",
              "   -0.11949019879102707,\n",
              "   0.04573759064078331,\n",
              "   -0.039388395845890045,\n",
              "   0.016664767637848854,\n",
              "   -0.09416531026363373,\n",
              "   0.002646752865985036,\n",
              "   0.06130775064229965,\n",
              "   -0.026503611356019974,\n",
              "   0.018113678321242332,\n",
              "   0.08654533326625824,\n",
              "   0.04287877306342125,\n",
              "   0.043060142546892166,\n",
              "   0.07581465691328049,\n",
              "   0.04889269918203354,\n",
              "   0.034893400967121124,\n",
              "   0.0451497957110405,\n",
              "   -0.06804709881544113,\n",
              "   -0.05401887372136116,\n",
              "   -0.0721496120095253,\n",
              "   -0.012850846163928509,\n",
              "   -0.03012373298406601,\n",
              "   -0.08693546056747437,\n",
              "   0.03761397302150726,\n",
              "   0.005205963738262653,\n",
              "   0.03189309313893318,\n",
              "   0.03624332696199417,\n",
              "   -0.03200143575668335,\n",
              "   0.01936865970492363,\n",
              "   -0.02629312127828598,\n",
              "   0.10936674475669861,\n",
              "   -0.04441343992948532,\n",
              "   -0.04647613689303398,\n",
              "   0.007706556469202042,\n",
              "   -0.10144337266683578,\n",
              "   -0.008220246993005276,\n",
              "   -0.0017734072171151638,\n",
              "   -0.05870841071009636,\n",
              "   -0.006713747046887875,\n",
              "   0.07228824496269226,\n",
              "   -0.018083155155181885,\n",
              "   0.027421755716204643,\n",
              "   0.04675666242837906,\n",
              "   -0.08214019984006882,\n",
              "   -0.010187932290136814,\n",
              "   0.04178706556558609,\n",
              "   -0.037161972373723984,\n",
              "   0.0417429581284523,\n",
              "   0.002395791932940483,\n",
              "   0.00509093701839447,\n",
              "   0.005412204656749964,\n",
              "   -0.03623897209763527,\n",
              "   0.09725603461265564,\n",
              "   -0.011394931934773922,\n",
              "   -0.01262357085943222,\n",
              "   -0.017719389870762825,\n",
              "   0.04122135415673256,\n",
              "   0.05076956748962402,\n",
              "   -0.04978521913290024,\n",
              "   0.042744651436805725,\n",
              "   0.022331465035676956,\n",
              "   -0.06190352514386177,\n",
              "   0.018821513280272484,\n",
              "   0.01563119888305664,\n",
              "   0.0615050233900547,\n",
              "   -0.010721988044679165,\n",
              "   0.033829379826784134,\n",
              "   -0.10797414183616638,\n",
              "   -0.029697809368371964,\n",
              "   0.0783938392996788,\n",
              "   -0.025114020332694054,\n",
              "   0.06776362657546997,\n",
              "   0.15931545197963715,\n",
              "   -0.054731231182813644,\n",
              "   0.0004526365373749286,\n",
              "   -0.015241960063576698,\n",
              "   -0.0600748248398304,\n",
              "   -0.008463186211884022,\n",
              "   0.11806917935609818,\n",
              "   0.01551804132759571,\n",
              "   -5.4256705794841764e-08,\n",
              "   0.005129444878548384,\n",
              "   -0.06502783298492432,\n",
              "   0.016800573095679283,\n",
              "   -0.07809610664844513,\n",
              "   -0.09873942285776138,\n",
              "   0.044442132115364075,\n",
              "   0.004161301534622908,\n",
              "   0.03335686773061752,\n",
              "   0.005272544454783201,\n",
              "   -0.004584016744047403,\n",
              "   -0.053456228226423264,\n",
              "   0.10699691623449326,\n",
              "   -0.03885119408369064,\n",
              "   -0.07109316438436508,\n",
              "   -0.0020108933094888926,\n",
              "   0.008540181443095207,\n",
              "   -0.07221715152263641,\n",
              "   -0.06412656605243683,\n",
              "   -0.004392132628709078,\n",
              "   -0.0037242709659039974,\n",
              "   0.006706934422254562,\n",
              "   0.0056530749425292015,\n",
              "   -0.06115344539284706,\n",
              "   -0.0013893505092710257,\n",
              "   0.056908443570137024,\n",
              "   -0.06346002221107483,\n",
              "   0.024949418380856514,\n",
              "   -0.08926759660243988,\n",
              "   -0.03538808971643448,\n",
              "   -0.048963647335767746,\n",
              "   -0.04419558495283127,\n",
              "   -0.02284601330757141,\n",
              "   0.00826614536345005,\n",
              "   -0.013669626787304878,\n",
              "   0.04127845913171768,\n",
              "   -0.007957162335515022,\n",
              "   -0.05079565942287445,\n",
              "   0.030999133363366127,\n",
              "   0.005054077133536339,\n",
              "   0.016059277579188347,\n",
              "   -0.02626553177833557,\n",
              "   0.04216117784380913,\n",
              "   -0.09939443320035934,\n",
              "   0.020290182903409004,\n",
              "   0.015490802936255932,\n",
              "   -0.006209907587617636,\n",
              "   -0.011281653307378292,\n",
              "   0.13119231164455414,\n",
              "   -0.07676759362220764,\n",
              "   -0.05016570910811424,\n",
              "   -0.01968848705291748,\n",
              "   -0.05552741512656212,\n",
              "   0.021635876968503,\n",
              "   -0.08060090243816376,\n",
              "   -0.05510620027780533,\n",
              "   0.05028380826115608,\n",
              "   -0.012758498080074787,\n",
              "   -0.01478415634483099,\n",
              "   0.012684883549809456,\n",
              "   0.019722631201148033,\n",
              "   -0.023332076147198677,\n",
              "   -0.016390739008784294,\n",
              "   0.026663022115826607,\n",
              "   -0.03598811849951744],\n",
              "  [-0.10735748708248138,\n",
              "   0.0038885215763002634,\n",
              "   -0.038562364876270294,\n",
              "   0.014719503931701183,\n",
              "   0.08038659393787384,\n",
              "   0.014318502508103848,\n",
              "   0.028026850894093513,\n",
              "   0.04254288971424103,\n",
              "   -0.032223571091890335,\n",
              "   -0.030651943758130074,\n",
              "   -0.007642779964953661,\n",
              "   0.03406111150979996,\n",
              "   0.048515427857637405,\n",
              "   0.02502608485519886,\n",
              "   -0.034776415675878525,\n",
              "   -0.08695026487112045,\n",
              "   -0.013628324493765831,\n",
              "   0.021323012188076973,\n",
              "   0.012863842770457268,\n",
              "   0.031897760927677155,\n",
              "   -0.030520116910338402,\n",
              "   -0.014095650985836983,\n",
              "   0.04554746299982071,\n",
              "   -0.062293220311403275,\n",
              "   -0.04691227152943611,\n",
              "   0.01989893987774849,\n",
              "   0.007855427451431751,\n",
              "   -0.008619507774710655,\n",
              "   0.014992729760706425,\n",
              "   -0.03961614891886711,\n",
              "   -0.02032404951751232,\n",
              "   -0.06935478001832962,\n",
              "   0.03719135746359825,\n",
              "   0.024802081286907196,\n",
              "   0.04485873505473137,\n",
              "   -0.0009336106013506651,\n",
              "   -0.009286215528845787,\n",
              "   -0.031127527356147766,\n",
              "   -0.08131735771894455,\n",
              "   -0.035112928599119186,\n",
              "   0.05261183902621269,\n",
              "   0.05108853802084923,\n",
              "   0.062898188829422,\n",
              "   -0.024493752047419548,\n",
              "   0.029249116778373718,\n",
              "   -0.11402200907468796,\n",
              "   0.06727380305528641,\n",
              "   -0.015728125348687172,\n",
              "   -0.006656285375356674,\n",
              "   0.0030667991377413273,\n",
              "   0.0391898937523365,\n",
              "   -0.034388218075037,\n",
              "   0.04329397901892662,\n",
              "   0.04312976822257042,\n",
              "   -0.044119495898485184,\n",
              "   0.07523764669895172,\n",
              "   -0.07753406465053558,\n",
              "   -0.007024148479104042,\n",
              "   0.029771016910672188,\n",
              "   -0.08101911842823029,\n",
              "   0.05385538563132286,\n",
              "   -0.02007611282169819,\n",
              "   0.02192099019885063,\n",
              "   -0.013872371055185795,\n",
              "   -0.004283830989152193,\n",
              "   -0.062307361513376236,\n",
              "   0.055573105812072754,\n",
              "   -0.015421035699546337,\n",
              "   -0.0006762677803635597,\n",
              "   0.02556181326508522,\n",
              "   -0.012914990074932575,\n",
              "   -0.0032014762982726097,\n",
              "   -0.0849742442369461,\n",
              "   -0.07156866788864136,\n",
              "   0.057817816734313965,\n",
              "   0.147885262966156,\n",
              "   0.10823872685432434,\n",
              "   -0.07250326126813889,\n",
              "   0.016280952841043472,\n",
              "   -0.09922455251216888,\n",
              "   -0.0030823328997939825,\n",
              "   0.08604451268911362,\n",
              "   0.09421603381633759,\n",
              "   0.04088019207119942,\n",
              "   0.055729012936353683,\n",
              "   0.0462154783308506,\n",
              "   -0.004135980736464262,\n",
              "   0.03165290504693985,\n",
              "   -0.02576218731701374,\n",
              "   -0.028972521424293518,\n",
              "   -2.7953507014899515e-05,\n",
              "   0.04689202830195427,\n",
              "   -0.047894902527332306,\n",
              "   0.021936651319265366,\n",
              "   -0.06469990313053131,\n",
              "   0.04602253437042236,\n",
              "   0.07004576176404953,\n",
              "   0.10626382380723953,\n",
              "   -0.044596992433071136,\n",
              "   0.012870661914348602,\n",
              "   0.08411262184381485,\n",
              "   0.13343270123004913,\n",
              "   -0.041740622371435165,\n",
              "   -0.0690520629286766,\n",
              "   -0.010352657176554203,\n",
              "   0.015028400346636772,\n",
              "   -0.006962434854358435,\n",
              "   0.10901113599538803,\n",
              "   -0.023682717233896255,\n",
              "   -0.003130939556285739,\n",
              "   -0.04210805520415306,\n",
              "   -0.0006535890279337764,\n",
              "   0.04707736894488335,\n",
              "   -0.009119088761508465,\n",
              "   -0.02787838503718376,\n",
              "   0.042034853249788284,\n",
              "   -0.06904593855142593,\n",
              "   0.01333009172230959,\n",
              "   -0.054184816777706146,\n",
              "   -0.023272238671779633,\n",
              "   -0.01452244445681572,\n",
              "   -0.0002536837419029325,\n",
              "   -0.030400000512599945,\n",
              "   -0.0019666291773319244,\n",
              "   -0.006562025286257267,\n",
              "   0.008296031504869461,\n",
              "   -0.10429776459932327,\n",
              "   2.1104629046701572e-33,\n",
              "   -0.002940029138699174,\n",
              "   0.0021241523791104555,\n",
              "   -0.015426501631736755,\n",
              "   -0.031618665903806686,\n",
              "   0.0012572786072269082,\n",
              "   -0.048740535974502563,\n",
              "   -0.03229716047644615,\n",
              "   0.025925355032086372,\n",
              "   -0.09274066239595413,\n",
              "   0.007411471102386713,\n",
              "   -0.08438119292259216,\n",
              "   -0.029384415596723557,\n",
              "   0.0442233681678772,\n",
              "   -0.08427930623292923,\n",
              "   -0.12833362817764282,\n",
              "   -0.0540289469063282,\n",
              "   0.024879461154341698,\n",
              "   0.00541708106175065,\n",
              "   0.004141917452216148,\n",
              "   -0.014856853522360325,\n",
              "   -0.015836281701922417,\n",
              "   -0.008155961520969868,\n",
              "   -0.03945862874388695,\n",
              "   -0.01738084852695465,\n",
              "   0.11941991746425629,\n",
              "   0.015067260712385178,\n",
              "   -0.05127640813589096,\n",
              "   0.0062522669322788715,\n",
              "   0.02898959442973137,\n",
              "   -0.001501298975199461,\n",
              "   -0.025727178901433945,\n",
              "   -0.038788266479969025,\n",
              "   -0.023603877052664757,\n",
              "   -0.008196261711418629,\n",
              "   0.08795084059238434,\n",
              "   -0.027879413217306137,\n",
              "   0.0324750691652298,\n",
              "   0.0419439934194088,\n",
              "   0.04584996774792671,\n",
              "   -0.03925187140703201,\n",
              "   0.049362871795892715,\n",
              "   0.022177549079060555,\n",
              "   -0.052242476493120193,\n",
              "   0.051447294652462006,\n",
              "   0.046639297157526016,\n",
              "   -0.014313094317913055,\n",
              "   0.02582980878651142,\n",
              "   0.07442318648099899,\n",
              "   -0.035213403403759,\n",
              "   -0.02594234235584736,\n",
              "   0.013700072653591633,\n",
              "   -0.07178022712469101,\n",
              "   0.006876388099044561,\n",
              "   0.04209226369857788,\n",
              "   -0.08211539685726166,\n",
              "   -0.020015953108668327,\n",
              "   -0.0632406547665596,\n",
              "   0.0014104217989370227,\n",
              "   0.026598893105983734,\n",
              "   0.05067271366715431,\n",
              "   0.04368225112557411,\n",
              "   0.01781025528907776,\n",
              "   -0.08276773989200592,\n",
              "   0.062303461134433746,\n",
              "   0.013764808885753155,\n",
              "   0.09127335250377655,\n",
              "   -0.03329667076468468,\n",
              "   -0.011019493453204632,\n",
              "   -0.04087920859456062,\n",
              "   0.09548274427652359,\n",
              "   -0.06878653168678284,\n",
              "   0.04489635303616524,\n",
              "   0.01148936152458191,\n",
              "   -0.010362058877944946,\n",
              "   0.0020625791512429714,\n",
              "   -0.035950660705566406,\n",
              "   -0.06918010115623474,\n",
              "   -0.01212471816688776,\n",
              "   0.1539154350757599,\n",
              "   0.06226348876953125,\n",
              "   -0.034909818321466446,\n",
              "   -0.005064180586487055,\n",
              "   -0.035221558064222336,\n",
              "   -0.01418999768793583,\n",
              "   -0.02964917942881584,\n",
              "   -0.11828897893428802,\n",
              "   -0.06369759142398834,\n",
              "   -0.05862514674663544,\n",
              "   0.025622230023145676,\n",
              "   0.05056552216410637,\n",
              "   0.03852665796875954,\n",
              "   -0.07940459251403809,\n",
              "   0.06519033014774323,\n",
              "   -0.08586318790912628,\n",
              "   -0.01903163641691208,\n",
              "   -4.281103773262723e-33,\n",
              "   -0.001484306761994958,\n",
              "   -0.017550187185406685,\n",
              "   -0.02594977430999279,\n",
              "   -0.0325649231672287,\n",
              "   -0.04421870782971382,\n",
              "   0.006695437245070934,\n",
              "   0.03803200647234917,\n",
              "   -0.04898069426417351,\n",
              "   -0.0032224131282418966,\n",
              "   -0.15346159040927887,\n",
              "   0.11948803812265396,\n",
              "   -0.03783252462744713,\n",
              "   -0.0070870062336325645,\n",
              "   -0.0038333283737301826,\n",
              "   0.01528710313141346,\n",
              "   0.017399050295352936,\n",
              "   -0.11669915914535522,\n",
              "   0.08646047860383987,\n",
              "   0.044974081218242645,\n",
              "   0.03225523978471756,\n",
              "   -0.04766368493437767,\n",
              "   -0.05131889507174492,\n",
              "   -0.02091217413544655,\n",
              "   0.07463178783655167,\n",
              "   -0.012214375659823418,\n",
              "   0.001728444010950625,\n",
              "   0.01581687666475773,\n",
              "   -0.05710873380303383,\n",
              "   0.08121122419834137,\n",
              "   0.1223844513297081,\n",
              "   -0.015276322141289711,\n",
              "   0.0885557010769844,\n",
              "   -0.08148617297410965,\n",
              "   0.03306422755122185,\n",
              "   -0.05599721893668175,\n",
              "   0.09060880541801453,\n",
              "   -0.03157152980566025,\n",
              "   -0.045958396047353745,\n",
              "   0.004853885620832443,\n",
              "   0.01944110542535782,\n",
              "   0.03391951322555542,\n",
              "   -0.06875459849834442,\n",
              "   0.03364870697259903,\n",
              "   0.017418013885617256,\n",
              "   0.045561447739601135,\n",
              "   0.12944163382053375,\n",
              "   -0.08122164011001587,\n",
              "   0.017011111602187157,\n",
              "   -0.08535915613174438,\n",
              "   -0.00018918493879027665,\n",
              "   0.05728603154420853,\n",
              "   0.03191131353378296,\n",
              "   -0.03299543261528015,\n",
              "   0.009085037745535374,\n",
              "   0.04220711439847946,\n",
              "   0.0805770605802536,\n",
              "   0.055628951638936996,\n",
              "   -0.012678949162364006,\n",
              "   -0.05688052996993065,\n",
              "   0.0016562415985390544,\n",
              "   0.06388929486274719,\n",
              "   0.03431481868028641,\n",
              "   -0.07624486833810806,\n",
              "   0.009166084229946136,\n",
              "   -0.034509964287281036,\n",
              "   -0.04861463978886604,\n",
              "   0.037813782691955566,\n",
              "   0.04236448183655739,\n",
              "   0.031227484345436096,\n",
              "   -0.07284761965274811,\n",
              "   0.031931322067976,\n",
              "   0.0033384677954018116,\n",
              "   0.01074107363820076,\n",
              "   -9.831667557591572e-05,\n",
              "   -0.038506414741277695,\n",
              "   -0.001991637982428074,\n",
              "   0.04092451184988022,\n",
              "   -0.03208782523870468,\n",
              "   0.020658506080508232,\n",
              "   0.00010397307778475806,\n",
              "   -0.028538623824715614,\n",
              "   0.07677721977233887,\n",
              "   -0.015670878812670708,\n",
              "   -0.012347402982413769,\n",
              "   0.03331902250647545,\n",
              "   0.038081780076026917,\n",
              "   0.007338436786085367,\n",
              "   0.09194423258304596,\n",
              "   -0.006066540721803904,\n",
              "   0.02602904476225376,\n",
              "   -0.02301441878080368,\n",
              "   -0.03510909900069237,\n",
              "   0.029248222708702087,\n",
              "   0.11099515110254288,\n",
              "   -0.03020104579627514,\n",
              "   -5.494394272886893e-08,\n",
              "   -0.01190826017409563,\n",
              "   -0.015866419300436974,\n",
              "   0.05261257290840149,\n",
              "   -0.05222030356526375,\n",
              "   -0.09207403659820557,\n",
              "   0.0423315167427063,\n",
              "   0.05191623419523239,\n",
              "   -0.0409906841814518,\n",
              "   -0.014306971803307533,\n",
              "   0.0017701939214020967,\n",
              "   0.03054874576628208,\n",
              "   0.0453311987221241,\n",
              "   0.017763370648026466,\n",
              "   -0.08079005032777786,\n",
              "   0.022065145894885063,\n",
              "   -0.02439742349088192,\n",
              "   -0.06830137223005295,\n",
              "   -0.04181620106101036,\n",
              "   -0.02779770828783512,\n",
              "   0.027723658829927444,\n",
              "   0.004449248779565096,\n",
              "   -0.005140453577041626,\n",
              "   -0.05322108790278435,\n",
              "   -0.00055086420616135,\n",
              "   0.03517943248152733,\n",
              "   -0.13291558623313904,\n",
              "   0.07168246060609818,\n",
              "   0.05831436812877655,\n",
              "   -0.0663919597864151,\n",
              "   -0.043642859905958176,\n",
              "   -0.007582078222185373,\n",
              "   -0.032139383256435394,\n",
              "   -0.023329544812440872,\n",
              "   0.10457934439182281,\n",
              "   0.08927394449710846,\n",
              "   -0.05994316190481186,\n",
              "   -0.0005253606359474361,\n",
              "   -0.043644651770591736,\n",
              "   0.01783502660691738,\n",
              "   0.02204359881579876,\n",
              "   -0.046953000128269196,\n",
              "   -0.07534940540790558,\n",
              "   0.024662382900714874,\n",
              "   -0.04634948447346687,\n",
              "   -0.005753046832978725,\n",
              "   0.044715799391269684,\n",
              "   0.006441834848374128,\n",
              "   0.08461226522922516,\n",
              "   0.01755153201520443,\n",
              "   0.0405629463493824,\n",
              "   -0.10696367919445038,\n",
              "   -0.0298293624073267,\n",
              "   0.02476103976368904,\n",
              "   -0.10608776658773422,\n",
              "   0.03370840102434158,\n",
              "   0.04557931795716286,\n",
              "   -0.007602259051054716,\n",
              "   -0.0075036003254354,\n",
              "   0.02337866835296154,\n",
              "   -0.014563250355422497,\n",
              "   0.011804011650383472,\n",
              "   -0.0370069295167923,\n",
              "   -0.01978946290910244,\n",
              "   -0.01743224635720253],\n",
              "  [-0.14880360662937164,\n",
              "   -0.10036742687225342,\n",
              "   0.012439753860235214,\n",
              "   0.02189810387790203,\n",
              "   0.11402064561843872,\n",
              "   -0.04351130127906799,\n",
              "   -0.019256655126810074,\n",
              "   0.10172118246555328,\n",
              "   0.08981458842754364,\n",
              "   -0.027726691216230392,\n",
              "   0.034215543419122696,\n",
              "   0.021211186423897743,\n",
              "   0.013262493535876274,\n",
              "   0.03394576534628868,\n",
              "   -0.06995510309934616,\n",
              "   -0.07295496761798859,\n",
              "   -0.020593145862221718,\n",
              "   0.04427909851074219,\n",
              "   -0.021754974499344826,\n",
              "   0.03404665365815163,\n",
              "   0.0015413459623232484,\n",
              "   -0.0514189712703228,\n",
              "   0.0006891710800118744,\n",
              "   0.01771991327404976,\n",
              "   -0.019588278606534004,\n",
              "   0.07839623093605042,\n",
              "   0.003121567191556096,\n",
              "   -0.03849296644330025,\n",
              "   -0.025192897766828537,\n",
              "   -0.08875665068626404,\n",
              "   -0.02256876230239868,\n",
              "   -0.029800469055771828,\n",
              "   -0.02776946872472763,\n",
              "   0.07755021750926971,\n",
              "   -0.05670370161533356,\n",
              "   0.052347660064697266,\n",
              "   0.06605289876461029,\n",
              "   -0.016424918547272682,\n",
              "   -0.02179548516869545,\n",
              "   -0.0530911423265934,\n",
              "   0.025270836427807808,\n",
              "   0.001975132618099451,\n",
              "   0.05640348419547081,\n",
              "   -0.005811103153973818,\n",
              "   0.026449501514434814,\n",
              "   -0.039661891758441925,\n",
              "   0.023214243352413177,\n",
              "   -0.015814127400517464,\n",
              "   -0.024787133559584618,\n",
              "   0.0029237656854093075,\n",
              "   0.02151351422071457,\n",
              "   -0.010050574317574501,\n",
              "   0.010709783993661404,\n",
              "   0.07991108298301697,\n",
              "   0.027283571660518646,\n",
              "   0.09520366042852402,\n",
              "   -0.08590850979089737,\n",
              "   0.0367477610707283,\n",
              "   -0.02699614316225052,\n",
              "   0.028589634224772453,\n",
              "   -0.010861440561711788,\n",
              "   0.000558186904527247,\n",
              "   0.016156159341335297,\n",
              "   0.004361296072602272,\n",
              "   0.057657308876514435,\n",
              "   0.016036102548241615,\n",
              "   0.0517968088388443,\n",
              "   0.025041814893484116,\n",
              "   0.02202684059739113,\n",
              "   0.05596036836504936,\n",
              "   -0.025908397510647774,\n",
              "   0.05905056372284889,\n",
              "   0.02149071730673313,\n",
              "   0.03609365224838257,\n",
              "   0.11177394539117813,\n",
              "   0.13846316933631897,\n",
              "   -0.012231248430907726,\n",
              "   -0.006141964811831713,\n",
              "   0.005856083240360022,\n",
              "   -0.011270756833255291,\n",
              "   -0.013906405307352543,\n",
              "   0.023756755515933037,\n",
              "   0.09573713690042496,\n",
              "   0.07061579078435898,\n",
              "   0.029552990570664406,\n",
              "   -0.006716814357787371,\n",
              "   0.06833020597696304,\n",
              "   0.03544061258435249,\n",
              "   -0.08207571506500244,\n",
              "   -0.008597016334533691,\n",
              "   0.014936055988073349,\n",
              "   0.04667280986905098,\n",
              "   -0.03411058709025383,\n",
              "   0.023031337186694145,\n",
              "   -0.058275431394577026,\n",
              "   0.084795281291008,\n",
              "   0.04588765650987625,\n",
              "   -0.01656830869615078,\n",
              "   0.010984351858496666,\n",
              "   -0.08260135352611542,\n",
              "   0.020893746986985207,\n",
              "   0.04692276194691658,\n",
              "   0.008981339633464813,\n",
              "   -0.03276340290904045,\n",
              "   -0.021280162036418915,\n",
              "   -0.12841086089611053,\n",
              "   -0.028732728213071823,\n",
              "   0.05259179323911667,\n",
              "   -0.041250891983509064,\n",
              "   -0.023499421775341034,\n",
              "   0.025267312303185463,\n",
              "   -0.018667742609977722,\n",
              "   0.048075150698423386,\n",
              "   -0.012946016155183315,\n",
              "   -0.024778835475444794,\n",
              "   -0.003761363448575139,\n",
              "   -0.03396905958652496,\n",
              "   -0.025931138545274734,\n",
              "   -0.021685710176825523,\n",
              "   -0.05254003405570984,\n",
              "   0.0131429024040699,\n",
              "   0.027657238766551018,\n",
              "   -0.04426250979304314,\n",
              "   0.07222796976566315,\n",
              "   -0.01691773347556591,\n",
              "   -0.06655796617269516,\n",
              "   -0.11520357429981232,\n",
              "   3.077804756610665e-33,\n",
              "   0.006572326645255089,\n",
              "   0.019586358219385147,\n",
              "   -0.015363799408078194,\n",
              "   0.10594926029443741,\n",
              "   -0.011331528425216675,\n",
              "   -0.08427582681179047,\n",
              "   -0.003765433793887496,\n",
              "   -0.022233903408050537,\n",
              "   -0.03245450183749199,\n",
              "   -0.011506922543048859,\n",
              "   0.004162931814789772,\n",
              "   -0.08765947818756104,\n",
              "   0.037308722734451294,\n",
              "   -0.06285494565963745,\n",
              "   -0.03287794440984726,\n",
              "   -0.007773811928927898,\n",
              "   0.003308011218905449,\n",
              "   -0.047699667513370514,\n",
              "   0.06317167729139328,\n",
              "   -0.10543425381183624,\n",
              "   0.04289894178509712,\n",
              "   0.06967145949602127,\n",
              "   0.024600276723504066,\n",
              "   -0.0454866886138916,\n",
              "   0.08093374222517014,\n",
              "   -0.006895073689520359,\n",
              "   -0.06201917678117752,\n",
              "   0.004485982935875654,\n",
              "   0.05343255773186684,\n",
              "   0.012705609202384949,\n",
              "   0.023125607520341873,\n",
              "   -0.05334997922182083,\n",
              "   -0.03860864043235779,\n",
              "   0.050218597054481506,\n",
              "   0.00871662050485611,\n",
              "   0.03892895579338074,\n",
              "   -0.03655824810266495,\n",
              "   -0.024432053789496422,\n",
              "   0.0087732607498765,\n",
              "   -0.01481164526194334,\n",
              "   0.007284939754754305,\n",
              "   -0.03588959202170372,\n",
              "   -0.04433552175760269,\n",
              "   0.022046862170100212,\n",
              "   0.06772267818450928,\n",
              "   -0.06218945235013962,\n",
              "   -0.029298558831214905,\n",
              "   0.06545177847146988,\n",
              "   -0.05187319964170456,\n",
              "   -0.05360010638833046,\n",
              "   0.04196956008672714,\n",
              "   -0.024831609800457954,\n",
              "   0.05614451691508293,\n",
              "   0.027209360152482986,\n",
              "   -0.07184486091136932,\n",
              "   -0.033155616372823715,\n",
              "   -0.05499383062124252,\n",
              "   -0.030865464359521866,\n",
              "   -0.06050695851445198,\n",
              "   0.04243721067905426,\n",
              "   0.05015703663229942,\n",
              "   0.025375908240675926,\n",
              "   -0.07614549994468689,\n",
              "   0.1024087443947792,\n",
              "   -0.04500707611441612,\n",
              "   0.019563762471079826,\n",
              "   -0.019780408591032028,\n",
              "   -0.05458245053887367,\n",
              "   -0.044443946331739426,\n",
              "   0.18904994428157806,\n",
              "   -0.04297627508640289,\n",
              "   -0.025448329746723175,\n",
              "   0.019612284377217293,\n",
              "   0.012129939161241055,\n",
              "   0.02147691510617733,\n",
              "   -0.013748995028436184,\n",
              "   -0.01942027360200882,\n",
              "   -0.0004603841807693243,\n",
              "   0.04925813525915146,\n",
              "   0.09957903623580933,\n",
              "   -0.021422063931822777,\n",
              "   -0.03441017493605614,\n",
              "   -0.030533116310834885,\n",
              "   0.0046221851371228695,\n",
              "   -0.09814435988664627,\n",
              "   -0.03567234426736832,\n",
              "   0.040456607937812805,\n",
              "   -0.005798358470201492,\n",
              "   0.03429151326417923,\n",
              "   0.04189671203494072,\n",
              "   0.09444738179445267,\n",
              "   -0.14814633131027222,\n",
              "   0.035907406359910965,\n",
              "   -0.006976785603910685,\n",
              "   0.004730788059532642,\n",
              "   -4.0524624078461107e-33,\n",
              "   0.05540470778942108,\n",
              "   -0.06031270697712898,\n",
              "   0.07327916473150253,\n",
              "   -0.05741516873240471,\n",
              "   -0.03781789913773537,\n",
              "   0.04732919856905937,\n",
              "   0.045589759945869446,\n",
              "   0.015562179498374462,\n",
              "   0.13821840286254883,\n",
              "   -0.03565247356891632,\n",
              "   0.08778983354568481,\n",
              "   0.009041329845786095,\n",
              "   -0.08075270056724548,\n",
              "   -0.05995233356952667,\n",
              "   -0.04507969692349434,\n",
              "   0.015140818431973457,\n",
              "   -0.08792542666196823,\n",
              "   -0.019725829362869263,\n",
              "   0.0779210552573204,\n",
              "   -0.013288107700645924,\n",
              "   -0.09214141219854355,\n",
              "   0.04515878111124039,\n",
              "   0.0703761950135231,\n",
              "   0.061047252267599106,\n",
              "   0.035530462861061096,\n",
              "   0.0340220183134079,\n",
              "   0.006982918828725815,\n",
              "   0.04866449534893036,\n",
              "   0.008542663417756557,\n",
              "   0.11002076417207718,\n",
              "   0.04026566073298454,\n",
              "   0.04038539156317711,\n",
              "   -0.06426626443862915,\n",
              "   0.041115399450063705,\n",
              "   -0.09448165446519852,\n",
              "   0.02279062010347843,\n",
              "   -0.042835697531700134,\n",
              "   -0.061446208506822586,\n",
              "   -0.018020838499069214,\n",
              "   -0.027420444414019585,\n",
              "   -0.017771637067198753,\n",
              "   -0.030836760997772217,\n",
              "   0.020154260098934174,\n",
              "   0.03361569717526436,\n",
              "   0.10047740489244461,\n",
              "   0.07574422657489777,\n",
              "   -0.025025004521012306,\n",
              "   0.023485491052269936,\n",
              "   0.045325685292482376,\n",
              "   -0.032102711498737335,\n",
              "   -0.026815809309482574,\n",
              "   0.0421147495508194,\n",
              "   -0.03940364718437195,\n",
              "   0.022595679387450218,\n",
              "   0.04599745571613312,\n",
              "   0.013563109561800957,\n",
              "   0.0007849184330552816,\n",
              "   -0.01100969035178423,\n",
              "   -0.019750457257032394,\n",
              "   -0.021776484325528145,\n",
              "   -0.00043466617353260517,\n",
              "   -0.09079157561063766,\n",
              "   -0.0404403954744339,\n",
              "   0.019382940605282784,\n",
              "   -0.02989761345088482,\n",
              "   -0.03939957916736603,\n",
              "   -0.03194386884570122,\n",
              "   0.1089223176240921,\n",
              "   0.02216787450015545,\n",
              "   0.021214723587036133,\n",
              "   0.035785652697086334,\n",
              "   0.01704665832221508,\n",
              "   0.009080561809241772,\n",
              "   -0.0249541774392128,\n",
              "   0.023186543956398964,\n",
              "   -0.028134213760495186,\n",
              "   -0.07037391513586044,\n",
              "   0.04439183697104454,\n",
              "   0.017675327137112617,\n",
              "   0.02298734150826931,\n",
              "   -0.0582892969250679,\n",
              "   0.07511969655752182,\n",
              "   -0.056920506060123444,\n",
              "   0.006409809459000826,\n",
              "   0.06607867777347565,\n",
              "   -0.028581973165273666,\n",
              "   0.07381682097911835,\n",
              "   0.18400487303733826,\n",
              "   -0.04079348221421242,\n",
              "   -0.0535496361553669,\n",
              "   -0.03137490153312683,\n",
              "   -0.045275598764419556,\n",
              "   0.04696492478251457,\n",
              "   0.03644246608018875,\n",
              "   0.031497523188591,\n",
              "   -4.601189829145369e-08,\n",
              "   -0.0004212597559671849,\n",
              "   0.008509382605552673,\n",
              "   0.0017766725504770875,\n",
              "   -0.06791186332702637,\n",
              "   -0.032786570489406586,\n",
              "   0.06920827925205231,\n",
              "   -0.015780402347445488,\n",
              "   0.057828061282634735,\n",
              "   -0.01876983419060707,\n",
              "   0.035931382328271866,\n",
              "   -0.1044415682554245,\n",
              "   0.08310437202453613,\n",
              "   0.05247712880373001,\n",
              "   -0.03831293433904648,\n",
              "   0.006039157975465059,\n",
              "   0.021621428430080414,\n",
              "   -0.06968782842159271,\n",
              "   -0.033631473779678345,\n",
              "   -0.002845487091690302,\n",
              "   -0.13578875362873077,\n",
              "   -0.025002846494317055,\n",
              "   -0.004508357495069504,\n",
              "   -0.033152010291814804,\n",
              "   0.019559716805815697,\n",
              "   -0.0017526556039229035,\n",
              "   0.004727906547486782,\n",
              "   0.0747806578874588,\n",
              "   -0.050878074020147324,\n",
              "   -0.10712683945894241,\n",
              "   -0.01798587664961815,\n",
              "   0.035575006157159805,\n",
              "   -0.08594158291816711,\n",
              "   -0.012773985974490643,\n",
              "   0.054699454456567764,\n",
              "   0.020526612177491188,\n",
              "   0.05305594205856323,\n",
              "   -0.006202980875968933,\n",
              "   0.027404431253671646,\n",
              "   0.03200054541230202,\n",
              "   0.029986390843987465,\n",
              "   -0.04037041589617729,\n",
              "   -0.028069578111171722,\n",
              "   -0.04223375767469406,\n",
              "   0.04887767881155014,\n",
              "   -0.006035483442246914,\n",
              "   -0.040655381977558136,\n",
              "   0.032889265567064285,\n",
              "   0.02231702022254467,\n",
              "   -0.011789998970925808,\n",
              "   -0.0703866258263588,\n",
              "   0.014345229603350163,\n",
              "   -0.02637356147170067,\n",
              "   -0.057847414165735245,\n",
              "   -0.07805703580379486,\n",
              "   -0.07448457181453705,\n",
              "   0.014506088569760323,\n",
              "   -0.027959996834397316,\n",
              "   0.006781268399208784,\n",
              "   0.013687336817383766,\n",
              "   -0.038066133856773376,\n",
              "   -0.008480686694383621,\n",
              "   -0.04419092833995819,\n",
              "   0.03921966999769211,\n",
              "   -0.0019414507551118731]],\n",
              " 'metadatas': [{'corpusid': 220647965, 'chunk': 0},\n",
              "  {'corpusid': 220647965, 'chunk': 1},\n",
              "  {'corpusid': 220647965, 'chunk': 2},\n",
              "  {'corpusid': 220647965, 'chunk': 3},\n",
              "  {'corpusid': 220647965, 'chunk': 4},\n",
              "  {'corpusid': 220647965, 'chunk': 5},\n",
              "  {'corpusid': 20044380, 'chunk': 0},\n",
              "  {'corpusid': 20044380, 'chunk': 1},\n",
              "  {'corpusid': 20044380, 'chunk': 2},\n",
              "  {'corpusid': 20044380, 'chunk': 3}],\n",
              " 'documents': ['Novel 3,4-diarylpyrazole as prospective anti-cancerous agents\\n\\n\\nVivek Pandey \\nCentre for Genetic Disorder\\nBanaras Hindu University\\nVaranasiU.PIndia\\n\\nGarima Tripathi \\nDepartment of Chemistry\\nT. N. B. College\\nTMBU\\nBhagalpurBiharIndia\\n\\nDhruv Kumar \\nAmity Institute of Molecular Medicine & Stem Cell Research\\nAmity University Uttar Pradesh\\nNoidaIndia\\n\\nAbhijeet Kumar \\nDepartment of Chemistry\\nSchool of Physical Sciences\\nMahatma Gandhi Central University\\nBiharIndia A R\\n\\nT I C L E I N F O\\n\\n\\nPawan K Dubey \\nCentre for Genetic Disorder\\nBanaras Hindu University\\nVaranasiU.PIndia\\n\\nNovel 3,4-diarylpyrazole as prospective anti-cancerous agents\\n10.1016/j.heliyon.2020.e04397Research articlePyrazole Molecular docking Apoptosis Reactive oxygen species MDA-MB-231 Chemistry Organic chemistry Pharmaceutical chemistry Biological sciences Cell biology Bioinformatics\\nA B S T R A C TCancer is a leading cause of death globally. Despite therapeutic advancements the mortality rate of cancer is continuously increasing. Thus, it is important to identify and design potential therapeutic agents which can specifically bind with most common targets of cancer and inhibit tumor progression. The present work discloses the potential therapeutic application of the novel 3,4-diaryl 1H-pyrazoles as prospective anti-cancerous agent. The in silico molecular docking studies performed with 3,4-disubstituted pyrazoles as ligand with targets including DNA, BCL-2 and F1-ATP Synthase revealed strong binding affinity with DNA (-7.5 kcal/mol), BCL-2 (-8.1 kcal/ mol) and F1-ATP Synthase (-7.2 kcal/mol). Furthermore, the in silico finding was validated with the in vitro cytotoxicity assay with human breast cancer cell line (MDA-MB-231). MDA-MB-231 cells treated with 3,4-diarylpyrazole resulted in an increase in annexin-V positive cells, production of reactive oxygen species (ROS), dissipation of the mitochondrial membrane potential and activation of caspase-3. Taken together, this study demonstrate that a novel synthesized 3,4-diarylpyrazoles, showed strong binding affinity against DNA, antiproliferative activity and executed apoptosis through ROS-dependent caspase-3-mediated mitochondrial intrinsic apoptotic pathway against MDA-MB-231 cells. These findings increase our understanding of the molecular mechanism (s) by which 3,4-diarylpyrazoles can exert their anticancer activity and may contribute towards development of novel therapeutic agent against breast cancer.\\n\\nIntroduction\\n\\nIn general, the naturally occurring heterocyclic compounds along with their synthetic analogues are found to display a vast variety of biological activities [1,2,3,4]. N-heterocycles such as functionalized pyroles, pyrimidines, pyridine, pyrazoles etc. are important bio-active heterocyclic scaffolds which are prevalent in nature and known to exhibit wide range of biological activities. The immense therapeutic importance of the N-heterocycles are evidenced by their presence as key pharmacophore in the major proportion of the commercially available and the best selling drugs prescribed to treat a vast range of diseases. Pyrazole derivatives have exhibited potential activity against a wide range of diseases such as anti-inflammatory, anti-bacterial, anti-cancerous, anti-viral and so on [2].\\n\\nSome of the marketed drugs containing pyrazole scaffolds have been shown in Figure 1. For example, pyrazole containing drug, Apixaban is one of the most prescribed anti-coagulant [5]. Likewise, Sildenafil is a very popular drug with brand name Viagra used to treat erectile dysfunction in male [6]. Celecoxib and Ruxolitinib are anti-inflammatory and anti-cancerous drugs respectively containing pyrazole scaffolds [7,8]. Apart from these examples, various other pyrazole based drugs are known in the literature and available in the market also [4]. In addition to that several research groups are involved in the synthesis of new pyrazole derivatives and exploration of their biological activities. For example, Liang and co-workers synthesized 4-Amino-(1H)-pyrazole derivatives and explored their efficacy as JAKs inhibitor comparable and some were found to be more potent than Ruxolitinib which is an approved anti-cancerous drugs [9]. Several other such novel pyrazoles derivatives are known to exhibit anti-cancerous activities [10]. Encouraged by the several such reports available in the literature about the effectiveness of pyrazole derivatives as potential anti-cancerous agents, the work reported herein explores the potential anti-cancerous activity of novel 3,4-diaryl pyrazoles.\\n\\n\\nMaterials and methods\\n\\n\\nSynthesis of 3,4-diarylpyrazoles\\n\\nThe 3,4-diarylpyrazoles was synthesized as described in detail earlier [4], following the reaction sequence presented in Figure 2 [4]. In detail, 3-trifloxychromone and triarylbismuths were used for the one-pot synthesis of 3,4-diarylpyrazoles as per described protocol [11,12,13]. Silica gel (100-200 mesh) was used for the isolation and purification of the final products. For thin layer chromatography, GF-254 silica gel (Merck) was used. In order to characterize the structure of the 3,4-diarylpyrazoles, JEOL-400 MHz (JNM ECS-400) spectrometer was used to record 1 H and 13 C NMR. Structures of the 3,4-diarylpyrazole were further confirmed through HRMS measurement using WATERS-Q-T of Premier-HAB213 and WATERS GCT Premier-CAB155 instruments. IR of the final products was measured using Bruker Vector 22 FTIR spectrometer. Standard drying procedures were followed to obtain dry solvents to be used for reaction and purification purposes.\\n\\n\\nIn silico molecular docking studies\\n\\nIn order to investigate the potential anti-cancerous activity of these aryalted pyrazoles, molecular docking studies were performed. During the period of docking, the DNA & proteins were considered to be rigid and ligands were considered to be flexible. The configuration file for the grid parameters was generated using Auto Grid [14]. The docking study was also used to predict the type of amino acids in the active site of the proteins which interact with the diarylpyrazoles. The results less than 1.0 Å in positional root-mean-square deviation (RMSD) were considered ideal and clustered together for finding the favourable binding. In order to perform the molecular docking study, the ligands i.e. 3,4-diarylpyrazoles were prepared using Chem Sketch [15]. These were further converted into the PDB format from the file format of sdf with the help of Open Babel software [16]. In order to simplify the further analysis, ligands were converted to PDBQT format using the graphical user interface version of Autodock Tool 4 (ADT) [14]. Similarly, the atomic coordinates of the DNA (PDB ID-1D66), BCL-2 (PDB ID-1G5M) and F1-ATP Synthase (PDB ID-1BSN) were downloaded from the RCSB PDB database. Before docking, charges, solvation parameters and fragmental volumes to the DNA, BCL-2 and F1-ATP Synthase were assigned using the Autodock Tool 4 (ADT). Further, to simplify the analysis, the DNA and proteins were also saved to PDBQT format.\\n\\n\\nBiological studies\\n\\nEncouraged by the results from molecular docking study which revealed considerable binding affinity of pyrazole derivative with the targets; this compound was further taken forward to investigate its efficacy as potential anti-cancerous agent. Chemicals used for these studies were procured from Sigma Chemical Co. (St. Louis, MO) unless stated otherwise. The cell line (MDA-MB-231; human breast cancer) was procured from National Centre for Cell Science (NCCS), Pune, India.',\n",
              "  \"Preparation of culture medium and working concentration of synthesized 3,4-diarylpyrazoles\\n\\nThe cells were maintained in high glucose containing Dulbecco's Modified Eagle Medium (DMEM) fortified with 10% fetal bovine serum (FBS) and antibiotics (100 units/mL penicillin and 100 μg/mL streptomycin). To get the working concentrations (2.5, 5, 10, 20, 40, 80, 120, 180 and 250 μM), stock solution (10 mM in DMSO) of synthesized 3,4diarylpyrazoles were diluted in a freshly prepared working DMEM. The osmolarity (290 AE 5 mOsmol) and pH 7.2 AE 0.10 of the culture medium did not change after adding 3,4-diarylpyrazoles. The culture medium without 3,4-diarylpyrazoles was used as control.\\n\\n\\nMTT assay\\n\\nTo examine cytotoxicity of synthesized 3,4-diarylpyrazoles, the colorimetric MTT metabolic assay were performed. In this study, MDA-     \\n\\n\\nAnalysis of ROS level\\n\\nTo examine whether apoptosis is mediated by ROS production, after  \\n\\n\\nAssessment of mitochondrial membrane potential\\n\\nThe mitochondrial membrane potential was analyzed using flurochrome reporter JC-1 (1μM) dye (5,5,6,6-tetra-chloro-1,1,3,3-tetraethyl-benz-imidazolo-carbocyanine iodide) (BD Bioscience) as per manufacture's instruction. In brief, cells were treated with 3,4-diarylpyrazoles (50 μM) for 24 h and then cells were incubated with JC-1 dye (1μM) for 10 min. Photographs were taken using an inverted fluorescence microscope (EVOS FL, Life technologies) and imaged first under red filter and then under green filter. Red fluorescence indicated cells with intact mitochondria while green fluorescence indicated cells with depolarized mitochondria. Values are expressed as total red CTCF from three replicates.\\n\\n\\nImmunocytochemistry of Bcl 2 and caspase 3\\n\\nThe immunocytochemistry was performed to detect the protein expression of anti-apoptotic protein Bcl 2 and apoptotic protein caspase3 in treated MDA-MB-231 cells. Briefly, 3,4-diarylpyrazoles (50 μM) treated MDA-MB-231 cells were fixed using 4% para-formaldehyde for 10 min, after PBS wash, cells were incubated with sodium citrate buffer for 30 min. To avoid background staining, the cells incubated with 5% blocking serum for 20 min. After PBS wash, cells were incubated with primary monoclonal antibodies Bcl 2 and caspase3 (1:200; Santacruze Biotechnology, Inc) overnight at 4 C. After incubation, cells were washed with PBS, and then incubate with texas red anti-mouse (1:800) secondary antibodies for 2h at 37 C. The photographs were captured using inverted fluorescence microscope (EVOS FL, LifeTechnologies).\\n\\n\\nReal-time PCR (quantitative qPCR)\\n\\nThe quantitative real-time PCR was performed to examine mRNA expression level of apoptotic genes (Caspase-3) and anti-apoptotic gene (Bcl 2 ) in 3,4-diarylpyrazoles (50 μM) treated MDA-MB-231 cells. The primer sequence of genes which used in this study was shown in Table 1. In brief, after 24 h s of treatment, MDA-MB-231 cells were harvested, total RNA was isolated through trizol reagent and reverse transcribed according to manufacturer's protocol (Thermo scientific). qPCR was then performed with SYBR green dye using Applied Biosystems™Real-Time PCR Instruments (Life Technologies, Carlsbad, CA) according to manufacturer's instruction. The changes in mRNA expression level were calculated using the Relative Quantity (RQ) equation, RQ ¼ 2-ΔΔ CT; where delta cycle threshold (ΔCT) denoted a difference in CT values between the target gene and GAPDH, and ΔΔCT signified the differences in the ΔCT value of treated sets compared to control.\\n\\n\\nStatistical analysis\\n\\nData obtained in this study were analyzed using SPSS statistical software (SPSS, Inc., Chicago, IL) and expressed as means AE SEM from three experiments. The ImageJ software (version 1.44 from the National Institute of Health, Bethesda, USA) was used to analyze intensity of fluorescence. A probability of p < 0.05 was considered as statistically significant.\\n\\n\\nResults and discussion\\n\\n\\nPreparation of 3,4-diarylpyrazoles\\n\\nThe 3,4-diarylpyrazoles used to study their prospective anticancerous properties were earlier synthesized by Kumar et. al [4]. following the reaction protocol shown in Figure 2. The synthesis involves the in situ formation of isoflavones under Pd-catalyzed reaction cross-coupling method performed with 3-trifloxychromone and triarylbismuth followed by reaction with hydrazine hydrate to furnish phenol substituted C-arylated pyrazoles in one pot operation in good to excellent yield.\\n\\n\\nMolecular docking studies\\n\\nAmong all the ligands, 2-(4-(naphthalen-2-yl)-1H-pyrazol-3-yl) phenol (1.6) showed strong binding affinity with DNA (-7.5 kcal/mol), BCL-2 (-8.1 kcal/mol) and F1-ATP Synthase (-7.2 kcal/mol) ( Table 2). Whereas the known BCL-2 inhibitor, Navitoclax showed -8.9 kcal/mol binding affinity with BCL-2 ( Figure 3), which is close to the diarylpyrazoles. The top ten docking result generated using Autodock-Vina, which showed best binding conformations of the ligands with receptors at their active site. In past few years, the molecular docking studies have extensively been used as an imperative tool in the area of drug discovery as it allows to investigate the probable non-covalent interactions between the ligands which are generally small molecules and macromolecules such as proteins, enzymes, DNA etc. as targets [18]. Using the command line version of Autodock Vina [19] molecular docking study was also performed in present case to realize the effect of functionalized pyrazoles on reactive oxygen species (ROS) production through their binding interaction with DNA, BCL-2 and F1-ATP Synthase ( Table 2). The 3D view of DNA, BCL-2 and F1-ATP Synthase with ligand interactions of the best confirmation created by ADT are shown in molecular surface representation ( Figure 3). As clearly shown in Figure 3, the establishment of important interactions between ligand and the amino acid residues which are involved in the catalytic mechanism of ROS generation through DNA, BCL-2 and F1-ATP Synthase. The protein-ligand complex is stabilized mainly by electrostatic interactions, hydrogen bonding and ionic interactions. The top docked confirmation generated by each docking showed well established interaction with one or more amino acids at the catalytic domains of DNA and proteins. The top-ranked pose with lowest binding affinities is generally used as a standard selection in most of the docking programs.\\n\\n\\nBiological studies\",\n",
              "  'Biological studies\\n\\nWe investigated whether the synthesized 3,4-diarylpyrazoles exert anticancer activity against breast cancer, and if so, via what mechanism. Breast cancer is one of the leading cancers in women world-wide. From last once decade, several medical interventions like surgery, chemotherapy and radiotherapy has been used to treat breast cancer [20,21]. Over the past several years, many of the chemical compounds having N-heterocycles such as pyridine, triazole, pyrazole, pyrimidines scaffolds along with O-heterocycles such as benzofuran, coumarin, chromones etc have increased significant notice because of their safety, efficacy, low-cost, and therapeutical potential against cancers. As shown in Figure 4, the metabolic activity of MDA-MB-231 cells were inhibited by 3,4-diarylpyrazoles in a dose dependent manner. We found half maximal inhibitory concentration (IC 50 ) of 3,4-diarylpyrazoles was 50 μM after 24h. Further, to examine whether the cytotoxic effect of 3,4-diarylpyrazoles in MDA-MB-231 cells was associated with apoptosis induction, various assays has been conducted. As shown in Figure 5A and 5B, the 3,4-diarylpyrazoles induces morphological features of apoptosis as it increased population of PI positive cells (early apoptosis) in AO/PI dual-staining assay and also increased the number of Annexin-V/PI positive cells (late apoptotic cells) in FACS assay as compared to respective control. Our results suggest that the synthesized compound possess tremendous capacity to inhibit proliferation and induction of apoptosis in MDA-MB-231 cells. Further, to assess the status of mitochondrial membrane potential (ΔΨm) in response to 3,4-diarylpyrazoles treatment, the cells were stained with JC-1 fluorochrome. High fluorescence intensity (higher JC-1 green fluroscence) was observed in treated cells which is indicative of depolarized ΔΨm state of mitochondria ( Figure 6). This result suggests that 3,4-diarylpyrazoles induced mitochondrial membrane depolarization may associated with increase of oxidative stress or/reactive oxygen species (ROS).\\n\\nIt has already been demonstrated that many cytotoxic agents/compounds inhibits growth of tumor through ROS dependent activation of apoptosis [22]. It has also been shown that increased level of ROS in cancer cells in response to phytochemicals eventually leads to cell death [23]. In present study, to examine whether 3,4-diarylpyrazoles induced apoptosis is mediated by production of ROS. Intracellular ROS level was analyzed in 3,4-diarylpyrazoles treated and untreated cells using a fluorescent probe DCFH-DA. As shown in Figure 7, significantly increased DCFDA fluorescence intensity was observed at 50 μM of 3, 4-diarylpyrazoles as compared to control (P < 0.05). Furthermore, the co-treatment of N-acetyl cysteine (NAC) which protects cells against ROS-induced damage, almost suppressed the 3,4-diarylpyrazoles induced ROS generation in MDA-MB-231 cells (Figure 7). Lower fluorescence intensity was observed in combination with NAC þ3,4-diarylpyrazoles as It is reported that apoptosis is mediated by mitochondrial ROS production which may activate both the intrinsic and extrinsic pathways of apoptosis [24,25]. In this study, our synthesized compound 3,4-diarylpyrazoles significantly induce ROS generation in MDA-MB-231 cells, indicate that 3,4-diarylpyrazoles induced ROS-dependent intrinsic pathway of apoptosis. Although, ROS-independent mechanism of cell death was also reported in some other types of cells [26,27] which suggests that some other intrinsic pathways are also involved to cell death because ROS is not always necessary to induce apoptosis.\\n\\nTo delineate the molecular mechanism of apoptosis, the expression level of BCL-2 and caspase-3 in 3,4-diarylpyrazoles treated MDA-MB-231 cells were examined by immunocytochemistry and real-time PCR. In general, the ratio of the pro-apoptotic gene (Bax) and anti-apoptotic gene (BCL-2) are detrimental factors for programmed cell death, including apoptosis. It is reported that caspase-3 is activated death proteases that  finally leads to DNA fragmentation [28]. As shown in Figure 8A, total corrected cell fluorescence intensity level of apoptotic Caspase-3 protein were found higher in 3,4-diarylpyrazoles treated UL cells as compared to level of BCL-2 protein. It suggest that 3,4-diarylpyrazoles significantly suppressed the expression Bcl 2 whereas it induced caspase-3 expression in treated MDA-MB-231 cells as compared to control ( Figure 8A). Further, quantitative RT-PCR analysis showed higher (p < 0.05) mRNA expression of pro-apoptotic gene caspase-3 whereas down regulation of Bcl 2 mRNA expression in 3,4-diarylpyrazoles treated MDA-MB-231 cells as compared to control ( Figure 8B). These results further suggest that 3, 4-diarylpyrazoles induces apoptosis in MDA-MB-231 cells via ROS-dependent caspase-3-mediated mitochondrial intrinsic apoptotic pathway (Figure 9).\\n\\n\\nConclusion\\n\\nThis study first time demonstrated that a novel synthesized 3,4-diarylpyrazoles, exerted strong binding affinity against DNA, potent antiproliferative activity and executed apoptosis through ROS-dependent caspase-3-mediated mitochondrial intrinsic apoptotic pathway against MDA-MB-231 cells. Our results may help to increase understanding of the molecular mechanisms by which 3,4-diarylpyrazoles inducing cell death in MDA-MB-231 cells and may contribute towards development of novel therapeutic agent against breast cancer.\\n\\n\\nDeclarations\\n\\nAuthor contribution statement P. Dubey and A. Kumar: Conceived and designed the experiments; Analyzed and interpreted the data; Wrote the paper.\\n\\nV. Pandey and D. Kumar: Performed the experiments; Analyzed and interpreted the data; Contributed reagents, materials, analysis tools or data.\\n\\nG. Tripathi: Performed the experiments; Analyzed and interpreted the data; Contributed reagents, materials, analysis tools or data; Wrote the paper.\\n\\n\\nFunding statement\\n\\nThis study was financially supported by Science and Engineering Research Board, Ministry of Science and Technology, Government of India (Grant No. SERB/YS/000819).',\n",
              "  \"Figure 1 .\\n1Examples of pyrazole scaffold in pre-existing drugs. 231 cell line was seeded (10,000 cells/well) in 24-well tissue culture plates for 24 h. After that cells were incubated with different concentrations of synthesized 3,4-diarylpyrazoles(2.5, 5, 10, 20, 40, 80, 120,    180 and 250 μM) for 24 h in a CO 2 incubator[17]. The working concentration of MTT (5 mg/mL) was prepared in PBS, and 10μL was added into each well; after 4 h incubation, 100μL of DMSO was used to dissolve formazan crystals. The absorbance intensity was taken using a microplate reader (BioRAD 680, USA) at 590 nm. The effective concentration (EC 50 value) were calculated with help of GraphPad prism 4.0 (GraphPad software, USA) using non-linear regression analysis. The experiments were repeated at least three times. The cell survival percentage was calculated using the following formula:%Control ¼ ½Mean O: D: of the Drug treated well =Mean O: D: of the control well Â 100 2.6. Acridine orange/propadium iodide (AO/PI) dual staining assay The AO/PI staining was used to examine morphological changes in MDA-MB-231 cells after 3,4-diarylpyrazoles treatment. In brief, MDA-MB-231cells (1Â10 6 ) were treated with or without an optimum effective dose of 3,4-diarylpyrazoles (50 μM) for 24 h in the CO 2 incubator. After washing in PBS, cells were further incubated with 100μL of AO/PI solution (1 part of 100 μg/mL of AO in PBS; 1 part of 100 μg/mL of PI in PBS). After 30 min incubation, cells were washed with PBS and morphology examined under a fluorescence microscope (EVOS FL, Life technologies). Viable cells exhibit intact membrane and produces green\\n\\nFigure 2 .\\n2Synthesis of 3,4-diaryl pyrazoles under one-pot operation involving cross-coupling.\\n\\n2. 7 .\\n7Apoptosis detection assay by annexin-V/PI To validate our AO/PI data, 3,4-diarylpyrazoles (50 μM) treated MDA-MB-231 cells were stained with Annexin V/PI and FACS analysis were carried out. In detail, after treatment, the MDA-MB-231 cells were incubated with 2μl of annexin V-FITC and 5μl of PI (100 μg/ml) in 200μl of binding buffer at room temperature for 15 min in the dark. After washing in 1X PBS, stained cells were examined by flow cytometry using BectoDickinson fluorescence-activated cell sorter (FACS) IV Calibur (Becton Dickinson, San Jose, CA, USA). The flow cytometry data were analyzed and plotted for annexin V-FITC using a Cell Quest software (Becton Dickinson).\\n\\n\\n24 h of 3,4-diarylpyrazoles (50 μM) treatment, the MDA-MB-231 cells were washed and incubated with 10μM dichlorofluorescein diacetate (DCFH-DA, Beyotime, Nantong, China) at 37 C for 30 min. Finally cells were washed 3 times with PBS and examine under fluorescence microscope (EVOS FL, Life technologies) as per manufacturer's instructions. The fluorescence of DCF was detected at an excitation wavelength of 488 nm and an emission wavelength of 525 nm using inverted fluorescence microscope (EVOS FL, Life technologies).\\n\\nFigure 3 .\\n3The 3D views of interaction of pyrazole (1.6) with DNA (A & B),BCL-2 (C & D), F1-ATP Synthase (E & F) and Navitoclax (Known BCL-2 inhibitor) with BCL-2 (G).\\n\\nFigure 4 .\\n4Cell viability of MDA-MB-231cells treated with 3,4-diarylpyrazoles. The bar diagram showing cell viability of MDA-MB-231cells after treatment with 0.0, 2.5, 5, 10, 20, 40, 80, 120, 180 and 250 μM of 3,4-diarylpyrazoles for 24 h. Data are normalized with control group (medium only) and presented as mean AE SEM.\\n\\nFigure 5 .\\n5AO/PI staining and flowcytometric analysis of MDA-MB-231cells treated with/without 3,4-diarylpyrazoles (A) The untreated cells are showing green fluorescence (AO)/representing live cells while treated cells showing red fluorescence (PI) representing dead cells (B) flow cytometry data are showing that treatment of 3,4-diarylpyrazoles (50 μM) induces early apoptotic as well as late apoptotic in MDA-MB-231 cells. (Bar¼ 200 μm). diarylpyrazoles alone suggest that 3,4-diarylpyrazoles induced ROS generation, which finally induced apoptosis in MDA-MB-231 cells. However, mechanism by which anti-cancer drugs may cause elevated ROS levels varies in response to treatment, such as disruption of the mitochondrial electron transport chain, changes in mitochondrial membrane potential and dysregulation of antioxidant enzymes.\\n\\nFigure 6 .Figure 7 .\\n67Representative photographs showing mitochondrial membrane potential in 3,4 diarylpyrazoles treatment in MDA-MB-231cells. Bar diagram shows CTCF analysis of fluorescence intensity of control and treated cells. Three independent experiments were conducted to confirm the results. (Bar ¼ 100 μm). Representative photographs showing intracellular ROS level in 3,4-diarylpyrazoles treatment in MDA-MB-231cells. Significant increased ROS level was noticed in treated MDA-MB-231 cells. The bar diagram shows CTCF analysis of fluorescence intensity of control and treated cells. Three independent experiments were conducted to confirm the results. (Bar ¼ 200 μm).\\n\\nFigure 8 .\\n8Effect MDA-MB-231 cells. The graph represents means AE SEMs from at least 3 separate experiments. The asterisk (**) indicates a significant P value the significance level in comparison to untreated group; **P < 0.05.\\n\\nFigure 9 .\\n9The proposed model diagram represented the molecular targets of 3,4-diarylpyrazoles on MDA-MB-231 cells. The 3,4-diarylpyrazoles induced ROS by disrupting mitochondrial membrane potential which release caspase-3 and thereby apoptosis.\\n\\nTable 2 .\\n2Results of molecular docking studies between 3,4-diarylpyrazoles and onco-proteins.S.No. \\nDNA \\nERBB2/HER2 \\nBCL2* \\nF1-ATPSYNTHASE \\n\\n1 \\nA r ¼ Ph \\n1.1 \\n\\n-6.6 \\n-8.2 \\n-6.7 \\n-6.4 \\n\\n2 \\nA r ¼ p-tolyl \\n1.2 \\n\\n-6.8 \\n-8.5 \\n-7.1 \\n-6.6 \\n\\n3 \\nA r ¼ 4-methoxyphenyl \\n1.3 \\n\\n-6.6 \\n-7.8 \\n-6.7 \\n-6.1 \\n\\n4 \\nA r ¼ 4-chlorophenyl \\n1.4 \\n\\n-6.2 \\n-8.4 \\n-6.8 \\n-6.2 \\n\\n5 \\nA r ¼ 3-methoxyphenyl \\n1.5 \\n\\n-6.4 \\n-7.5 \\n-6.9 \\n-6.4 \\n\\n6 \\nA r ¼ 2-naphthyl \\n1.6 \\n\\n-7.5 \\n-9.8 \\n-8.1 \\n-7.2 \\n\\n7 \\nA r ¼ 4-fluorophenyl \\n1.7 \\n\\n-6.7 \\n-8.3 \\n-6.9 \\n-6.5 \\n\\nTable 1 .\\n1List of used primers sequence.fluorescence while apoptotic cells show bright orange color which indicates dead cells. The photographs were captured using inverted fluorescence microscope (EVOS FL, Life technologies at 20x magnification).Gene \\nForward Primer\",\n",
              "  'Competing interest statementThe authors declare no conflict of interest.Additional informationSuport of Prof. M.L.N. Rao and research lab facilities at IIT-Kanpur is dully acknowledged.\\nModern advances in heterocyclic chemistry in drug discovery. P T Alexandria, R P Robinson, Y M Fobian, D C Blakemore, L H Jones, O Fadeyi, Org. Biomol. Chem. 14P.T. Alexandria, R.P. Robinson, Y.M. Fobian, D.C. Blakemore, L.H. Jones, O. Fadeyi, Modern advances in heterocyclic chemistry in drug discovery, Org. Biomol. Chem. 14 (2016) 6611-6637.\\n\\nRecent advances in bioactive pyrazoles. S G Kucukguzel, S Senkardes, Eur. J. Med. Chem. 9S.G. Kucukguzel, S. Senkardes, Recent advances in bioactive pyrazoles, Eur. J. Med. Chem. 9 (2015) 77-86.\\n\\nA concise synthesis of pyrazole analogues of combretastatin A1 as potent anti-tubulin agents. R Zaninetti, S V Cortese, S Aprile, A Massarotti, P L Canonico, G Sorba, G Grosa, A A Genazzani, T Pirali, ChemMedChem. 8R. Zaninetti, S.V. Cortese, S. Aprile, A. Massarotti, P.L. Canonico, G. Sorba, G. Grosa, A.A. Genazzani, T. Pirali, A concise synthesis of pyrazole analogues of combretastatin A1 as potent anti-tubulin agents, ChemMedChem 8 (2013) 633-643.\\n\\nPot-economic synthesis of diarylpyrazoles and pyrimidines involving Pd-catalyzed cross-coupling of 3-trifloxychromone and triarylbismuth. A Kumar, M L N Rao, J. Chem. Sci. 130165A. Kumar, M.L.N. Rao, Pot-economic synthesis of diarylpyrazoles and pyrimidines involving Pd-catalyzed cross-coupling of 3-trifloxychromone and triarylbismuth, J. Chem. Sci. 130 (2018) 165.\\n\\nApixaban: a new player in the anticoagulant class. R Agrawal, P Jain, S N Dikshit, Curr. Drug Targets. 13R. Agrawal, P. Jain, S.N. Dikshit, Apixaban: a new player in the anticoagulant class, Curr. Drug Targets 13 (2012) 863-875.\\n\\nSildenafil in the treatment of erectile dysfunction: an overview of the clinical evidence. K Hatzimouratidis, Clin. Interv. Aging. 4K. Hatzimouratidis, Sildenafil in the treatment of erectile dysfunction: an overview of the clinical evidence, Clin. Interv. Aging 4 (2006) 403-414.\\n\\nCelecoxib: considerations regarding its potential disease-modifying properties in osteoarthritis. Manon C Zweers, N Tineke, Jo€ De Boer, W J El Van Roon, Floris Pjg Bijlsma, Simon C Lafeber, Mastbergen, Arthritis Res. Ther. 13239Manon C. Zweers, Tineke N. de Boer, Jo€ el van Roon, W.J. Bijlsma, Floris PJG. Lafeber, Simon C. Mastbergen, Celecoxib: considerations regarding its potential disease-modifying properties in osteoarthritis, Arthritis Res. Ther. 13 (2011) 239.\\n\\nRuxolitinib: the first FDA approved therapy for the treatment of myelofibrosis. J Mascarenhas, R Hoffman, Clin. Canc. Res. 18J. Mascarenhas, R. Hoffman, Ruxolitinib: the first FDA approved therapy for the treatment of myelofibrosis, Clin. Canc. Res. 18 (2012) 3008-3014.\\n\\nDesign, synthesis, and antitumor evaluation of 4-amino-(1H)-pyrazole derivatives as JAKs inhibitors. X Liang, J Zang, M Zhu, Q Gao, B Wang, W Xu, Y Zhang, ACS Med. Chem. Lett. 7X. Liang, J. Zang, M. Zhu, Q. Gao, B. Wang, W. Xu, Y. Zhang, Design, synthesis, and antitumor evaluation of 4-amino-(1H)-pyrazole derivatives as JAKs inhibitors, ACS Med. Chem. Lett. 7 (2016) 950-955.\\n\\nnovel anticancer fused pyrazole derivatives as EGFR and VEGFR-2 dual TK inhibitors. M Nashwa, G Saleh, Marwa, M El-Gazzar, Hala, Rana A Aly, Othman, Front. Chem. 7917Nashwa M. Saleh, G. Marwa, El-Gazzar, M. Hala, Aly and Rana A. Othman, novel anticancer fused pyrazole derivatives as EGFR and VEGFR-2 dual TK inhibitors, Front. Chem. 7 (2020) 917.\\n\\nStructure-activity relationships of targeted Ru II (η 6 -pcymene) anticancer complexes with flavonol-derived ligands. A Kurzwernhart, W Kandioller, S B€ Achler, C Bartel, S Martic, M Buczkowska, G Mühlgassner, A M Jakupec, H.-B Kraatz, P J Bednarski, V B Arion, D Marko, B K Keppler, C G Hartinger, J. Med. Chem. 55A. Kurzwernhart, W. Kandioller, S. B€ achler, C. Bartel, S. Martic, M. Buczkowska, G. Mühlgassner, A.M. Jakupec, H.-B. Kraatz, P.J. Bednarski, V.B. Arion, D. Marko, B.K. Keppler, C.G. Hartinger, Structure-activity relationships of targeted Ru II (η 6 -p- cymene) anticancer complexes with flavonol-derived ligands, J. Med. Chem. 55 (2012) 10512-10522.\\n\\nEfficient synthesis of ratiometric fluorescent nucleosides featuring 3-hydroxychromone nucleobases. M Spadafora, V Y Postupalenko, V V Shvadchak, A S Klymchenko, Y Mely, A Burger, R Benhida, Tetrahedron. 65M. Spadafora, V.Y. Postupalenko, V.V. Shvadchak, A.S. Klymchenko, Y. Mely, A. Burger, R. Benhida, Efficient synthesis of ratiometric fluorescent nucleosides featuring 3-hydroxychromone nucleobases, Tetrahedron 65 (2009) 7809-7816.\\n\\nPd-catalyzed cross-coupling study of bi-functional 3-bromo-4-trifloxycoumarins with triarylbismuth reagents. M L N Rao, A Kumar, Tetrahedron. 71M.L.N. Rao, A. Kumar, Pd-catalyzed cross-coupling study of bi-functional 3-bromo- 4-trifloxycoumarins with triarylbismuth reagents, Tetrahedron 71 (2015) 5137-5147.',\n",
              "  \"Software news and updates AutoDock4 and AutoDockTools4: automated docking with selective receptor flexibility. G M Morris, H Ruth, W Lindstrom, M F Sanner, R K Belew, D S Goodsell, A J Olson, J. Comput. Chem. 30G.M. Morris, H. Ruth, W. Lindstrom, M.F. Sanner, R.K. Belew, D.S. Goodsell, A.J. Olson, Software news and updates AutoDock4 and AutoDockTools4: automated docking with selective receptor flexibility, J. Comput. Chem. 30 (2009) 2785-2791.\\n\\n. A D Hunter, ChemSketch 1.0 (freewareA.D. Hunter, ACD/ChemSketch 1.0 (freeware);\\n\\nACD/ChemSketch 2.0 and its tautomers, dictionary, and 3D plug-ins. ACD/ChemSketch 2.0 and its tautomers, dictionary, and 3D plug-ins;\\n\\nACD/HNMR 2.0; ACD/CNMR 2.0. J. Chem. Educ. ACD/HNMR 2.0; ACD/CNMR 2.0, J. Chem. Educ. (1997).\\n\\nOpen Babel, An Open chemical toolbox. N M O&apos;boyle, M Banck, C A James, C Morley, T Vandermeersch, G R Hutchison, J. Cheminf. N.M. O'Boyle, M. Banck, C.A. James, C. Morley, T. Vandermeersch, G.R. Hutchison, Open Babel, An Open chemical toolbox, J. Cheminf. (2011).\\n\\nEndoplasmic reticulum stress as a novel cellular response to di (2-ethylhexyl) phthalate exposure. A Peropadre, F F Paloma, J M Perez Martin, O Herrero, M J Hazen, Toxicol. Vitro. 30A. Peropadre, F.F. Paloma, J.M. Perez Martin, O. Herrero, M.J. Hazen, Endoplasmic reticulum stress as a novel cellular response to di (2-ethylhexyl) phthalate exposure, Toxicol. Vitro 30 (2015) 281-287.\\n\\nStructure-based virtual screening. Q Li, S Shah, Methods in Molecular Biology. Q. Li, S. Shah, Structure-based virtual screening, in: Methods in Molecular Biology, 2017.\\n\\nSoftware news and update AutoDock Vina: improving the speed and accuracy of docking with a new scoring function, efficient optimization, and multithreading. O Trott, A J Olson, J. Comput. Chem. 31O. Trott, A.J. Olson, Software news and update AutoDock Vina: improving the speed and accuracy of docking with a new scoring function, efficient optimization, and multithreading, J. Comput. Chem. 31 (2010) 455-461.\\n\\nTiming of radiotherapy in breast cancer conserving treatment, Canc. M G Ruo, Redda1 R Verna, A Guarneri, G L Sannazzari, Treat Rev. 28M.G. Ruo, Redda1 R. Verna, A. Guarneri, G.L. Sannazzari, Timing of radiotherapy in breast cancer conserving treatment, Canc. Treat Rev. 28 (2002) 5-10.\\n\\nMicroRNA mediated drug resistance in breast cancer. K R Kutanzi, O V Yurchenko, F A Beland, C Vasyl&apos;f, I P Pogribny, Clin. Epigenet. 2171K.R. Kutanzi, O.V. Yurchenko, F.A. Beland, C. Vasyl'F, I.P. Pogribny, MicroRNA mediated drug resistance in breast cancer, Clin. Epigenet. 2 (2011) 171.\\n\\nCellular thiols and reactive oxygen species in druginduced apoptosis. W Davis, Z E Ronai, K D Tew, J. Pharmacol. Exp. Therapeut. 2961W. Davis, Z.e. Ronai, K.D. Tew, Cellular thiols and reactive oxygen species in drug- induced apoptosis, J. Pharmacol. Exp. Therapeut. 296 (1) (2001) 1-6.\\n\\n/ Fas) gene in response to DNA damage by anticancer drugs. M Müller, S Wilder, D Bannasch, D Israeli, K Lehlbach, M Li-Weber, S L Friedman, P R Galle, W Stremmel, M Oren, J. Exp. Med. 18811p53 activates the CD95 (APO-1M. Müller, S. Wilder, D. Bannasch, D. Israeli, K. Lehlbach, M. Li-Weber, S.L. Friedman, P.R. Galle, W. Stremmel, M. Oren, p53 activates the CD95 (APO-1/ Fas) gene in response to DNA damage by anticancer drugs, J. Exp. Med. 188 (11) (1998) 2033-2045.\\n\\nActivation of apoptosis signalling pathways by reactive oxygen species. M Redza-Dutordoir, D A Averill-Bates, BBA-Mol. Cell Res. 1863M. Redza-Dutordoir, D.A. Averill-Bates, Activation of apoptosis signalling pathways by reactive oxygen species, BBA-Mol. Cell Res. 1863 (2016) 2977-2992.\\n\\nReactive oxygen species and cancer paradox: to promote or to suppress? Free Radic. S Galadari, A Rahman, S Pallichankandy, F Thayyullathil, Biol. Med. 104S. Galadari, A. Rahman, S. Pallichankandy, F. Thayyullathil, Reactive oxygen species and cancer paradox: to promote or to suppress? Free Radic. Biol. Med. 104 (2017) 144-164.\\n\\nInduction of ROS independent DNA damage by curcumin leads to G2/M cell cycle arrest and apoptosis in human papillary thyroid carcinoma BCPAP cells. L Zhang, X Cheng, Y Gao, J Bao, H Guan, R Lu, Y Sun, Food Funct. 7L. Zhang, X. Cheng, Y. Gao, J. Bao, H. Guan, R. Lu, Y. Sun, Induction of ROS independent DNA damage by curcumin leads to G2/M cell cycle arrest and apoptosis in human papillary thyroid carcinoma BCPAP cells, Food Funct. 7 (2016) 315-325.\\n\\nEffect of ascorbic acid and X-irradiation on HL-60 human leukemia cells: the kinetics of reactive oxygen species. S Terashima, Y Hosokawa, H Yoshino, M Yamaguchi, T Nakamura, Oncol. Rep. 30S. Terashima, Y. Hosokawa, H. Yoshino, M. Yamaguchi, T. Nakamura, Effect of ascorbic acid and X-irradiation on HL-60 human leukemia cells: the kinetics of reactive oxygen species, Oncol. Rep. 30 (2013) 2653-2658.\\n\\nEmerging roles of caspase-3 in apoptosis. A G Porter, R U Janicke, Cell Death Differ. 62A.G. Porter, R.U. Janicke, Emerging roles of caspase-3 in apoptosis, Cell Death Differ. 6 (2) (1999) 99-104.\",\n",
              "  'Human SOD1 before Harboring the Catalytic Metal SOLUTION STRUCTURE OF COPPER-DEPLETED, DISULFIDE-REDUCED FORM * □ S\\nPublished, JBC Papers in Press, November 14, 2005,\\n\\nLucia Banci \\nMagnetic Resonance Center\\nUniversity of Florence\\nVia Luigi Sacconi 650019Sesto Fiorentino, Italy\\n\\nDepartment of Chemistry\\nUniversity of Florence\\nVia della Lastruccia 350019Sesto Fiorentino, Italy\\n\\nIvano Bertini \\nMagnetic Resonance Center\\nUniversity of Florence\\nVia Luigi Sacconi 650019Sesto Fiorentino, Italy\\n\\nFrancesca §1 \\nDepartment of Chemistry\\nUniversity of Florence\\nVia della Lastruccia 350019Sesto Fiorentino, Italy\\n\\nCantini \\nMagnetic Resonance Center\\nUniversity of Florence\\nVia Luigi Sacconi 650019Sesto Fiorentino, Italy\\n\\nNicola D&apos;amelio \\nMagnetic Resonance Center\\nUniversity of Florence\\nVia Luigi Sacconi 650019Sesto Fiorentino, Italy\\n\\nDepartment of Chemistry\\nData Bank, Research Collaboratory for Structural Bioinformatics\\nUniversity of Siena\\nVia Aldo Moro53100SienaItaly\\n\\nRutgers University\\nNew BrunswickNJ\\n\\nElena Gaggelli \\nDepartment of Chemistry\\nData Bank, Research Collaboratory for Structural Bioinformatics\\nUniversity of Siena\\nVia Aldo Moro53100SienaItaly\\n\\nRutgers University\\nNew BrunswickNJ\\n\\nHuman SOD1 before Harboring the Catalytic Metal SOLUTION STRUCTURE OF COPPER-DEPLETED, DISULFIDE-REDUCED FORM * □ S\\nPublished, JBC Papers in Press, November 14, 2005,10.1074/jbc.M506497200Received for publication, June 15, 2005, and in revised form, October 3, 2005\\nSOD1 has to undergo several post-translational modifications before reaching its mature form. The protein requires insertion of zinc and copper atoms, followed by the formation of a conserved S-S bond between Cys-57 and Cys-146 (human numbering), which makes the protein fully active. In this report an NMR structural investigation of the reduced SH-SH form of thermostable E,Zn-as-SOD1 (E is empty; as is C6A, C111S) is reported, characterizing the protein just before the last step leading to the mature form. The structure is compared with that of the oxidized S-S form as well as with that of the yeast SOD1 complexed with its copper chaperone, CCS. Local conformational rearrangements upon disulfide bridge reduction are localized in the region near Cys-57 that is completely exposed to the solvent in the present structure, at variance with the oxidized forms. There is a local disorder around Cys-57 that may serve for protein-protein recognition and may possibly be involved in intermolecular S-S bonds in familial amyotrophic lateral sclerosis-related SOD1 mutants. The structure allows us to further discuss the copper loading mechanism in SOD1.\\n\\nCu,Zn-SOD 2 is a very efficient enzyme that catalyzes the dismutation of superoxide to oxygen and hydrogen peroxide (1). It is a dimer in all eukaryotes, whereas in prokaryotes it exists as either a dimer or a monomer. The mature, active form of SOD1 contains, in each subunit, a copper ion essential for catalysis and a zinc ion that has primarily a structural role.\\n\\nIn the active site, a narrow channel is present that is large enough to admit only superoxide, water, and small anions and ligands such as imidazole and peroxynitrite. In the lining of the channel there is a positively charged side chain from an arginine residue (Arg-143, human numbering). The cationic residue generates an electrostatic gradient proposed to be responsible for steering the superoxide anion toward the active site. Site-directed mutagenesis of residue 143 to a neutral or ani-onic residue produced proteins with dramatically reduced or abolished activity (2).\\n\\nA peculiar feature of the mature form of SOD1 is the presence of a kinetically stable disulfide bond. Once formed, the disulfide bond is maintained in the reducing cytoplasmic environment where the majority of the SOD1 inside the cell is located. The presence of intramolecular disulfide bonds is common in secreted proteins, where their primary purpose is for protein stabilization. However, disulfide bonds are rare in intracellular proteins because of the highly reducing environment and low concentration of dioxygen in the cytosol (3,4). It has been previously shown that intramolecular disulfide bonds in intracellular proteins (like SOD1) can play more than just a structural role and have functional significance (3). The disulfide bond between Cys-57 and Cys-146 is fully conserved in all SOD1 structures. This bond links loop IV, which contains Cys-57, with strand ␤8, containing Cys-146. The linkage of the secondary structure elements contributes to the stabilization of the SOD1 fold.\\n\\nThe active form of the protein is obtained when one zinc and one copper ion per subunit are bound. Although nothing is known on how the protein acquires the zinc ion, a description of the insertion of copper is available. The process is aided by a copper metallochaperone called CCS (copper chaperone for superoxide dismutase) (5,6). Yeast mutants lacking CCS express a form of SOD1 protein that is essentially apo for copper but contains zinc (5,7). However, a CCS-independent activation of mammalian SOD1 is possible (8). A mechanism of the copper transfer from CCS to SOD1 in yeast has been proposed based on the x-ray structure of the heterodimeric CCS-SOD1 complex (9). In this picture, SOD1 first loads zinc ions from an unknown source, and subsequently copper transfer takes place through the formation of an intermolecular S-S bond, involving Cys-57 of SOD1 and a Cys of CCS. The disulfidereduced, copper-depleted form of SOD1 therefore represents the state of the protein before its complex formation with copper(I) CCS (10).\\n\\nAdditionally, this protein state may play a relevant role in pathological conditions. The presence of exposed cysteine residues may modulate its ability to interact with other macromolecules and to give rise to large molecular aggregates (11). Aggregates containing SOD1 have been found in some fALS patients and in some fALS SOD1 transgenic mice and rats (12,13). The fALS pathology has been related to mutations in the protein Cu,Zn-SOD1, some of which have a tendency to form fibrils (14). An increasing number of pathological states involve protein aggregates, which represent the final state of biomolecules with altered folding properties and/or structural conformations (15). To avoid interference effects arising from Cys-6 and Cys-111 (i.e. the two other Cys present in SOD1 that are in the reduced state), we studied the thermostable mutant of human SOD1 (C6A, C111S), which is structurally and functionally equivalent to the wild type protein (16 -18) and is commonly used in the study of fALS mutants to prevent intermolecular aggregation (4, 19 -21).\\n\\nCurrently, there are a huge number of structures of SOD1 (Ͼ50, ϳ10 of which of human SOD1) deposited in the Protein Data Bank in various states (22)(23)(24)(25), including several mutants, but all with the disulfide bond intact. Therefore, the structural characterization of E,Zn-SOD1 with the disulfide bond reduced that is reported here (hereafter E,Zn-hasSOD1 SH-SH ) is particularly meaningful as it is the first structure of SOD1 with the disulfide bond broken. It represents the step before the protein acquires copper but also a state that is emerging to be relevant in the fALS-related SOD1 mutants (11,26). Therefore, it provides useful new data on the features of SOD1.\\n\\n\\nMATERIALS AND METHODS',\n",
              "  'MATERIALS AND METHODS\\n\\nSample Preparation-Human asSOD1 (hasSOD1) was expressed in the Escherichia coli TOPP1 (Stratagene) or BL21(DE3) strain. The 15 N and 15 N, 13 C, 2 H-labeled protein samples were obtained as previously reported (25). The protein was isolated and purified according to previously published protocols (16). The triple-labeled dimeric hasSOD1 contained ϳ70% 2 H. Metal ions were removed according to published protocols (27), and zinc reconstitution was obtained as described in Ref. 17. The reduction of disulfide bridge was accomplished by addition of dithiothreitol. Fully reduced and empty copper protein hasSOD1 (E,Zn-hasSOD1 SH-SH ) was prepared under a nitrogen atmosphere in an anaerobic chamber. The thiol-disulfide status of purified hasSOD1 was determined by chemical modification with the thiol-specific reagent 4-acetamide-4Ј-maleimidylstilbene-2,2Ј-disulfonic acid (Molecular Probes, Inc.) (10).\\n\\nThe NMR samples of the E,Zn-hasSOD1 SH-SH were in 20 mM sodium phosphate buffer, pH 5, 90% H 2 O/10% D 2 O. The final protein concentration ranges between 1 and 1.5 mM in the presence of 20 mM dithiothreitol. ϳ0.6 ml of sample was loaded into high quality NMR tubes that were capped with latex serum caps in the Vacuum Atmospheres chamber.\\n\\nNMR Measurements and Structure Calculation-The NMR spectra were acquired on Avance 900, 800, 600, and 500 Bruker spectrometers. All of the triple resonance probes used were equipped with pulsed field gradients along the z-axis. The 800-and 500-MHz spectrometers were equipped with a triple resonance cryoprobe.\\n\\nThe NMR experiments, used for the backbone and the aliphatic side chain resonances assignment and for obtaining structural restraints, recorded on 2 H/ 13 C/ 15 N/, 13 C/ 15 N-enriched and unlabeled E,Zn-hasSOD1 SH-SH samples, are summarized in supplemental Table S1. The 1 H, 13 C, and 15 N resonance assignments of E,Zn-hasSOD1 SH-SH are reported in supplemental Table S2. In total, 88% of the 1 H, 98% of the 15 N, and 80% of the 13 C have been assigned. The combined variation of the 15 N and 1 H amide chemical shifts of E,Zn-hasSOD1 SH-SH , Cu,Zn-hasSOD1 SS , and E,Zn-E133QM2-asSOD1 SS are shown in supplemental Fig. S1. 1 H-1 H nuclear Overhauser enhancement spectroscopy (NOESY) experiment (80-ms mixing time) (28) was carried out to identify connectivities involving histidines of the binding sites at 298 K. All the spectra were collected at 298 K, processed using the standard Bruker software (XWINNMR), and analyzed through the CARA program (29).\\n\\nAn automated CANDID approach combined with DYANA torsion angle dynamics algorithm (30) was used to assign the ambiguous NOE cross-peaks and to have a preliminary protein structure. Structure calculations were then performed through iterative cycles of DYANA (31) followed by restrained energy minimization with AMBER 5.0 (32) applied to each member of the final DYANA family. The assessment of the structures was evaluated using the program PROCHECK-NMR (33). The root mean square deviation values for the backbone and heavy atoms of one subunit are shown in supplemental Fig. S2. 15 N R 1 , R 2 , and steady-state heteronuclear NOEs, which can provide information on internal mobility as well as on the overall protein tumbling rate, were measured with pulse sequences as described by Farrow et al. (34). R 2 were measured using a refocusing time of 450 ms. In all experiments the water signal was suppressed with the \"water flip back\" scheme (35). Average R 1 , R 2 , and 1 H- 15 N NOE values of 0.81 Ϯ 0.10, 23.7 Ϯ 2.9, and 0.82 Ϯ 0.10 s Ϫ1 are found, respectively, at 600 MHz. A correlation time of 17.3 Ϯ 2.3 ns was estimated from the R 2 /R 1 ratio (the experimental relaxation rates are shown in supplemental Fig. S3).\\n\\nThe exchange with the solvent of the backbone amide proton was investigated through a series of 1 H-15 N heteronuclear single quantum correlation (HSQC) experiments performed over 4 days on the E,Zn-hasSOD1 SH-SH protein previously frozen dry and then dissolved in D 2 O. Exchange with the bulk solvent in D 2 O solution was measured following the intensity of amide proton moieties in 15 N-HSQC spectra. The amide protons can be grouped in three groups: fast exchanging (the peak disappears in less than 20 min), intermediate exchanging (the peak is still present after 20 min but disappears within 10 h), and slow exchanging (the peak persists after 10 h). The behavior indicates not only solvent exposure but also the presence of hydrogen bonding, which may slow down the exchange of the exposed amide protons. This is the case for almost all residues located in regions having a defined secondary structure, as they are involved in extensive H-bond networks that stabilize the ␤-barrel structure typical of this protein.\\n\\nCopper(I) Insertion-The copper-depleted protein was added with the copper(I) complex (Cu(CH 3 CN) 4 ) ϩ in a 1:1 ratio in an anaerobic chamber. The protein was then washed extensively in the presence of phosphate buffer 20 mM at pH 5. The metal uptake of the protein was established through one-dimensional NMR spectrum in the region of imidazole NHs (10 -16 ppm).\\n\\n\\nRESULTS AND DISCUSSION\\n\\nThe structure of the copper-depleted thermostable form (C6A, C111S) of human SOD1 with the disulfide bond reduced (E,Zn-hasSOD1 SH-SH ) has been determined in solution by NMR with high resolution (root mean square deviation 0.66 Ϯ 0.14 and 1.15 Ϯ 0.13 for the backbone and heavy atom, respectively, considering the segments involved in the ␤-barrel). The structure quality and the statistics on constraint violations for the final family is shown in Table 1. The protein is in a dimeric state and completely maintains the global fold of mature SOD1 (Fig. 1). Surprisingly, the breakage of the disulfide bond does not significantly destabilize the protein structure. This can suggest a functional rather than structural role for this highly conserved moiety. In addition, the backbone conformation of the region at the dimer interface (residues 53-55, 149 -151) is very similar to that found in x-ray (22,36) and solution structures (25) of human and yeast Cu,Zn-SOD1 SS . Local conformational rearrangements are found essentially only in the immediate surrounding of the two cysteines (Cys-57 and -146) involved in the intrasubunit disulfide bond. This protein region, and in particular residue 57, becomes extremely more solvent accessible with respect to the oxidized protein (Fig. 2). The solvent exposure of cysteine 57 is a peculiarity that is never observed in SOD1 structures, not even in the apo form of the protein (Fig. 2 and supplemental Fig. S4). The presence of the disulfide bond, in fact, prevents solvent exposure by forcing the Cys-57 to reach the Cys-146 that is located within the protein core. Solvent exposure is also confirmed by H 2 O/D 2 O solvent exchange measurements. The region spatially close to the cysteine residues responsible for the disulfide bridge (residues 57 and 146) appears more exposed to the solvent. This was not observed in the structures of Cu,Zn-hasSOD1 SS and E,Zn-E133QM2-asSOD1 SS .',\n",
              "  'Interestingly, the metal binding ligands in E,Zn-hasSOD1 SH-SH have a conformation very close to that found in Cu,Zn-hasSOD1 SS and in the monomeric copper-depleted form (24,25), indicating that the protein is ready to harbor the catalytic metal. In the active site channel of Cu,Zn-SOD1 SS , located between the electrostatic loop VII (121-142) and loop IV (49 -84), some H-bonds form a highly conserved network that is important for generating the optimal electrostatic field necessary to attract and drive the substrate toward the copper site. In the present structure most of the H-bonds are maintained with the exceptions of the copper ligand His-48 hydrogen bonded to Cys-57 via the carbonyl oxygen of Gly-61 and the catalytically important side chain of Arg-143. This is also assessed by the chemical shift perturbation of the amide proton of Gly-61 (supplemental Fig. S1). The total solvent exposure of Cys-57 located in loop IV is also reflected in an increased solvent exposure of residues 48 and 146, which are buried in Cu,Zn-hSOD1 SS . Accordingly, the relative amide protons display fast hydrogen/deuterium exchange in E,Zn-hasSOD1 SH-SH (data not shown).\\n\\nThe conformation of Arg-143 is extremely sensitive to the position of Cys-57 (37)(38)(39). Arg-143 is located at the end of the electrostatic loop (loop VII) whose conformation greatly affects the enzymatic activity. In the dimeric Cu,Zn-hasSOD1 SS solution structure, Cys-57 is quite rigid along with loop VII. As a consequence, the side chain of Arg-143 is highly ordered and is in the optimal conformation to interact with the  The number of meaningful constraints for each class is reported in parentheses. c Medium range distance constraints are those between residues (i,iϩ2), (i,iϩ3), (i,iϩ4), and (i,iϩ5). d As it results from the Ramachandran plot analysis.  superoxide anion bound to copper. In the monomeric Cu,Zn-E133QM2-asSOD1 SS species, loop IV containing Cys-57 experiences conformational equilibria and Arg-143 moves away from copper, determining a sizable drop in SOD activity (40). In the present structure, loop IV is highly disordered due to the S-S breaking resulting in extensive disorder of the Arg-143 side chain. Consequently, the overall electrostatic field is highly perturbed. The comparison of our solution structure with that of monomeric E,ZnE133QM2-asSOD1 SS (24) reveals that the observed disorder of residue Arg-143 is not a consequence of the absence of copper but of the relaxation of loop IV due to the S-S breaking. The breakage of the disulfide bond results in a perturbed distribution of the electrostatic charges in the active site channel essential to guide the superoxide ion in the catalytic site. Two mechanisms have been proposed for copper uptake and consequent activation of human SOD1; the first requires the presence of CCS (9,41), and the second involves the reduced form of glutathione (8). The first mechanism, described in yeast, is based on the x-ray structure of the heterodimeric yeast CCS-SOD1 complex. Although a direct comparison with the yeast E,Zn-SOD1 SH-SH is not possible because of the lack of its structure in the Protein Data Bank, the present solution structure shares some specific features with that of the yeast mutant H48F SOD1 in the heterodimeric complex with CCS (9) (Fig. 3). All SOD1 residues involved in the interaction with domain II of CCS (residues 51, 114, 152) experience similar conformation in E,Zn-hasSOD1 SH-SH regardless of the differences in the primary sequence. It has been proposed that in yeast the presence of Pro-142 and Pro-144 in place of Ser-142 and Leu-144 may account for the loss of ability to acquire copper independently of CCS (8). Because yeast E,Zn-SOD1 SH-SH has been found to be monomeric (10), we can speculate that these differences in the primary sequence destabilize the structure of this region, leading to monomerization (Leu-144 is inserted in the hydrophobic part of the protein called the plug of the barrel). The correct conformation for acquiring copper could be restored only by the formation of the heterodimeric complex with CCS facilitated by interaction with domain III. However, the complexity of the issue is demonstrated by the increased propensity to monomerization displayed by a subgroup of fALS mutants (for example G85R, D125H, L38V) that have the mutation far away from the dimer interface (11,19,26).\\n\\nIn the crystal structure of the heterodimeric complex, an intermolecular disulfide bond between Cys-57 of SOD1 and Cys-229 of CCS was found. If the present solution structure is superimposed on the yeast H48F SOD1 in the heterodimeric complex, a similar distance of the sulfur atom of Cys-57 from that of Cys-229 of CCS is found (0.20 nm in the heterodimer complex and 0.28 nm in the superimposed solution structure of E,Zn-hasSOD1 SH-SH ). This feature demonstrates that Cys-57 has a favorable conformation to form the same disulfide bond. Also, in the heterodimer complex the side chain of Arg-143 is far away from the copper site and is hydrogen bonded to the backbone of Ala-230 of yeast CCS; the mobility of Arg-143 could have the role of making its side chain available for this kind of interaction.\\n\\nOn the basis of the present structure we can speculate that the copper ion could enter in its site with the assistance of Cys-57, which, being close to the copper site, could constitute a ligand in a hypothetical intermediate (Fig. 4). Indeed, the high affinity of copper for thiol groups and the complete exposure of the side chain of Cys-57 could be significant . Proposed mechanism for acquisition of copper by E,Zn-hSOD1 SH-SH (displayed as an ellipsoid). Copper and zinc are represented as spheres (dark and light gray, respectively) and their sites as empty circles. Thiol groups are shown as small spheres. A hypothetical mechanism for acquisition of copper ion from CCS or glutathione (GSH) complex (shown as a thick wavy line) could take place through the formation of a metalbridged intermediate. for the incorporation of copper ion into the protein, regardless of whether this occurs through the interaction with CCS (9,41) or with other copper-containing species (e.g. copper-glutathione complex) (8).\\n\\nIndependently of the mechanism by which the copper enters the site in vivo, we have observed that E,Zn-hasSOD1 SH-SH is able to acquire copper(I) in vitro. The acquisition of copper can, in fact, be easily monitored by one-dimensional NMR spectrum of imidazole NH protons. Supplemental Fig. S5 shows the signals originated by copper binding; the figure also shows that the spectra of Cu(I),Zn-hasSOD1 SH-SH and Cu(I),Zn-hSOD1 SS are very similar (and different from E,Zn-hasSOD1 SH-SH ), thus demonstrating that the metal enters correctly in its site. Subsequent oxidation of the disulfide bond would allow the formation of the hydrogen bond network constraining the side chain of Arg-143, making the protein fully functioning.',\n",
              "  \"Exposed cysteines can also easily form interprotein disulfide bonds, thus giving rise to aggregates. The status of the disulfide bond in fALS SOD1 mutants has not been extensively characterized, although in all crystal structures of fALS mutants such bonds are intact (42)(43)(44). However, the disulfide bonds in fALS mutants are more susceptible to reduction than wild type SOD1 (26), and it has been suggested that in some fALS mutants, fibrils are formed through intersubunit S-S bonds (11). This implies that under cellular disulfide-reducing conditions at physiological pH and temperature, the fALS SOD1 mutants are more prone to aggregation. The present structure provides an important piece of information for understanding the copper uptake mechanism in SOD1 and for the debate about which is the relevant form of SOD1 for the onset of the fALS pathology.\\n\\n\\nstructure elements (residues 1-153) ␤1 (3-6), ␤2 (15-21), ␤3 (29-37), B4 (41-48), ␤5 (85-89), ␤6 (94-101), ␤7 (116-121), ␣1 (131-137), ␤8 (143-152) a REM indicates the energy minimized family of 30 structures, ͗REM͘ is the energy-minimized mean structure obtained from the coordinates of the individual REM structures. b\\n\\nFIGURE 1 .\\n1Solution structure of E,Zn-hasSOD1 SH-SH . The side chains of Cys-57 and Cys-146 are shown as sticks.\\n\\nFIGURE 2 .\\n2Solvent exposure of Cys-57 in the structures of one subunit of thermostable E,Zn-hasSOD1 SH-SH (left), Cu,Zn-hasSOD1 SS (right; Protein Data Bank code 1SOS), and wild type E,E-hSOD1 SS (center; Protein Data Bank code 1HL4). The surface of Cys-57 is shown in lighter color.\\n\\nFIGURE 4\\n4FIGURE 4. Proposed mechanism for acquisition of copper by E,Zn-hSOD1 SH-SH (displayed as an ellipsoid). Copper and zinc are represented as spheres (dark and light gray, respectively) and their sites as empty circles. Thiol groups are shown as small spheres. A hypothetical mechanism for acquisition of copper ion from CCS or glutathione (GSH) complex (shown as a thick wavy line) could take place through the formation of a metalbridged intermediate.\\n\\nFIGURE 3 .\\n3Structure superimposition of E,Zn-hasSOD1 SH-SH (light gray) and yeast SOD1 interacting with yeast CCS (black). Some key residues are highlighted (gray for E,Zn-hasSOD1 SH-SH and black for yeast SOD1): the carbonyl of Ala-230 involved in an H-bond with Arg-143 in the yeast complex, Cys-229 and Cys-231 belonging to CCS, Cys-57 and Cys-146 belonging to SOD1 structures.\\n\\nTABLE 1\\n1Statistical analysis of the energy-minimized family of conformers and the mean structure of E, Zn-hasSOD1 SH-SHREM a \\n(REM) a \\n\\n. I Fridovich, Adv. Enzymol. Relat. Areas Mol. Biol. 58Fridovich, I. (1986) Adv. Enzymol. Relat. Areas Mol. Biol. 58, 61-97\\n\\n. L Banci, I Bertini, C Luchinat, R A Hallewell, J. Am. Chem. Soc. 110Banci, L., Bertini, I., Luchinat, C., and Hallewell, R. A. (1988) J. Am. Chem. Soc. 110, 3629 -3633\\n\\n. C S Sevier, C A Kaiser, Nat. Rev. Mol. Cell. Biol. 3Sevier, C. S., and Kaiser, C. A. (2002) Nat. Rev. Mol. Cell. Biol. 3, 836 -847\\n\\n. J S Valentine, P A Doucette, S Z Potter, Annu. Rev. Biochem. Valentine, J. S., Doucette, P. A., and Potter, S. Z. (2004) Annu. Rev. Biochem.\\n\\n. V C Culotta, L W Klomp, J Strain, R L Casareno, B Krems, J D Gitlin, J. Biol. Chem. 272Culotta, V. C., Klomp, L. W., Strain, J., Casareno, R. L., Krems, B., and Gitlin, J. D. (1997) J. Biol. Chem. 272, 23469 -23472\\n\\n. M E Portnoy, P J Schmidt, R S Rogers, V C Culotta, Mol. Genet. Genomics. 265Portnoy, M. E., Schmidt, P. J., Rogers, R. S., and Culotta, V. C. (2001) Mol. Genet. Genomics 265, 873-882\\n\\n. P C Wong, D Waggoner, J R Subramaniam, L Tessarollo, T B Bartnikas, V C Culotta, D L Price, J Rothstein, J D Gitlin, Proc. Natl. Acad. Sci. U. S. A. 97Wong, P. C., Waggoner, D., Subramaniam, J. R., Tessarollo, L., Bartnikas, T. B., Cu- lotta, V. C., Price, D. L., Rothstein, J., and Gitlin, J. D. (2000) Proc. Natl. Acad. Sci. U. S. A. 97, 2886 -2891\\n\\n. M C Carroll, J B Girouard, J L Ulloa, J R Subramaniam, P C Wong, J S Valentine, V C Culotta, Proc. Natl. Acad. Sci. U. S. A. 101Carroll, M. C., Girouard, J. B., Ulloa, J. L., Subramaniam, J. R., Wong, P. C., Valentine, J. S., and Culotta, V. C. (2004) Proc. Natl. Acad. Sci. U. S. A. 101, 5964 -5969\\n\\n. A L Lamb, A S Torres, T V O&apos;halloran, A C Rosenzweig, Nat. Struct. Biol. 8Lamb, A. L., Torres, A. S., O'Halloran, T. V., and Rosenzweig, A. C. (2001) Nat. Struct. Biol. 8, 751-755\\n\\n. Y Furukawa, A S Torres, T V Halloran, EMBO J. 23Furukawa, Y., Torres, A. S., and O'Halloran, T. V. (2004) EMBO J. 23, 2872-2881\\n\\n. Y Furukawa, T V Halloran, J. Biol. Chem. 280Furukawa, Y., and O'Halloran, T. V. (2005) J. Biol. Chem. 280, 17266 -17274\\n\\n. J Wang, H Slunt, V Gonzales, D Fromholt, M Coonfield, N G Copeland, N A Jenkins, D R Borchelt, Hum. Mol. Genet. 12Wang, J., Slunt, H., Gonzales, V., Fromholt, D., Coonfield, M., Copeland, N. G., Jenkins, N. A., and Borchelt, D. R. (2003) Hum. Mol. Genet. 12, 2753-2764\"]}"
            ]
          },
          "execution_count": 17,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "### Connect to the Chroma DB\n",
        "abstract_store = chroma.get_collection(name=\"specter_abstracts\")\n",
        "lang_collect = chroma.get_collection(name=\"fulltext\")\n",
        "lang_collect.peek()"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 18,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "['Training, data collection, and clinical assessment Before each survey round, which occurred approximately twice per year, all trachoma graders (integrated eye care workers) participated in a 5 to 7-day training [12,13,16]. Graders were required to attend training regardless of the number or date of their previous trainings. Training methods used for these surveys pre-date but were very similar to those used in more current trachoma surveys [17]. Training consisted of classroom sessions to understand the definitions of all 5 World Health Organization (WHO) simplified trachoma signs through slide-deck review, and field-based sessions focused on providing graders experience in the range of presentations trachoma signs can take [18]. Graders were then required to pass a slide-based exam, which included images of all 5 simplified trachoma signs, including approximately 5 images of TS. Graders who passed the slide exam then moved onto a field reliability exam where they were required to grade 50 eyes under field conditions. The \"gold standard\" for the field reliability exam was the consensus grade of 3 expert graders, typically ophthalmologists, ocular surgeons, or trachoma managers [16,19]. Those graders who scored an interobserver agreement of ≥ 84 % and kappa ≥ 0.7, compared to the consensus grade of the 3 expert graders for the sign TF joined the survey team [16]. Although the field reliability exam was based on grading TF, graders were also trained on how to identify and grade TS using slide training, slide examination, and field-based practice within TT surgery campaigns that were specifically planned to accommodate the graders\\' practice for these surveys.\\n\\nThe primary aim of these surveys was to estimate TF among children ages 1 to 9 years, but all study participants were examined for all 5 WHO simplified signs of trachoma, including TS. Both eyes of all study participants were examined for TS, which was defined as the presence of scarring in the tarsal conjunctiva (scars are easily visible as white lines, bands, or sheets [fibrosis] in the tarsal conjunctiva) [18]. Trachoma grading was performed using binocular loupes with x2.5 magnification under sunlight or using a flashlight. Multiple grading teams worked in each district, and supervisors traveling with grading teams spot-checked grading, particularly during the first 2 weeks of each survey.\\n\\nData collection took place on Samsung Galaxy tablets running custom-built software [20]. Data recorders were information technology or biostatistics college graduates who took a 4 to 5-day training and passed a data collection exam using the tablets. All household surveys were conducted in the Amharic language.\\n\\n\\nStatistical analysis\\n\\nAll TS prevalence estimates were weighted to account for the probability of sampling at both sampling stages. Furthermore, all TS prevalence estimates reported, district-and region-level, were age-sex adjusted with post-stratification weighting using 5-year age-sex bands from the survey census population [21,22]. Confidence intervals (CI) were calculated using Taylor linearization and survey procedures (svy package) in Stata, version 15.1 (STATA Corporation, College Station TX, USA), to account for household and village level clustering. Most analyses were focused on individuals ages 15 years and older (standard age group used to estimate TT prevalence), individuals younger than 15 years, and individuals ages 1 to 9 years (standard age group for estimating TF prevalence) [8]. The prevalence of TS concurrent with TF and TI was also estimated. Chi-squared tests were used to evaluate TS prevalence between age group and sex. A Spearman correlation test was performed to examine relationships between TS and TT, TS and TF and TS and TI at the district level. Clustering by household and survey cluster was assessed by calculating the intraclass correlation coefficient (ICC) of TS prevalence with a random-effects model. District-level prevalence was grouped into the categories of < 5 %; 5.0-9.9 %; 10-14.9 %; > 15 % based on the overall TS prevalence distribution.\\n\\n\\nResults\\n\\nA total of 276,068 individuals were surveyed between 2010 and 2015 in all 152 programmatic districts across the 10 zones of the Amhara region. Among those enumerated, 75.5 % (208,510) of individuals were present and examined for clinical signs of trachoma. The majority of examined individuals, 52.9 % (110,211 respondents), were ages 15 years and older, 35.2 % (73,406 respondents) were children ages 1 to 9 years, and 53.5 % (111,577 respondents) were female.\\n\\nThe regional TS prevalence among all ages was 8.2 % (95 % CI: 7.7-8.6 %), and the TS prevalence among individuals ages 15 years and older was 12.6 % (95 % CI: 12.0-13.3 %). The range of TS prevalence by district (n = 152) in this age group was right-skewed and ranged from 0.6 % (95 % CI: 0.1-3.4 %) to 36.9 % (95 % CI: 27.2-47.8) (Fig. 1). Each of the 10 zones had at least one district with a TS prevalence in the highest category (TS ≥ 15 %), and Waghemra and South Gondar in the central part of the region had the heaviest burden of the high prevalence districts (Fig. 2). The district prevalence of TS was moderately correlated with the prevalence of TT (Spearman correlation [r] = 0.31, P < 0.001 (Fig. 3)).\\n\\nAmong individuals, the prevalence of TS increased significantly with age (P < 0.0001); the highest prevalence was observed in adults ages 56 to 60 years ( (Fig. 4). Among individuals ages 15 years and older, TS prevalence among females and males was 15.2 % (95 % CI: 11.5-19.7 %) and 14.6 (95 % CI: 12.6-16.9 %), respectively; the difference was not statistically significant (P = 0.73). Among individuals with TT in this age group, 955 (23.3 %) individuals had concurrent TS, and the individual-level correlation between TT and TS was low (r = 0.0275; P < 0.001).There was some evidence of clustering at the household level with an ICC of 0.24 (95 % CI: 0.23-0.26) and little evidence of clustering at the village level with an ICC of 0.08 (95 % CI: 0.07-0.10). At the household level, 79.0 % of households did not have any adults with TS, but in 9.3 % of households all of the adults had TS. The remaining 11.7 % of households had at least one but not all adults with TS.\\n\\nThe regional TS prevalence among individuals ages 1 to 15 years and individuals ages 1 to 9 years was 2.2 % (95 % CI: 1.8-2.8 %) and 1.1 % (95 % CI: 1.0-1.2 %) respectively. Moderate correlations were observed between district prevalence TS and district prevalence of TF (r = 0.33, P < 0.0001) and TI (r = 0.52, P < 0.0001) among individuals ages 1 to 15 years. Among individuals ages 1 to 15 years, the prevalence of TS increased with age (P < 0.001) reaching as high as 10.6 % in 15-yearold males (Fig. 5). Overall, however, the TS prevalence did not differ by sex (P = 0.40) in this age group. Among individuals ages 1 to 15 years with TS, 24.3 % also had TF and 4.8 % also had TI. In this age group, the prevalence of concurrent TS and active trachoma (TF and/or TI) was 0.6 % (95 % CI: 0.4-0.9 %), while the prevalence of concurrent TS and TF was 0.2 % (95 % CI: 0.04-0.9 %) (Fig. 6).\\n\\n\\nDiscussion',\n",
              " 'Discussion\\n\\nThe results of these 152 population-based surveys suggest that the Amhara region has had a long history of intense trachoma exposure, and that a large population remains at risk for developing TT. Future monitoring of incident TT will likely be needed even after the elimination as a public health problem threshold for TT has     [13]. Furthermore, as part of TT surgical campaigns throughout the region, over 408,000 surgeries were conducted on TT patients between 2007 and 2015 [13]. Despite this, impact surveys conducted after these interventions demonstrated that many districts still had high levels of TF and C. trachomatis infection and that all but 4 districts were still above the elimination as a public health problem threshold for TT [12,13]. This continued high prevalence of TT observed in many districts may in part be due to incident TT as a result of years of progressive scarring. We found that districts with high levels of TT were more likely to have a concurrently high level of TS. Future surgical services will most likely be needed the most in districts with a high TT burden with concurrently high TS prevalence.\\n\\nIn trachoma-endemic communities, scarring starts to develop in childhood and continues to progress with age. Scarring is thought to be driven by repeated infection with C. trachomatis, whereby a higher cumulative burden of infection would lead to increased scarring [24]. It is also believed that scarring may be a result of persistent inflammation [10]. Evidence for this mechanism comes from cohort studies by which progressive scarring was observed in the absence of C. trachomatis infection [10,25]. Within adults with scarring in Amhara, 23 % had observed scarring progression over a 2-year period [10]. Within our surveys, we found that the prevalence of TS increased with age, reaching as high as a quarter of the population among those ages 60 to 64 years. Given the size of Amhara, this represents a large population at risk for scarring progression and incident TT. The Trachoma Control Program in Amhara and in other hyperendemic regions will need to develop surveillance systems to detect and operate incident TT cases. Given the within-household clustering of scarring observed in this report, the Program may expect incident TT to cluster within households as well. All communities within Amhara have received on average 5 years of the SAFE strategy including annual MDA with azithromycin. Azithromycin is an antibiotic shown to be effective against C. trachomatis and thought to also have antiinflammatory properties [26,27]. It is possible therefore that annual azithromycin distribution will reduce the progression of scarring and possibly the incidence of TT in this population.\\n\\nIn this trachoma endemic setting, TS was detectable in children, although at relatively low levels (≤ 2 %). This low TS prevalence may have been due to the annual antibiotic distribution in communities throughout the region over the last 5 years. Although pre-SAFE TS data from children in Amhara has not been previously published, in neighboring trachoma hyperendemic South Sudan, the prevalence of TS prior to the start of the SAFE strategy was 19.3 % among children ages 1 to 9 years and 31.7 % among children ages 10 to 14 years, measured using the same WHO simplified system [28]. In Guinea Bissau, a country where the TF prevalence pre-SAFE would be considered mesoendemic (22 %), the prevalence of TS among children ages 0 to 5 years and ages 6 to 10 years was 2.7 and 2.8 % respectively [29]. Although comparisons across various settings is difficult due to differences in the assessment of scarring, it appears that the level of scarring observed in Amhara among children is not reflective of a trachoma hyperendemic setting. However, children are clearly still at risk for developing scarring in Amhara since the prevalence of TF, TI, and C. trachomatis infection among children ages 1 to 9 years was still 25.9, 5.5, and 5.7 % respectively, and a statistically significant association was observed between age and scarring, even in this limited age range [12,13]. Prior studies have demonstrated that the presence of TI is associated with the incidence and progression of scarring [4,10,24,30,31]. We observed that approximately 25 % of children with TS also had TF, and nearly 5 % also had TI, therefore this subset in Amhara may be at particular risk for scarring progression. Although longitudinal studies are needed to know the degree to which scarring is still developing and progressing in Amhara, clearly, further enhanced efforts are needed to reduce the levels of inflammatory trachoma in these communities in order to reducing the burden of future scarring.\\n\\nThese surveys had several limitations. Survey teams graded TS in the field using the WHO simplified grading scheme, and therefore it was not possible to grade scarring severity. As graders were trained to identify bands or sheets that were \"easily visible,\" it is likely that mild scarring would be missed in this definition. Therefore, our TS estimates may have underestimated the total scarring burden in Amhara, and the correlation between TT and TS may have been attenuated. Evidence for this comes from a study in Amhara showing that 70 % of participants without TT, and 95 % of those with TT had had some degree of scarring using a detailed grading scheme [32]. In research settings, grading using conjunctival photographs has been common and could possibly be considered for programmatic use to allow for more granular grading of scarring. Although the field reliability test was focused on the trachoma sign TF, TS grading was integrated with the survey training, and the trainees had a chance to practice in the field along with TT identification practice that was integrated with TT surgery campaigns. This study was cross-sectional, and therefore could not measure incident or progression of TS, which would be useful indicators for trachoma programs. Currently programmatic monitoring globally no longer includes TS prevalence as an indicator, and longitudinal studies focused on scarring are uncommon [3,7,10,14,24]. The inclusion of TS estimation within programmatic surveys should be reconsidered as scarring data has recently been used to better understand the history of trachoma within treatment-naïve endemic regions and has further been used to help make MDA treatment decisions under select circumstances [6,33]. More data on scarring can also provide modelers with empirical data to improve models for the natural history of trachoma and in forecasting incident TT [34,35].\\n\\nOverall, a considerable percentage of adults have TS in these trachoma-endemic districts in Amhara, which suggests both a long history of C. trachomatis infection and the possibility of continued development of incident cases of TT in the future. Including TS data in the surveillence of trachoma may be useful for the global elimination program. Trachoma programs serving formerly highly endemic areas should continue to develop context-appropriate monitoring systems for incident TT for the foreseeable future. \\n\\nFig. 1\\n1District-level TS prevalence among individuals ages 15 years and older, Amhara, Ethiopia, 2010 to 2015 Fig. 2 District-level TS prevalence among individuals ages 15 years and older, Amhara, Ethiopia, 2010 to 2015\\n\\nFig. 3\\n3Scatterplot of district-level TS and TT prevalence, Amhara, Ethiopia, 2010 to 2015. N=152 districts, r=0.31\\n\\nFig. 4\\n4Prevalence of TS by sex and age group, Amhara, Ethiopia, 2010 to 2015\\n\\nFig. 5\\n5Prevalence of TS by sex and ages 1 to 15 years, Amhara, Ethiopia, 2010 to 2015 Fig. 6 Active trachoma with TS prevalence in ages 1 to 15, Amhara, Ethiopia, 2010 to 2015\\n\\n\\nAbbreviations CI: Confidence interval; ICC: Intraclass correlation coefficient; MDA: Mass drug administration; SAFE: Surgery, Antibiotic, Facial cleanliness, and Environmental improvement; R: Spearman correlation; TF: Trachomatous inflammationfollicular; TI: Trachomatous inflammation-intense; TS: Trachomatous scarring; TT: Trachomatous trichiasis; WHO: World Health Organization.',\n",
              " \"These results led to the scale-up of the full SAFE strategy in all districts of Amhara. At scale, the Trachoma Control Program was delivering approximately 15 million doses of antibiotic per year as part of its MDA campaignsAstale et al. BMC Ophthalmology \\n(2021) 21:213 been reached. It is promising, however, that children \\nages 1 to 15 years, many born after the MDA program \\nbegan, have a TS prevalence of 2 %. Currently, program-\\nmatic monitoring of scarring is rare, and TS prevalence \\nas a district-level indicator is no longer included in most \\ntrachoma survey methodologies. Future longitudinal re-\\nsearch is needed to determine whether scarring data can \\nhelp programs plan future needs for TT services. \\nEarly district-level surveys conducted in Amhara in \\n2001 and 2003 demonstrated that trachoma was hyper-\\nendemic in 5 pilot districts, with a TF prevalence among \\nchildren ages 1 to 9 years ranging between 49 and 90 % \\nand TT among individuals ages 15 years and older ran-\\nging between 2.0 and 5.4 % [15]. In 2006, a zonal-level \\nbaseline trachoma survey demonstrated both a high TF \\n(32.7 %) and TT (6.2 %) prevalence region-wide [23]. \\n\\nAcknowledgementsWe are grateful to all study participants who volunteered their time to participate in the study. We appreciate and thank the dedicated field teams, drivers and supervisors who persevered through difficult conditions to compete the data collection process, many of whom participated in all surveys. We are also grateful for the collaboration with the Lions-Carter Center Sight-First Initiative.Authors' contributionsFundingFunding for the survey in summer 2013 was provided by USAIDs ENVISION project led by RTI International.Availability of data and materials All data generated or analyzed during this study are included in this published article.DeclarationsEthics approval and consent to participate Survey protocols were approved by Amhara National Regional State Health Bureau, and by Emory University Institutional Review Board under protocol 079-2006. All methods were carried out in accordance with relevant national guidelines and regulations. The study objectives and procedures were explained to the head of households or adult respondents. Due to high illiteracy rate among the population, Institutional Review Board approval was obtained for informed oral consent or assent. Informed oral consent or assent was obtained according to the principles of the Declaration of Helsinki.Consent for publicationNot applicable.Competing interestsThe authors declare that they have no competing interests.\\nBacterial infection in scarring trachoma. V H Hu, P Massae, H A Weiss, C Chevallier, J J Onyango, I A Afwamba, D C Mabey, R L Bailey, M J Burton, Invest Ophthalmol Vis Sci. 525Hu VH, Massae P, Weiss HA, Chevallier C, Onyango JJ, Afwamba IA, Mabey DC, Bailey RL, Burton MJ. Bacterial infection in scarring trachoma. Invest Ophthalmol Vis Sci. 2011;52(5):2181-6.\\n\\nIn vivo confocal microscopy of trachoma in relation to normal tarsal conjunctiva. V H Hu, P Massae, H A Weiss, I A Cree, P Courtright, D C Mabey, R L Bailey, M J Burton, Ophthalmology. 1184Hu VH, Massae P, Weiss HA, Cree IA, Courtright P, Mabey DC, Bailey RL, Burton MJ. In vivo confocal microscopy of trachoma in relation to normal tarsal conjunctiva. Ophthalmology. 2011;118(4):747-54.\\n\\nIncidence estimates of late stages of trachoma among women in a hyperendemic area of central Tanzania. B Munoz, Aron J Turner, V West, S , Trop Med Int Health. 211Munoz B, Aron J, Turner V, West S. Incidence estimates of late stages of trachoma among women in a hyperendemic area of central Tanzania. Trop Med Int Health. 1997;2(11):1030-8.\\n\\nProgression of active trachoma to scarring in a cohort of Tanzanian children. S K West, B Munoz, H Mkocha, Y H Hsieh, M C Lynch, Ophthalmic Epidemiol. 82-3West SK, Munoz B, Mkocha H, Hsieh YH, Lynch MC. Progression of active trachoma to scarring in a cohort of Tanzanian children. Ophthalmic Epidemiol. 2001;8(2-3):137-44.\\n\\nAge, sex, and cohort effects in a longitudinal study of trachomatous scarring. M A Wolle, B Munoz, H Mkocha, S K West, Invest Ophthalmol Vis Sci. 502Wolle MA, Munoz B, Mkocha H, West SK. Age, sex, and cohort effects in a longitudinal study of trachomatous scarring. Invest Ophthalmol Vis Sci. 2009;50(2):592-6.\\n\\nClinical signs of trachoma are prevalent among Solomon Islanders who have no persistent markers of prior infection with Chlamydia trachomatis. R Butcher, O Sokana, K Jack, L Sui, C Russell, A Last, D L Martin, M J Burton, A W Solomon, Dcw Mabey, Wellcome Open Res. 314Butcher R, Sokana O, Jack K, Sui L, Russell C, Last A, Martin DL, Burton MJ, Solomon AW, Mabey DCW, et al. Clinical signs of trachoma are prevalent among Solomon Islanders who have no persistent markers of prior infection with Chlamydia trachomatis. Wellcome Open Res. 2018;3:14.\\n\\nProgression of scarring trachoma in Tanzanian children: a four-year cohort study. A M Ramadhani, T Derrick, D Macleod, P Massae, E Mafuru, A Malisa, K Mbuya, C H Roberts, W Makupa, T Mtuy, PLoS Negl Trop Dis. 1387638Ramadhani AM, Derrick T, Macleod D, Massae P, Mafuru E, Malisa A, Mbuya K, Roberts CH, Makupa W, Mtuy T, et al. Progression of scarring trachoma in Tanzanian children: a four-year cohort study. PLoS Negl Trop Dis. 2019;13(8): e0007638.\\n\\nWHO alliance for the global elimination of trachoma by 2020: progress report on elimination of trachoma. Wkly Epidemiol Rec. 9429WHO. WHO alliance for the global elimination of trachoma by 2020: progress report on elimination of trachoma, 2018. Wkly Epidemiol Rec. 2019;94(29):317-28.\\n\\nThe prevalence of trachomatous trichiasis in Oman (Oman eye study. R Khandekar, A J Mohammed, Ophthalmic Epidemiol. 145Khandekar R, Mohammed AJ. The prevalence of trachomatous trichiasis in Oman (Oman eye study 2005). Ophthalmic Epidemiol. 2007;14(5):267-72.\\n\\nPathogenesis of progressive scarring trachoma in Ethiopia and Tanzania and its implications for disease control: two cohort studies. M J Burton, S N Rajak, V H Hu, A Ramadhani, E Habtamu, P Massae, Z Tadesse, K Callahan, P M Emerson, P T Khaw, PLoS Negl Trop Dis. 953763Burton MJ, Rajak SN, Hu VH, Ramadhani A, Habtamu E, Massae P, Tadesse Z, Callahan K, Emerson PM, Khaw PT, et al. Pathogenesis of progressive scarring trachoma in Ethiopia and Tanzania and its implications for disease control: two cohort studies. PLoS Negl Trop Dis. 2015;9(5):e0003763.\",\n",
              " 'Risk factors for incidence of trachomatous scarring in a cohort of women in low endemic district. R Karani, M Wolle, H Mkocha, B Munoz, S K West, Br J Ophthalmol. 1024Karani R, Wolle M, Mkocha H, Munoz B, West SK. Risk factors for incidence of trachomatous scarring in a cohort of women in low endemic district. Br J Ophthalmol. 2018;102(4):419-23.\\n\\nOcular chlamydia trachomatis infection under the surgery, antibiotics, facial cleanliness, and environmental improvement strategy in Amhara. S D Nash, Aep Stewart, M Zerihun, E Sata, D Gessese, B Melak, T Endeshaw, M Chanyalew, A Chernet, B Bayissasse, Clin Infect Dis. 6712Nash SD, Stewart AEP, Zerihun M, Sata E, Gessese D, Melak B, Endeshaw T, Chanyalew M, Chernet A, Bayissasse B, et al. Ocular chlamydia trachomatis infection under the surgery, antibiotics, facial cleanliness, and environmental improvement strategy in Amhara, Ethiopia, 2011-2015. Clin Infect Dis. 2018; 67(12):1840-1846.\\n\\nProgress to eliminate trachoma as a public health problem in Amhara National Regional State, Ethiopia: results of 152 population-based surveys. Aep Stewart, M Zerihun, D Gessese, B Melak, E Sata, A W Nute, T Astale, T Endeshaw, T Teferi, Z Tadesse, Am J Trop Med Hyg. 1016Stewart AEP, Zerihun M, Gessese D, Melak B, Sata E, Nute AW, Astale T, Endeshaw T, Teferi T, Tadesse Z, et al. Progress to eliminate trachoma as a public health problem in Amhara National Regional State, Ethiopia: results of 152 population-based surveys. Am J Trop Med Hyg. 2019;101(6):1286-95.\\n\\nNatural history of trachomatous scarring in The Gambia: results of a 12-year longitudinal follow-up. R J Bowman, B Jatta, B Cham, R L Bailey, H Faal, M Myatt, A Foster, G J Johnson, Ophthalmology. 10812Bowman RJ, Jatta B, Cham B, Bailey RL, Faal H, Myatt M, Foster A, Johnson GJ. Natural history of trachomatous scarring in The Gambia: results of a 12- year longitudinal follow-up. Ophthalmology. 2001;108(12):2219-24.\\n\\nEvaluation of three years of the SAFE strategy (Surgery, Antibiotics, Facial cleanliness and Environmental improvement) for trachoma control in five districts of Ethiopia hyperendemic for trachoma. J Ngondi, T Gebre, E B Shargie, L Adamu, Y Ejigsemahu, T Teferi, M Zerihun, B Ayele, V Cevallos, J King, Trans R Soc Trop Med Hyg. 10310Ngondi J, Gebre T, Shargie EB, Adamu L, Ejigsemahu Y, Teferi T, Zerihun M, Ayele B, Cevallos V, King J. Evaluation of three years of the SAFE strategy (Surgery, Antibiotics, Facial cleanliness and Environmental improvement) for trachoma control in five districts of Ethiopia hyperendemic for trachoma. Trans R Soc Trop Med Hyg. 2009;103(10):1001-10.\\n\\nPrevalence of trachoma at subdistrict level in Ethiopia: determining when to stop mass azithromycin distribution. J D King, T Teferi, E A Cromwell, M Zerihun, J M Ngondi, M Damte, F Ayalew, Z Tadesse, T Gebre, A Mulualem, PLoS Negl Trop Dis. 832732King JD, Teferi T, Cromwell EA, Zerihun M, Ngondi JM, Damte M, Ayalew F, Tadesse Z, Gebre T, Mulualem A, et al. Prevalence of trachoma at sub- district level in Ethiopia: determining when to stop mass azithromycin distribution. PLoS Negl Trop Dis. 2014;8(3):e2732.\\n\\nThe global trachoma mapping project: methodology of a 34-country population-based study. A W Solomon, A L Pavluck, P Courtright, A Aboe, L Adamu, W Alemayehu, M Alemu, N D Alexander, A B Kello, B Bero, Ophthalmic Epidemiol. 223Solomon AW, Pavluck AL, Courtright P, Aboe A, Adamu L, Alemayehu W, Alemu M, Alexander ND, Kello AB, Bero B, et al. The global trachoma mapping project: methodology of a 34-country population-based study. Ophthalmic Epidemiol. 2015;22(3):214-25.\\n\\nA simple system for the assessment of trachoma and its complications. B Thylefors, C R Dawson, B R Jones, S West, H R Taylor, Bull World Health Org. 654477Thylefors B, Dawson CR, Jones BR, West S, Taylor HR. A simple system for the assessment of trachoma and its complications. Bull World Health Org. 1987;65(4):477.\\n\\nTrachoma prevalence remains below threshold in five districts after stopping mass drug administration: results of five surveillance surveys within a hyperendemic setting in Amhara. S D Nash, Aep Stewart, T Astale, E Sata, M Zerihun, D Gessese, B Melak, G Ayenew, Z Ayele, B Bayissasse, Ethiopia. Trans R Soc Trop Med Hyg. 11212Nash SD, Stewart AEP, Astale T, Sata E, Zerihun M, Gessese D, Melak B, Ayenew G, Ayele Z, Bayissasse B, et al. Trachoma prevalence remains below threshold in five districts after stopping mass drug administration: results of five surveillance surveys within a hyperendemic setting in Amhara, Ethiopia. Trans R Soc Trop Med Hyg. 2018;112(12):538-45.\\n\\nA novel electronic data collection system for large-scale surveys of neglected tropical diseases. J D King, J Buolamwini, E A Cromwell, A Panfel, T Teferi, M Zerihun, B Melak, J Watson, Z Tadesse, D Vienneau, PloS One. 8974570King JD, Buolamwini J, Cromwell EA, Panfel A, Teferi T, Zerihun M, Melak B, Watson J, Tadesse Z, Vienneau D, et al. A novel electronic data collection system for large-scale surveys of neglected tropical diseases. PloS One. 2013;8(9):e74570.\\n\\nBurden of trachoma in five counties of eastern Equatoria state, south Sudan: results from population-based surveys. A M Sanders, Aep Stewart, S Makoy, J J Chebet, P Magok, A Kuol, C Blauvelt, R Lako, J Rumunu, E K Callahan, PLoS Negl Trop Dis. 1165658Sanders AM, Stewart AEP, Makoy S, Chebet JJ, Magok P, Kuol A, Blauvelt C, Lako R, Rumunu J, Callahan EK, et al. Burden of trachoma in five counties of eastern Equatoria state, south Sudan: results from population-based surveys. PLoS Negl Trop Dis. 2017;11(6):e0005658.',\n",
              " 'Progress toward elimination of trachoma as a public health problem in seven localities in the republic of Sudan: results from population-based surveys. A M Sanders, Z Abdalla, B E Elshafie, M Elsanosi, A W Nute, N Aziz, E K Callahan, S D Nash, Am J Trop Med Hyg. 1016Sanders AM, Abdalla Z, Elshafie BE, Elsanosi M, Nute AW, Aziz N, Callahan EK, Nash SD. Progress toward elimination of trachoma as a public health problem in seven localities in the republic of Sudan: results from population-based surveys. Am J Trop Med Hyg. 2019;101(6):1296-302.\\n\\nIntegrating an NTD with one of \"The big three\": combined malaria and trachoma survey in Amhara Region of Ethiopia. P M Emerson, J Ngondi, E Biru, P M Graves, Y Ejigsemahu, T Gebre, T Endeshaw, A Genet, A W Mosher, M Zerihun, PLoS Negl Trop Dis. 23197Emerson PM, Ngondi J, Biru E, Graves PM, Ejigsemahu Y, Gebre T, Endeshaw T, Genet A, Mosher AW, Zerihun M, et al. Integrating an NTD with one of \"The big three\": combined malaria and trachoma survey in Amhara Region of Ethiopia. PLoS Negl Trop Dis. 2008;2(3):e197.\\n\\nConstant ocular infection with Chlamydia trachomatis predicts risk of scarring in children in Tanzania. M A Wolle, B E Munoz, H Mkocha, S K West, Ophthalmology. 1162Wolle MA, Munoz BE, Mkocha H, West SK. Constant ocular infection with Chlamydia trachomatis predicts risk of scarring in children in Tanzania. Ophthalmology. 2009;116(2):243-7.\\n\\nIncidence and progression of trachomatous scarring in a cohort of children in a formerly hyper-endemic district of Tanzania. M S Kashaf, B E Munoz, H Mkocha, M A Wolle, F Naufal, S K West, PLoS Negl Trop Dis. 14108708Kashaf MS, Munoz BE, Mkocha H, Wolle MA, Naufal F, West SK. Incidence and progression of trachomatous scarring in a cohort of children in a formerly hyper-endemic district of Tanzania. PLoS Negl Trop Dis. 2020; 14(10):e0008708.\\n\\nRandomised controlled trial of single-dose azithromycin in treatment of trachoma. R L Bailey, P Arullendran, H C Whittle, D C Mabey, Lancet. 3428869Bailey RL, Arullendran P, Whittle HC, Mabey DC. Randomised controlled trial of single-dose azithromycin in treatment of trachoma. Lancet. 1993; 342(8869):453-6.\\n\\nAnti-inflammatory effects of macrolides-an underappreciated benefit in the treatment of community-acquired respiratory tract infections and chronic inflammatory pulmonary conditions?. G W Amsden, J Antimicrob Chemother. 551Amsden GW. Anti-inflammatory effects of macrolides-an underappreciated benefit in the treatment of community-acquired respiratory tract infections and chronic inflammatory pulmonary conditions? J Antimicrob Chemother. 2005;55(1):10-21.\\n\\nThe burden of trachoma in Ayod County of Southern Sudan. J D King, J Ngondi, G Gatpan, B Lopidia, S Becknell, P M Emerson, PLoS Negl Trop Dis. 29299King JD, Ngondi J, Gatpan G, Lopidia B, Becknell S, Emerson PM. The burden of trachoma in Ayod County of Southern Sudan. PLoS Negl Trop Dis. 2008;2(9):e299.\\n\\nRisk factors for active trachoma and ocular Chlamydia trachomatis infection in treatment-naive trachomahyperendemic communities of the Bijagos Archipelago. A R Last, S E Burr, H A Weiss, E M Harding-Esch, E Cassama, M Nabicassa, D C Mabey, M J Holland, R L Bailey, PLoS Negl Trop Dis. 862900Last AR, Burr SE, Weiss HA, Harding-Esch EM, Cassama E, Nabicassa M, Mabey DC, Holland MJ, Bailey RL. Risk factors for active trachoma and ocular Chlamydia trachomatis infection in treatment-naive trachoma- hyperendemic communities of the Bijagos Archipelago, Guinea Bissau. PLoS Negl Trop Dis. 2014;8(6):e2900.\\n\\nWhat clinical signs are critical in evaluating the intervention in trachoma. C R Dawson, Mr, T Daghfous, R Juster, J Schachter, Chlamyidal Infections. WR BCambridgeCambridge Universtify PressDawson CR MR, Daghfous T, Juster R, Schachter J. What clinical signs are critical in evaluating the intervention in trachoma? In: WR B, editor. Chlamyidal Infections. Cambridge: Cambridge Universtify Press; 1990. p. 271-278.\\n\\nBlinding trachoma: systematic review of rates and risk factors for progressive disease. A M Ramadhani, T Derrick, M J Holland, M J Burton, PLoS Negl Trop Dis. 1084859Ramadhani AM, Derrick T, Holland MJ, Burton MJ. Blinding trachoma: systematic review of rates and risk factors for progressive disease. PLoS Negl Trop Dis. 2016;10(8):e0004859.\\n\\nWorld Health Organization. WHO. Report of the 4th global scientific meeting on trachoma. Geneva; GenevaWHO. Report of the 4th global scientific meeting on trachoma, Geneva, 27- 29 November 2018. Geneva: World Health Organization; 2019.\\n\\nConjunctival scarring, corneal pannus and Herbert\\'s pits in adolescent children in trachoma-endemic populations of the Solomon Islands and Vanuatu. R Butcher, J Tagabasoe, J Manemaka, A Bong, M Garae, L Daniel, C Roberts, B L Handley, V H Hu, E M Harding-Esch, Clin Infect Dis. Epub ahead of printButcher R, Tagabasoe J, Manemaka J, Bong A, Garae M, Daniel L, Roberts C, Handley BL, Hu VH, Harding-Esch EM, et al. Conjunctival scarring, corneal pannus and Herbert\\'s pits in adolescent children in trachoma-endemic populations of the Solomon Islands and Vanuatu. Clin Infect Dis. 2020. [Epub ahead of print].\\n\\nThe development of an agestructured model for trachoma transmission dynamics, pathogenesis and control. M Gambhir, M G Basanez, M J Burton, A W Solomon, R L Bailey, M J Holland, I M Blake, C A Donnelly, I Jabr, D C Mabey, PLoS Negl Trop Dis. 36462Gambhir M, Basanez MG, Burton MJ, Solomon AW, Bailey RL, Holland MJ, Blake IM, Donnelly CA, Jabr I, Mabey DC, et al. The development of an age- structured model for trachoma transmission dynamics, pathogenesis and control. PLoS Negl Trop Dis. 2009;3(6):e462.',\n",
              " \"Estimating the future impact of a multi-pronged intervention strategy on ocular disease sequelae caused by trachoma: a modeling study. M Gambhir, N C Grassly, M J Burton, A W Solomon, H R Taylor, D C Mabey, I M Blake, M G Basanez, Ophthalmic Epidemiol. 226Gambhir M, Grassly NC, Burton MJ, Solomon AW, Taylor HR, Mabey DC, Blake IM, Basanez MG. Estimating the future impact of a multi-pronged intervention strategy on ocular disease sequelae caused by trachoma: a modeling study. Ophthalmic Epidemiol. 2015;22(6):394-402.\\n\\nPublisher's Note. Publisher's Note\\n\\nSpringer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations. Springer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations.\"]"
            ]
          },
          "execution_count": 18,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "### Extract Text from corpus ids\n",
        "seperated_text = lang_collect.get(\n",
        "    ids = ['234488766-1','234488766-2','234488766-3','234488766-4','234488766-5','234488766-6'],\n",
        "    include=[\"embeddings\", \"documents\"]\n",
        ").get('documents')\n",
        "seperated_text"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 21,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'Training, data collection, and clinical assessment Before each survey round, which occurred approximately twice per year, all trachoma graders (integrated eye care workers) participated in a 5 to 7-day training [12,13,16]. Graders were required to attend training regardless of the number or date of their previous trainings. Training methods used for these surveys pre-date but were very similar to those used in more current trachoma surveys [17]. Training consisted of classroom sessions to understand the definitions of all 5 World Health Organization (WHO) simplified trachoma signs through slide-deck review, and field-based sessions focused on providing graders experience in the range of presentations trachoma signs can take [18]. Graders were then required to pass a slide-based exam, which included images of all 5 simplified trachoma signs, including approximately 5 images of TS. Graders who passed the slide exam then moved onto a field reliability exam where they were required to grade 50 eyes under field conditions. The \"gold standard\" for the field reliability exam was the consensus grade of 3 expert graders, typically ophthalmologists, ocular surgeons, or trachoma managers [16,19]. Those graders who scored an interobserver agreement of ≥ 84 % and kappa ≥ 0.7, compared to the consensus grade of the 3 expert graders for the sign TF joined the survey team [16]. Although the field reliability exam was based on grading TF, graders were also trained on how to identify and grade TS using slide training, slide examination, and field-based practice within TT surgery campaigns that were specifically planned to accommodate the graders\\' practice for these surveys.\\n\\nThe primary aim of these surveys was to estimate TF among children ages 1 to 9 years, but all study participants were examined for all 5 WHO simplified signs of trachoma, including TS. Both eyes of all study participants were examined for TS, which was defined as the presence of scarring in the tarsal conjunctiva (scars are easily visible as white lines, bands, or sheets [fibrosis] in the tarsal conjunctiva) [18]. Trachoma grading was performed using binocular loupes with x2.5 magnification under sunlight or using a flashlight. Multiple grading teams worked in each district, and supervisors traveling with grading teams spot-checked grading, particularly during the first 2 weeks of each survey.\\n\\nData collection took place on Samsung Galaxy tablets running custom-built software [20]. Data recorders were information technology or biostatistics college graduates who took a 4 to 5-day training and passed a data collection exam using the tablets. All household surveys were conducted in the Amharic language.\\n\\n\\nStatistical analysis\\n\\nAll TS prevalence estimates were weighted to account for the probability of sampling at both sampling stages. Furthermore, all TS prevalence estimates reported, district-and region-level, were age-sex adjusted with post-stratification weighting using 5-year age-sex bands from the survey census population [21,22]. Confidence intervals (CI) were calculated using Taylor linearization and survey procedures (svy package) in Stata, version 15.1 (STATA Corporation, College Station TX, USA), to account for household and village level clustering. Most analyses were focused on individuals ages 15 years and older (standard age group used to estimate TT prevalence), individuals younger than 15 years, and individuals ages 1 to 9 years (standard age group for estimating TF prevalence) [8]. The prevalence of TS concurrent with TF and TI was also estimated. Chi-squared tests were used to evaluate TS prevalence between age group and sex. A Spearman correlation test was performed to examine relationships between TS and TT, TS and TF and TS and TI at the district level. Clustering by household and survey cluster was assessed by calculating the intraclass correlation coefficient (ICC) of TS prevalence with a random-effects model. District-level prevalence was grouped into the categories of < 5 %; 5.0-9.9 %; 10-14.9 %; > 15 % based on the overall TS prevalence distribution.\\n\\n\\nResults\\n\\nA total of 276,068 individuals were surveyed between 2010 and 2015 in all 152 programmatic districts across the 10 zones of the Amhara region. Among those enumerated, 75.5 % (208,510) of individuals were present and examined for clinical signs of trachoma. The majority of examined individuals, 52.9 % (110,211 respondents), were ages 15 years and older, 35.2 % (73,406 respondents) were children ages 1 to 9 years, and 53.5 % (111,577 respondents) were female.\\n\\nThe regional TS prevalence among all ages was 8.2 % (95 % CI: 7.7-8.6 %), and the TS prevalence among individuals ages 15 years and older was 12.6 % (95 % CI: 12.0-13.3 %). The range of TS prevalence by district (n = 152) in this age group was right-skewed and ranged from 0.6 % (95 % CI: 0.1-3.4 %) to 36.9 % (95 % CI: 27.2-47.8) (Fig. 1). Each of the 10 zones had at least one district with a TS prevalence in the highest category (TS ≥ 15 %), and Waghemra and South Gondar in the central part of the region had the heaviest burden of the high prevalence districts (Fig. 2). The district prevalence of TS was moderately correlated with the prevalence of TT (Spearman correlation [r] = 0.31, P < 0.001 (Fig. 3)).\\n\\nAmong individuals, the prevalence of TS increased significantly with age (P < 0.0001); the highest prevalence was observed in adults ages 56 to 60 years ( (Fig. 4). Among individuals ages 15 years and older, TS prevalence among females and males was 15.2 % (95 % CI: 11.5-19.7 %) and 14.6 (95 % CI: 12.6-16.9 %), respectively; the difference was not statistically significant (P = 0.73). Among individuals with TT in this age group, 955 (23.3 %) individuals had concurrent TS, and the individual-level correlation between TT and TS was low (r = 0.0275; P < 0.001).There was some evidence of clustering at the household level with an ICC of 0.24 (95 % CI: 0.23-0.26) and little evidence of clustering at the village level with an ICC of 0.08 (95 % CI: 0.07-0.10). At the household level, 79.0 % of households did not have any adults with TS, but in 9.3 % of households all of the adults had TS. The remaining 11.7 % of households had at least one but not all adults with TS.\\n\\nThe regional TS prevalence among individuals ages 1 to 15 years and individuals ages 1 to 9 years was 2.2 % (95 % CI: 1.8-2.8 %) and 1.1 % (95 % CI: 1.0-1.2 %) respectively. Moderate correlations were observed between district prevalence TS and district prevalence of TF (r = 0.33, P < 0.0001) and TI (r = 0.52, P < 0.0001) among individuals ages 1 to 15 years. Among individuals ages 1 to 15 years, the prevalence of TS increased with age (P < 0.001) reaching as high as 10.6 % in 15-yearold males (Fig. 5). Overall, however, the TS prevalence did not differ by sex (P = 0.40) in this age group. Among individuals ages 1 to 15 years with TS, 24.3 % also had TF and 4.8 % also had TI. In this age group, the prevalence of concurrent TS and active trachoma (TF and/or TI) was 0.6 % (95 % CI: 0.4-0.9 %), while the prevalence of concurrent TS and TF was 0.2 % (95 % CI: 0.04-0.9 %) (Fig. 6).\\n\\n\\nDiscussion Discussion\\n\\nThe results of these 152 population-based surveys suggest that the Amhara region has had a long history of intense trachoma exposure, and that a large population remains at risk for developing TT. Future monitoring of incident TT will likely be needed even after the elimination as a public health problem threshold for TT has     [13]. Furthermore, as part of TT surgical campaigns throughout the region, over 408,000 surgeries were conducted on TT patients between 2007 and 2015 [13]. Despite this, impact surveys conducted after these interventions demonstrated that many districts still had high levels of TF and C. trachomatis infection and that all but 4 districts were still above the elimination as a public health problem threshold for TT [12,13]. This continued high prevalence of TT observed in many districts may in part be due to incident TT as a result of years of progressive scarring. We found that districts with high levels of TT were more likely to have a concurrently high level of TS. Future surgical services will most likely be needed the most in districts with a high TT burden with concurrently high TS prevalence.\\n\\nIn trachoma-endemic communities, scarring starts to develop in childhood and continues to progress with age. Scarring is thought to be driven by repeated infection with C. trachomatis, whereby a higher cumulative burden of infection would lead to increased scarring [24]. It is also believed that scarring may be a result of persistent inflammation [10]. Evidence for this mechanism comes from cohort studies by which progressive scarring was observed in the absence of C. trachomatis infection [10,25]. Within adults with scarring in Amhara, 23 % had observed scarring progression over a 2-year period [10]. Within our surveys, we found that the prevalence of TS increased with age, reaching as high as a quarter of the population among those ages 60 to 64 years. Given the size of Amhara, this represents a large population at risk for scarring progression and incident TT. The Trachoma Control Program in Amhara and in other hyperendemic regions will need to develop surveillance systems to detect and operate incident TT cases. Given the within-household clustering of scarring observed in this report, the Program may expect incident TT to cluster within households as well. All communities within Amhara have received on average 5 years of the SAFE strategy including annual MDA with azithromycin. Azithromycin is an antibiotic shown to be effective against C. trachomatis and thought to also have antiinflammatory properties [26,27]. It is possible therefore that annual azithromycin distribution will reduce the progression of scarring and possibly the incidence of TT in this population.\\n\\nIn this trachoma endemic setting, TS was detectable in children, although at relatively low levels (≤ 2 %). This low TS prevalence may have been due to the annual antibiotic distribution in communities throughout the region over the last 5 years. Although pre-SAFE TS data from children in Amhara has not been previously published, in neighboring trachoma hyperendemic South Sudan, the prevalence of TS prior to the start of the SAFE strategy was 19.3 % among children ages 1 to 9 years and 31.7 % among children ages 10 to 14 years, measured using the same WHO simplified system [28]. In Guinea Bissau, a country where the TF prevalence pre-SAFE would be considered mesoendemic (22 %), the prevalence of TS among children ages 0 to 5 years and ages 6 to 10 years was 2.7 and 2.8 % respectively [29]. Although comparisons across various settings is difficult due to differences in the assessment of scarring, it appears that the level of scarring observed in Amhara among children is not reflective of a trachoma hyperendemic setting. However, children are clearly still at risk for developing scarring in Amhara since the prevalence of TF, TI, and C. trachomatis infection among children ages 1 to 9 years was still 25.9, 5.5, and 5.7 % respectively, and a statistically significant association was observed between age and scarring, even in this limited age range [12,13]. Prior studies have demonstrated that the presence of TI is associated with the incidence and progression of scarring [4,10,24,30,31]. We observed that approximately 25 % of children with TS also had TF, and nearly 5 % also had TI, therefore this subset in Amhara may be at particular risk for scarring progression. Although longitudinal studies are needed to know the degree to which scarring is still developing and progressing in Amhara, clearly, further enhanced efforts are needed to reduce the levels of inflammatory trachoma in these communities in order to reducing the burden of future scarring.\\n\\nThese surveys had several limitations. Survey teams graded TS in the field using the WHO simplified grading scheme, and therefore it was not possible to grade scarring severity. As graders were trained to identify bands or sheets that were \"easily visible,\" it is likely that mild scarring would be missed in this definition. Therefore, our TS estimates may have underestimated the total scarring burden in Amhara, and the correlation between TT and TS may have been attenuated. Evidence for this comes from a study in Amhara showing that 70 % of participants without TT, and 95 % of those with TT had had some degree of scarring using a detailed grading scheme [32]. In research settings, grading using conjunctival photographs has been common and could possibly be considered for programmatic use to allow for more granular grading of scarring. Although the field reliability test was focused on the trachoma sign TF, TS grading was integrated with the survey training, and the trainees had a chance to practice in the field along with TT identification practice that was integrated with TT surgery campaigns. This study was cross-sectional, and therefore could not measure incident or progression of TS, which would be useful indicators for trachoma programs. Currently programmatic monitoring globally no longer includes TS prevalence as an indicator, and longitudinal studies focused on scarring are uncommon [3,7,10,14,24]. The inclusion of TS estimation within programmatic surveys should be reconsidered as scarring data has recently been used to better understand the history of trachoma within treatment-naïve endemic regions and has further been used to help make MDA treatment decisions under select circumstances [6,33]. More data on scarring can also provide modelers with empirical data to improve models for the natural history of trachoma and in forecasting incident TT [34,35].\\n\\nOverall, a considerable percentage of adults have TS in these trachoma-endemic districts in Amhara, which suggests both a long history of C. trachomatis infection and the possibility of continued development of incident cases of TT in the future. Including TS data in the surveillence of trachoma may be useful for the global elimination program. Trachoma programs serving formerly highly endemic areas should continue to develop context-appropriate monitoring systems for incident TT for the foreseeable future. \\n\\nFig. 1\\n1District-level TS prevalence among individuals ages 15 years and older, Amhara, Ethiopia, 2010 to 2015 Fig. 2 District-level TS prevalence among individuals ages 15 years and older, Amhara, Ethiopia, 2010 to 2015\\n\\nFig. 3\\n3Scatterplot of district-level TS and TT prevalence, Amhara, Ethiopia, 2010 to 2015. N=152 districts, r=0.31\\n\\nFig. 4\\n4Prevalence of TS by sex and age group, Amhara, Ethiopia, 2010 to 2015\\n\\nFig. 5\\n5Prevalence of TS by sex and ages 1 to 15 years, Amhara, Ethiopia, 2010 to 2015 Fig. 6 Active trachoma with TS prevalence in ages 1 to 15, Amhara, Ethiopia, 2010 to 2015\\n\\n\\nAbbreviations CI: Confidence interval; ICC: Intraclass correlation coefficient; MDA: Mass drug administration; SAFE: Surgery, Antibiotic, Facial cleanliness, and Environmental improvement; R: Spearman correlation; TF: Trachomatous inflammationfollicular; TI: Trachomatous inflammation-intense; TS: Trachomatous scarring; TT: Trachomatous trichiasis; WHO: World Health Organization. These results led to the scale-up of the full SAFE strategy in all districts of Amhara. At scale, the Trachoma Control Program was delivering approximately 15 million doses of antibiotic per year as part of its MDA campaignsAstale et al. BMC Ophthalmology \\n(2021) 21:213 been reached. It is promising, however, that children \\nages 1 to 15 years, many born after the MDA program \\nbegan, have a TS prevalence of 2 %. Currently, program-\\nmatic monitoring of scarring is rare, and TS prevalence \\nas a district-level indicator is no longer included in most \\ntrachoma survey methodologies. Future longitudinal re-\\nsearch is needed to determine whether scarring data can \\nhelp programs plan future needs for TT services. \\nEarly district-level surveys conducted in Amhara in \\n2001 and 2003 demonstrated that trachoma was hyper-\\nendemic in 5 pilot districts, with a TF prevalence among \\nchildren ages 1 to 9 years ranging between 49 and 90 % \\nand TT among individuals ages 15 years and older ran-\\nging between 2.0 and 5.4 % [15]. In 2006, a zonal-level \\nbaseline trachoma survey demonstrated both a high TF \\n(32.7 %) and TT (6.2 %) prevalence region-wide [23]. \\n\\nAcknowledgementsWe are grateful to all study participants who volunteered their time to participate in the study. We appreciate and thank the dedicated field teams, drivers and supervisors who persevered through difficult conditions to compete the data collection process, many of whom participated in all surveys. We are also grateful for the collaboration with the Lions-Carter Center Sight-First Initiative.Authors\\' contributionsFundingFunding for the survey in summer 2013 was provided by USAIDs ENVISION project led by RTI International.Availability of data and materials All data generated or analyzed during this study are included in this published article.DeclarationsEthics approval and consent to participate Survey protocols were approved by Amhara National Regional State Health Bureau, and by Emory University Institutional Review Board under protocol 079-2006. All methods were carried out in accordance with relevant national guidelines and regulations. The study objectives and procedures were explained to the head of households or adult respondents. Due to high illiteracy rate among the population, Institutional Review Board approval was obtained for informed oral consent or assent. Informed oral consent or assent was obtained according to the principles of the Declaration of Helsinki.Consent for publicationNot applicable.Competing interestsThe authors declare that they have no competing interests.\\nBacterial infection in scarring trachoma. V H Hu, P Massae, H A Weiss, C Chevallier, J J Onyango, I A Afwamba, D C Mabey, R L Bailey, M J Burton, Invest Ophthalmol Vis Sci. 525Hu VH, Massae P, Weiss HA, Chevallier C, Onyango JJ, Afwamba IA, Mabey DC, Bailey RL, Burton MJ. Bacterial infection in scarring trachoma. Invest Ophthalmol Vis Sci. 2011;52(5):2181-6.\\n\\nIn vivo confocal microscopy of trachoma in relation to normal tarsal conjunctiva. V H Hu, P Massae, H A Weiss, I A Cree, P Courtright, D C Mabey, R L Bailey, M J Burton, Ophthalmology. 1184Hu VH, Massae P, Weiss HA, Cree IA, Courtright P, Mabey DC, Bailey RL, Burton MJ. In vivo confocal microscopy of trachoma in relation to normal tarsal conjunctiva. Ophthalmology. 2011;118(4):747-54.\\n\\nIncidence estimates of late stages of trachoma among women in a hyperendemic area of central Tanzania. B Munoz, Aron J Turner, V West, S , Trop Med Int Health. 211Munoz B, Aron J, Turner V, West S. Incidence estimates of late stages of trachoma among women in a hyperendemic area of central Tanzania. Trop Med Int Health. 1997;2(11):1030-8.\\n\\nProgression of active trachoma to scarring in a cohort of Tanzanian children. S K West, B Munoz, H Mkocha, Y H Hsieh, M C Lynch, Ophthalmic Epidemiol. 82-3West SK, Munoz B, Mkocha H, Hsieh YH, Lynch MC. Progression of active trachoma to scarring in a cohort of Tanzanian children. Ophthalmic Epidemiol. 2001;8(2-3):137-44.\\n\\nAge, sex, and cohort effects in a longitudinal study of trachomatous scarring. M A Wolle, B Munoz, H Mkocha, S K West, Invest Ophthalmol Vis Sci. 502Wolle MA, Munoz B, Mkocha H, West SK. Age, sex, and cohort effects in a longitudinal study of trachomatous scarring. Invest Ophthalmol Vis Sci. 2009;50(2):592-6.\\n\\nClinical signs of trachoma are prevalent among Solomon Islanders who have no persistent markers of prior infection with Chlamydia trachomatis. R Butcher, O Sokana, K Jack, L Sui, C Russell, A Last, D L Martin, M J Burton, A W Solomon, Dcw Mabey, Wellcome Open Res. 314Butcher R, Sokana O, Jack K, Sui L, Russell C, Last A, Martin DL, Burton MJ, Solomon AW, Mabey DCW, et al. Clinical signs of trachoma are prevalent among Solomon Islanders who have no persistent markers of prior infection with Chlamydia trachomatis. Wellcome Open Res. 2018;3:14.\\n\\nProgression of scarring trachoma in Tanzanian children: a four-year cohort study. A M Ramadhani, T Derrick, D Macleod, P Massae, E Mafuru, A Malisa, K Mbuya, C H Roberts, W Makupa, T Mtuy, PLoS Negl Trop Dis. 1387638Ramadhani AM, Derrick T, Macleod D, Massae P, Mafuru E, Malisa A, Mbuya K, Roberts CH, Makupa W, Mtuy T, et al. Progression of scarring trachoma in Tanzanian children: a four-year cohort study. PLoS Negl Trop Dis. 2019;13(8): e0007638.\\n\\nWHO alliance for the global elimination of trachoma by 2020: progress report on elimination of trachoma. Wkly Epidemiol Rec. 9429WHO. WHO alliance for the global elimination of trachoma by 2020: progress report on elimination of trachoma, 2018. Wkly Epidemiol Rec. 2019;94(29):317-28.\\n\\nThe prevalence of trachomatous trichiasis in Oman (Oman eye study. R Khandekar, A J Mohammed, Ophthalmic Epidemiol. 145Khandekar R, Mohammed AJ. The prevalence of trachomatous trichiasis in Oman (Oman eye study 2005). Ophthalmic Epidemiol. 2007;14(5):267-72.\\n\\nPathogenesis of progressive scarring trachoma in Ethiopia and Tanzania and its implications for disease control: two cohort studies. M J Burton, S N Rajak, V H Hu, A Ramadhani, E Habtamu, P Massae, Z Tadesse, K Callahan, P M Emerson, P T Khaw, PLoS Negl Trop Dis. 953763Burton MJ, Rajak SN, Hu VH, Ramadhani A, Habtamu E, Massae P, Tadesse Z, Callahan K, Emerson PM, Khaw PT, et al. Pathogenesis of progressive scarring trachoma in Ethiopia and Tanzania and its implications for disease control: two cohort studies. PLoS Negl Trop Dis. 2015;9(5):e0003763. Risk factors for incidence of trachomatous scarring in a cohort of women in low endemic district. R Karani, M Wolle, H Mkocha, B Munoz, S K West, Br J Ophthalmol. 1024Karani R, Wolle M, Mkocha H, Munoz B, West SK. Risk factors for incidence of trachomatous scarring in a cohort of women in low endemic district. Br J Ophthalmol. 2018;102(4):419-23.\\n\\nOcular chlamydia trachomatis infection under the surgery, antibiotics, facial cleanliness, and environmental improvement strategy in Amhara. S D Nash, Aep Stewart, M Zerihun, E Sata, D Gessese, B Melak, T Endeshaw, M Chanyalew, A Chernet, B Bayissasse, Clin Infect Dis. 6712Nash SD, Stewart AEP, Zerihun M, Sata E, Gessese D, Melak B, Endeshaw T, Chanyalew M, Chernet A, Bayissasse B, et al. Ocular chlamydia trachomatis infection under the surgery, antibiotics, facial cleanliness, and environmental improvement strategy in Amhara, Ethiopia, 2011-2015. Clin Infect Dis. 2018; 67(12):1840-1846.\\n\\nProgress to eliminate trachoma as a public health problem in Amhara National Regional State, Ethiopia: results of 152 population-based surveys. Aep Stewart, M Zerihun, D Gessese, B Melak, E Sata, A W Nute, T Astale, T Endeshaw, T Teferi, Z Tadesse, Am J Trop Med Hyg. 1016Stewart AEP, Zerihun M, Gessese D, Melak B, Sata E, Nute AW, Astale T, Endeshaw T, Teferi T, Tadesse Z, et al. Progress to eliminate trachoma as a public health problem in Amhara National Regional State, Ethiopia: results of 152 population-based surveys. Am J Trop Med Hyg. 2019;101(6):1286-95.\\n\\nNatural history of trachomatous scarring in The Gambia: results of a 12-year longitudinal follow-up. R J Bowman, B Jatta, B Cham, R L Bailey, H Faal, M Myatt, A Foster, G J Johnson, Ophthalmology. 10812Bowman RJ, Jatta B, Cham B, Bailey RL, Faal H, Myatt M, Foster A, Johnson GJ. Natural history of trachomatous scarring in The Gambia: results of a 12- year longitudinal follow-up. Ophthalmology. 2001;108(12):2219-24.\\n\\nEvaluation of three years of the SAFE strategy (Surgery, Antibiotics, Facial cleanliness and Environmental improvement) for trachoma control in five districts of Ethiopia hyperendemic for trachoma. J Ngondi, T Gebre, E B Shargie, L Adamu, Y Ejigsemahu, T Teferi, M Zerihun, B Ayele, V Cevallos, J King, Trans R Soc Trop Med Hyg. 10310Ngondi J, Gebre T, Shargie EB, Adamu L, Ejigsemahu Y, Teferi T, Zerihun M, Ayele B, Cevallos V, King J. Evaluation of three years of the SAFE strategy (Surgery, Antibiotics, Facial cleanliness and Environmental improvement) for trachoma control in five districts of Ethiopia hyperendemic for trachoma. Trans R Soc Trop Med Hyg. 2009;103(10):1001-10.\\n\\nPrevalence of trachoma at subdistrict level in Ethiopia: determining when to stop mass azithromycin distribution. J D King, T Teferi, E A Cromwell, M Zerihun, J M Ngondi, M Damte, F Ayalew, Z Tadesse, T Gebre, A Mulualem, PLoS Negl Trop Dis. 832732King JD, Teferi T, Cromwell EA, Zerihun M, Ngondi JM, Damte M, Ayalew F, Tadesse Z, Gebre T, Mulualem A, et al. Prevalence of trachoma at sub- district level in Ethiopia: determining when to stop mass azithromycin distribution. PLoS Negl Trop Dis. 2014;8(3):e2732.\\n\\nThe global trachoma mapping project: methodology of a 34-country population-based study. A W Solomon, A L Pavluck, P Courtright, A Aboe, L Adamu, W Alemayehu, M Alemu, N D Alexander, A B Kello, B Bero, Ophthalmic Epidemiol. 223Solomon AW, Pavluck AL, Courtright P, Aboe A, Adamu L, Alemayehu W, Alemu M, Alexander ND, Kello AB, Bero B, et al. The global trachoma mapping project: methodology of a 34-country population-based study. Ophthalmic Epidemiol. 2015;22(3):214-25.\\n\\nA simple system for the assessment of trachoma and its complications. B Thylefors, C R Dawson, B R Jones, S West, H R Taylor, Bull World Health Org. 654477Thylefors B, Dawson CR, Jones BR, West S, Taylor HR. A simple system for the assessment of trachoma and its complications. Bull World Health Org. 1987;65(4):477.\\n\\nTrachoma prevalence remains below threshold in five districts after stopping mass drug administration: results of five surveillance surveys within a hyperendemic setting in Amhara. S D Nash, Aep Stewart, T Astale, E Sata, M Zerihun, D Gessese, B Melak, G Ayenew, Z Ayele, B Bayissasse, Ethiopia. Trans R Soc Trop Med Hyg. 11212Nash SD, Stewart AEP, Astale T, Sata E, Zerihun M, Gessese D, Melak B, Ayenew G, Ayele Z, Bayissasse B, et al. Trachoma prevalence remains below threshold in five districts after stopping mass drug administration: results of five surveillance surveys within a hyperendemic setting in Amhara, Ethiopia. Trans R Soc Trop Med Hyg. 2018;112(12):538-45.\\n\\nA novel electronic data collection system for large-scale surveys of neglected tropical diseases. J D King, J Buolamwini, E A Cromwell, A Panfel, T Teferi, M Zerihun, B Melak, J Watson, Z Tadesse, D Vienneau, PloS One. 8974570King JD, Buolamwini J, Cromwell EA, Panfel A, Teferi T, Zerihun M, Melak B, Watson J, Tadesse Z, Vienneau D, et al. A novel electronic data collection system for large-scale surveys of neglected tropical diseases. PloS One. 2013;8(9):e74570.\\n\\nBurden of trachoma in five counties of eastern Equatoria state, south Sudan: results from population-based surveys. A M Sanders, Aep Stewart, S Makoy, J J Chebet, P Magok, A Kuol, C Blauvelt, R Lako, J Rumunu, E K Callahan, PLoS Negl Trop Dis. 1165658Sanders AM, Stewart AEP, Makoy S, Chebet JJ, Magok P, Kuol A, Blauvelt C, Lako R, Rumunu J, Callahan EK, et al. Burden of trachoma in five counties of eastern Equatoria state, south Sudan: results from population-based surveys. PLoS Negl Trop Dis. 2017;11(6):e0005658. Progress toward elimination of trachoma as a public health problem in seven localities in the republic of Sudan: results from population-based surveys. A M Sanders, Z Abdalla, B E Elshafie, M Elsanosi, A W Nute, N Aziz, E K Callahan, S D Nash, Am J Trop Med Hyg. 1016Sanders AM, Abdalla Z, Elshafie BE, Elsanosi M, Nute AW, Aziz N, Callahan EK, Nash SD. Progress toward elimination of trachoma as a public health problem in seven localities in the republic of Sudan: results from population-based surveys. Am J Trop Med Hyg. 2019;101(6):1296-302.\\n\\nIntegrating an NTD with one of \"The big three\": combined malaria and trachoma survey in Amhara Region of Ethiopia. P M Emerson, J Ngondi, E Biru, P M Graves, Y Ejigsemahu, T Gebre, T Endeshaw, A Genet, A W Mosher, M Zerihun, PLoS Negl Trop Dis. 23197Emerson PM, Ngondi J, Biru E, Graves PM, Ejigsemahu Y, Gebre T, Endeshaw T, Genet A, Mosher AW, Zerihun M, et al. Integrating an NTD with one of \"The big three\": combined malaria and trachoma survey in Amhara Region of Ethiopia. PLoS Negl Trop Dis. 2008;2(3):e197.\\n\\nConstant ocular infection with Chlamydia trachomatis predicts risk of scarring in children in Tanzania. M A Wolle, B E Munoz, H Mkocha, S K West, Ophthalmology. 1162Wolle MA, Munoz BE, Mkocha H, West SK. Constant ocular infection with Chlamydia trachomatis predicts risk of scarring in children in Tanzania. Ophthalmology. 2009;116(2):243-7.\\n\\nIncidence and progression of trachomatous scarring in a cohort of children in a formerly hyper-endemic district of Tanzania. M S Kashaf, B E Munoz, H Mkocha, M A Wolle, F Naufal, S K West, PLoS Negl Trop Dis. 14108708Kashaf MS, Munoz BE, Mkocha H, Wolle MA, Naufal F, West SK. Incidence and progression of trachomatous scarring in a cohort of children in a formerly hyper-endemic district of Tanzania. PLoS Negl Trop Dis. 2020; 14(10):e0008708.\\n\\nRandomised controlled trial of single-dose azithromycin in treatment of trachoma. R L Bailey, P Arullendran, H C Whittle, D C Mabey, Lancet. 3428869Bailey RL, Arullendran P, Whittle HC, Mabey DC. Randomised controlled trial of single-dose azithromycin in treatment of trachoma. Lancet. 1993; 342(8869):453-6.\\n\\nAnti-inflammatory effects of macrolides-an underappreciated benefit in the treatment of community-acquired respiratory tract infections and chronic inflammatory pulmonary conditions?. G W Amsden, J Antimicrob Chemother. 551Amsden GW. Anti-inflammatory effects of macrolides-an underappreciated benefit in the treatment of community-acquired respiratory tract infections and chronic inflammatory pulmonary conditions? J Antimicrob Chemother. 2005;55(1):10-21.\\n\\nThe burden of trachoma in Ayod County of Southern Sudan. J D King, J Ngondi, G Gatpan, B Lopidia, S Becknell, P M Emerson, PLoS Negl Trop Dis. 29299King JD, Ngondi J, Gatpan G, Lopidia B, Becknell S, Emerson PM. The burden of trachoma in Ayod County of Southern Sudan. PLoS Negl Trop Dis. 2008;2(9):e299.\\n\\nRisk factors for active trachoma and ocular Chlamydia trachomatis infection in treatment-naive trachomahyperendemic communities of the Bijagos Archipelago. A R Last, S E Burr, H A Weiss, E M Harding-Esch, E Cassama, M Nabicassa, D C Mabey, M J Holland, R L Bailey, PLoS Negl Trop Dis. 862900Last AR, Burr SE, Weiss HA, Harding-Esch EM, Cassama E, Nabicassa M, Mabey DC, Holland MJ, Bailey RL. Risk factors for active trachoma and ocular Chlamydia trachomatis infection in treatment-naive trachoma- hyperendemic communities of the Bijagos Archipelago, Guinea Bissau. PLoS Negl Trop Dis. 2014;8(6):e2900.\\n\\nWhat clinical signs are critical in evaluating the intervention in trachoma. C R Dawson, Mr, T Daghfous, R Juster, J Schachter, Chlamyidal Infections. WR BCambridgeCambridge Universtify PressDawson CR MR, Daghfous T, Juster R, Schachter J. What clinical signs are critical in evaluating the intervention in trachoma? In: WR B, editor. Chlamyidal Infections. Cambridge: Cambridge Universtify Press; 1990. p. 271-278.\\n\\nBlinding trachoma: systematic review of rates and risk factors for progressive disease. A M Ramadhani, T Derrick, M J Holland, M J Burton, PLoS Negl Trop Dis. 1084859Ramadhani AM, Derrick T, Holland MJ, Burton MJ. Blinding trachoma: systematic review of rates and risk factors for progressive disease. PLoS Negl Trop Dis. 2016;10(8):e0004859.\\n\\nWorld Health Organization. WHO. Report of the 4th global scientific meeting on trachoma. Geneva; GenevaWHO. Report of the 4th global scientific meeting on trachoma, Geneva, 27- 29 November 2018. Geneva: World Health Organization; 2019.\\n\\nConjunctival scarring, corneal pannus and Herbert\\'s pits in adolescent children in trachoma-endemic populations of the Solomon Islands and Vanuatu. R Butcher, J Tagabasoe, J Manemaka, A Bong, M Garae, L Daniel, C Roberts, B L Handley, V H Hu, E M Harding-Esch, Clin Infect Dis. Epub ahead of printButcher R, Tagabasoe J, Manemaka J, Bong A, Garae M, Daniel L, Roberts C, Handley BL, Hu VH, Harding-Esch EM, et al. Conjunctival scarring, corneal pannus and Herbert\\'s pits in adolescent children in trachoma-endemic populations of the Solomon Islands and Vanuatu. Clin Infect Dis. 2020. [Epub ahead of print].\\n\\nThe development of an agestructured model for trachoma transmission dynamics, pathogenesis and control. M Gambhir, M G Basanez, M J Burton, A W Solomon, R L Bailey, M J Holland, I M Blake, C A Donnelly, I Jabr, D C Mabey, PLoS Negl Trop Dis. 36462Gambhir M, Basanez MG, Burton MJ, Solomon AW, Bailey RL, Holland MJ, Blake IM, Donnelly CA, Jabr I, Mabey DC, et al. The development of an age- structured model for trachoma transmission dynamics, pathogenesis and control. PLoS Negl Trop Dis. 2009;3(6):e462. Estimating the future impact of a multi-pronged intervention strategy on ocular disease sequelae caused by trachoma: a modeling study. M Gambhir, N C Grassly, M J Burton, A W Solomon, H R Taylor, D C Mabey, I M Blake, M G Basanez, Ophthalmic Epidemiol. 226Gambhir M, Grassly NC, Burton MJ, Solomon AW, Taylor HR, Mabey DC, Blake IM, Basanez MG. Estimating the future impact of a multi-pronged intervention strategy on ocular disease sequelae caused by trachoma: a modeling study. Ophthalmic Epidemiol. 2015;22(6):394-402.\\n\\nPublisher\\'s Note. Publisher\\'s Note\\n\\nSpringer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations. Springer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations.'"
            ]
          },
          "execution_count": 21,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "### Combine Seperated Text to input into models\n",
        "def combine_full_text_embeddings(text_embedding):\n",
        "    final_doc = []\n",
        "    for i in range(len(text_embedding)):\n",
        "        final_doc.append(text_embedding[i])\n",
        "    return \" \".join(final_doc)\n",
        "full_text_example1 = combine_full_text_embeddings(seperated_text)\n",
        "full_text_example1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 982
        },
        "id": "VEhFwE_UT2kh",
        "outputId": "107fbcf3-b777-4b0c-c48a-25aa6133abfb"
      },
      "outputs": [
        {
          "data": {
            "text/html": [
              "\n",
              "  <div id=\"df-2436cb5c-0caa-439f-a031-e7f14c616c22\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th>corpusid</th>\n",
              "      <th>externalids</th>\n",
              "      <th>url</th>\n",
              "      <th>title</th>\n",
              "      <th>authors</th>\n",
              "      <th>venue</th>\n",
              "      <th>publicationvenueid</th>\n",
              "      <th>year</th>\n",
              "      <th>referencecount</th>\n",
              "      <th>citationcount</th>\n",
              "      <th>influentialcitationcount</th>\n",
              "      <th>isopenaccess</th>\n",
              "      <th>s2fieldsofstudy</th>\n",
              "      <th>publicationtypes</th>\n",
              "      <th>publicationdate</th>\n",
              "      <th>journal</th>\n",
              "      <th>updated</th>\n",
              "      <th>id</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0</td>\n",
              "      <td>52273203</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/39063bcd...</td>\n",
              "      <td>Large conformational changes of a highly dynam...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"51285510\\\", \\\"name\\\": \\\"Isa...</td>\n",
              "      <td>Communications Biology</td>\n",
              "      <td>069e05e7-ca35-41d0-a8c7-bdc9ec6a82af</td>\n",
              "      <td>2018.0</td>\n",
              "      <td>28</td>\n",
              "      <td>13</td>\n",
              "      <td>0</td>\n",
              "      <td>True</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...</td>\n",
              "      <td>{JournalArticle}</td>\n",
              "      <td>2018-09-03</td>\n",
              "      <td>{\"name\": \"Communications Biology\", \"pages\": nu...</td>\n",
              "      <td>2023-03-31T01:28:37.897Z</td>\n",
              "      <td>5692596</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1</td>\n",
              "      <td>23708908</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/6ec7c156...</td>\n",
              "      <td>Drug sensitivity profiling identifies potentia...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"5699001\\\", \\\"name\\\": \\\"H. K...</td>\n",
              "      <td>OncoTarget</td>\n",
              "      <td>106ecc86-1801-4b36-a05c-d36db69a968c</td>\n",
              "      <td>2017.0</td>\n",
              "      <td>46</td>\n",
              "      <td>24</td>\n",
              "      <td>0</td>\n",
              "      <td>True</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...</td>\n",
              "      <td>{JournalArticle}</td>\n",
              "      <td>2017-10-31</td>\n",
              "      <td>{\"name\": \"Oncotarget\", \"pages\": \"97516 - 97527...</td>\n",
              "      <td>2023-04-05T06:18:15.837Z</td>\n",
              "      <td>5692609</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>2</td>\n",
              "      <td>13232625</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/c8566272...</td>\n",
              "      <td>Men's memory for women's sexual-interest and r...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"4425040\\\", \\\"name\\\": \\\"T. T...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>2011.0</td>\n",
              "      <td>36</td>\n",
              "      <td>13</td>\n",
              "      <td>0</td>\n",
              "      <td>True</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Psychology\\\", \\\"source\\\": \\...</td>\n",
              "      <td>{Review}</td>\n",
              "      <td>2011-09-01</td>\n",
              "      <td>{\"name\": \"Applied Cognitive Psychology\", \"page...</td>\n",
              "      <td>2023-04-05T08:18:02.240Z</td>\n",
              "      <td>5692611</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>3</td>\n",
              "      <td>73484844</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/d4c9b2fa...</td>\n",
              "      <td>The Reactive Plasticity of Hippocampal Ionotro...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"2225106\\\", \\\"name\\\": \\\"A. M...</td>\n",
              "      <td>International Journal of Molecular Sciences</td>\n",
              "      <td>8506a01a-40b8-4e6f-bbb8-ce2492139c15</td>\n",
              "      <td>2019.0</td>\n",
              "      <td>97</td>\n",
              "      <td>15</td>\n",
              "      <td>1</td>\n",
              "      <td>True</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...</td>\n",
              "      <td>{Review,JournalArticle}</td>\n",
              "      <td>2019-02-27</td>\n",
              "      <td>{\"name\": \"International Journal of Molecular S...</td>\n",
              "      <td>2023-05-20T17:20:54.230Z</td>\n",
              "      <td>5692615</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>4</td>\n",
              "      <td>229159752</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/7f13abe2...</td>\n",
              "      <td>High Mortality Rate in Adult COVID-19 Inpatien...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"7685173\\\", \\\"name\\\": \\\"Saee...</td>\n",
              "      <td>Journal of Multidisciplinary Healthcare</td>\n",
              "      <td>12e5e550-28bd-401c-a348-f347956deec4</td>\n",
              "      <td>2020.0</td>\n",
              "      <td>37</td>\n",
              "      <td>17</td>\n",
              "      <td>1</td>\n",
              "      <td>True</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Medicine\\\", \\\"source\\\": \\\"s...</td>\n",
              "      <td>{JournalArticle,Review}</td>\n",
              "      <td>2020-12-01</td>\n",
              "      <td>{\"name\": \"Journal of Multidisciplinary Healthc...</td>\n",
              "      <td>2023-05-23T13:34:07.581Z</td>\n",
              "      <td>5692620</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>5</td>\n",
              "      <td>10041098</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/92b26dee...</td>\n",
              "      <td>NADH-dependent metabolism of nitric oxide in a...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"1792255\\\", \\\"name\\\": \\\"A. I...</td>\n",
              "      <td>Planta</td>\n",
              "      <td>66507f0b-12a1-4058-a35c-73029853d5b9</td>\n",
              "      <td>2004.0</td>\n",
              "      <td>30</td>\n",
              "      <td>145</td>\n",
              "      <td>3</td>\n",
              "      <td>False</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...</td>\n",
              "      <td>{JournalArticle}</td>\n",
              "      <td>2004-01-22</td>\n",
              "      <td>{\"name\": \"Planta\", \"pages\": \"95-102\", \"volume\"...</td>\n",
              "      <td>2023-04-05T15:01:41.273Z</td>\n",
              "      <td>5692625</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>6</td>\n",
              "      <td>19472375</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/2c682757...</td>\n",
              "      <td>Aspects of posttraumatic stress disorder in lo...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"2000616\\\", \\\"name\\\": \\\"A. D...</td>\n",
              "      <td>Journal of cancer survivorship</td>\n",
              "      <td>649b7693-cb10-4892-b8a2-65b9b51fcd50</td>\n",
              "      <td>2016.0</td>\n",
              "      <td>33</td>\n",
              "      <td>18</td>\n",
              "      <td>0</td>\n",
              "      <td>False</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Psychology\\\", \\\"source\\\": \\...</td>\n",
              "      <td>{JournalArticle}</td>\n",
              "      <td>2016-02-26</td>\n",
              "      <td>{\"name\": \"Journal of Cancer Survivorship\", \"pa...</td>\n",
              "      <td>2023-04-05T19:07:12.948Z</td>\n",
              "      <td>5692626</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>7</th>\n",
              "      <td>7</td>\n",
              "      <td>219603447</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/9b9659a4...</td>\n",
              "      <td>Multicenter assessment of the Brain Injury Gui...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"145776250\\\", \\\"name\\\": \\\"Ab...</td>\n",
              "      <td>Trauma Surgery &amp; Acute Care Open</td>\n",
              "      <td>d40f33e8-6f79-4e91-a62e-acf29f9a3e2f</td>\n",
              "      <td>2020.0</td>\n",
              "      <td>35</td>\n",
              "      <td>11</td>\n",
              "      <td>0</td>\n",
              "      <td>True</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Medicine\\\", \\\"source\\\": \\\"s...</td>\n",
              "      <td>{JournalArticle,Review}</td>\n",
              "      <td>2020-05-01</td>\n",
              "      <td>{\"name\": \"Trauma Surgery &amp; Acute Care Open\", \"...</td>\n",
              "      <td>2023-04-07T01:56:26.282Z</td>\n",
              "      <td>5692637</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>8</td>\n",
              "      <td>15826244</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/314f8d8a...</td>\n",
              "      <td>Mycobacterium haemophilum as the Initial Prese...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"2060804071\\\", \\\"name\\\": \\\"T...</td>\n",
              "      <td>Case Reports in Rheumatology</td>\n",
              "      <td>884ab217-3965-4a74-90d1-003d0b378d6d</td>\n",
              "      <td>2014.0</td>\n",
              "      <td>24</td>\n",
              "      <td>8</td>\n",
              "      <td>0</td>\n",
              "      <td>True</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Medicine\\\", \\\"source\\\": \\\"s...</td>\n",
              "      <td>{JournalArticle}</td>\n",
              "      <td>2014-01-12</td>\n",
              "      <td>{\"name\": \"Case Reports in Rheumatology\", \"page...</td>\n",
              "      <td>2023-04-07T01:15:17.667Z</td>\n",
              "      <td>5692638</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>9</th>\n",
              "      <td>9</td>\n",
              "      <td>2311509</td>\n",
              "      <td>{\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...</td>\n",
              "      <td>https://www.semanticscholar.org/paper/7a0b4601...</td>\n",
              "      <td>Up-regulation and subcellular localization of ...</td>\n",
              "      <td>{\"{\\\"authorId\\\": \\\"14726800\\\", \\\"name\\\": \\\"Hua...</td>\n",
              "      <td>BMC Cancer</td>\n",
              "      <td>b482a99f-ef53-4694-aec6-32ad78e0d716</td>\n",
              "      <td>2010.0</td>\n",
              "      <td>58</td>\n",
              "      <td>44</td>\n",
              "      <td>0</td>\n",
              "      <td>True</td>\n",
              "      <td>{\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...</td>\n",
              "      <td>{JournalArticle}</td>\n",
              "      <td>2010-07-06</td>\n",
              "      <td>{\"name\": \"BMC Cancer\", \"pages\": \"356 - 356\", \"...</td>\n",
              "      <td>2023-04-07T01:28:15.493Z</td>\n",
              "      <td>5692639</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2436cb5c-0caa-439f-a031-e7f14c616c22')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-2436cb5c-0caa-439f-a031-e7f14c616c22 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-2436cb5c-0caa-439f-a031-e7f14c616c22');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ],
            "text/plain": [
              "   Unnamed: 0   corpusid                                        externalids  \\\n",
              "0           0   52273203  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "1           1   23708908  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "2           2   13232625  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "3           3   73484844  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "4           4  229159752  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "5           5   10041098  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "6           6   19472375  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "7           7  219603447  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "8           8   15826244  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "9           9    2311509  {\"ACL\": null, \"DBLP\": null, \"ArXiv\": null, \"MA...   \n",
              "\n",
              "                                                 url  \\\n",
              "0  https://www.semanticscholar.org/paper/39063bcd...   \n",
              "1  https://www.semanticscholar.org/paper/6ec7c156...   \n",
              "2  https://www.semanticscholar.org/paper/c8566272...   \n",
              "3  https://www.semanticscholar.org/paper/d4c9b2fa...   \n",
              "4  https://www.semanticscholar.org/paper/7f13abe2...   \n",
              "5  https://www.semanticscholar.org/paper/92b26dee...   \n",
              "6  https://www.semanticscholar.org/paper/2c682757...   \n",
              "7  https://www.semanticscholar.org/paper/9b9659a4...   \n",
              "8  https://www.semanticscholar.org/paper/314f8d8a...   \n",
              "9  https://www.semanticscholar.org/paper/7a0b4601...   \n",
              "\n",
              "                                               title  \\\n",
              "0  Large conformational changes of a highly dynam...   \n",
              "1  Drug sensitivity profiling identifies potentia...   \n",
              "2  Men's memory for women's sexual-interest and r...   \n",
              "3  The Reactive Plasticity of Hippocampal Ionotro...   \n",
              "4  High Mortality Rate in Adult COVID-19 Inpatien...   \n",
              "5  NADH-dependent metabolism of nitric oxide in a...   \n",
              "6  Aspects of posttraumatic stress disorder in lo...   \n",
              "7  Multicenter assessment of the Brain Injury Gui...   \n",
              "8  Mycobacterium haemophilum as the Initial Prese...   \n",
              "9  Up-regulation and subcellular localization of ...   \n",
              "\n",
              "                                             authors  \\\n",
              "0  {\"{\\\"authorId\\\": \\\"51285510\\\", \\\"name\\\": \\\"Isa...   \n",
              "1  {\"{\\\"authorId\\\": \\\"5699001\\\", \\\"name\\\": \\\"H. K...   \n",
              "2  {\"{\\\"authorId\\\": \\\"4425040\\\", \\\"name\\\": \\\"T. T...   \n",
              "3  {\"{\\\"authorId\\\": \\\"2225106\\\", \\\"name\\\": \\\"A. M...   \n",
              "4  {\"{\\\"authorId\\\": \\\"7685173\\\", \\\"name\\\": \\\"Saee...   \n",
              "5  {\"{\\\"authorId\\\": \\\"1792255\\\", \\\"name\\\": \\\"A. I...   \n",
              "6  {\"{\\\"authorId\\\": \\\"2000616\\\", \\\"name\\\": \\\"A. D...   \n",
              "7  {\"{\\\"authorId\\\": \\\"145776250\\\", \\\"name\\\": \\\"Ab...   \n",
              "8  {\"{\\\"authorId\\\": \\\"2060804071\\\", \\\"name\\\": \\\"T...   \n",
              "9  {\"{\\\"authorId\\\": \\\"14726800\\\", \\\"name\\\": \\\"Hua...   \n",
              "\n",
              "                                         venue  \\\n",
              "0                       Communications Biology   \n",
              "1                                   OncoTarget   \n",
              "2                                          NaN   \n",
              "3  International Journal of Molecular Sciences   \n",
              "4      Journal of Multidisciplinary Healthcare   \n",
              "5                                       Planta   \n",
              "6               Journal of cancer survivorship   \n",
              "7             Trauma Surgery & Acute Care Open   \n",
              "8                 Case Reports in Rheumatology   \n",
              "9                                   BMC Cancer   \n",
              "\n",
              "                     publicationvenueid    year  referencecount  \\\n",
              "0  069e05e7-ca35-41d0-a8c7-bdc9ec6a82af  2018.0              28   \n",
              "1  106ecc86-1801-4b36-a05c-d36db69a968c  2017.0              46   \n",
              "2                                   NaN  2011.0              36   \n",
              "3  8506a01a-40b8-4e6f-bbb8-ce2492139c15  2019.0              97   \n",
              "4  12e5e550-28bd-401c-a348-f347956deec4  2020.0              37   \n",
              "5  66507f0b-12a1-4058-a35c-73029853d5b9  2004.0              30   \n",
              "6  649b7693-cb10-4892-b8a2-65b9b51fcd50  2016.0              33   \n",
              "7  d40f33e8-6f79-4e91-a62e-acf29f9a3e2f  2020.0              35   \n",
              "8  884ab217-3965-4a74-90d1-003d0b378d6d  2014.0              24   \n",
              "9  b482a99f-ef53-4694-aec6-32ad78e0d716  2010.0              58   \n",
              "\n",
              "   citationcount  influentialcitationcount  isopenaccess  \\\n",
              "0             13                         0          True   \n",
              "1             24                         0          True   \n",
              "2             13                         0          True   \n",
              "3             15                         1          True   \n",
              "4             17                         1          True   \n",
              "5            145                         3         False   \n",
              "6             18                         0         False   \n",
              "7             11                         0          True   \n",
              "8              8                         0          True   \n",
              "9             44                         0          True   \n",
              "\n",
              "                                     s2fieldsofstudy         publicationtypes  \\\n",
              "0  {\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...         {JournalArticle}   \n",
              "1  {\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...         {JournalArticle}   \n",
              "2  {\"{\\\"category\\\": \\\"Psychology\\\", \\\"source\\\": \\...                 {Review}   \n",
              "3  {\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...  {Review,JournalArticle}   \n",
              "4  {\"{\\\"category\\\": \\\"Medicine\\\", \\\"source\\\": \\\"s...  {JournalArticle,Review}   \n",
              "5  {\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...         {JournalArticle}   \n",
              "6  {\"{\\\"category\\\": \\\"Psychology\\\", \\\"source\\\": \\...         {JournalArticle}   \n",
              "7  {\"{\\\"category\\\": \\\"Medicine\\\", \\\"source\\\": \\\"s...  {JournalArticle,Review}   \n",
              "8  {\"{\\\"category\\\": \\\"Medicine\\\", \\\"source\\\": \\\"s...         {JournalArticle}   \n",
              "9  {\"{\\\"category\\\": \\\"Biology\\\", \\\"source\\\": \\\"s2...         {JournalArticle}   \n",
              "\n",
              "  publicationdate                                            journal  \\\n",
              "0      2018-09-03  {\"name\": \"Communications Biology\", \"pages\": nu...   \n",
              "1      2017-10-31  {\"name\": \"Oncotarget\", \"pages\": \"97516 - 97527...   \n",
              "2      2011-09-01  {\"name\": \"Applied Cognitive Psychology\", \"page...   \n",
              "3      2019-02-27  {\"name\": \"International Journal of Molecular S...   \n",
              "4      2020-12-01  {\"name\": \"Journal of Multidisciplinary Healthc...   \n",
              "5      2004-01-22  {\"name\": \"Planta\", \"pages\": \"95-102\", \"volume\"...   \n",
              "6      2016-02-26  {\"name\": \"Journal of Cancer Survivorship\", \"pa...   \n",
              "7      2020-05-01  {\"name\": \"Trauma Surgery & Acute Care Open\", \"...   \n",
              "8      2014-01-12  {\"name\": \"Case Reports in Rheumatology\", \"page...   \n",
              "9      2010-07-06  {\"name\": \"BMC Cancer\", \"pages\": \"356 - 356\", \"...   \n",
              "\n",
              "                    updated       id  \n",
              "0  2023-03-31T01:28:37.897Z  5692596  \n",
              "1  2023-04-05T06:18:15.837Z  5692609  \n",
              "2  2023-04-05T08:18:02.240Z  5692611  \n",
              "3  2023-05-20T17:20:54.230Z  5692615  \n",
              "4  2023-05-23T13:34:07.581Z  5692620  \n",
              "5  2023-04-05T15:01:41.273Z  5692625  \n",
              "6  2023-04-05T19:07:12.948Z  5692626  \n",
              "7  2023-04-07T01:56:26.282Z  5692637  \n",
              "8  2023-04-07T01:15:17.667Z  5692638  \n",
              "9  2023-04-07T01:28:15.493Z  5692639  "
            ]
          },
          "execution_count": 6,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "attributes_csv = starter_data + \"/final.csv\"\n",
        "attributes_df = pd.read_csv(attributes_csv)\n",
        "attributes_df.head(10)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "BNzkll_5CcPD",
        "outputId": "1237b7ad-e0a0-4bce-be75-ead818c99022"
      },
      "outputs": [
        {
          "name": "stdout",
          "output_type": "stream",
          "text": [
            "In spite of many years of work by a multitude of organizations, a clear and simple standard for software engineering and management requirements and a method to assess their applicability to projects of various types and sizes remains elusive. From IEEE to CMMI to NASA's NPR 7150.2, there is no shortage of sources of information providing various types of requirements and standards for software engineering. Even a book on software project management for “dummies” approaches 400 pages. Wading through this information can dizzy the mind of even the most experienced software engineer; the newbie just trying to “do the right thing” will probably give up, open a text editor and start coding. This lack of clarity and simplicity perhaps goes a long way towards explaining why, in spite of this large body of work, there remains such an incredible variability in the knowledge and application of software engineering discipline not only from one organization to the next, but between groups within the same organization, or even between individual developers in the same group! Surely at least the basics of what should be done and why those things should be done can be conveyed in less than a novel-sized volume. There must be some timeless principles that cut across structured and object-oriented techniques, waterfall and agile methods, and CMMI and NASA standards. To properly interpret software engineering requirements and approaches and successfully (and selectively) apply them, one must first understand them at a fundamental level and how they can benefit the project. This paper will make an admittedly bold and brash attempt to boil it all down into something anyone can understand, hopefully resulting in a brief reference — a type of lens through which existing standards can be more practically viewed.\n"
          ]
        }
      ],
      "source": [
        "# sch = SemanticScholar()\n",
        "\n",
        "# sch = SemanticScholar(api_key='jvN8XitMY59Wtoqm3FZnc94qRHnsKw3z3hLNh7XB')\n",
        "\n",
        "# paper = sch.get_paper('d0bc1501ae6f54dd16534e651d90d2aeeeb1cfc1')  #'10.1093/mind/lix.236.433')\n",
        "\n",
        "# paper.referenceCount #number of references in this one, can help us create some sort of pagerank type algo\n",
        "# paper.citationCount #number of papers that cited this one\n",
        "# paper.externalIds\n",
        "# text_example = paper.abstract\n",
        "# # paper.embedding\n",
        "# print(text_example)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 22,
      "metadata": {},
      "outputs": [
        {
          "data": {
            "text/plain": [
              "'\\nRelation between quantity and quality of peri-coronary epicardial adipose tissue and its underlying hemodynamically significant coronary stenosis\\n\\n\\nYu Du \\n† \\nLin Yang \\nYan Liu \\nBangguo Yang \\nSai Lv \\nChenping Hu \\nYong Zhu \\nHongkai Zhang \\nQian Ma \\nZhijian Wang \\nYuyang Liu \\nDongmei Shi \\nYingxin Zhao \\nLei Xu \\nYujie Zhou \\nRelation between quantity and quality of peri-coronary epicardial adipose tissue and its underlying hemodynamically significant coronary stenosis\\n10.1186/s12872-020-01499-wR E S E A R C H A R T I C L E Open Access\\nBackground: We aimed to investigate the association of lesion-specific epicardial adipose tissue (EAT) volume and density with the presence of myocardial ischemia.Methods:We enrolled 45 patients (55 lesions) with known or suspected coronary artery disease who underwent coronary computed tomography angiography (CTA) followed by invasive fractional flow reserve (FFR) assessment within 30 days. EAT volume (index) and density in patient-, vessel-and lesion-level were measured on CTA images. Lesion-specific ischemia was defined as a lesion with stenosis diameter > 90% or FFR ≤0.80. Multivariate analysis determined the independent association of EAT parameters with lesion-specific ischemia. Results: Mean age of the patients was 60 years, and 75% were male. Overall, 55.6% of patients had ischemic lesions and a mean FFR baseline value of 0.82 ± 0.10. Total EAT volume index was significantly higher in patients with functionally or anatomically significant stenosis. Specifically, peri-lesion EAT volume index, not the density, was positively correlated with lesion-specific ischemia independent of luminal stenosis and plaque characteristics (hazard ratio 1.56, 95% confidence interval 1.04-2.33, P = 0.032; per 0.1 ml/m 2 increase). Moreover, peri-lesion EAT volume was negatively correlated with lesion FFR values, whereas total EAT volume was positively correlated with fat accumulation and glucose metabolism. In addition, there was no association of EAT volume or density with myocardial ischemia in vessel-level analysis.Conclusions: Lesion-specific EAT volume index, but not density, seems positively and independently associated with myocardial ischemia, while its incremental diagnostic value of lesion-specific ischemia should be further investigated.\\n\\nBackground\\n\\nEpicardial adipose tissue (EAT) is a specific fat depot between the myocardium and the visceral pericardium, mainly surrounding major epicardial coronary arteries or within the myocardium [1]. Numerous studies showed that the accumulation of EAT is closely associated with the presence and severity of coronary artery disease (CAD), myocardial ischemia, plaque vulnerability, and major adverse cardiovascular events (MACE) [2][3][4][5]. However, not all patients with increased EAT volume develop CAD and vice versa.\\n\\nTo further explain the relation between EAT and CAD, the following two factors are attracting much attention. The first one is peri-coronary EAT, because there is close cross-talk between this focal metaboloactive EAT and its underlying coronary wall via a paracrine pathway [6]. External administration of interleukin 1β or monocyte chemotactic protein 1 to porcine coronary arteries led to coronary wall inflammation and atherosclerosis formation [7,8], while surgical removal of peri-coronary EAT ameliorated the progression of coronary atherosclerosis in the pigs [9]. The second important factor is EAT quality, shown as fat attenuation on computed tomography (CT), which reveals adipocyte lipid content and size, reflecting a metabolic response to cardiovascular risk factors [10]. Antonopoulos et al. found that the peri-coronary EAT attenuation index can expose vascular inflammation, subclinical coronary atherosclerosis, and vulnerable plaque [11] and even independently predict cardiac mortality [12].\\n\\nAccordingly, it is theoretically feasible to determine the hemodynamic significance and vulnerability of coronary stenosis by simply measuring peri-coronary EAT quantity or quality via coronary computed tomography angiography (CTA) in high-risk CAD patients before hospitalization. Therefore, in this hypothesis-generating study, we aimed to investigate the association of lesionspecific EAT volume or attenuation with the functional significance of coronary stenosis on a hospital-based Chinese population.\\n\\n\\nMethods\\n\\n\\nStudy patients\\n\\nWe retrospectively screened 134 consecutive patients who underwent coronary CTA and invasive fractional flow reserve (FFR) assessment for known or suspected CAD between June 2015 and June 2018 in our center. Exclusion criteria included the presence of acute coronary syndrome (ACS) within the past 30 days; a history of myocardial revascularization or old myocardial infarction; presence of co-morbidities (i.e., heart failure, atrial fibrillation, insulin-dependent diabetes, severe lipoprotein disorders, chronic liver or renal disease, inflammatory disease, malignancy); coronary CTA not performed within 30 days before FFR evaluation or technical failure to measure FFR or EAT parameters; presence of threevessel disease, left main or chronic totally occluded lesions, or any intermediate lesions on main branch not measured by FFR. In addition, in lesion-level analysis, the following FFR-measured lesions were excluded: ostial lesions; lesions in a distal segment of a vessel; bifurcation lesions; side-branch lesions; extensively calcified or tortuous lesions. The study protocol conformed to the Declaration of Helsinki and was approved by the ethics committee of Beijing Anzhen Hospital of Capital Medical University. Written informed consent was obtained from each patient.\\n\\n\\nCT imaging\\n\\nCT imaging was performed using a dual-source 128slice CT scanner (Somatom Definition Flash, 280-ms rotation, 2 × 128 × 0.6 collimation; Siemens Healthcare, Forchheim, Germany). Contrast-enhanced scanning and reconstruction protocols were as previously described [13]. In this study, we enrolled only patients with contrast-enhanced scans obtained at 120-kV tube voltage and with a z-axis scan range at least from the pulmonary artery bifurcation to the ventricular apex. We obtained the location, length, volume, and classification of coronary plaque on CTA images. Plaque was classified as calcified or non-calcified. Calcified plaque referred to lesions composed exclusively of structures with a CT attenuation higher than that of the contrast-enhanced coronary lumen in at least two planes. Non-calcified plaque was defined as lesions clearly assignable to the vessel wall (in at least two views) with an attenuation lower than that of the contrast-enhanced lumen.\\n\\nEAT volume and density quantification EAT volume and density were quantified offline from CTA images using cardiac risk analysis software (Siemens, Erlangen, Germany) and a computer workstation (MMWP, Siemens Healthcare). CT attenuation range of EAT was set between − 190 Hounsfield Unit (HU) and − 30 HU. Assessment of total EAT (ranging from the bifurcation of the pulmonary artery to the end slice of the left ventricular apex, from the outer wall of the myocardium to the visceral pericardium) volumes were according to region of interests (ROIs) on all axial imaging planes that were semi-automatically delineated (every 5 mm interval), and then manually adjusted to modify and track the silhouette of the EAT on every slice (Fig. 1a) [14]. Peri-vessel EAT (veEAT) (surrounding from the ostium to the end of each major epicardial coronary artery within the interventricular or atrioventricular groove, respectively) volumes were manually delineated according to ROIs on all axial imaging slices (every 5 mm interval) (Fig. 1b) [15]. While, peri-lesion EAT (leEAT) was a part of veEAT, which covered the vessel segment containing the full length of study coronary lesion, and leEAT volumes were calculated by ROIs on each axial imaging slice (every 1 mm interval) (Fig. 1c) [16]. Volume index was defined as volume (ml) divided by body surface area (BSA, m 2 ).\\n\\nWe also obtained leEAT density by using manual measurement, defined as the mean value of three or more circular ROIs (each ROI area = 0.03 cm 2 ) surrounding the cross-sectional slice of the study lesion with the most severe diameter stenosis (Fig. 1d, d\\'). The location of ROIs was required to be more than 1 mm from the adventitia of the coronary artery to reduce the influence of luminal contrast on leEAT density (Fig. 1e) [14]. Correspondingly, the veEAT density-defined as EAT CT attenuation of the proximal and distal reference segments adjacent to the stenotic segment and free from stenosis-was measured and used as the reference veEAT density. All EAT parameter measurements were performed repeatedly by an experienced CT reader, who was blinded to the clinical characteristics of the patients and the functional significance of the coronary stenosis. Intra-observer reproducibility was excellent [intraclass correlation coefficient 0.95, 95% confidence interval (CI) 0.90-0.97, P < 0.001].\\n\\n\\nCoronary angiography and fractional flow reserve\\n\\nAll patients underwent coronary angiography according to standard clinical protocols. CAD was diagnosed by the presence of at least one coronary artery stenosed > 50% in diameter stenosis on coronary angiography. Stenosis causing 30-90% coronary lumen reduction were evaluated by invasive FFR. A lesion with luminal stenosis ≥75% was defined as significantly stenosis. A lesion diameter with > 90% stenosis or FFR ≤0.80 indicated hemodynamic significance. Invasive FFR was measured using a coronary pressure wire (Volcano, Rancho Cordova, CA, USA or St Jude Medical, Minneapolis, MN, USA), as described previously [17]. In this study, maximum hyperemia was induced by intravenous administration of adenosine triphosphate (140-180 μg/kg/ min) via the forearm vein. The decision for revascularization was left to the operator\\'s discretion.\\n\\n\\nStatistical analysis\\n\\nContinuous variables were expressed as means ± standard deviation, or the median (lower quartile, upper quartile), where indicated. Mean values were compared using Student\\'s t test and median values using the Mann-Whitney U test. Categorical variables were expressed as percentages and analyzed using the chi-square test or Fisher\\'s exact test where appropriate. The correlations of EAT parameters with clinical characteristics were evaluated using Spearman\\'s correlation analysis. Univariate and multivariate analyses were performed to determine the independent predictors for hemodynamically significant coronary stenosis. P < 0.05 was considered to indicate statistical significance. All statistical analyses were performed using SPSS 22.0 software (IBM, Armonk, NY, USA). According to the most severe diameter stenotic cross-sectional CTA image (e), peri-lesion EAT density was the average CT attenuation of perilesion EAT. EAT, epicardial adipose tissue; CTA, computed tomography angiography; ROI, region of interest\\n\\n\\nResults\\n\\nA total of 45 patients were enrolled in patient-level analysis. The study flow chart is shown in Supplementary Figure S1. As shown in Table 1, the mean age of the patients was 60 years, 75% were male, and 67% had one-vessel disease. The mean time between coronary CTA and invasive angiography and FFR was 11 days. In all, 68 lesions had a diameter stenosed ≥30%. Among them, FFR was performed on 65 (96%) lesions. Three lesions with diameters stenosis > 90% were not measured by \\n\\n\\nEAT volume and myocardial ischemia in patient-level analysis\\n\\nPatients were divided into non-ischemia and ischemia groups based on the presence of at least one hemodynamically relevant lesion (44 and 56%, LAD left anterior descending, LCX left circumflex, RCA right coronary artery, EAT epicardial adipose tissue, veEAT peri-vessel epicardial adipose tissue, leEAT perilesion epicardial adipose tissue, FFR fractional flow reserve a indicated FFR was not measured in three lesions of the ischemia group because of angiographically severe stenosis respectively). The ischemic patients were slightly older than the non-ischemic patients, had more extensive and severe CAD, and half underwent PCI (Table 1). As shown in Fig. 2a, the total EAT volume was nonsignificantly greater in ischemic patients than nonischemic patients (102.0 vs 84.0 ml, P = 0.077), whereas this pattern became statistically significant regarding the indexed total EAT volume (53.4 vs 47.6 ml/m 2 , P = 0.032). Patients were also divided into non-significant and significant stenosis groups according to the presence of at least one significantly stenotic lesion (58 and 42%, respectively). Similarly, as shown in Fig. 2b, compared with non-significant stenosis patients, significant stenosis patients had a slightly higher total EAT volume (102.0 vs 85.9 ml, P = 0.061) and a markedly higher indexed total EAT volume (53.4 vs 47.6 ml/m 2 , P = 0.046). The relation of EAT volume and myocardial ischemia in a dose-dependent manner was not further investigated because most of the patients had at least one ischemic or significant stenosis (91 and 95%, respectively).\\n\\nIn addition, we found that the total EAT volume was positively correlated with the body mass index (r = 0.381, P = 0.010), fasting blood glucose level (r = 0.341, P = 0.022), and glycated serum protein level (r = 0.434, P = 0.003). Moreover, the indexed total EAT volume was still positively correlated with the fasting blood glucose (r = 0.330, P = 0.027) and glycated serum protein (r = 0.466, P = 0.001) (Fig. 2c). Glycated serum protein was also positively correlated with the total veEAT volume (r = 0.308, P = 0.040) and its index value (r = 0.293, P = 0.051) (Fig. 2d).\\n\\n\\nEAT volume or density and myocardial ischemia in lesionlevel analysis\\n\\nAll study lesions were divided into non-ischemic and ischemic lesions, the baseline characteristics of which are shown in Table 2. Compared with non-ischemia lesions, ischemic lesions had more severe diameter stenosis and greater plaque length. Pre-procedural FFR values for ischemic lesions were significantly lower than those of the non-ischemic lesions (0.74 ± 0.08 vs 0.89 ± 0.05, P < 0.001). Ischemic lesions accounted for 46% of all lesions that underwent PCI. None of the non-ischemic lesions were treated. As shown in Table 2, compared with non-ischemic lesions, ischemic lesions had higher leEAT volumes (0.64 vs 0.50 ml, P = 0.051) and a significantly higher leEAT volume index (0.34 vs 0.27 ml/m 2 , P = 0.045). This pattern, however, was not apparent for the veEAT volume (6.89 vs 7.20 ml, P = 0.607) or veEAT volume index (3.83 vs 3.80 ml/m 2 , P = 0.508). Subgroup analysis showed that, except for the ischemic lesions (P = 0.045), the leEAT volume index was not significantly different with respect to proximal lesions, LAD, calcified plaque, or diameter stenosed ≥75% (all P > 0.05) (Fig. 3a). Furthermore, as shown in Supplementary Table S1, the independent lesion-level predictors for ischemia were diameter stenosis of ≥75% [hazard ratio (HR) Fig. 2 EAT volume and myocardial ischemia in patient-level analysis. Total EAT volume (index) in patient with functionally (a) or anatomically significant stenosis (b). Correlations of glycated serum protein with total EAT volume index (c) and total veEAT volume index (d). EAT, epicardial adipose tissue; veEAT, peri-vessel epicardial adipose tissue 14.56, 95% CI 2.93-72.24, P = 0.001] and the leEAT volume index (per 0.1 ml/m 2 increase) (HR 1.56, 95% CI 1.04-2.33, P = 0.032), after adjusting for LAD, proximal segment, plaque length, plaque calcification, and reference vessel diameter.\\n\\nWe also evaluated EAT density between non-ischemic and ischemic lesions, which showed no significant differences with regard to leEAT density (− 76.54 ± 12.71 vs − 78.23 ± 10.44 HU, P = 0.600), veEAT density of the proximal reference vessel (− 71.94 ± 9.60 vs − 74.00 ± 11.63 HU, P = 0.475)], or the distal reference vessel (− 90.40 ± 9.65 vs − 91.38 ± 14.04 HU, P = 0.763) (Fig. 3b). Additionally, we found that baseline FFR values negatively correlated with the leEAT volume (r = − 0.282, P = 0.043) (Fig. 3c) or the leEAT volume index (r = − 0.281, P = 0.044) (Fig. 3d).\\n\\n\\nDiscussion\\n\\nWe performed a proof-of-concept study to evaluate the associations of EAT volume and density with myocardial ischemia in patient-, vessel-, and lesion-level analyses, in a high-risk Chinese population. First, we found that the total EAT volume index was markedly increased in patients with functionally or anatomically significant stenosis. Second, leEAT volume index was markedly higher for ischemic lesions, furthermore, this index was a lesion-specific predictor of hemodynamic relevance, independent of a variety of lesion-level parameters. Third, the total EAT volume was positively correlated with fat accumulation and glucose metabolism, whereas the leEAT volume was negatively correlated with lesion FFR values. Fourth, we did not observe any significant differences in CT attenuation between non-ischemic and ischemic lesions in both lesion-or vessel-level analyses.\\n\\nThe significant correlations of the total EAT volume with functionally or anatomically significant CAD have been widely observed [2,3], although not reported in a recent study by Muthalaly et al. with moderately severe CAD patients [18] or even in the prospective, large-scale CORE320 multicenter study [19]. Likewise, Romijn et al. found an independent relation between EAT volume and myocardial ischemia, but the diagnostic performance for identifying hemodynamically significant CAD was not improved when considering coronary artery calcification [20]. A recent meta-analysis that included more than 40,000 subjects at low to intermediate risk of cardiovascular disorders showed that EAT volume was an independent predictor of obstructive stenosis, significant stenosis, myocardial ischemia, and MACE, irrespective of traditional cardiovascular risk factors [21]. In line with this meta-analysis, we found that indexed total EAT volume was significantly increased in patients with functionally or anatomically significant stenosis. The above conflicting results might be attributed, at least in part, to a diversity of populations [21], more importantly illustrate a complex relation between EAT and the pathogenesis of CAD, which is hardly explain by the overall EAT volume alone.\\n\\nRecent evidence suggests a potentially active role of peri-coronary EAT in the pathogenesis of coronary atherosclerosis [22]. It was reported that peri-coronary segment EAT quantity was related to plaque size and composition, independent of cardiovascular risk factors and overall EAT volume [16,23]. Also, this coronary segment-specific EAT volume was associated with luminal stenosis severity, the presence of reversible perfusion defects, and the culprit lesion [15,24,25]. Unlike the above studies, we performed a lesion-specific analysis and found a close relation between indexed leEAT volume and the presence of lesion ischemia. Furthermore, every 0.1 ml/m 2 increase in the leEAT volume index was independently associated with a 1.6-fold increased risk of lesion ischemia, irrespective of lesion stenosis or plaque features. Although the predicting value was relatively small, our results showed the feasibility to predict lesion-specific ischemia using leEAT volume index, which appeared to have a promising prospective considering a large and increasing number of patients referred to undergo coronary CTA examination for suspected CAD. We did not, however, observe a markedly increased leEAT volume index in significantly luminal stenosis or non-calcified plaque, which has been reported elsewhere [24]. Mahabadi et al., however, also found no difference between non-calcified and calcified plaque in terms of leEAT volume [16]. Additionally, a vessel-level analysis showed no associations of veEAT volume and ischemia, which might support the hypothesis that the function of EAT varies depending on its particular depot. In line with this notion, we found the total EAT volume in patient-or vessel-level was significantly associated with fat accumulation and glucose metabolism, while the leEAT volume was associated with FFR values. More specifically, a dysfunctional secretion profile and M2 macrophage accumulation in lesion-level analysis were reported to be higher for peri-coronary EAT near stenotic coronary segments than for those near non-stenotic segments [26,27]. Another possible explanation is that the metabolic activities of veEAT and leEAT, shown as EAT CT attenuation, were not consistent or remained unchanged. Markedly increased peri-coronary EAT CT attenuation was observed in stenotic segments, lesions with 18 F-sodium fluoride uptake on positron emission tomography and culprit lesions [28][29][30]. In this study, however, there were no significant differences between ischemic and non-ischemic lesions in terms of peri-coronary EAT and peri-vessel EAT attenuation. In line with our findings, Hell et al. found that EAT volume, but not density, was associated with myocardial ischemia in patients suspected of having CAD [31]. Also, Balcer et al. reported that the peri-coronary EAT volume, but not attenuation, was independently correlated with culprit lesions [25].\\n\\nThis study had some limitations. First, the study\\'s sample size was relatively small, which impeded the performance of more subgroup or exploratory analyses. Second, because of the relatively strict inclusion and exclusion criteria, extrapolating our conclusions to other populations should be cautious. Third, based on the cross-sectional design, no causal relation could be confirmed, although multivariate regression analysis was performed. Fourth, methods for measuring the pericoronary fat quality varies among studies, which might lead to non-comparable results among them. In addition, ROIs of veEAT and leEAT were manually, not automatically, delineated due to software limitations.\\n\\n\\nConclusion\\n\\nWe found that the indexed lesion-specific EAT volume appeared to be positively associated with lesion ischemia, independent of luminal stenosis and plaque characteristics. Hence, the peri-coronary lesion EAT volume index might be a promising marker for identifying lesionspecific ischemia, although its incremental diagnostic value when considering other risk factors should be further investigated.\\n\\n\\nSupplementary information\\n\\nSupplementary information accompanies this paper at https://doi.org/10. 1186/s12872-020-01499-w.\\n\\nAdditional file 1: Figure S1. Study flow-chart. Table S1. Lesion-level predictors for myocardial ischemia in univariate and multivariate analyses. \\n\\nFig. 1\\n1Measurements of EAT quantity and quality on coronary CTA images. Cross-sectional image for volume measurement of total EAT (a), perivessel EAT (b), and peri-lesion EAT (c). Location and length of study coronary lesion were double checked using CTA (c) and corresponding angiographic images (c′). ROIs are stained red. The narrowest location in diameter stenosis of study lesion were determined in (d) and (d\\').\\n\\nFig. 3\\n3The leEAT volume or density and lesion-specific characteristics. (a) The leEAT volume index and lesion-specific characteristics. (b) EAT density and lesion ischemia. Correlations of baseline FFR values with leEAT volume (c) and its indexed value (d). leEAT, peri-lesion epicardial adipose tissue; LAD, left anterior descending coronary artery; DS, diameter stenosis; EAT, epicardial adipose tissue; FFR, fractional flow reserve\\n\\n\\nAbbreviations EAT: Epicardial adipose tissue; CAD: Coronary artery disease; MACE: Major adverse cardiovascular events; CT: Computed tomography; CTA: Computed tomography angiography; FFR: Fractional flow reserve; ACS: Acute coronary syndrome; ROI: Regions of interest; HU: Hounsfield unit; veEAT: Peri-vessel EAT; leEAT: Peri-lesion EAT; BSA: Body surface area; CI: Confidence interval; LAD: Left anterior descending; PCI: Percutaneous coronary intervention; HR: Hazard ratio\\n\\nTable 1\\n1Baseline characteristics of study patients according to the presence of ischemia lesion BMI body mass index, BSA body surface area, PAD peripheral artery disease, CAD coronary artery disease, LVEF left ventricular ejection fraction, EAT epicardial adipose tissue, ACEI angiotensin-converting enzyme inhibitor, ARB Angiotensin II receptor blockerTotal (n = 45) \\nNon-ischemia patient(n = 20) \\nIschemia patient (n = 25) \\nP value \\n\\nClinical variable \\n\\nAge (years) \\n59.6 ± 4.7 \\n58.2 ± 4.8 \\n60.7 ± 4.4 \\n0.073 \\n\\nMale \\n34 (75.6) \\n13 (65.0) \\n21 (84.0) \\n0.176 \\n\\nBody weight (kg) \\n73.8 ± 11.1 \\n74.7 ± 13.5 \\n73.1 ± 9.0 \\n0.649 \\n\\nHeight (m) \\n1.69 ± 0.08 \\n1.69 ± 0.09 \\n1.70 ± 0.07 \\n0.689 \\n\\nBMI (kg/m 2 ) \\n25.7 ± 2.8 \\n26.1 ± 3.2 \\n25.4 ± 2.4 \\n0.382 \\n\\nBSA (m 2 ) \\n1.86 ± 0.18 \\n1.86 ± 0.21 \\n1.85 ± 0.14 \\n0.818 \\n\\nHypertension \\n26 (57.8) \\n10 (50.0) \\n16 (64.0) \\n0.379 \\n\\nDiabetes mellitus \\n13 (28.9) \\n3 (15.0) \\n10 (40.0) \\n0.100 \\n\\nHypercholesteromia \\n15 (33.3) \\n6 (30.0) \\n9 (36.0) \\n0.757 \\n\\nCurrent smoking \\n17 (37.8) \\n5 (25.0) \\n12 (48.0) \\n0.135 \\n\\nPAD or stroke \\n3 (6.7) \\n2 (10.0) \\n1 (4.0) \\n0.577 \\n\\nFamily history of CAD \\n5 (11.1) \\n2 (10.0) \\n3 (12.0) \\n1.000 \\n\\nLVEF (%) \\n65.2 ± 5.3 \\n66.4 ± 5.6 \\n64.2 ± 5.0 \\n0.181 \\n\\nProcedural variable \\n\\nTime interval (day) \\n11.3 ± 8.4 \\n12.9 ± 9.2 \\n10.1 ± 7.7 \\n0.304 \\n\\nNo. of diseased vessels \\n0.005 \\n\\n0-vessel disease a \\n2 (4.4) \\n2 (10.0) \\n0 \\n\\n1-vessel disease \\n30 (66.7) \\n17 (85.0) \\n13 (52.0) \\n\\n2-vessel disease \\n7 (15.6) \\n1 (5.0) \\n6 (24.0) \\n\\n3-vessel disease \\n6 (13.3) \\n0 \\n6 (24.0) \\n\\nNo. of lesions per-patient \\n1.51 ± 0.73 \\n1.15 ± 0.37 \\n1.80 ± 0.82 \\n0.003 \\n\\nNo. of ischemia lesions per-patient \\n0.67 ± 0.71 \\n0 \\n1.20 ± 0.50 \\n< 0.001 \\n\\nNo. of significantly stenotic lesions per-patient \\n0.47 ± 0.59 \\n0.15 ± 0.37 \\n0.72 ± 0.61 \\n0.001 \\n\\nNo. of significantly stenotic patients \\n19 (42.2) \\n3 (15.0) \\n16 (64.0) \\n0.002 \\n\\nCoronary intervention \\n13 (28.9) \\n0 \\n13 (52.0) \\n< 0.001 \\n\\nNo. of stents \\n0.36 ± 0.61 \\n0 \\n0.64 ± 0.70 \\n< 0.001 \\n\\nLength of stents \\n9.04 ± 15.66 \\n0 \\n16.3 ± 18.1 \\n< 0.001 \\n\\nEATparameter \\n\\nTotal EAT volume (ml) \\n89.6 (73.4, 122.8) \\n84.0 (64.7, 110.5) \\n102.0 (74.7, 139.8) \\n0.077 \\n\\nTotal EAT volume index (ml/m 2 ) \\n49.8 (40.5, 64.4) \\n47.6 (35.1, 55.4) \\n53.4 (43.1, 71.6) \\n0.032 \\n\\nMedications on discharge \\n\\nAspirin \\n42 (93.3) \\n18 (90.0) \\n24 (96.0) \\n0.577 \\n\\nClopidogrel \\n27 (60.0) \\n11 (55.0) \\n16 (64.0) \\n0.760 \\n\\nStatin \\n37 (82.2) \\n15 (75.0) \\n22 (88.0) \\n0.435 \\n\\nACEI/ARB \\n18 (40.0) \\n6 (30.0) \\n12 (48.0) \\n0.359 \\n\\nNitrates \\n28 (62.2) \\n8 (40.0) \\n20 (80.0) \\n0.012 \\n\\na \\n\\nindicated at least one lesion with angiographic diameter stenosis of 30-49% in major epicardial coronary artery \\n\\n\\nTable 2\\n2Baseline characteristics of study lesions according to the presence of ischemiaTotal (n = 55) \\nNon-ischemia lesion (n = 29) \\nIschemia lesion (n = 26) \\nP value \\n\\nLesion parameter \\n\\nLocation \\n0.350 \\n\\nLAD \\n39 (70.9) \\n18 (62.1) \\n21 (80.8) \\n\\nLCX \\n4 (7.3) \\n3 (10.3) \\n1 (3.8) \\n\\nRCA \\n12 (21.8) \\n8 (27.6) \\n4 (15.4) \\n\\nSegment \\n0.584 \\n\\nProximal \\n19 (34.5) \\n9 (31.0) \\n10 (38.5) \\n\\nMiddle \\n36 (65.5) \\n20 (69.2) \\n16 (61.5) \\n\\nReference vessel diameter (mm) \\n\\nProximal \\n3.9 ± 0.5 \\n4.0 ± 0.5 \\n3.9 ± 0.5 \\n0.341 \\n\\nDistal \\n3.0 ± 0.5 \\n3.1 ± 0.6 \\n2.8 ± 0.4 \\n0.069 \\n\\nStenosis severity \\n< 0.001 \\n\\n30~49% \\n6 (10.9) \\n6 (20.7) \\n0 \\n\\n50~74% \\n27 (49.1) \\n18 (62.1) \\n9 (34.6) \\n\\n≥ 75% \\n22 (40.0) \\n5 (17.2) \\n17 (65.4) \\n\\nPlaque length (mm) \\n14.2 (11.1, 20.2) \\n12.9 (10.8, 16.7) \\n19.3 (12.0, 21.5) \\n0.028 \\n\\nPlaque volume (mm 3 ) \\n139.0 (81.0, 192.0) \\n118.0 (76.0, 161.0) \\n156.5 (84.8, 220.3) \\n0.159 \\n\\nPlaque type \\n0.405 \\n\\nNon-calcified \\n21 (38.2) \\n13 (44.8) \\n8 (30.8) \\n\\nCalcified \\n34 (61.8) \\n16 (55.2) \\n18 (69.2) \\n\\nEAT parameter \\n\\nveEAT volume (ml) \\n7.14 (4.93, 9.18) \\n7.20 (4.98, 8.54) \\n6.89 (4.75, 10.57) \\n0.607 \\n\\nveEAT volume index (ml/m 2 ) \\n3.80 (2.62, 4.70) \\n3.80 (2.67, 4.26) \\n3.83 (2.38, 5.87) \\n0.508 \\n\\nleEAT volume (ml) \\n0.57 (0.35, 0.72) \\n0.50 (0.32, 0.66) \\n0.64 (0.37, 1.07) \\n0.051 \\n\\nleEAT volume index (ml/m 2 ) \\n0.30 (0.20, 0.40) \\n0.27 (0.19, 0.35) \\n0.34 (0.20, 0.53) \\n0.045 \\n\\nveEAT density (HU) \\n\\nProximal reference vessel \\n−72.91 ± 10.56 \\n−71.94 ± 9.60 \\n−74.00 ± 11.63 \\n0.475 \\n\\nDistal reference vessel \\n−90.86 ± 11.82 \\n−90.40 ± 9.65 \\n−91.38 ± 14.04 \\n0.763 \\n\\nleEAT density (HU) \\n−77.32 ± 11.63 \\n−76.54 ± 12.71 \\n−78.23 ± 10.44 \\n0.600 \\n\\nProcedural variable \\n\\nPre-procedural FFR a \\n0.82 ± 0.10 \\n0.89 ± 0.05 \\n0.74 ± 0.08 \\n< 0.001 \\n\\nCoronary intervention \\n12 (21.8) \\n0 \\n12 (46.2) \\n< 0.001 \\n\\nNo. of stents per-lesion \\n0.24 ± 0.47 \\n0 \\n0.50 ± 0.58 \\n< 0.001 \\n\\nLength of stents \\n6.38 ± 12.91 \\n0 \\n13.50 ± 16.12 \\n< 0.001 \\n\\nDiameter of stents \\n0.72 ± 1.38 \\n0 \\n1.52 ± 1.69 \\n< 0.001 \\n\\n\\nAcknowledgementsNot applicable.(DFL20150601) (for Principal Investigator: Yujie Zhou) and Mission plan (SML20180601) (for Principal Investigator: Yujie Zhou), Beijing Municipal Health Commission \"Project of Science and Technology Innovation Center\" (PXM2019_026272_000006) (PXM2019_026272_000005) (for Principal Investigator: Yujie Zhou). The funding bodies had no role in the design of the study and collection, analysis, and interpretation of data and in writing the manuscript.Availability of data and materialsThe data that support the findings of this study are available on request from the corresponding author.Ethics approval and consent to participateThe study protocol conformed to the Declaration of Helsinki and was approved by the ethics committee of Beijing Anzhen Hospital of Capital Medical University. Written informed consent was obtained from each patient. The data that support the findings of this study are available on request from the corresponding author.Consent for publicationNot applicable.Competing interestsAll authors have no potential conflict of interest to disclose.Author details1\\nEpicardial adipose tissue: anatomic, biomolecular and clinical relationships with the heart. G Iacobellis, D Corradi, A M Sharma, Nat Clin Pract Cardiovasc Med. 210Iacobellis G, Corradi D, Sharma AM. Epicardial adipose tissue: anatomic, biomolecular and clinical relationships with the heart. Nat Clin Pract Cardiovasc Med. 2005;2(10):536-43.\\n\\nEpicardial adipose tissue thickness and its association with the presence and severity of coronary artery disease in clinical setting: a cross-sectional observational study. S K Sinha, R Thakur, M J Jha, A Goel, V Kumar, A Kumar, V Mishra, C M Varma, V Krishna, A K Singh, J Clin Med Res. 85Sinha SK, Thakur R, Jha MJ, Goel A, Kumar V, Kumar A, Mishra V, Varma CM, Krishna V, Singh AK, et al. Epicardial adipose tissue thickness and its association with the presence and severity of coronary artery disease in clinical setting: a cross-sectional observational study. J Clin Med Res. 2016; 8(5):410-9.\\n\\nEpicardial fat volume and concurrent presence of both myocardial ischemia and obstructive coronary artery disease. R Nakazato, D Dey, V Y Cheng, H Gransar, P J Slomka, S W Hayes, L E Thomson, J D Friedman, J K Min, D S Berman, Atherosclerosis. 2212Nakazato R, Dey D, Cheng VY, Gransar H, Slomka PJ, Hayes SW, Thomson LE, Friedman JD, Min JK, Berman DS. Epicardial fat volume and concurrent presence of both myocardial ischemia and obstructive coronary artery disease. Atherosclerosis. 2012;221(2):422-6.\\n\\nAssociation of Epicardial Adipose Tissue and High-Risk Plaque Characteristics: A Systematic Review and Meta-Analysis. N Nerlekar, A J Brown, R G Muthalaly, A Talman, T Hettige, J D Cameron, Dtl W, J Am Heart Assoc. 68Nerlekar N, Brown AJ, Muthalaly RG, Talman A, Hettige T, Cameron JD, DTL W. Association of Epicardial Adipose Tissue and High-Risk Plaque Characteristics: A Systematic Review and Meta-Analysis. J Am Heart Assoc. 2017;6(8).\\n\\nAssociation of epicardial fat with cardiovascular risk factors and incident myocardial infarction in the general population: the Heinz Nixdorf recall study. A A Mahabadi, M H Berg, N Lehmann, H Kalsch, M Bauer, K Kara, N Dragano, S Moebus, K H Jockel, R Erbel, J Am Coll Cardiol. 6113Mahabadi AA, Berg MH, Lehmann N, Kalsch H, Bauer M, Kara K, Dragano N, Moebus S, Jockel KH, Erbel R, et al. Association of epicardial fat with cardiovascular risk factors and incident myocardial infarction in the general population: the Heinz Nixdorf recall study. J Am Coll Cardiol. 2013;61(13): 1388-95.\\n\\nThe interplay between adipose tissue and the cardiovascular system: is fat always bad?. I Akoumianakis, C Antoniades, Cardiovasc Res. 1139Akoumianakis I, Antoniades C. The interplay between adipose tissue and the cardiovascular system: is fat always bad? Cardiovasc Res. 2017;113(9): 999-1008.\\n\\nChronic treatment with interleukin-1 beta induces coronary intimal lesions and vasospastic responses in pigs in vivo. The role of platelet-derived growth factor. H Shimokawa, A Ito, Y Fukumoto, T Kadokami, R Nakaike, M Sakata, T Takayanagi, K Egashira, A Takeshita, J Clin Invest. 973Shimokawa H, Ito A, Fukumoto Y, Kadokami T, Nakaike R, Sakata M, Takayanagi T, Egashira K, Takeshita A. Chronic treatment with interleukin-1 beta induces coronary intimal lesions and vasospastic responses in pigs in vivo. The role of platelet-derived growth factor. J Clin Invest. 1996;97(3): 769-76.\\n\\nRho-kinase is involved in macrophage-mediated formation of coronary vascular lesions in pigs in vivo. K Miyata, H Shimokawa, T Kandabashi, T Higo, K Morishige, Y Eto, K Egashira, K Kaibuchi, A Takeshita, Arterioscler Thromb Vasc Biol. 2011Miyata K, Shimokawa H, Kandabashi T, Higo T, Morishige K, Eto Y, Egashira K, Kaibuchi K, Takeshita A. Rho-kinase is involved in macrophage-mediated formation of coronary vascular lesions in pigs in vivo. Arterioscler Thromb Vasc Biol. 2000;20(11):2351-8.\\n\\nEpicardial adipose excision slows the progression of porcine coronary atherosclerosis. M L Mckenney, K A Schultz, J H Boyd, J P Byrd, M Alloosh, S D Teague, A A Arce-Esquivel, J N Fain, M H Laughlin, H S Sacks, J Cardiothorac Surg. 92McKenney ML, Schultz KA, Boyd JH, Byrd JP, Alloosh M, Teague SD, Arce- Esquivel AA, Fain JN, Laughlin MH, Sacks HS, et al. Epicardial adipose excision slows the progression of porcine coronary atherosclerosis. J Cardiothorac Surg. 2014;9:2.\\n\\nRelation between cardiovascular disease risk factors and epicardial adipose tissue density on cardiac computed tomography in patients at high risk of cardiovascular events. B T Franssens, H M Nathoe, T Leiner, Y Van Der Graaf, F L Visseren, Eur J Prev Cardiol. 246Franssens BT, Nathoe HM, Leiner T, van der Graaf Y, Visseren FL. Relation between cardiovascular disease risk factors and epicardial adipose tissue density on cardiac computed tomography in patients at high risk of cardiovascular events. Eur J Prev Cardiol. 2017;24(6):660-70.\\n\\nDetecting human coronary inflammation by imaging perivascular fat. A S Antonopoulos, F Sanna, N Sabharwal, S Thomas, E K Oikonomou, L Herdman, M Margaritis, C Shirodaria, A M Kampoli, I Akoumianakis, Sci Transl Med. 9398Antonopoulos AS, Sanna F, Sabharwal N, Thomas S, Oikonomou EK, Herdman L, Margaritis M, Shirodaria C, Kampoli AM, Akoumianakis I, et al. Detecting human coronary inflammation by imaging perivascular fat. Sci Transl Med. 2017;9(398).\\n\\nNon-invasive detection of coronary inflammation using computed tomography and prediction of residual cardiovascular risk (the CRISP CT study): a post-hoc analysis of prospective outcome data. E K Oikonomou, M Marwan, M Y Desai, J Mancio, A Alashi, Hutt Centeno, E Thomas, S Herdman, L Kotanidis, C P Thomas, K E , Lancet (London). 392Oikonomou EK, Marwan M, Desai MY, Mancio J, Alashi A, Hutt Centeno E, Thomas S, Herdman L, Kotanidis CP, Thomas KE, et al. Non-invasive detection of coronary inflammation using computed tomography and prediction of residual cardiovascular risk (the CRISP CT study): a post-hoc analysis of prospective outcome data. Lancet (London). 2018;392(10151): 929-39.\\n\\nSCCT guidelines on radiation dose and dose-optimization strategies in cardiovascular CT. S S Halliburton, S Abbara, M Y Chen, R Gentry, M Mahesh, G L Raff, L J Shaw, J Hausleiter, J Cardiovasc Comput Tomogr. 54Halliburton SS, Abbara S, Chen MY, Gentry R, Mahesh M, Raff GL, Shaw LJ, Hausleiter J. SCCT guidelines on radiation dose and dose-optimization strategies in cardiovascular CT. J Cardiovasc Comput Tomogr. 2011;5(4):198- 224.\\n\\nComparison of epicardial adipose tissue radiodensity threshold between contrast and non-contrast enhanced computed tomography scans: a cohort study of derivation and validation. L Xu, Y Xu, R Coulden, E Sonnex, S Hrybouski, I Paterson, C Butler, Atherosclerosis. 275Xu L, Xu Y, Coulden R, Sonnex E, Hrybouski S, Paterson I, Butler C. Comparison of epicardial adipose tissue radiodensity threshold between contrast and non-contrast enhanced computed tomography scans: a cohort study of derivation and validation. Atherosclerosis. 2018;275:74-9.\\n\\nIncreased regional epicardial fat volume associated with reversible myocardial ischemia in patients with suspected coronary artery disease. T Khawaja, C Greer, S R Thadani, T S Kato, K Bhatia, D Shimbo, A Kontak, S Bokhari, A J Einstein, P C Schulze, J Nucl Cardiol. 222Khawaja T, Greer C, Thadani SR, Kato TS, Bhatia K, Shimbo D, Kontak A, Bokhari S, Einstein AJ, Schulze PC. Increased regional epicardial fat volume associated with reversible myocardial ischemia in patients with suspected coronary artery disease. J Nucl Cardiol. 2015;22(2):325-33.\\n\\nAssociation of pericoronary fat volume with atherosclerotic plaque burden in the underlying coronary artery: a segment analysis. A A Mahabadi, N Reinsch, N Lehmann, J Altenbernd, H Kalsch, R M Seibel, R Erbel, S Mohlenkamp, Atherosclerosis. 2111Mahabadi AA, Reinsch N, Lehmann N, Altenbernd J, Kalsch H, Seibel RM, Erbel R, Mohlenkamp S. Association of pericoronary fat volume with atherosclerotic plaque burden in the underlying coronary artery: a segment analysis. Atherosclerosis. 2010;211(1):195-9.\\n\\nMeasurement of fractional flow reserve to assess the functional severity of coronary-artery stenoses. N H Pijls, B De Bruyne, K Peels, P H Van Der Voort, H J Bonnier, Jkjj Bartunek, J J Koolen, N Engl J Med. 33426Pijls NH, De Bruyne B, Peels K, Van Der Voort PH, Bonnier HJ, Bartunek JKJJ, Koolen JJ. Measurement of fractional flow reserve to assess the functional severity of coronary-artery stenoses. N Engl J Med. 1996; 334(26):1703-8.\\n\\nEpicardial adipose tissue and myocardial ischemia assessed by computed tomography perfusion imaging and invasive fractional flow reserve. R G Muthalaly, N Nerlekar, D T Wong, J D Cameron, S K Seneviratne, B S Ko, J Cardiovasc Comput Tomogr. 111Muthalaly RG, Nerlekar N, Wong DT, Cameron JD, Seneviratne SK, Ko BS. Epicardial adipose tissue and myocardial ischemia assessed by computed tomography perfusion imaging and invasive fractional flow reserve. J Cardiovasc Comput Tomogr. 2017;11(1):46-53.\\n\\nLack of association between epicardial fat volume and extent of coronary artery calcification, severity of coronary artery disease, or presence of myocardial perfusion abnormalities in a diverse, symptomatic patient population: results from the CORE320 multicenter study. Y Tanami, M Jinzaki, S Kishi, M Matheson, A L Vavere, C E Rochitte, M Dewey, M Y Chen, M E Clouse, C Cox, Circ Cardiovasc Imaging. 832676Tanami Y, Jinzaki M, Kishi S, Matheson M, Vavere AL, Rochitte CE, Dewey M, Chen MY, Clouse ME, Cox C, et al. Lack of association between epicardial fat volume and extent of coronary artery calcification, severity of coronary artery disease, or presence of myocardial perfusion abnormalities in a diverse, symptomatic patient population: results from the CORE320 multicenter study. Circ Cardiovasc Imaging. 2015;8(3):e002676.\\n\\nIncremental diagnostic value of epicardial adipose tissue for the detection of functionally relevant coronary artery disease. M A Romijn, I Danad, M J Bakkum, W J Stuijfzand, I I Tulevski, G A Somsen, A A Lammertsma, C Van Kuijk, P M Van De Ven, J K Min, Atherosclerosis. 2421Romijn MA, Danad I, Bakkum MJ, Stuijfzand WJ, Tulevski II, Somsen GA, Lammertsma AA, van Kuijk C, van de Ven PM, Min JK, et al. Incremental diagnostic value of epicardial adipose tissue for the detection of functionally relevant coronary artery disease. Atherosclerosis. 2015;242(1): 161-6.\\n\\nEpicardial adipose tissue volume assessed by computed tomography and coronary artery disease: a systematic review and meta-analysis. J Mancio, D Azevedo, F Saraiva, A I Azevedo, G Pires-Morais, A Leite-Moreira, I Falcao-Pires, N Lunet, N Bettencourt, Eur Heart J Cardiovasc Imaging. 195Mancio J, Azevedo D, Saraiva F, Azevedo AI, Pires-Morais G, Leite-Moreira A, Falcao-Pires I, Lunet N, Bettencourt N. Epicardial adipose tissue volume assessed by computed tomography and coronary artery disease: a systematic review and meta-analysis. Eur Heart J Cardiovasc Imaging. 2018; 19(5):490-7.\\n\\nPerivascular adipose tissue as a cause of atherosclerosis. S N Verhagen, F L Visseren, Atherosclerosis. 2141Verhagen SN, Visseren FL. Perivascular adipose tissue as a cause of atherosclerosis. Atherosclerosis. 2011;214(1):3-10.\\n\\nCoronary perivascular adipose tissue characteristics are related to atherosclerotic plaque size and composition. A post-mortem study. S N Verhagen, A Vink, Y Van Der Graaf, F L Visseren, Atherosclerosis. 2251Verhagen SN, Vink A, van der Graaf Y, Visseren FL. Coronary perivascular adipose tissue characteristics are related to atherosclerotic plaque size and composition. A post-mortem study. Atherosclerosis. 2012;225(1):99-104.\\n\\nSegmental peri-coronary epicardial adipose tissue volume and coronary plaque characteristics. M Hassan, K Said, H Rizk, F Elmogy, M Donya, M Houseni, M Yacoub, Eur Heart J Cardiovasc Imaging. 1710Hassan M, Said K, Rizk H, ElMogy F, Donya M, Houseni M, Yacoub M. Segmental peri-coronary epicardial adipose tissue volume and coronary plaque characteristics. Eur Heart J Cardiovasc Imaging. 2016;17(10):1169-77.\\n\\nPericoronary fat volume but not attenuation differentiates culprit lesions in patients with myocardial infarction. B Balcer, I Dykun, T Schlosser, M Forsting, T Rassaf, A A Mahabadi, Atherosclerosis. 276Balcer B, Dykun I, Schlosser T, Forsting M, Rassaf T, Mahabadi AA. Pericoronary fat volume but not attenuation differentiates culprit lesions in patients with myocardial infarction. Atherosclerosis. 2018;276:182-8.\\n\\nSecretion of adipocytokines by perivascular adipose tissue near stenotic and non-stenotic coronary artery segments in patients undergoing CABG. S N Verhagen, M P Buijsrogge, A Vink, L A Van Herwerden, Y Van Der Graaf, F L Visseren, Atherosclerosis. 2331Verhagen SN, Buijsrogge MP, Vink A, van Herwerden LA, van der Graaf Y, Visseren FL. Secretion of adipocytokines by perivascular adipose tissue near stenotic and non-stenotic coronary artery segments in patients undergoing CABG. Atherosclerosis. 2014;233(1):242-7.\\n\\nAssociation between omentin-1 expression in human epicardial adipose tissue and coronary atherosclerosis. Y Du, Ji Q Cai, L Huang, F Lai, Y Liu, Y Yu, J Han, B Zhu, E Zhang, J , Cardiovasc Diabetol. 1590Du Y, Ji Q, Cai L, Huang F, Lai Y, Liu Y, Yu J, Han B, Zhu E, Zhang J, et al. Association between omentin-1 expression in human epicardial adipose tissue and coronary atherosclerosis. Cardiovasc Diabetol. 2016;15:90.\\n\\nCT attenuation of Pericoronary adipose tissue in Normal versus atherosclerotic coronary segments as defined by intravascular ultrasound. M Marwan, M Hell, A Schuhback, S Gauss, D Bittner, T Pflederer, S Achenbach, J Comput Assist Tomogr. 415Marwan M, Hell M, Schuhback A, Gauss S, Bittner D, Pflederer T, Achenbach S. CT attenuation of Pericoronary adipose tissue in Normal versus atherosclerotic coronary segments as defined by intravascular ultrasound. J Comput Assist Tomogr. 2017;41(5):762-7.\\n\\nPericardial fat inflammation correlates with coronary artery disease. M Konishi, S Sugiyama, Y Sato, S Oshima, K Sugamura, T Nozaki, K Ohba, J Matsubara, H Sumida, Y Nagayoshi, Atherosclerosis. 2132Konishi M, Sugiyama S, Sato Y, Oshima S, Sugamura K, Nozaki T, Ohba K, Matsubara J, Sumida H, Nagayoshi Y, et al. Pericardial fat inflammation correlates with coronary artery disease. Atherosclerosis. 2010;213(2):649-55.\\n\\nPeri-coronary adipose tissue density is associated with (18) F-sodium fluoride coronary uptake in stable patients with high-risk plaques. J Kwiecinski, D Dey, S Cadet, S E Lee, Y Otaki, P T Huynh, M K Doris, E Eisenberg, M Yun, M A Jansen, J Am Coll Cardiol Img. 1210Kwiecinski J, Dey D, Cadet S, Lee SE, Otaki Y, Huynh PT, Doris MK, Eisenberg E, Yun M, Jansen MA, et al. Peri-coronary adipose tissue density is associated with (18) F-sodium fluoride coronary uptake in stable patients with high-risk plaques. J Am Coll Cardiol Img. 2019;12(10):2000-10.\\n\\nEpicardial adipose tissue volume but not density is an independent predictor for myocardial ischemia. M M Hell, X Ding, M Rubeaux, P Slomka, H Gransar, D Terzopoulos, S Hayes, M Marwan, S Achenbach, D S Berman, J Cardiovasc Comput Tomogr. 102Hell MM, Ding X, Rubeaux M, Slomka P, Gransar H, Terzopoulos D, Hayes S, Marwan M, Achenbach S, Berman DS, et al. Epicardial adipose tissue volume but not density is an independent predictor for myocardial ischemia. J Cardiovasc Comput Tomogr. 2016;10(2):141-9.\\n\\nPublisher\\'s Note. Publisher\\'s Note\\n\\nSpringer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations. Springer Nature remains neutral with regard to jurisdictional claims in published maps and institutional affiliations.\\n'"
            ]
          },
          "execution_count": 22,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "full_doc_example = full_text['text'][8]\n",
        "full_doc_example"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "id": "kMd8xm1MGFfi"
      },
      "outputs": [],
      "source": [
        "openai_api_key = \"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\"\n",
        "\n",
        "llm = OpenAI(temperature=0.5, openai_api_key = openai_api_key)\n",
        "\n",
        "text_splitter = CharacterTextSplitter(chunk_size=2000, chunk_overlap=0)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 23,
      "metadata": {
        "id": "KuC9k5l0G4-P"
      },
      "outputs": [
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "2023-07-20 21:27:01 ip-172-31-90-198 sentence_transformers.SentenceTransformer[1829] INFO Load pretrained SentenceTransformer: t5-large\n",
            "2023-07-20 21:27:01 ip-172-31-90-198 sentence_transformers.SentenceTransformer[1829] WARNING No sentence-transformers model found with name /home/ubuntu/.cache/torch/sentence_transformers/t5-large. Creating a new one with MEAN pooling.\n",
            "2023-07-20 21:27:36 ip-172-31-90-198 sentence_transformers.SentenceTransformer[1829] INFO Use pytorch device: cpu\n",
            "2023-07-20 21:27:36 ip-172-31-90-198 chromadb.telemetry.posthog[1829] INFO Anonymized telemetry enabled. See https://docs.trychroma.com/telemetry for more information.\n",
            "2023-07-20 21:27:37 ip-172-31-90-198 clickhouse_connect.driver.ctypes[1829] INFO Successfully imported ClickHouse Connect C data optimizations\n",
            "2023-07-20 21:27:37 ip-172-31-90-198 clickhouse_connect.json_impl[1829] INFO Using python library for writing JSON byte strings\n"
          ]
        },
        {
          "data": {
            "application/vnd.jupyter.widget-view+json": {
              "model_id": "8b388c950035439da2772f79dc201c60",
              "version_major": 2,
              "version_minor": 0
            },
            "text/plain": [
              "Batches:   0%|          | 0/1 [00:00<?, ?it/s]"
            ]
          },
          "metadata": {},
          "output_type": "display_data"
        },
        {
          "name": "stderr",
          "output_type": "stream",
          "text": [
            "Asking to truncate to max_length but no maximum length is provided and the model has no predefined maximum length. Default to no truncation.\n",
            "2023-07-20 21:27:54 ip-172-31-90-198 openai[1829] INFO error_code=None error_message='Request failed due to server shutdown' error_param=None error_type=server_error message='OpenAI API error received' stream_error=False\n",
            "2023-07-20 21:27:54 ip-172-31-90-198 langchain.llms.openai[1829] WARNING Retrying langchain.llms.openai.completion_with_retry.<locals>._completion_with_retry in 4.0 seconds as it raised APIError: Request failed due to server shutdown {\n",
            "  \"error\": {\n",
            "    \"message\": \"Request failed due to server shutdown\",\n",
            "    \"type\": \"server_error\",\n",
            "    \"param\": null,\n",
            "    \"code\": null\n",
            "  }\n",
            "}\n",
            " 500 {'error': {'message': 'Request failed due to server shutdown', 'type': 'server_error', 'param': None, 'code': None}} {'Date': 'Thu, 20 Jul 2023 21:27:54 GMT', 'Content-Type': 'application/json', 'Content-Length': '141', 'Connection': 'keep-alive', 'access-control-allow-origin': '*', 'openai-model': 'text-davinci-003', 'openai-organization': 'user-4ce7xbchph7z3holbhpge0c4', 'openai-processing-ms': '3369', 'openai-version': '2020-10-01', 'strict-transport-security': 'max-age=15724800; includeSubDomains', 'x-ratelimit-limit-requests': '200', 'x-ratelimit-limit-tokens': '150000', 'x-ratelimit-remaining-requests': '198', 'x-ratelimit-remaining-tokens': '149743', 'x-ratelimit-reset-requests': '14m20.824s', 'x-ratelimit-reset-tokens': '102ms', 'x-request-id': '81faa6144204b981c2314107e8cebac2', 'CF-Cache-Status': 'DYNAMIC', 'Server': 'cloudflare', 'CF-RAY': '7e9e4a5c39911ff1-IAD', 'alt-svc': 'h3=\":443\"; ma=86400'}.\n"
          ]
        },
        {
          "data": {
            "text/plain": [
              "' This study surveyed over 276,000 individuals in the Amhara region of Ethiopia to estimate the prevalence of trachoma among children aged 1-9 years. Trachoma graders underwent a 5-7 day training program prior to the survey, which included classroom sessions and field-based practice. Data collection was conducted using Samsung Galaxy tablets and the Amharic language, and results were weighted, age-sex adjusted, and tested for correlation between trachoma and other diseases. Intraclass correlation coefficient was also calculated to assess clustering.'"
            ]
          },
          "execution_count": 23,
          "metadata": {},
          "output_type": "execute_result"
        }
      ],
      "source": [
        "## Use text_splitter to split documents\n",
        "## Use Documents package to format text\n",
        "text = text_splitter.split_text(full_text_example1)\n",
        "docs = [Document(page_content=t) for t in text[:3]]\n",
        "embedding_function = SentenceTransformerEmbeddings(model_name=\"t5-large\")\n",
        "vectordb = Chroma.from_documents(docs, embedding_function)\n",
        "\n",
        "chain = load_summarize_chain(llm, chain_type=\"map_reduce\")\n",
        "chain.run(docs)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "### Query to extract full text from postgres\n",
        "def requested_papers(courpusId_list, summary=False):\n",
        "    if summary:\n",
        "        table_name = 'abstracts_encodings'\n",
        "    else:\n",
        "        table_name = 'fulltext'\n",
        "\n",
        "    client_request = text(f'''\n",
        "        SELECT empty, corpusId, text, id FROM {table_name} ''')\n",
        "\n",
        "    with engine.connect() as conn: \n",
        "        query = conn.execute(client_request).filter(fulltext.corpusid.in_(corpusID_list))\n",
        "\n",
        "    solution_papers = pd.DataFrame(query.fetchall())\n",
        "    return solution_papers"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "### Huggingface transformer abstract summarization T5 and LED\n",
        "def hugging_face_summarization(model, tokenizer, dataset):\n",
        "    abstract_holder = []\n",
        "    for i in dataset['abstracts']:\n",
        "        tokenized_abstract = tokenizer([\"Summarize\" + i], truncation=True, padding=True, max_length=True)\n",
        "        output_encodings = model.genertae(tokenized_abstract.input_ids,\n",
        "                                           num_beams = 4,\n",
        "                                           top_k = 10,\n",
        "                                           top_p = 80,\n",
        "                                           no_repeat_ngram_size = 3,\n",
        "                                           max_length = 100,\n",
        "                                           min_length = 50)\n",
        "        abstract_summary = [tokenizer.decode(w, skip_special_tokens=True, clean_up_tokenization_spaces=True) for w in output_encodings]\n",
        "        abstract_holder.append(abstract_summary)\n",
        "    return abstract_holder"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": [
        "user_input = 'Phrenic nerve palsy'\n",
        "input_string =  \" & \".join(user_input.split(\" \"))\n",
        "\n",
        "sql = f'''\n",
        "SELECT\n",
        "    attributes.title,\n",
        "    abstract, \n",
        "    corpusid             \n",
        "FROM abstract_encodings_test4\n",
        "JOIN attributes using (corpusid)\n",
        "WHERE ts_left_750 @@ websearch_to_tsquery('english','{input_string}')\n",
        "order by attributes.referencecount desc, attributes.publicationdate, attributes.citationcount desc\n",
        "limit 5\n",
        "\n",
        "'''\n",
        "with engine.connect() as conn:\n",
        "    query = conn.execute(text(sql))\n",
        "    \n",
        "df = pd.DataFrame(query.fetchall())\n",
        "df"
      ]
    },
    {
      "attachments": {},
      "cell_type": "markdown",
      "metadata": {},
      "source": [
        "#### Question & Answer"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [
        {
          "ename": "ValidationError",
          "evalue": "1 validation error for OpenAI\n__root__\n  Did not find openai_api_key, please add an environment variable `OPENAI_API_KEY` which contains it, or pass  `openai_api_key` as a named parameter. (type=value_error)",
          "output_type": "error",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValidationError\u001b[0m                           Traceback (most recent call last)",
            "Cell \u001b[0;32mIn[16], line 4\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mlangchain\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mvectorstores\u001b[39;00m \u001b[39mimport\u001b[39;00m Chroma\n\u001b[1;32m      2\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mlangchain\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39membeddings\u001b[39;00m \u001b[39mimport\u001b[39;00m OpenAIEmbeddings\n\u001b[0;32m----> 4\u001b[0m qa \u001b[39m=\u001b[39m VectorDBQA\u001b[39m.\u001b[39mfrom_chain_type(llm\u001b[39m=\u001b[39mOpenAI(), chain_type\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mreduce\u001b[39m\u001b[39m\"\u001b[39m, vectorstore\u001b[39m=\u001b[39mlang_collect)\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/langchain/load/serializable.py:74\u001b[0m, in \u001b[0;36mSerializable.__init__\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m     73\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__init__\u001b[39m(\u001b[39mself\u001b[39m, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs: Any) \u001b[39m-\u001b[39m\u001b[39m>\u001b[39m \u001b[39mNone\u001b[39;00m:\n\u001b[0;32m---> 74\u001b[0m     \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49m\u001b[39m__init__\u001b[39;49m(\u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m     75\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_lc_kwargs \u001b[39m=\u001b[39m kwargs\n",
            "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/pydantic/main.py:341\u001b[0m, in \u001b[0;36mpydantic.main.BaseModel.__init__\u001b[0;34m()\u001b[0m\n",
            "\u001b[0;31mValidationError\u001b[0m: 1 validation error for OpenAI\n__root__\n  Did not find openai_api_key, please add an environment variable `OPENAI_API_KEY` which contains it, or pass  `openai_api_key` as a named parameter. (type=value_error)"
          ]
        }
      ],
      "source": [
        "from langchain.vectorstores import Chroma\n",
        "from langchain.embeddings import OpenAIEmbeddings\n",
        "\n",
        "qa = VectorDBQA.from_chain_type(llm=OpenAI(), chain_type=\"map_reduce\", vectorstore=lang_collect)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {},
      "outputs": [],
      "source": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.11"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
