{"cells":[{"cell_type":"code","source":"import pandas as _hex_pandas\nimport datetime as _hex_datetime\nimport json as _hex_json","execution_count":null,"metadata":{},"outputs":[]},{"cell_type":"code","source":"hex_scheduled = _hex_json.loads(\"false\")","outputs":[],"execution_count":null,"metadata":{}},{"cell_type":"code","source":"hex_user_email = _hex_json.loads(\"\\\"example-user@example.com\\\"\")","outputs":[],"execution_count":null,"metadata":{}},{"cell_type":"code","source":"hex_run_context = _hex_json.loads(\"\\\"logic\\\"\")","outputs":[],"execution_count":null,"metadata":{}},{"cell_type":"code","source":"hex_timezone = _hex_json.loads(\"\\\"UTC\\\"\")","outputs":[],"execution_count":null,"metadata":{}},{"cell_type":"code","source":"hex_project_id = _hex_json.loads(\"\\\"1e7b90bc-1996-4336-8653-50313b6d2f75\\\"\")","outputs":[],"execution_count":null,"metadata":{}},{"cell_type":"code","source":"hex_color_palette = _hex_json.loads(\"[\\\"#4C78A8\\\",\\\"#F58518\\\",\\\"#E45756\\\",\\\"#72B7B2\\\",\\\"#54A24B\\\",\\\"#EECA3B\\\",\\\"#B279A2\\\",\\\"#FF9DA6\\\",\\\"#9D755D\\\",\\\"#BAB0AC\\\"]\")","outputs":[],"execution_count":null,"metadata":{}},{"cell_type":"code","source":"import pandas as pd\nfrom sqlalchemy import create_engine\nfrom sqlalchemy.sql import text\nfrom pandas import json_normalize\nfrom scipy.spatial.distance import cdist\nimport sys\npd.set_option('display.max_colwidth', None)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"!pip install -U huggingface_hub\n!pip install -U chromadb\n!pip install -U transformers\n!pip install -U tokenizers\nfrom transformers import (\n    AutoTokenizer,\n    pipeline,\n    AutoModelForSeq2SeqLM,\n    AutoModel,\n    BigBirdForQuestionAnswering,\n    BigBirdPegasusForConditionalGeneration,\n)\nimport os\nimport tensorflow as tf\nimport torch\nimport ast","metadata":{},"execution_count":null,"outputs":[{"data":{"text/plain":"Requirement already satisfied: huggingface_hub in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (0.11.1)\nCollecting huggingface_hub\n  Obtaining dependency information for huggingface_hub from https://files.pythonhosted.org/packages/7f/c4/adcbe9a696c135578cabcbdd7331332daad4d49b7c43688bc2d36b3a47d2/huggingface_hub-0.16.4-py3-none-any.whl.metadata\n  Downloading huggingface_hub-0.16.4-py3-none-any.whl.metadata (12 kB)\nRequirement already satisfied: filelock in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from huggingface_hub) (3.8.0)\nRequirement already satisfied: fsspec in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from huggingface_hub) (2022.11.0)\nRequirement already satisfied: requests in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from huggingface_hub) (2.26.0)\nRequirement already satisfied: tqdm>=4.42.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from huggingface_hub) (4.65.0)\nRequirement already satisfied: pyyaml>=5.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from huggingface_hub) (6.0)\nRequirement already satisfied: typing-extensions>=3.7.4.3 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from huggingface_hub) (4.4.0)\nRequirement already satisfied: packaging>=20.9 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from huggingface_hub) (23.1)\nRequirement already satisfied: urllib3<1.27,>=1.21.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests->huggingface_hub) (1.26.7)\nRequirement already satisfied: certifi>=2017.4.17 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests->huggingface_hub) (2021.10.8)\nRequirement already satisfied: charset-normalizer~=2.0.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests->huggingface_hub) (2.0.6)\nRequirement already satisfied: idna<4,>=2.5 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests->huggingface_hub) (3.2)\nDownloading huggingface_hub-0.16.4-py3-none-any.whl (268 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m268.8/268.8 kB\u001b[0m \u001b[31m29.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25h\u001b[33mDEPRECATION: dropbox 11.25.0 has a non-standard dependency specifier stone>=2.*. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of dropbox or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063\u001b[0m\u001b[33m\n\u001b[0mInstalling collected packages: huggingface_hub\n  Attempting uninstall: huggingface_hub\n    Found existing installation: huggingface-hub 0.11.1\n    Uninstalling huggingface-hub-0.11.1:\n      Successfully uninstalled huggingface-hub-0.11.1\nSuccessfully installed huggingface_hub-0.16.4\n\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\nCollecting chromadb\n  Obtaining dependency information for chromadb from https://files.pythonhosted.org/packages/5f/86/599e1482723dea206ad269aece4e4907f4a981840fcb6c731d39f79098ed/chromadb-0.4.5-py3-none-any.whl.metadata\n  Downloading chromadb-0.4.5-py3-none-any.whl.metadata (6.8 kB)\nCollecting requests>=2.28 (from chromadb)\n  Obtaining dependency information for requests>=2.28 from https://files.pythonhosted.org/packages/70/8e/0e2d847013cb52cd35b38c009bb167a1a26b2ce6cd6965bf26b47bc0bf44/requests-2.31.0-py3-none-any.whl.metadata\n  Downloading requests-2.31.0-py3-none-any.whl.metadata (4.6 kB)\nCollecting pydantic<2.0,>=1.9 (from chromadb)\n  Obtaining dependency information for pydantic<2.0,>=1.9 from https://files.pythonhosted.org/packages/02/bd/2c6a34eeffcc437fca7d5c4b6cf7745fcc806842de5fced482d4cdba55f0/pydantic-1.10.12-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata\n  Downloading pydantic-1.10.12-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (149 kB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m149.3/149.3 kB\u001b[0m \u001b[31m18.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hCollecting chroma-hnswlib==0.7.2 (from chromadb)\n  Downloading chroma-hnswlib-0.7.2.tar.gz (31 kB)\n  Installing build dependencies ... \u001b[?25l-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \bdone\n\u001b[?25h  Getting requirements to build wheel ... \u001b[?25l-\b \bdone\n\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25l-\b \bdone\n\u001b[?25hCollecting fastapi<0.100.0,>=0.95.2 (from chromadb)\n  Obtaining dependency information for fastapi<0.100.0,>=0.95.2 from https://files.pythonhosted.org/packages/73/eb/03b691afa0b5ffa1e93ed34f97ec1e7855c758efbdcfb16c209af0b0506b/fastapi-0.99.1-py3-none-any.whl.metadata\n  Downloading fastapi-0.99.1-py3-none-any.whl.metadata (23 kB)\nCollecting uvicorn[standard]>=0.18.3 (from chromadb)\n  Obtaining dependency information for uvicorn[standard]>=0.18.3 from https://files.pythonhosted.org/packages/79/96/b0882a1c3f7ef3dd86879e041212ae5b62b4bd352320889231cc735a8e8f/uvicorn-0.23.2-py3-none-any.whl.metadata\n  Downloading uvicorn-0.23.2-py3-none-any.whl.metadata (6.2 kB)\nRequirement already satisfied: numpy>=1.21.6 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from chromadb) (1.23.4)\nCollecting posthog>=2.4.0 (from chromadb)\n  Downloading posthog-3.0.1-py2.py3-none-any.whl (37 kB)\nCollecting typing-extensions>=4.5.0 (from chromadb)\n  Obtaining dependency information for typing-extensions>=4.5.0 from https://files.pythonhosted.org/packages/ec/6b/63cc3df74987c36fe26157ee12e09e8f9db4de771e0f3404263117e75b95/typing_extensions-4.7.1-py3-none-any.whl.metadata\n  Downloading typing_extensions-4.7.1-py3-none-any.whl.metadata (3.1 kB)\nCollecting pulsar-client>=3.1.0 (from chromadb)\n  Obtaining dependency information for pulsar-client>=3.1.0 from https://files.pythonhosted.org/packages/f1/c2/e71786bd365a4ad33993aa7aa80e6e4c0e586bb0286bb92e5aeedd9a7a90/pulsar_client-3.2.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata\n  Downloading pulsar_client-3.2.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.0 kB)\nCollecting onnxruntime>=1.14.1 (from chromadb)\n  Obtaining dependency information for onnxruntime>=1.14.1 from https://files.pythonhosted.org/packages/50/d5/f156d808c9cc59e7a8b87843a5313719d01eec7d3b17ca3f6b6f0fbee0f0/onnxruntime-1.15.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata\n  Downloading onnxruntime-1.15.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.0 kB)\nRequirement already satisfied: tokenizers>=0.13.2 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from chromadb) (0.13.2)\nCollecting pypika>=0.48.9 (from chromadb)\n  Downloading PyPika-0.48.9.tar.gz (67 kB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m67.3/67.3 kB\u001b[0m \u001b[31m20.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25h  Installing build dependencies ... \u001b[?25l-\b \b\\\b \b|\b \bdone\n\u001b[?25h  Getting requirements to build wheel ... \u001b[?25l-\b \bdone\n\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25l-\b \bdone\n\u001b[?25hRequirement already satisfied: tqdm>=4.65.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from chromadb) (4.65.0)\nCollecting overrides>=7.3.1 (from chromadb)\n  Downloading overrides-7.3.1-py3-none-any.whl (17 kB)\nRequirement already satisfied: importlib-resources in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from chromadb) (5.10.0)\nCollecting starlette<0.28.0,>=0.27.0 (from fastapi<0.100.0,>=0.95.2->chromadb)\n  Obtaining dependency information for starlette<0.28.0,>=0.27.0 from https://files.pythonhosted.org/packages/58/f8/e2cca22387965584a409795913b774235752be4176d276714e15e1a58884/starlette-0.27.0-py3-none-any.whl.metadata\n  Downloading starlette-0.27.0-py3-none-any.whl.metadata (5.8 kB)\nCollecting coloredlogs (from onnxruntime>=1.14.1->chromadb)\n  Downloading coloredlogs-15.0.1-py2.py3-none-any.whl (46 kB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m46.0/46.0 kB\u001b[0m \u001b[31m13.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hRequirement already satisfied: flatbuffers in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from onnxruntime>=1.14.1->chromadb) (2.0.7)\nRequirement already satisfied: packaging in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from onnxruntime>=1.14.1->chromadb) (23.1)\nRequirement already satisfied: protobuf in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from onnxruntime>=1.14.1->chromadb) (3.20.3)\nRequirement already satisfied: sympy in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from onnxruntime>=1.14.1->chromadb) (1.12)\nRequirement already satisfied: six>=1.5 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from posthog>=2.4.0->chromadb) (1.15.0)\nCollecting monotonic>=1.5 (from posthog>=2.4.0->chromadb)\n  Downloading monotonic-1.6-py2.py3-none-any.whl (8.2 kB)\nRequirement already satisfied: backoff>=1.10.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from posthog>=2.4.0->chromadb) (2.1.2)\nRequirement already satisfied: python-dateutil>2.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from posthog>=2.4.0->chromadb) (2.8.2)\nRequirement already satisfied: certifi in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from pulsar-client>=3.1.0->chromadb) (2021.10.8)\nRequirement already satisfied: charset-normalizer<4,>=2 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests>=2.28->chromadb) (2.0.6)\nRequirement already satisfied: idna<4,>=2.5 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests>=2.28->chromadb) (3.2)\nRequirement already satisfied: urllib3<3,>=1.21.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests>=2.28->chromadb) (1.26.7)\nRequirement already satisfied: click>=7.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (8.0.1)\nCollecting h11>=0.8 (from uvicorn[standard]>=0.18.3->chromadb)\n  Downloading h11-0.14.0-py3-none-any.whl (58 kB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.3/58.3 kB\u001b[0m \u001b[31m16.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hCollecting httptools>=0.5.0 (from uvicorn[standard]>=0.18.3->chromadb)\n  Obtaining dependency information for httptools>=0.5.0 from https://files.pythonhosted.org/packages/28/7c/5ddc99737fb141bd9077f45af23e9d3c83496b4c04bf463e4e72f57043cd/httptools-0.6.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata\n  Downloading httptools-0.6.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.6 kB)\nCollecting python-dotenv>=0.13 (from uvicorn[standard]>=0.18.3->chromadb)\n  Downloading python_dotenv-1.0.0-py3-none-any.whl (19 kB)\nRequirement already satisfied: pyyaml>=5.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from uvicorn[standard]>=0.18.3->chromadb) (6.0)\nCollecting uvloop!=0.15.0,!=0.15.1,>=0.14.0 (from uvicorn[standard]>=0.18.3->chromadb)\n  Downloading uvloop-0.17.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (4.2 MB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m4.2/4.2 MB\u001b[0m \u001b[31m104.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m\n\u001b[?25hCollecting watchfiles>=0.13 (from uvicorn[standard]>=0.18.3->chromadb)\n  Downloading watchfiles-0.19.0-cp37-abi3-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.3 MB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.3/1.3 MB\u001b[0m \u001b[31m100.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hCollecting websockets>=10.4 (from uvicorn[standard]>=0.18.3->chromadb)\n  Downloading websockets-11.0.3-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (129 kB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m129.7/129.7 kB\u001b[0m \u001b[31m31.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hRequirement already satisfied: zipp>=3.1.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from importlib-resources->chromadb) (3.10.0)\nRequirement already satisfied: anyio<5,>=3.4.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from starlette<0.28.0,>=0.27.0->fastapi<0.100.0,>=0.95.2->chromadb) (3.6.2)\nCollecting humanfriendly>=9.1 (from coloredlogs->onnxruntime>=1.14.1->chromadb)\n  Downloading humanfriendly-10.0-py2.py3-none-any.whl (86 kB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m86.8/86.8 kB\u001b[0m \u001b[31m25.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hRequirement already satisfied: mpmath>=0.19 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from sympy->onnxruntime>=1.14.1->chromadb) (1.3.0)\nRequirement already satisfied: sniffio>=1.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from anyio<5,>=3.4.0->starlette<0.28.0,>=0.27.0->fastapi<0.100.0,>=0.95.2->chromadb) (1.3.0)\nDownloading chromadb-0.4.5-py3-none-any.whl (402 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m402.8/402.8 kB\u001b[0m \u001b[31m73.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading fastapi-0.99.1-py3-none-any.whl (58 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m58.4/58.4 kB\u001b[0m \u001b[31m19.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading onnxruntime-1.15.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.9 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.9/5.9 MB\u001b[0m \u001b[31m104.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m\n\u001b[?25hDownloading pulsar_client-3.2.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (5.3 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m5.3/5.3 MB\u001b[0m \u001b[31m104.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m\n\u001b[?25hDownloading pydantic-1.10.12-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m106.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading requests-2.31.0-py3-none-any.whl (62 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m62.6/62.6 kB\u001b[0m \u001b[31m18.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading typing_extensions-4.7.1-py3-none-any.whl (33 kB)\nDownloading httptools-0.6.0-cp39-cp39-manylinux_2_5_x86_64.manylinux1_x86_64.manylinux_2_17_x86_64.manylinux2014_x86_64.whl (432 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m432.5/432.5 kB\u001b[0m \u001b[31m58.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading starlette-0.27.0-py3-none-any.whl (66 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m67.0/67.0 kB\u001b[0m \u001b[31m18.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hDownloading uvicorn-0.23.2-py3-none-any.whl (59 kB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m59.5/59.5 kB\u001b[0m \u001b[31m18.7 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hBuilding wheels for collected packages: chroma-hnswlib, pypika\n  Building wheel for chroma-hnswlib (pyproject.toml) ... \u001b[?25l-\b \b\\\b \b|\b \b/\b \b-\b \b\\\b \b|\b \b/\b \b-\b \bdone\n\u001b[?25h  Created wheel for chroma-hnswlib: filename=chroma_hnswlib-0.7.2-cp39-cp39-linux_x86_64.whl size=2283566 sha256=9c71fecbb2e2b6a111c7cbbaee5ebd6b48a6544cf6683da424a14e7ec28a2fbe\n  Stored in directory: /home/hexuser/.cache/pip/wheels/ba/4f/be/e63b0303c720e68e7f78debdc7e2032d63b0d80353ba8d79e1\n  Building wheel for pypika (pyproject.toml) ... \u001b[?25l-\b \bdone\n\u001b[?25h  Created wheel for pypika: filename=PyPika-0.48.9-py2.py3-none-any.whl size=53723 sha256=69b83d9fa6dc5d221c7f7d74544e72791da80ef0095263fd3219074d7db8b1fc\n  Stored in directory: /home/hexuser/.cache/pip/wheels/f7/02/64/d541eac67ec459309d1fb19e727f58ecf7ffb4a8bf42d4cfe5\nSuccessfully built chroma-hnswlib pypika\n\u001b[33mDEPRECATION: dropbox 11.25.0 has a non-standard dependency specifier stone>=2.*. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of dropbox or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063\u001b[0m\u001b[33m\n\u001b[0mInstalling collected packages: pypika, monotonic, websockets, uvloop, typing-extensions, requests, python-dotenv, pulsar-client, overrides, humanfriendly, httptools, h11, chroma-hnswlib, watchfiles, uvicorn, starlette, pydantic, posthog, coloredlogs, onnxruntime, fastapi, chromadb\n  Attempting uninstall: websockets\n    Found existing installation: websockets 9.1\n    Uninstalling websockets-9.1:\n      Successfully uninstalled websockets-9.1\n  Attempting uninstall: typing-extensions\n    Found existing installation: typing_extensions 4.4.0\n    Uninstalling typing_extensions-4.4.0:\n      Successfully uninstalled typing_extensions-4.4.0\n  Attempting uninstall: requests\n    Found existing installation: requests 2.26.0\n    Uninstalling requests-2.26.0:\n      Successfully uninstalled requests-2.26.0\n  Attempting uninstall: pydantic\n    Found existing installation: pydantic 1.8.2\n    Uninstalling pydantic-1.8.2:\n      Successfully uninstalled pydantic-1.8.2\n\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\npython-kernel-startup 0.1.0 requires cryptography<4.0.0,>=3.3.2, but you have cryptography 40.0.2 which is incompatible.\npython-kernel-startup 0.1.0 requires snowflake-connector-python==3.0.0, but you have snowflake-connector-python 3.0.4 which is incompatible.\npython-kernel-startup 0.1.0 requires vegafusion==1.4.0rc2, but you have vegafusion 1.4.0rc5 which is incompatible.\npython-kernel-startup 0.1.0 requires vegafusion-python-embed==1.4.0rc2, but you have vegafusion-python-embed 1.4.0rc5 which is incompatible.\nredshift-connector 2.0.908 requires requests<2.28.1,>=2.23.0, but you have requests 2.31.0 which is incompatible.\nspacy 3.1.3 requires pydantic!=1.8,!=1.8.1,<1.9.0,>=1.7.4, but you have pydantic 1.10.12 which is incompatible.\nthinc 8.0.10 requires pydantic!=1.8,!=1.8.1,<1.9.0,>=1.7.4, but you have pydantic 1.10.12 which is incompatible.\nweb3 5.29.2 requires websockets<10,>=9.1, but you have websockets 11.0.3 which is incompatible.\u001b[0m\u001b[31m\n\u001b[0mSuccessfully installed chroma-hnswlib-0.7.2 chromadb-0.4.5 coloredlogs-15.0.1 fastapi-0.99.1 h11-0.14.0 httptools-0.6.0 humanfriendly-10.0 monotonic-1.6 onnxruntime-1.15.1 overrides-7.3.1 posthog-3.0.1 pulsar-client-3.2.0 pydantic-1.10.12 pypika-0.48.9 python-dotenv-1.0.0 requests-2.31.0 starlette-0.27.0 typing-extensions-4.7.1 uvicorn-0.23.2 uvloop-0.17.0 watchfiles-0.19.0 websockets-11.0.3\n\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\nRequirement already satisfied: transformers in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (4.25.1)\nCollecting transformers\n  Obtaining dependency information for transformers from https://files.pythonhosted.org/packages/21/02/ae8e595f45b6c8edee07913892b3b41f5f5f273962ad98851dc6a564bbb9/transformers-4.31.0-py3-none-any.whl.metadata\n  Downloading transformers-4.31.0-py3-none-any.whl.metadata (116 kB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m116.9/116.9 kB\u001b[0m \u001b[31m15.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hRequirement already satisfied: filelock in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from transformers) (3.8.0)\nRequirement already satisfied: huggingface-hub<1.0,>=0.14.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from transformers) (0.16.4)\nRequirement already satisfied: numpy>=1.17 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from transformers) (1.23.4)\nRequirement already satisfied: packaging>=20.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from transformers) (23.1)\nRequirement already satisfied: pyyaml>=5.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from transformers) (6.0)\nRequirement already satisfied: regex!=2019.12.17 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from transformers) (2023.5.5)\nRequirement already satisfied: requests in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from transformers) (2.31.0)\nRequirement already satisfied: tokenizers!=0.11.3,<0.14,>=0.11.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from transformers) (0.13.2)\nCollecting safetensors>=0.3.1 (from transformers)\n  Downloading safetensors-0.3.1-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (1.3 MB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.3/1.3 MB\u001b[0m \u001b[31m46.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n\u001b[?25hRequirement already satisfied: tqdm>=4.27 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from transformers) (4.65.0)\nRequirement already satisfied: fsspec in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from huggingface-hub<1.0,>=0.14.1->transformers) (2022.11.0)\nRequirement already satisfied: typing-extensions>=3.7.4.3 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from huggingface-hub<1.0,>=0.14.1->transformers) (4.7.1)\nRequirement already satisfied: charset-normalizer<4,>=2 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests->transformers) (2.0.6)\nRequirement already satisfied: idna<4,>=2.5 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests->transformers) (3.2)\nRequirement already satisfied: urllib3<3,>=1.21.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests->transformers) (1.26.7)\nRequirement already satisfied: certifi>=2017.4.17 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests->transformers) (2021.10.8)\nDownloading transformers-4.31.0-py3-none-any.whl (7.4 MB)\n\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7.4/7.4 MB\u001b[0m \u001b[31m84.9 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n\u001b[?25h\u001b[33mDEPRECATION: dropbox 11.25.0 has a non-standard dependency specifier stone>=2.*. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of dropbox or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063\u001b[0m\u001b[33m\n\u001b[0mInstalling collected packages: safetensors, transformers\n  Attempting uninstall: transformers\n    Found existing installation: transformers 4.25.1\n    Uninstalling transformers-4.25.1:\n      Successfully uninstalled transformers-4.25.1\nSuccessfully installed safetensors-0.3.1 transformers-4.31.0\n\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\nRequirement already satisfied: tokenizers in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (0.13.2)\nCollecting tokenizers\n  Downloading tokenizers-0.13.3-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (7.8 MB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m7.8/7.8 MB\u001b[0m \u001b[31m78.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m00:01\u001b[0m\n\u001b[?25h\u001b[33mDEPRECATION: dropbox 11.25.0 has a non-standard dependency specifier stone>=2.*. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of dropbox or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063\u001b[0m\u001b[33m\n\u001b[0mInstalling collected packages: tokenizers\n  Attempting uninstall: tokenizers\n    Found existing installation: tokenizers 0.13.2\n    Uninstalling tokenizers-0.13.2:\n      Successfully uninstalled tokenizers-0.13.2\nSuccessfully installed tokenizers-0.13.3\n\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"},"execution_count":null,"metadata":{},"output_type":"execute_result"}]},{"cell_type":"code","source":"!pip install psycopg2-binary\n!pip install rouge-score\n!pip install tensorflow\nimport numpy as np\nfrom rouge_score import rouge_scorer\nimport yaml","metadata":{},"execution_count":null,"outputs":[{"data":{"text/plain":"Collecting psycopg2-binary\n  Downloading psycopg2_binary-2.9.6-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.0 MB)\n\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.0/3.0 MB\u001b[0m \u001b[31m78.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m:00:01\u001b[0m\n\u001b[?25h\u001b[33mDEPRECATION: dropbox 11.25.0 has a non-standard dependency specifier stone>=2.*. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of dropbox or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063\u001b[0m\u001b[33m\n\u001b[0mInstalling collected packages: psycopg2-binary\nSuccessfully installed psycopg2-binary-2.9.6\n\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\nCollecting rouge-score\n  Downloading rouge_score-0.1.2.tar.gz (17 kB)\n  Preparing metadata (setup.py) ... \u001b[?25l-\b \bdone\n\u001b[?25hRequirement already satisfied: absl-py in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from rouge-score) (1.1.0)\nRequirement already satisfied: nltk in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from rouge-score) (3.7)\nRequirement already satisfied: numpy in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from rouge-score) (1.23.4)\nRequirement already satisfied: six>=1.14.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from rouge-score) (1.15.0)\nRequirement already satisfied: click in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from nltk->rouge-score) (8.0.1)\nRequirement already satisfied: joblib in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from nltk->rouge-score) (1.2.0)\nRequirement already satisfied: regex>=2021.8.3 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from nltk->rouge-score) (2023.5.5)\nRequirement already satisfied: tqdm in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from nltk->rouge-score) (4.65.0)\nBuilding wheels for collected packages: rouge-score\n  Building wheel for rouge-score (setup.py) ... \u001b[?25l-\b \b\\\b \b|\b \b/\b \bdone\n\u001b[?25h  Created wheel for rouge-score: filename=rouge_score-0.1.2-py3-none-any.whl size=24935 sha256=4f423d6a36c75c435674d05664a889255582244655c9c7751c56b46b6b689488\n  Stored in directory: /home/hexuser/.cache/pip/wheels/9b/3d/39/09558097d3119ca0a4d462df68f22c6f3c1b345ac63a09b86e\nSuccessfully built rouge-score\n\u001b[33mDEPRECATION: dropbox 11.25.0 has a non-standard dependency specifier stone>=2.*. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of dropbox or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063\u001b[0m\u001b[33m\n\u001b[0mInstalling collected packages: rouge-score\nSuccessfully installed rouge-score-0.1.2\n\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\nRequirement already satisfied: tensorflow in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (2.12.0)\nRequirement already satisfied: absl-py>=1.0.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (1.1.0)\nRequirement already satisfied: astunparse>=1.6.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (1.6.3)\nRequirement already satisfied: flatbuffers>=2.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (2.0.7)\nRequirement already satisfied: gast<=0.4.0,>=0.2.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (0.4.0)\nRequirement already satisfied: google-pasta>=0.1.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (0.2.0)\nRequirement already satisfied: grpcio<2.0,>=1.24.3 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (1.50.0)\nRequirement already satisfied: h5py>=2.9.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (3.1.0)\nRequirement already satisfied: jax>=0.3.15 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (0.3.23)\nRequirement already satisfied: keras<2.13,>=2.12.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (2.12.0)\nRequirement already satisfied: libclang>=13.0.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (14.0.1)\nRequirement already satisfied: numpy<1.24,>=1.22 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (1.23.4)\nRequirement already satisfied: opt-einsum>=2.3.2 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (3.3.0)\nRequirement already satisfied: packaging in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (23.1)\nRequirement already satisfied: protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (3.20.3)\nRequirement already satisfied: setuptools in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (65.6.3)\nRequirement already satisfied: six>=1.12.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (1.15.0)\nRequirement already satisfied: tensorboard<2.13,>=2.12 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (2.12.1)\nRequirement already satisfied: tensorflow-estimator<2.13,>=2.12.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (2.12.0)\nRequirement already satisfied: termcolor>=1.1.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (1.1.0)\nRequirement already satisfied: typing-extensions>=3.6.6 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (4.7.1)\nRequirement already satisfied: wrapt<1.15,>=1.11.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (1.12.1)\nRequirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorflow) (0.26.0)\nRequirement already satisfied: wheel<1.0,>=0.23.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from astunparse>=1.6.0->tensorflow) (0.38.4)\nRequirement already satisfied: scipy>=1.5 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from jax>=0.3.15->tensorflow) (1.10.1)\nRequirement already satisfied: etils[epath] in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from jax>=0.3.15->tensorflow) (0.9.0)\nRequirement already satisfied: google-auth<3,>=1.6.3 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorboard<2.13,>=2.12->tensorflow) (2.17.1)\nRequirement already satisfied: google-auth-oauthlib<1.1,>=0.5 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorboard<2.13,>=2.12->tensorflow) (1.0.0)\nRequirement already satisfied: markdown>=2.6.8 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorboard<2.13,>=2.12->tensorflow) (3.3.4)\nRequirement already satisfied: requests<3,>=2.21.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorboard<2.13,>=2.12->tensorflow) (2.31.0)\nRequirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorboard<2.13,>=2.12->tensorflow) (0.7.0)\nRequirement already satisfied: tensorboard-plugin-wit>=1.6.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorboard<2.13,>=2.12->tensorflow) (1.8.0)\nRequirement already satisfied: werkzeug>=1.0.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from tensorboard<2.13,>=2.12->tensorflow) (2.0.2)\nRequirement already satisfied: cachetools<6.0,>=2.0.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.13,>=2.12->tensorflow) (5.3.0)\nRequirement already satisfied: pyasn1-modules>=0.2.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.13,>=2.12->tensorflow) (0.2.8)\nRequirement already satisfied: rsa<5,>=3.1.4 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from google-auth<3,>=1.6.3->tensorboard<2.13,>=2.12->tensorflow) (4.7.2)\nRequirement already satisfied: requests-oauthlib>=0.7.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from google-auth-oauthlib<1.1,>=0.5->tensorboard<2.13,>=2.12->tensorflow) (1.3.0)\nRequirement already satisfied: charset-normalizer<4,>=2 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests<3,>=2.21.0->tensorboard<2.13,>=2.12->tensorflow) (2.0.6)\nRequirement already satisfied: idna<4,>=2.5 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests<3,>=2.21.0->tensorboard<2.13,>=2.12->tensorflow) (3.2)\nRequirement already satisfied: urllib3<3,>=1.21.1 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests<3,>=2.21.0->tensorboard<2.13,>=2.12->tensorflow) (1.26.7)\nRequirement already satisfied: certifi>=2017.4.17 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests<3,>=2.21.0->tensorboard<2.13,>=2.12->tensorflow) (2021.10.8)\nRequirement already satisfied: importlib_resources in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from etils[epath]->jax>=0.3.15->tensorflow) (5.10.0)\nRequirement already satisfied: zipp in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from etils[epath]->jax>=0.3.15->tensorflow) (3.10.0)\nRequirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from pyasn1-modules>=0.2.1->google-auth<3,>=1.6.3->tensorboard<2.13,>=2.12->tensorflow) (0.4.8)\nRequirement already satisfied: oauthlib>=3.0.0 in /home/hexuser/.cache/pypoetry/virtualenvs/python-kernel-OtKFaj5M-py3.9/lib/python3.9/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib<1.1,>=0.5->tensorboard<2.13,>=2.12->tensorflow) (3.1.1)\n\u001b[33mDEPRECATION: dropbox 11.25.0 has a non-standard dependency specifier stone>=2.*. pip 23.3 will enforce this behaviour change. A possible replacement is to upgrade to a newer version of dropbox or contact the author to suggest that they release a version with a conforming dependency specifiers. Discussion can be found at https://github.com/pypa/pip/issues/12063\u001b[0m\u001b[33m\n\u001b[0m\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.2.1\u001b[0m\n\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"},"execution_count":null,"metadata":{},"output_type":"execute_result"}]},{"cell_type":"code","source":"# Create engine to connect to database\nengine = create_engine(f'postgresql://postgres:qRd71PqwOsbv62WAvboR@database-1.cuaho2dof33c.us-east-1.rds.amazonaws.com:5432/postgres')\n\nprint(engine.table_names())\n\ntable_name = 'fulltext'\n\nsql = text(f''' \n    SELECT empty, corpusid, text, id FROM {table_name} LIMIT 5;\n''')\n\nwith engine.connect() as conn: \n    query = conn.execute(sql)\n\nfull_text = pd.DataFrame(query.fetchall())\nfull_text.head()","metadata":{},"execution_count":null,"outputs":[{"data":{"text/plain":"<ipython-input-16-2dedf3350ce2>:4: SADeprecationWarning: The Engine.table_names() method is deprecated and will be removed in a future release.  Please refer to Inspector.get_table_names(). (deprecated since: 1.4)\n  print(engine.table_names())\n['abstract_encodings_test4', 'abstracts', 'abstracts_encodings', 'abstracts_encodings_test3', 'assayTerms', 'assayterms', 'attributes', 'fulltext', 'authors', 'copy_test']\n"},"execution_count":null,"metadata":{},"output_type":"execute_result"},{"data":{"application/vnd.hex.export+parquet":{"success":true,"exportKey":"604dd43f-9bf2-4f53-b2ad-a57886ed9dc6/1e7b90bc-1996-4336-8653-50313b6d2f75/exports/6734f1ca-3327-4c7c-a68a-9cd75754768e"},"text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>empty</th>\n      <th>corpusid</th>\n      <th>text</th>\n      <th>id</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>5640</td>\n      <td>12934035</td>\n      <td>\\nMolecular Tools for the Selective Detection of Nine Diatom Species Biomarkers of Various Water Quality Levels Elongation Factor Gene (eEF1-a) Diatom Species Probe Name Probe Sequence 5' → 3' Achnanthidium minutissimum (Kützing) Czarnecki\\n\\n\\nMolecular Tools for the Selective Detection of Nine Diatom Species Biomarkers of Various Water Quality Levels Elongation Factor Gene (eEF1-a) Diatom Species Probe Name Probe Sequence 5' → 3' Achnanthidium minutissimum (Kützing) Czarnecki\\nSupplementary Information\\nTable S1. Sequence of the oligonucleotide probes developed in this work. The probe's classification is based on the two marker genes (eEF1-a and SIT), on the diatom species and also on the position within the target DNA sequence (sense strand or anti-sense strand).Antisense strand AchMinEFas01 ATACCGACAGAGTTTCCGGGTCCGGCAG AchMinEFas03 ACAGGACAGTATCCAACCTTGAGAACTCCGGG AchMinEFas04 CGCGGTACGGGAAAAGATAACAGGGCAGTAT AchMinEFas05 CAACCTTCTGTCCTCCAGTCTTCTTCGATTGC AchMinEFas02 GCAACCTTGGCGGTACGGCAGAAAATAACAG Craticula halophila (Grunow ex Van Heurck) Mann Sense strand CraHalEFs02 TATTGCCAAGGACGAGAAGGTACAGCC Antisense strand CraHalEFas05 GGCTGGACCTTCTCTTCCTTGGCAATA CraHalEFas01 CTTCCACAAAATCTTGGTCATCTTGCA Cyclotella meneghiniana Kützing Sense strand CycMenEFs01 CTTGTGGAAGATGGGCAAGAAGACTGGAGG CycMenEFs02 ACTCCTGTTGTCTTCTGCCGTACTGCCAA CycMenEFs03 AAGATGAGAAGGTTGAGCCAGGTGACA\\n\\nSurirella angusta Kützing\\n\\nSense strand   Table S3. Collection of DNA sequences of diatom marker genes obtained in this work.\\nSurAngEFs01 ACAGGACAGTATCCAACCTTGAGAACTCCGGG SurAngEFs04 GAAGGTTCTTGACTCCGCCGGACCAGGAA SurAngEFs05 TACTGCCCTGTTATCTTTTCCCGTACCGCG SurAngEFs06 GGAAGCAATCGAAGAAGACTGGAGGACAGAAG Antisense strand SurAngEFas01 GGCTGTACCTTCTCGTCCTTGGCAATA SurAngEFas02 GCTGTACCTTCTCGTCCTTGGCAATAC SurAngEFas03 AATCTTGGTCATCTTGCAGGCAACCTTGGC SurAngEFas06 GCAACCTTGGCGGTACGGCAGAAAATAACAG SurAngEFas07 AACAGGGCAGTATCCCGGCTTGAGGACAC SurAngEFas04 CGCGGTACGGGAAAAGATAACAGGGCAGTAT SurAngEFas05 CAACCTTCTGTCCTCCAGTCTTCTTCGATTGC Bacillariophyceae OrNavNavSSEFs01 AGCGACCAAAGCGGTGAAAGACTTGATGG OrNavNavSSEFs02 CCAGGGTTGACCTTCTCATCCTTGGCAATACC OrNavNavSSEFas01 TTTCACCGCTTTGGTCGCTGTTCAGGAG OrNavNavSSEFas02 CATCAAGTCTTTCACCGCTTTGGTCGCTG OrNavNavSSEFas03 GTCGCTGTTCAGGAGCACCCCGGTGT ClBacNavSSEFs01 GCAAGATGACCAAGATTCTCTGGAAGCAGTCC ClBacNavSSEFs02 GTTATCTTCTCTCGTACCGCCAAGGTTGCCT ClBacNavSSEFs03 GCCCGGTTACTGTCCTGTTATCTTCTCTCGTAC ClBacNavSSEFs04 GACTGCTTCCAGAGAATCTTGGTCATCTTG ClBacNavSSEFs05 AGGCAACCTTGGCGGTACGAGAGAAGATAA ClBacNavSSEFs06 TACGAGAGAAGATAACAGGACAGTAACCGGGC ClBacNavSSEFs08 CAAGAAGACCGGTGGTGCCAAGGTTGACAA ClBacNavSSEFs09 CCCTGGTGACATCATCTACCTTGAAAAGGAAGG ClBacNavSSEFs010 TCAACCTTGGCACCACCGGTCTTCTTGGAC ClBacNavSSEFs011 GATAACAGGACAGTAACCGGGCTTGAGGACACC ClBacNavSSEFs012 CAACCTTGGCGGTACGAGAGAAGATAACAGGA ClBacNavSSEFs013 ATCTTGGTCATCTTGCAGGCAACCTTGGC\\n\\nDiatom Species Silicic Acid Transporter Gene (SIT) [Sequence 5' → 3']\\n\\nCraticula halophila   \\n\\n\\nDiatom Species Elongation Factor Gene (eEF1-a) [Sequence 5' → 3']\\n\\nAchnanthidium minutissimum  \\n\\nS2 Table S1 .\\nS2S1Mayamaea atomus var. permitis (Hustedt) Bruder &amp; Medlin Cont.Sense strand \\n\\nMayPerEFs01 \\nTTATTTTCTGCCGTACCGCCAAGGTTGC \\nMayPerEFs02 \\nCTGTCGGTATGTCCATCAGGGGTATTGCCAAG \\nMayPerEFs03 \\nAGAAGACCGGAGGTGCCAAGTTGGAAAATCCC \\n\\nAntisense \\nstrand \\n\\nMayPerEFas02 ATGTCTCCAGGCTGTACCTTCTCGTCCTT \\nMayPerEFas03 TTATCGACCTTGGCACCTCCCGTCTTCTT \\nMayPerEFas04 ATACCGACAGAGTTTCCGGGTCCGGCAG \\nMayPerEFas05 ACAGGACAGTATCCAACCTTGAGAACTCCGGG \\nMayPerEFas06 GCAACCTTGGCGGTACGGCAGAAAATAACAG \\n\\nNavicula veneta \\nKützing \\n\\nSense strand \\nNavVenEFs07 \\nTATTGCCAAGGACGAGAAGGTACAGCC \\n\\nAntisense \\nstrand \\n\\nNavVenEFas07 GTCAACCTTGGCACCTCCAGTCTTCTT \\nNavVenEFas08 GACCTTCTCGTCCTTTCCAATACCCTTGA \\nNavVenEFas03 GGCTTGAGGACACCGGGATGCTCTTGGAC \\nNavVenEFas02 AAGGCAGTGAACGACTTGACGGGCTTGAG \\nNavVenEFas04 GGGCAGTATCCGGGCTTGAGGACACCA \\nNavVenEFas01 GGCTGTACCTTCTCGTCCTTGGCAATA \\nNavVenEFas05 GCTGTACCTTCTCGTCCTTGGCAATAC \\nNavVenEFas09 CTTCCACAAAATCTTGGTCATCTTGCA \\n\\nNitzschia dissipata \\n(Kützing) Grunow \\n\\nSense strand \\n\\nNitDisEFs02 \\nTTATTTTCTGCCGTACCGCCAAGGTTGC \\nNitDisEFs03 \\nCTGTCGGTATGTCCATCAGGGGTATTGCCAAG \\nNitDisEFs04 \\nAGAAGACCGGAGGTGCCAAGTTGGAAAATCCC \\n\\nAntisense \\nstrand \\n\\nNitDisEFas04 \\nGGCTGGACCTTCTCTTCCTTGGCAATA \\nNitDisEFas13 \\nATGTCTCCAGGCTGTACCTTCTCGTCCTT \\nNitDisEFas14 \\nTTATCGACCTTGGCACCTCCCGTCTTCTT \\nNitDisEFas15 \\nAAGGCAGTGAACGACTTGACGGGCTTGAG \\nNitDisEFas06 \\nATACCGACAGAGTTTCCGGGTCCGGCAG \\nNitDisEFas16 \\nGGCTGTACCTTCTCGTCCTTGGCAATA \\nNitDisEFas07 \\nACAGGACAGTATCCAACCTTGAGAACTCCGGG \\nNitDisEFas08 \\nGCAACCTTGGCGGTACGGCAGAAAATAACAG \\nElongation Factor Gene (eEF1-a) \\n\\nDiatom Species \\nProbe Name \\nProbe Sequence 5' → 3' \\n\\nNitzschia palea \\n(Kützing) W Smith \\n\\nSense strand \\n\\nNitPalEFs02 \\nAAGGACGAAAAGGTCAACCCCGGAGACATCATC \\n\\nNitPalEFs03 \\nCTCTGTTGGTATGTCCGTCAAGGGTATTGGAAAGG \\n\\nNitPalEFs04 \\nCCTCAAGCCTGGATACTGTCCCATCATTTTCTC \\n\\nNitPalEFs05 \\nTTTCACCGCTTTGGTCGCTGTTCAGGAG \\n\\nNitPalEFs06 \\nCATCAAGTCTTTCACCGCTTTGGTCGCTG \\n\\nNitPalEFs07 \\nGTCGCTGTTCAGGAGCACCCCGGTGT \\n\\nAntisense strand \\n\\nNitPalEFas06 \\nACTGGGAAAGTTCTGGGGGGTTGTCGACCTT \\n\\nNitPalEFas07 \\nGAAAATGATGGGACAGTATCCAGGCTTGAGGACAC \\n\\nNitPalEFas08 \\nGTCTCCGGGGTTGACCTTTTCGTCCTTTCCAATA \\n\\nNitPalEFas01 \\nGACCTTCTCGTCCTTTCCAATACCCTTGA \\n\\nNitPalEFas02 \\nGTCAACCTTGGCACCTCCAGTCTTCTT \\n\\nNitPalEFas04 \\nGGGCAGTATCCGGGCTTGAGGACACCA \\n\\nNitPalEFas05 \\nCAACCTTGGCACCTCCAGTCTTCTTCGA \\n\\nNitPalEFas09 \\nCGCGGTACGGGAAAAGATAACAGGGCAGTAT \\n\\nNitPalEFas09 \\nCAACCTTCTGTCCTCCAGTCTTCTTCGATTGC \\n\\n\\n\\nTable S1 .\\nS1Cont.Silicic Acid Transporter Gene (SIT) \\n\\nDiatom Species \\nProbe Name \\nProbe Sequence 5' → 3' \\n\\nCraticula halophila \\n(Grunow ex Van Heurck) Mann \\n\\nSense strand \\n\\nCraHalSITs09 \\nCATTCTTCTCCGTTACTGCCTGTTGCTC \\n\\nCraHalSITs04 \\nGCGGTTACCCTCTTCTTTGTCTTCATGGCCAT \\n\\nCraHalSITs05 \\nGTTACTCTATATGCCCTCTTCCATGGCCAGACT \\n\\nCraHalSITs06 \\nGTCTTCATGGCCATTGTTGGTATGCTTGAGGGA \\n\\nCraHalSITs01 \\nTCGCCGTTACTCTATATGCCCTCTTCCATG \\n\\nCraHalSITs02 \\nTGTGGCAAAAATGACCGAGGAAGAGCGTG \\n\\nAntisense \\n\\nstrand \\n\\nCraHalSITas08 \\nCCAACGGCGTCTCTGTATTCCTCTTCTT \\n\\nCraHalSITas04 \\nATGGAAGAGGGCATATAGAGTAACGGCGAAAG \\n\\nCraHalSITas05 \\nCATGAAGACAAAGAAGAGGGTAACCGCGGC \\n\\nCraHalSITas06 \\nTTAGGGACGCCGGGCCACATGGTAGTC \\n\\nEolimna minima \\n(Grunow) Lange-Bertalot \\n\\nSense strand \\nEolMinSITs02 \\nGATTCAGATCGCCTTCTTCGCCGTCACC \\n\\nAntisense \\n\\nstrand \\n\\nEolMinSITas01 \\nGCGAAGAAGGCGATCTGAATCGCCTCGAGCAT \\n\\nEolMinSITas02 \\nGGTGACGGCGAAGAAGGCGATCTGAATC \\n\\nNavicula veneta \\nKützing \\n\\nSense strand \\n\\nNavVenSITs05 \\nGCATTGCTGCTATCTTGTACAGATGTTCTTCG \\n\\nNavVenSITs01 \\nCCAACGGCGTCTCTGTATTCCTCTTCTT \\n\\nNavVenSITs02 \\nCATTCTTCTCCGTTACTGCCTGTTGCTC \\n\\nNavVenSITs03 \\nTCGCCGTTACTCTATATGCCCTCTTCCATG \\n\\nNavVenSITs04 \\nTGTGGCAAAAATGACCGAGGAAGAGCGTG \\n\\nAntisense \\n\\nstrand \\n\\nNavVenSITas05 \\nTGGCCATGGAAGAGGGCATATAGAGTAACGG \\n\\nNavVenSITas06 \\nGGCGAAGAACATCTGTACAAGATAGCAGCAAT \\n\\nNavVenSITas07 \\nATGAAGACAAAGAAGAGGGTAACCGCGGCA \\n\\nNavVenSITas09 \\nCACATGGTGGTCTGTCCATTGAACAGGG \\n\\nNavVenSITas10 \\nCCAAAGAAGAAGATGTTTTGCATAGTGTT \\n\\nNitzschia dissipata \\n(Kützing) Grunow \\n\\nSense strand \\n\\nNitDisSITs06/AmpCofSITas05 \\nCACCATGAGCACTCTCTTCAACGAACAAACCAC \\n\\nNitDisSITs07/AmpCofSITs03 \\nAATCTCTTCTTCTGGGGACGATGCCTCTTCTCT \\n\\nNitDisSITs08/AmpCofSITas09 \\nGCTGTTGCCAAGATCCGCGAGTCGGAA \\n\\nAntisense \\n\\nstrand \\n\\nNitDisSITas02/AmpCofSITas02 GCAAAAGCACACGACGGCGAAAGAGAAG \\n\\nNitDisSITas03/AmpCofSITas03 AGAAGAAGAGATTGGCCACGGCACTGC \\n\\nNitDisSITas04/AmpCofSITas04 AAGCACAATTCCCTTGGGGATTCCCTCCCA \\n\\nNitzschia palea \\n(Kützing) W Smith \\n\\nAntisense \\n\\nstrand \\nNitPalSITas11 \\nGAGCAACAGGCAGTAACGGAGAAGAATG \\n\\nBacillariophyceae \\n\\nOrRaphSITs01 \\nTTGCCGACATAGACTTCATCACGCTGGA \\n\\nOrRaphexNaviculalesSITs02 \\nGGAACACCCTCCCACATGGTGGTCTGTCCA \\n\\nProbes classification \\n\\nProbe producing either weak or no signal \\nSpecies-specific probe \\nProbe recognizing two species \\nProbe recognizing three species \\nProbe recognizing four species \\nProbe recognizing Order level and Genus level diatoms \\nProbe recognizing Class level diatoms (Bacillariophyceae) \\n\\n\\nTable S2 .\\nS2List of oligonucleotides spotted on the MicroAqua-array-02. Multiple names indicate probes whose sequence is shared by more than one diatom species.MicroAqua-array-02 \\nProbe Name \\n\\nCy3-Marker \\nBuffer \\nEmpty \\nAmpCofTag7F \\nAmpCofTag25 \\nNitPalEFas02/NavVenEFas07 \\nSurAngEFas03/AmpCofEFas10 \\nOrRaphexNaviculalesSITs02 \\nAmpCofSITas08 \\nClBacNavSSEFs09 \\nNavVenSITas07 \\nNitDisSITs08/AmpCofSITas09 \\nSurAngEFas07 \\nLambda \\nAmpCofP20Tag41F \\nNitVitTag7R \\nNavVenEFas05/SurAngEFas02/ \\nAmpCofEFas09 \\nMayPerEFas03/NitDisEFas14 \\nEolMinSITas02 \\nAmpCofSITas07 \\nClBacNavSSEFs08 \\nNavVenSITas06 \\nNitDisSITs07/AmpCofSITs03 \\nSurAngEFas05/AchMinEFas05/ \\nNitPalEFas10 \\nPOSITIVE_25_dT_12.5 μM \\nNitDisTag46R \\nAmpCofTagP29 \\nNitDisEFas04/CraHalEFas05 \\nNavVenEFas05/SurAngEFas02/ \\nAmpCofEFas09 \\nNitVitSIT01 \\nAmpCofSITas06 \\nClBacNavSSEFs03 \\nNavVenSITas05 \\nNitDisSITs06/AmpCofSITas05 \\nSurAngEFas04/AchMinEFas04/ \\nNitPalEFasS09 \\nPOSITIVE_25_dT_25 μM \\nCraHalTag23R \\nAmpCofP20Tag221R \\nCraHalEFs02/NavVenEFs07 \\nMayPerEFas02/NitDisEFas13 \\nEolMinSITs02 \\nAmpCofSITs08 \\nClBacNavSSEFs012 \\nNavVenSITs05 \\n\\nNitDisSITas04/AmpCofSITas04 \\nSurAngEFs06 \\nPOSITIVE_25_dT_50 μM \\nAmpCofTag46F \\nNavVenTag55 \\nNavVenEFas03 \\nSurAngEFs01/NitDisEFs16/ \\nNavVenEFas01 \\nAmpCofTag10 \\nAmpCofSITs07 \\nClBacNavSSEFs011 \\nNavVenSITs04/CraHalSITs02 \\nNitDisSITas03/AmpCofSITas03 \\nSurAngEFs05 \\nDunGS05_25_dT_dT_12.5 μM \\nNitDisTag23F \\nNitVitTag27F \\nAmpCofEFas03 \\nCycMenEFs03 \\nEolMinSITas01 \\nAmpCofSITs06 \\nClBacNavSSEFs010 \\nNavVenSITs03/CraHalSITs01 \\nNitDisSITas02/AmpCofSITas02 \\nSurAngEFs04 \\nDunGS05_25_dT_dT_25 μM \\nNitPalTag36F \\nNitVitTag7 \\nNavVenEFas02/NitDisEFas15 \\nNitPalEFas01/NavVenEFas08 \\nNitPalSITas11 \\nClBacNavSSEFs06 \\nCraHalSITas06 \\nOrNavNavSSEFs02 \\nNitDisEFas08/SurAngEFas06/ \\nAchMinEFas02/MayPerEFas06 \\nNitPalEFas08 \\nClBacNavSSEFs03 \\nDunGS05_25_dT_dT_50 μM \\nAmpCofTag5R \\nAmpCofEFas02 \\nCycMenEFs02 \\nNavVenSITas10 \\nNavVenSITs01/CraHalSITas08 \\nCrahalSITas05 \\nOrNavNavSSEFs01 \\nNitDisEFas07/SurAngEFs01/ \\nAchMinEFas03/MayPerEFas05 \\nNitPalEFas07 \\nDunGS02_25_dT_dT_12.5 μM \\n\\nNitDisTag67R \\nNitPalTag23 \\nNavVenEFas09/AmpCofEFas01/ \\nCraHalEFas01 \\nNitPalEFas05/AmpCofEFas11 \\nOrRaphSITs01 \\nClBacNavSSEFs05 \\nCraHalSITas04 \\nClBacNavSSEFs013 \\nNitDisEFas06/AchMinEFas01/ \\nMayPerEFas04 \\nNitPalEFas06 \\nClBacNavSSEFs02 \\nDunGS02_25_dT_dT_25 μM \\nAmpCofP20F \\nAmpCofTag \\nCycMenEFs01 \\nNavVenSITas09 \\nCraHalSITs06 \\nOrNavNavSSEFas03/NitPalEFs07 \\nNitDisEFs04/MayPerEFs03 \\nNitPalEFs04 \\nDunGS02_25_dT_dT_50 μM \\nNitDisTag67F \\nNitDisTag7R \\nSurAngEFs01/NitDisEFs16/ \\nNavVenEFas01 \\nNitPalEFas04/NavVenEFas04 \\nAmpCofTag09 \\nClBacNavSSEFs04 \\nCraHalSITs05 \\nOrNavNavSSEFas02/NitPalEFs06 \\nNitDisEFs03/MayPerEFs02 \\nNitPalEFs03 \\nClBacNavSSEFs01 \\neuk 1209 \\nAmpCofTag29R \\nAmpCofP29Tag54R \\nAmpCof29Tag \\nCraHalSITs04 \\nOrNavNavSSEFas01/NitPalEFs05 \\nNitDisEFs02/MayPerEFs01 \\nNitPalEFs02 \\nNavVenSITs02/CraHalSITs09 \\neuk 328 \\n\\n\\nTable S3 .\\nS3Cont.Diatom Species \\nSilicic Acid Transporter Gene (SIT) \\n[Sequence 5' → 3'] \\n\\nNitzschia dissipata \\n\\n1 GAATTCTCCGGACTGTTGCATGCTTCCTACCTCCTTGAAATGATGGTTTGCAGCATGGCT \\nCTTAAGAGGCCTGACAACGTACGAAGGATGGAGGAACTTTACTACCAAACGTCGTACCGA \\n\\n61 GGACAACCCAAGGTTAGCAACGAACCCCCTCGCAGTGCCGTGGCCAATCTCTTCTTCTGG \\nCCTGTTGGGTTCCAATCGTTGCTTGGGGGAGCGTCACGGCACCGGTTAGAGAAGAAGACC \\n\\n121 GGACGATGCCTCTTCTCTTTCGCCGTCGTGTGCTTTTGCGCGGCCGTCACCATGAGCACT \\nCCTGCTACGGAGAAGAGAAAGCGGCAGCACACGAAAACGCGCCGGCAGTGGTACTCGTGA \\n\\n181 CTCTTCAACGAACAAACCACCGTCTGGGAGGGAATCCCCAAGGGAATTGTGCTTGTCCTT \\nGAGAAGTTGCTTGTTTGGTGGCAGACCCTCCCTTAGGGGTTCCCTTAACACGAACAGGAA \\n\\n241 TTCGTGGTGTTGTTCTGCGTCGTTGGTTTGTTGGAGGCCATGCAAATCGCCTTTTTCGCT \\nAAGCACCACAACAAGACGCAGCAACCAAACAACCTCCGGTACGTTTAGCGGAAAAAGCGA \\n\\n301 GTTGCCAAGATCCGCGAGTCGGAACG \\nCAACGGTTCTAGGCGCTCAGAATTGC \\n\\nNitzschia palea \\n\\n1 GCAACTTTGGTGGCCCTGTTCAATGGACAGACCACCATGTGGGAGGGTGTTCCCAACGGT \\nCGTTGAAACCACCGGGACAAGTTACCTGTCTGGTGGTACACCCTCCCACAAGGGTTGCCA \\n\\n61 GTCGCCGTCCTTGTTTTCTTCCTTCTTATGGCGGTCGTAGGAATGTTGGAAGGAATGCAA \\nCAGCGGCAGGAACAAAAGAAGGAAGAATACCGCCAGCATCCTTACAACCTTCCTTACGTT \\n\\n121 ATCGCCTTTTTNTCTGTATCCNGAATCCCCGAGGATGAACGTGGNGACTCCTNGANTGCC \\nTAGCGGAAAAANAGACATAGGNCTTAGGGGCTCCTACTTGCACCNCTGAGGANCTNACGG \\n\\n181 AAGAAAGTANGCAA \\nTTCTTTCATNCGTT \\n\\n\\nTable S3 .\\nS3Cont.\\n\\nTable S3 .\\nS3Cont.Diatom Species \\nElongation Factor Gene (eEF1-a) \\n[Sequence 5' → 3'] \\n\\nEolimna minima \\n\\n1 CACAAGAAGGTCCTTGACTCGGCCGGACCAGGAAACTCGGTCGGCATGTCCATCAAGGGT \\nGTGTTCTTCCAGGAACTGAGCCGGCCTGGTCCTTTGAGCCAGCCGTACAGGTAGTTCCCA \\n\\n61 ATTGCCAAGGACGAGAAGGTACAGCCTGGAGACATTATCTACCTTGAAAAGGAAGGTCAC \\nTAACGGTTCCTGCTCTTCCATGTCGGACCTCTGTAATAGATGGAACTTTTCCTTCCAGTG \\n\\n121 CTCAAGCCCGTCAAGTCGTTCACTGCCTTTGTCGCCGTCCAAGAGCATCCCGGTGTCCTC \\nGAGTTCGGGCAGTTCAGCAAGTGACGGAAACAGCGGCAGGTTCTCGTAGGGCCACAGGAG \\n\\n181 AAGCCTGGTTACTGCCCTGTCATTTTCTCTCGTACCGCCAAGGTGGCCTGCAAGATGACC \\nTTCGGACCAATGACGGGACAGTAAAAGAGAGCATGGCGGTTCCACCGGACGTTCTACTGG \\n\\n241 AAGATTTTGTGGAAGCAGTCCAAGAAGACGGGAGGTGCCAAGGTCGATAACCCTCCCGAA \\nTTCTAAAACACCTTCGTCAGGTTCTTCTGCCCTCCACGGTTCCAGCTATTGGGAGGGCTT \\n\\n301 CTGTCTCAGTTTGAATCCGCCGAGGTCGAATTCGAACCCACTGCCC \\nGACAGAGTCAAACTTAGGCGGCTCCAGCTTAAGCTTGGGTGACGGG \\n\\nMayamaea permitis \\n\\n1 CAAGAAGGTCCTTGAATCCGCCGGACCCGGTAACTCCGTTGGTCTTTCCATCAAGGGTAT \\nGTTCTTCCAGGAACTTAGGCGGCCTGGGCCATTGAGGCAACCAGAAAGGTAGTTCCCATA \\n\\n61 TGCCAAGGATGAGAAGGTCAACCCTGGTGACATCATCTACCTTGAAAAGGAAGGTGAACT \\nACGGTTCCTACTCTTCCAGTTGGGACCACTGTAGTAGATGGAACTTTTCCTTCCACTTGA \\n\\n121 CAAGCCCATCAAGTCTTTCACCGCTTTGGTCGCTGTTCAGGAGCACCCCGGTGTCCTCAA \\nGTTCGGGTAGTTCAGAAAGTGGCGAAACCAGCGACAAGTCCTCGTGGGGCCACAGGAGTT \\n\\n181 GCCCGGTTACTGTCCAGTCATCTTCTCTCGTACCGCCAAGGTTGCCTGCAAGATGACCAA \\nCGGGCCAATGACAGGNCAGTAGAAGAGAGCATGGCGGTTCCAACGGACGTTCTACTGGTT \\n\\n241 GATCCTCTGGAAGCAGTCCAAGAAGACCGGTGGTGCCAAGGTTGACAACCCTCCGGAACT \\nCTAGGAGACCTTCGTCAGGTTCTTCTGGCCACCACGGTTCCAACTGTTGGGAGGCCTTGA \\n\\n301 CTCGCAGTT \\nGAGCGTCAA \\n\\nNavicula veneta \\n\\n1 GACTCGGCCGGACCAGGAAACTCGGTCGGCATGTCCATCAAGGGTATTGCCAAGGACGAG \\nCTGAGCCGGCCTGGTCCTTTGAGCCAGCCGTACAGGTAGTTCCCATAACGGTTCCTGCTC \\n\\n61 AAGGTACAGCCTGGAGACATTATCTACCTTGAAAAGGAAGGTCACCTCAAGCCCGTCAAG \\nTTCCATGTCGGACCTCTGTAATAGATGGAACTTTTCCTTCCAGTGGAGTTCGGGCAGTTC \\n\\n121 TCGTTCACTGCCTTTGTCGCCGTCCAAGAGCATCCCGGTGTCCTCAAGCCTGGTTACTGC \\nAGCAAGTGACGGAAACAGCGGCAGGTTCTCGTAGGGCCACAGGAGTTCGGACCAATGACG \\n\\n181 CCCGTCATTTTCTCTCGTACCGCCAAGGTGGCCTGCAAGATGACCAAGATTTTGTGGAAG \\nGGGCAGTAAAAGAGAGCATGGCGGTTCCACCGGACGTTCTACTGGTTCTAAAACACCTTC \\n\\n241 CAGTCCAAGAAGACGGGAGGTGCCAAGGTCGATAACCCTCCCGAACTGTCTCAGTTTGAA \\nGTCAGGTTCTTCTGCCCTCCACGGTTCCAGCTATTGGGAGGGCTTGACAGAGTCAAACTT \\n\\n301 TCCGCCGAGGTCGAATTCGAACCCACTGCCNNCCTTGTTCTTGGAACCT \\nAGGCGGCTCCAGCTTAAGCTTGGGTGACGGNNGGAACAAGAACCTTGGA \\n\\n</td>\n      <td>4227</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>5665</td>\n      <td>283904</td>\n      <td>\\nIntroduction Clinical review: Sepsis and septic shock -the potential of gene arrays R E V I E W\\n\\n\\nHector R Wong hector.wong@cchmc.org \\nDivision of Critical Care Medicine\\nDepartment of Pediatrics\\nCincinnati Children's Hospital Medical Center\\nCincinnati Children's Hospital Research Foundation\\nUniversity of Cincinnati College of Medicine\\n45229-3039CincinnatiOHUSA\\n\\nIntroduction Clinical review: Sepsis and septic shock -the potential of gene arrays R E V I E W\\n*Correspondence: Wong Critical Care 2012, 16:204\\nOver the past decade several investigators have applied microarray technology and related bioinformatic approaches to clinical sepsis and septic shock, thus allowing for an assessment of how, or if, this branch of genomic medicine has meaningfully impacted the sepsis fi eld. Th is review will fi rst provide an overview of the gene microarray approach, including limitations and study design considerations. Subsequently, the review will focus on the potential translational application of microarray data and genome-wide expression profi ling to the sepsis fi eld. Four broad areas will be discussed: genome-level understanding of sepsis, biomarker discovery, gene expression-based identifi cation of septic shock subclasses, and discovery of novel targets and pathways.Technology, approaches, and limitationsMicroarray-related technology, approaches, and limitations have been extensively reviewed elsewhere [1-5], and will be summarized below. Notably, there is now an emerging technology, RNA sequencing (RNA-Seq)[6], that has potentially intriguing applications for the fi eld, but will not be further discussed as there are no RNA-Seq data specifi cally related to sepsis.Th e fundamental technical innovation of microarray technology is the ability to simultaneously measure mRNA abundance of thousands of transcripts (transcriptomics). Th e technique generally involves reverse transcription of RNA into cDNA, with the inclusion of a labeling molecule for detection. Th e labeled cDNA (targets) is subsequently applied to a support surface arrayed with nucleotide sequences corresponding to specifi c genes (probes). Th e probes and targets hybridize via standard nucleic acid interactions and the amount of hybridization refl ects the abundance of a specifi c mRNA species. Th e supporting surface is subsequently washed and scanned to provide raw mRNA abundance data. An important limitation of transcriptomics is that it solely provides a 'snapshot' of steady-state mRNA abundance. Th e degree of mRNA abundance is infl uenced by multiple factors, and does not provide any direct information about gene end products (proteins), nor post-translational modifi ers of protein function, such as phosphorylation or glycation.One major consideration in designing a microarray experiment involves the RNA source. Ideally, the RNA source should be relatively homogenous and closely represent the disease/condition biology of interest. For example, the discovery of neutrophil gelatinase-associated lipocalin as a biomarker for acute kidney injury included microarray-based analysis of kidneys fromAbstractOver the past decade several investigators have applied microarray technology and related bioinformatic approaches to clinical sepsis and septic shock, thus allowing for an assessment of how, or if, this branch of genomic medicine has meaningfully impacted the fi eld of sepsis research. The ability to simultaneously and effi ciently measure the steady-state mRNA abundance of thousands of transcripts from a given tissue source (that is, 'transcriptomics') has provided an unprecedented opportunity to gain a broader, genomelevel 'picture' of complex and heterogeneous clinical syndromes such as sepsis. A trancriptomic approach to sepsis and septic shock is technically challenging on multiple levels, but nonetheless modest, tangible advances are being realized. These include a genomelevel understanding of the complexity of sepsis and septic shock, identifi cation of novel candidate pathways and targets for potential intervention, discovery of novel, candidate diagnostic and stratifi cation biomarkers, and the ability to stratify patients into clinically relevant, expression-based subclasses. The challenges moving forward include robust validation studies, standardization of technical approaches, standardization and further development of analytical algorithms, and large-scale collaborations.\\n\\nIntroduction\\n\\nOver the past decade several investigators have applied microarray technology and related bioinformatic approaches to clinical sepsis and septic shock, thus allowing for an assessment of how, or if, this branch of genomic medicine has meaningfully impacted the sepsis fi eld. Th is review will fi rst provide an overview of the gene microarray approach, including limitations and study design considerations. Subsequently, the review will focus on the potential translational application of microarray data and genome-wide expression profi ling to the sepsis fi eld. Four broad areas will be discussed: genome-level understanding of sepsis, biomarker discovery, gene expression-based identifi cation of septic shock subclasses, and discovery of novel targets and pathways.\\n\\n\\nTechnology, approaches, and limitations\\n\\nMicroarray-related technology, approaches, and limitations have been extensively reviewed elsewhere [1][2][3][4][5], and will be summarized below. Notably, there is now an emerging technology, RNA sequencing (RNA-Seq) [6], that has potentially intriguing applications for the fi eld, but will not be further discussed as there are no RNA-Seq data specifi cally related to sepsis.\\n\\nTh e fundamental technical innovation of microarray technology is the ability to simultaneously measure mRNA abundance of thousands of transcripts (transcriptomics). Th e technique generally involves reverse transcription of RNA into cDNA, with the inclusion of a labeling molecule for detection. Th e labeled cDNA (targets) is subsequently applied to a support surface arrayed with nucleotide sequences corresponding to specifi c genes (probes). Th e probes and targets hybridize via standard nucleic acid interactions and the amount of hybridization refl ects the abundance of a specifi c mRNA species. Th e supporting surface is subsequently washed and scanned to provide raw mRNA abundance data. An important limitation of transcriptomics is that it solely provides a 'snapshot' of steady-state mRNA abundance. Th e degree of mRNA abundance is infl uenced by multiple factors, and does not provide any direct information about gene end products (proteins), nor post-translational modifi ers of protein function, such as phosphorylation or glycation.\\n\\nOne major consideration in designing a microarray experiment involves the RNA source. Ideally, the RNA source should be relatively homogenous and closely represent the disease/condition biology of interest. For example, the discovery of neutrophil gelatinase-associated lipocalin as a biomarker for acute kidney injury included microarray-based analysis of kidneys from\\n\\n\\nAbstract\\n\\nOver the past decade several investigators have applied microarray technology and related bioinformatic approaches to clinical sepsis and septic shock, thus allowing for an assessment of how, or if, this branch of genomic medicine has meaningfully impacted the fi eld of sepsis research. The ability to simultaneously and effi ciently measure the steady-state mRNA abundance of thousands of transcripts from a given tissue source (that is, 'transcriptomics') has provided an unprecedented opportunity to gain a broader, genomelevel 'picture' of complex and heterogeneous clinical syndromes such as sepsis. A trancriptomic approach to sepsis and septic shock is technically challenging on multiple levels, but nonetheless modest, tangible advances are being realized. These include a genomelevel understanding of the complexity of sepsis and septic shock, identifi cation of novel candidate pathways and targets for potential intervention, discovery of novel, candidate diagnostic and stratifi cation biomarkers, and the ability to stratify patients into clinically relevant, expression-based subclasses. The challenges moving forward include robust validation studies, standardization of technical approaches, standardization and further development of analytical algorithms, and large-scale collaborations.\\n\\nrodents subjected to renal ischemia [7]. Most of the studies described below have used the blood compartment as the RNA source. Reliance on the blood compartment has obvious limitations with regard to specifi c organ perturbations in clinical sepsis, but also refl ects the practical limitations of tissue sampling in clinical research and does provide a broad picture of a systemic response. Blood-derived RNA can come from either whole blood (a mixed population of blood cells), or following the isolation of specifi c blood cells. Th e whole-blood approach facilitates the procurement of samples from multiple centers, without the requirement for cell separation expertise, and has the potential to provide a comprehensive picture. However, the whole blood approach has the potential to confound data interpre ta tion due to heterogeneous blood cell populations. Th e cell-specifi c RNA approach provides a more homogenous RNA source, but has the potential to miss biologically relevant expression signatures from cells that are excluded from the experimental approach. For example, a study that focuses exclusively on peripheral blood mononuclear cells will not account for the potentially important response of neutrophils.\\n\\nAnother important consideration in designing a microarray experiment involves the reference (control) group to which gene expression in the population of interest will be compared. For example, if one is interested in studying gene expression patterns in sepsis, relative to a normal state, then comparisons to normal controls is appropriate. In contrast, if one is interested in discovering gene expression patterns that distinguish sepsis from 'sterile infl ammation' , then a more appropriate control group would consist of patients who are not infected, but meet criteria for systemic infl ammatory response (SIRS).\\n\\nTh e heterogeneity and complexity that characterize clinical sepsis present an important challenge to clinical microarray studies. From one perspective, one could say that the comprehensive nature of a microarray approach is ideally suited for studying such a heterogeneous and complex syndrome. From another perspective, the hetero geneity and complexity are potentially profound confounders for data interpretation. Accordingly, it is critical that microarray data be interpreted in the context of robust clinical/biological data that can infl uence gene expression patterns. Th ese include, but are not limited to, race, gender, age, co-morbidities, infecting pathogen class, state of immune competence, and therapy.\\n\\nAnalysis of microarray data is an evolving and complex fi eld. A universal initial step involves data normalization, which allows valid comparisons across samples by reducing technical variations not directly related to biological variation [5]. A typical next step involves statistical compari sons across groups of interest using either parametric or non-parametric analysis of variance. Unfortunately, there is no clear consensus as to which statistical test is most appropriate for a given data set, and it is particularly troubling that lists of 'diff erentially regulated genes' , from the same data set, can substantially vary based on the statistical test [8,9]. Regardless of what statistical test one uses, it is imperative that the statistical test incorporates corrections for multiple comparisons to account for a substantially high risk of false positives. One common fi lter that is applied to microarray data involves an expression fi lter that compares mRNA abundance of specifi c gene probes in one cohort versus a reference cohort. Expression fi lters are useful to assess 'magnitude of eff ect' and to reduce the number of comparisons for a subsequent statistical test, but they are not valid substitutes for formal statistical testing. Finally, there is the issue of statistical power in microarray experiments, which can be calculated, but is dependent on assumptions that can be diffi cult to derive objectively [10]. In general, a heterogeneous study cohort will require substantially more independent samples, compared to a more homogenous cohort. Th e statistical tests described above typically yield large lists of diff erentially regulated genes, thus leaving one with the challenge of assigning biological meaning to these gene lists. One approach to data interpretation involves the generation of 'heat maps' , which statistically cluster genes and samples based on similarity of expression. Heat maps provide a broad picture of gene expression patterns and allow for the discovery of disease 'subclasses' based on diff erential gene expression [11]. Another approach to viewing large microarray data sets involves the generation of gene expression 'mosaics' based on a 'self-organizing map' algorithm [12,13]. Th ese gene expression mosaics provide microarray data with a 'face' that is recognizable via intuitive pattern recognition, and were recently applied to allocate patients with septic shock into clinically relevant subclasses [14,15].\\n\\nBeyond these global assessments of gene expression patterns there exist a number of public and proprietary databases allowing for the assignment of biological function to gene lists. Th ese databases examine uploaded gene lists and determine whether the gene list is enriched for genes that are biologically related, based on the established literature. Th e outputs from these databases range from generic (for example, 'immune response') to specifi c (for example, 'antigen presentation') biological processes. Furthermore, the outputs from these databases provide an estimate of signifi cance (P-values) indicating how likely a gene list is enriched for a given biological function by chance alone. Th e level of signifi cance is directly proportional to the number of genes in the list that correspond to the given biological function, and indirectly proportional to the total number of genes in the list. A related approach to assigning biological meaning to gene lists involves the generation of gene networks based on known, direct and indirect, interactions between genes [16,17].\\n\\n\\nGenome-level understanding of sepsis\\n\\nMicroarray-based expression profi ling has provided an unprecedented opportunity to gain a broader, genomelevel 'picture' of complex and heterogeneous clinical syndromes such as sepsis. In addition, this genome-level approach has the potential to reduce investigator bias, and thus increase discovery capability, in as much as all genes are potentially interrogated, rather than a specifi c set of genes chosen by the investigator based on a priori and potentially biased assumptions.\\n\\nMany of the fundamental physiologic and biologic principles of the sepsis paradigms are derived from experiments involving human volunteers subjected to intravenous endotoxin challenge [18][19][20][21]. More recently, the genome-level response during experimental human endotoxemia has been studied using microarray technology [16,22,23]. Talwar and colleagues [22] compared eight volunteers challenged with intravenous endotoxin to four controls challenged with saline. Mononuclear cell-specifi c RNA was obtained at four diff erent time points after endotoxin challenge and analyzed via microarray. As expected, a large number of transcripts related to infl ammation and innate immunity were substantially up-regulated in response to endotoxin challenge. Interest ingly, the peak transcriptomic response to the single endotoxin challenge occurred within 6 hours and mRNA levels generally returned to control levels within 24 hours. Th e investigators also reported endotoxinmediated diff erential regulation of over 100 genes not typically associated with acute infl ammation (for example, cathepsin H, sialidase 1, UDP-glucose dehydrogenase, zinc fi nger protein 266, homeo box B2). Finally, and of relevance to subsequent sections of this review, endotoxin challenge also led to repression of several gene programs directly related to adaptive immunity (for example, interleukin-7 receptor, T cell receptor α locus, zeta-chain T cell receptor associated protein kinase 70 kDa, T cell receptor γ locus).\\n\\nCalvano and colleagues [16] also studied normal volunteers subjected to a single endotoxin challenge, but applied a (then) novel approach to microarray data analysis centered on knowledge-based interactive gene networks. Again, the maximal up-regulation of gene networks corresponding to infl ammation and innate immunity occurred at approximately 6 hours after the endotoxin challenge, and generally returned to baseline by 24 hours. Perhaps the most interesting fi nding from this network-centered analysis, however, was the widespread and early repression of gene networks related to mitochondrial energy production (for example, NADH dehydrogenase 1, pyruvate dehydrogenase, ATP synthase) and protein synthesis (ribosomal protein L3, ribosomal protein S8, eukaryotic translation initiation factor). Tang and colleagues [24] have corroborated the repression of mitochondrial energy production-related genes in a study focused on neutrophil-specifi c gene expression in critically ill patients with sepsis.\\n\\nTh e human endotoxemia studies described above provide a highly controlled and reproducible experimental setting to explore sepsis biology at the level of the entire transcriptome, but as with all sepsis models, this model does not fully replicate the complex and heterogeneous syndrome seen at the bedside following infection with live microbes [25]. Consequently, several investigators have attempted microarray-based studies in critically ill patients with sepsis and septic shock. Th ese studies present considerable experimental challenges due to the inherent heterogeneity of clinical sepsis and septic shock. Nonetheless, several studies have provided novel insight into the overall genome-level response to sepsis [9,17,24,[26][27][28][29][30][31][32][33][34]. A common theme across many of these studies is the massive up-regulation of infl ammation-and innate immunity-related genes in patients with sepsis and septic shock. Th ese observations are not intrinsically novel, but they are consistent with the long-standing sepsis paradigms centered on a hyperactive infl ammatory response, and thus provide an important layer of biological plausibility with regard to overall microarray data output in the context of clinical sepsis.\\n\\nAnother common paradigm in the sepsis fi eld involves a two-phase model consisting of an initial hyper-infl amma tory phase followed by a compensatory anti-infl amma tory phase, but this has been recently challenged, in large part due to the multiple failures of interventional clinical trials founded on this paradigm [35][36][37]. Recently, Tang and colleagues [3] conducted a formal systematic review of a carefully selected group of microarray-based human sepsis studies. A major conclusion of this systematic review is that, in aggregate, the transcriptome-level data do not consistently separate sepsis into distinct proand anti-infl ammatory phases. Th is conclusion has been questioned [38], but is supported by several recent cytokine-and infl ammatory mediator-based studies in clinical and experimental sepsis [39][40][41].\\n\\nAnother prevailing paradigm in the sepsis fi eld involves the concept of immune-paralysis, which frames sepsis as more of an adaptive immune problem (rather than just an overactive innate immune system) and the inability to adequately clear infection [42,43]. Recently, this paradigm was elegantly corroborated in mice subjected to sepsis and rescued by administration of IL-7, an anti-apoptotic cytokine essential for lymphocyte survival and expansion [44,45]. As mentioned previously, studies in human volun teers challenged with endotoxin revealed early repres sion of gene programs related to adaptive immunity [22]. In studies focused on mononuclear cellspecifi c expression profi les, Tang and colleagues [30,31] have also reported early repression of adaptive immunity genes in patients with sepsis. Finally, multiple studies in children with septic shock have reported, and validated, early and persistent repression of adaptive immunityrelated gene programs (for example, genes corresponding to the T cell receptor) [9,11,14,15,17,[32][33][34]. Th us, the concept of adaptive immune dysfunction as an early and prominent feature of clinical sepsis and septic shock seems to be well supported by the available genome-wide expression data.\\n\\nDevelopmental age is thought to be a major contributor to sepsis heterogeneity. Recently, a microarray-based study in children with septic shock corroborated this concept at the genomic level [46]. Four developmental age groups of children were compared based on wholeblood-derived gene expression profi les. Children in the 'neonate' group (&lt;28 days of age) demonstrated a unique expression profi le relative to older children. For example, children in the neonate group demonstrated widespread repression of genes corresponding to the triggering receptor expressed on myeloid cells 1 (TREM1) pathway. TREM1 is critical for amplifi cation of the infl ammatory response to microbial products and there has been recent interest in blockade of the TREM1 signaling pathway in septic shock [47]. Th e observation that TREM1 signaling may not be relevant in neonates with septic shock illustrates how some potential therapeutic strategies for septic shock may not have biological plausibility in certain developmental age groups.\\n\\n\\nBiomarker discovery\\n\\nA daily conundrum in the intensive care unit is the ability to distinguish which patients that meet criteria for SIRS are infected, and which patients with SIRS are not infected. Accordingly, there are ongoing eff orts to discover diagnostic biomarkers for sepsis (SIRS secondary to infection), and microarray approaches have the potential to enhance these eff orts. Several investigators have reported genome-level signatures that can distinguish patients with SIRS from patients with sepsis [26,29,31,48]. A substantial amount of work, including validation, remains to be done in order to leverage these datasets into clinically applicable diagnostic biomarkers, but the datasets nonetheless provide a foundation for the derivation and development of diagnostic biomarkers for sepsis.\\n\\nInvestigators have also applied microarray technology to address other important clinical challenges directly related to infection. Cobb and colleagues [49,50] have reported an expression signature (the 'ribonucleogram') having the potential to predict ventilator-associated pneumonia in critically ill blunt trauma patients up to 4 days before traditional clinical recognition. Similarly, Ramilo and colleagues [51] have reported expression signatures that can distinguish infl uenza A infection from bacterial infection, and Escherichia coli infection from Staphylococcus aureus infection, in hospitalized febrile children. In contrast, Tang and colleagues [30] were unable to defi ne organism-specifi c gene expression signatures (Gram positive versus Gram negative bacteria) in critically ill adults with sepsis.\\n\\nAnother aspect of biomarker development in sepsis surrounds stratifi cation biomarkers, particularly to predict outcome. Th eoretically, any gene that is consistently diff erentially regulated between survivors and nonsurvivors in a microarray dataset may warrant further investigation as a potential outcome biomarker. For example, a microarray study by Pachot and colleagues [27,52] identifi ed CX3CR1 (fractalkine receptor) as a potential outcome biomarker in sepsis. Similarly, Nowak and colleagues [53] have leveraged microarray data to identify chemokine (C-C motif ) ligand 4 (CCL4) as an outcome biomarker in children with septic shock. Both candidate stratifi cation biomarkers, however, require further validation.\\n\\nIL-8 has emerged as a robust stratifi cation biomarker in children with septic shock [54], and the rationale for pursuing it stemmed directly from microarray-based studies identifying IL-8 as one of the more highly expressed genes in pediatric non-survivors of septic shock, compared to survivors [34]. Subsequent studies demonstrated that serum IL-8 protein levels, measured within 24 hours of presentation to the intensive care unit with septic shock, could predict survival in pediatric septic shock with a probability of 95% [54]. Th e ability of IL-8 to serve as a stratifi cation biomarker was subsequently validated in a completely independent cohort of children with septic shock. Consequently, it has been proposed that IL-8 could be used in future pediatric septic shock interventional trials as a means to exclude patients having a high likelihood of survival with standard care, as a means of improving the risk-to-benefi t ratio of a given intervention. Th is type of stratifi cation strategy would be particularly applicable for an intervention that carries more than minimal risk. Interestingly, it appears that IL-8-based stratifi cation may not perform in a similarly robust manner in adults with septic shock [55], thus providing another example of how developmental age contributes to septic shock heterogeneity.\\n\\nCurrently, there is an ongoing eff ort to derive and validate a multi-biomarker sepsis outcome risk model in pediatric septic shock. Th e foundation of this eff ort is the relatively unbiased selection of a panel of candidate outcome biomarkers using microarray data from a large cohort of children with septic shock [56,57].\\n\\n\\nGene-expression-based identifi cation of septic shock subclasses\\n\\nViewing septic shock as a highly heterogeneous syndrome implies the existence of 'disease subclasses' , in an analogous manner to that encountered in the oncology fi eld [37]. Recently, there has been an attempt to identify septic shock subclasses in children based on genomewide expression profi ling [11]. Complete microarray data from a large cohort of children with septic shock, representing the fi rst 24 hours of admission, were used to identify septic shock subclasses. A heat map of over 6,000 diff erentially regulated genes was generated using an unsupervised clustering algorithm. Patients were then classifi ed into one of three subclasses (subclasses ' A' , 'B' , or 'C') based on statistically similar gene expression patterns, as determined by the fi rst and second order branching patterns of the heat map. Subsequently, the clinical database was mined to determine if there were any phenotypic diff erences between the three subclasses. Patients in subclass A had a signifi cantly higher level of illness severity as measured by mortality, organ failure, and illness severity score.\\n\\nTh e gene expression patterns that distinguished the subclasses were distilled to a 100-gene expression signature by conducting a leave-one-out cross-validation procedure and selecting the 100 genes having the greatest subclass prediction capability. Th ese 100 genes were then uploaded to a gene expression database that identifi ed enrichment for genes corresponding to adaptive immunity, glucocorticoid receptor signaling, and the peroxisome proliferator-activated receptor-α signaling pathway. Of note, the genes corresponding to these functional annotations were generally repressed in the subclass of patients with the higher level of illness severity (that is, subclass A patients).\\n\\nIn a subsequent study, the expression patterns of the 100 subclass-defi ning genes were depicted using visually intuitive gene expression mosaics and shown to a panel of clinicians with no formal bioinformatic training and blinded to the actual patient subclasses (Figure 1). Th e clinicians were able to allocate patients into the respective subclasses with a high degree of sensitivity and specifi city [15]. Th e ability to identify a subclass of children with a higher illness severity was further corrobor ated when the gene-expression-based subclassi fication strategy was applied to a separate validation cohort of children with septic shock [14]. Collectively, these studies demonstrate the feasibility of subclassifying patients with septic shock, in a clinically relevant manner, based on the expression patterns of a discrete set of genes having relevance to sepsis biology. Th e availability of clinical microfl uidics [58] and digital mRNA measurement technology [59] may allow for clinical feasibility of measuring the 100 class-defi ning genes in a timely manner that is suitable to direct patient care or for clinical trial stratifi cation.\\n\\n\\nDiscovery of novel targets and pathways\\n\\nTh e potential to interrogate the entire genome in a relatively unbiased manner provides an opportunity to discover previously unrecognized, or unconsidered, targets and pathways relevant to sepsis biology. Th is is a daunting task in the context of a highly heterogeneous syndrome such as clinical sepsis, and the many unavoidable confounding factors inherent to clinical sepsis microarray studies. Nonetheless, several studies illustrate the potential of genome-wide expression profi ling in the discovery of novel targets and pathways.\\n\\nFor example, using a combination of expression profi ling and in vitro approaches, Pathan and colleagues [60] have identifi ed interleukin-6 as a major contributor to myocardial depression in patients with meningococcal sepsis. Th is is a particularly intriguing and robust study because the study population is relatively homogeneous (that is, exclusively patients with meningococcal) compared to the majority of sepsis microarray studies that have enrolled patients with heterogeneous sepsis etiologies.\\n\\nIn one of the earliest clinical sepsis microarray studies, Pachot and colleagues [27] identifi ed a set of genes diff eren tially regulated between survivors and non-survivors. Th e gene most highly expressed in survivors, relative to non-survivors, was that encoding the chemokine receptor CX3CR1 (fractalkine receptor). In a subsequent validation study, these same investigators provided further evidence supporting the novel concept that dys regu lation of CX3CR1 in monocytes contributes to immuneparalysis in human sepsis [52]. Th ese studies further demonstrate the potential to discover novel pathways through discovery-oriented expression profi ling.\\n\\nSeveral studies in children with septic shock have documented early and persistent repression of gene programs directly related to zinc homeostasis, in combina tion with low serum zinc concentrations [9,11,17,32,34]. Since normal zinc homeostasis is absolutely critical to normal immune function [61], these obser vations have raised the possibility of zinc supplementation as a potentially safe and low cost therapeutic strategy in clinical septic shock and other forms of critical illness [62][63][64]. Importantly, Knoell and colleagues [65,66] have independently corroborated that zinc supplementation is a highly benefi cial strategy in experimental sepsis. Additional studies by Knoell and colleagues [67] have corroborated decreased plasma zinc concentrations in patients with sepsis, and that low plasma zinc concen trations correlate with higher illness severity. Furthermore, plasma zinc concentrations correlate inversely with mono cyte expression of the zinc transporter gene SLC39A8 (also know as ZIP8) [67,68]. Interestingly, microarray-based studies in children with septic shock have reported high levels of SLC39A8 expression in nonsurvivors, relative to survivors [34]. Despite the intriguing convergence of these data from independent laboratories, the safety and effi cacy of zinc supplementation in clinical sepsis remains to be directly demonstrated and is a current area of active investigation. One consideration for these studies will be the incorporation of trancriptomic analyses to determine if zinc supplementation infl uences the zinc-related gene repression patterns described above.\\n\\nIn the aforementioned studies involving children with septic shock, metalloproteinase (MMP)-8 has consistently been the highest expressed gene in patients with septic shock, relative to normal controls [9,11,17,[32][33][34]46]. In addition, MMP-8 is more highly expressed in patients with septic shock compared to patients with sepsis, and in septic shock non-survivors compared to septic shock survivors [69]. MMP-8 is also known as neutrophil collagenase because it is a neutrophil-derived protease that cleaves collagen in the extracellular matrix, but MMP-8 is also known to have other cellular sources and non-extracellular matrix substrates, including chemokines and cytokines [70]. Th e consistently high level of expression of MMP-8 in clinical septic shock recently stimulated the formal study of MMP-8 in experimental sepsis. Th ese studies demonstrated that either genetic abla tion of MMP-8 or pharmacologic inhibition of MMP-8 activity confers a signifi cant survival advantage in a murine model of sepsis [69]. While these studies require further development and validation, the fi ndings are intriguing given that there exist a number of drugs to eff ectively inhibit MMP-8 activity in the clinical setting [71].\\n\\n\\nConclusion\\n\\nDespite the tremendous methodological challenges that come with translational research involving humans with sepsis, microarray technology and complex bioinformatic approaches are beginning to provide novel insights into this complex syndrome. Progress, albeit slow, has been realized with regard to our understanding of the genomelevel response during sepsis, the identifi cation of potential novel targets and pathways, discovery of candidate diagnostic and stratifi cation biomarkers, and the possibility of clinically relevant and clinically feasible gene-expression-based subclassifi cation. Th e challenges ahead include robust validation studies, standardization of technical approaches, standardization and further development of analytical algorithms, and large scale collaborations.\\n\\n\\nAbbreviations\\n\\nIL, interleukin; MMP, metalloproteinase; SIRS, systemic infl ammatory response syndrome; TREM1, triggering receptor expressed on myeloid cells 1.\\n\\n\\nCompeting interests\\n\\nThe Cincinnati Children's Hospital Research Foundation and the author have a provisional patent for the use of interleukin-8 as a stratifi cation biomarker for septic shock.\\n\\nFigure 1 .\\n1Examples of gene expression mosaics for individual patients in septic shock subclasses A, B, and C, respectively[14,15]. The expression mosaics represent the expression patterns of the same 100 class-defi ning genes corresponding to adaptive immunity, glucocorticoid receptor signaling, and the peroxisome proliferator-activated receptor-α signaling pathway. The color bar on the right depicts the relative intensity of gene expression. These individual patient mosaics have not been previously published.\\nAcknowledgementsSupported by grants from the National Institutes of Health: RO1 GM064619, 1RC1HL100474, and R01 GM096994.\\n. Jd: Christie, Microarrays, Crit Care Med. 3312SupplChristie JD: Microarrays. Crit Care Med 2005, 33(12 Suppl):S449-452.\\n\\nMicroarray technology: an array of opportunities. D Gershon, Nature. 416Gershon D: Microarray technology: an array of opportunities. Nature 2002, 416:885-891.\\n\\nGenome-wide transcription profi ling of human sepsis: a systematic review. B M Tang, S J Huang, A S Mclean, Crit Care. 14237Tang BM, Huang SJ, McLean AS: Genome-wide transcription profi ling of human sepsis: a systematic review. Crit Care 2010, 14:R237.\\n\\nMicroarrays, deep sequencing and the true measure of the transcriptome. J H Malone, B Oliver, BMC Biol. 934Malone JH, Oliver B: Microarrays, deep sequencing and the true measure of the transcriptome. BMC Biol 2011, 9:34.\\n\\nExpectations, validity, and reality in gene expression profi ling. K Kim, S O Zakharkin, Allison Db, J Clin Epidemiol. 63Kim K, Zakharkin SO, Allison DB: Expectations, validity, and reality in gene expression profi ling. J Clin Epidemiol 2010, 63:950-959.\\n\\nRNA sequencing: advances, challenges and opportunities. F Ozsolak, P M Milos, Nat Rev Genet. 12Ozsolak F, Milos PM: RNA sequencing: advances, challenges and opportunities. Nat Rev Genet 2011, 12:87-98.\\n\\nIdentifi cation of neutrophil gelatinase-associated lipocalin as a novel early urinary biomarker for ischemic renal injury. J Mishra, Q Ma, A Prada, M Mitsnefes, K Zahedi, J Yang, J Barasch, P Devarajan, J Am Soc Nephrol. 14Mishra J, Ma Q, Prada A, Mitsnefes M, Zahedi K, Yang J, Barasch J, Devarajan P: Identifi cation of neutrophil gelatinase-associated lipocalin as a novel early urinary biomarker for ischemic renal injury. J Am Soc Nephrol 2003, 14:2534-2543.\\n\\nMicroarray data analysis: from disarray to consolidation and consensus. D B Allison, X Cui, G P Page, M Sabripour, Nat Rev Genet. 7Allison DB, Cui X, Page GP, Sabripour M: Microarray data analysis: from disarray to consolidation and consensus. Nat Rev Genet 2006, 7:55-65.\\n\\nGenomics of Pediatric SIRS/Septic Shock Investigators: Validating the genomic signature of pediatric septic shock. N Cvijanovich, T P Shanley, R Lin, G L Allen, N J Thomas, P Checchia, N Anas, R J Freishtat, M Monaco, K Odoms, B Sakthivel, H R Wong, Physiol Genomics. 34Cvijanovich N, Shanley TP, Lin R, Allen GL, Thomas NJ, Checchia P, Anas N, Freishtat RJ, Monaco M, Odoms K, Sakthivel B, Wong HR; Genomics of Pediatric SIRS/Septic Shock Investigators: Validating the genomic signature of pediatric septic shock. Physiol Genomics 2008, 34:127-134.\\n\\nMicroarray experimental design: power and sample size considerations. M C Yang, J J Yang, R A Mcindoe, J X She, Physiol Genomics. 16Yang MC, Yang JJ, McIndoe RA, She JX: Microarray experimental design: power and sample size considerations. Physiol Genomics 2003, 16:24-28.\\n\\nIdentifi cation of pediatric septic shock subclasses based on genome-wide expression profi ling. H R Wong, N Cvijanovich, R Lin, G L Allen, N J Thomas, D F Willson, R J Freishtat, N Anas, K Meyer, P A Checchia, M Monaco, K Odom, T P Shanley, BMC Med. 734Wong HR, Cvijanovich N, Lin R, Allen GL, Thomas NJ, Willson DF, Freishtat RJ, Anas N, Meyer K, Checchia PA, Monaco M, Odom K, Shanley TP: Identifi cation of pediatric septic shock subclasses based on genome-wide expression profi ling. BMC Med 2009, 7:34.\\n\\nGene Expression Dynamics Inspector (GEDI): for integrative analysis of expression profi les. G S Eichler, S Huang, D E Ingber, Bioinformatics. 19Eichler GS, Huang S, Ingber DE: Gene Expression Dynamics Inspector (GEDI): for integrative analysis of expression profi les. Bioinformatics 2003, 19:2321-2322.\\n\\nTowards a holistic, yet genecentered analysis of gene expression profi les: a case study of human lung cancers. Y Guo, G S Eichler, Y Feng, D E Ingber, S Huang, J Biomed Biotechnol. 69141Guo Y, Eichler GS, Feng Y, Ingber DE, Huang S: Towards a holistic, yet gene- centered analysis of gene expression profi les: a case study of human lung cancers. J Biomed Biotechnol 2006, 2006:69141.\\n\\nValidation of a gene expression-based subclassifi cation strategy for pediatric septic shock. H R Wong, N Z Cvijanovich, G L Allen, N J Thomas, R J Freishtat, N Anas, K Meyer, P A Checchia, R Lin, T P Shanley, M T Bigham, D S Wheeler, L A Doughty, K Tegtmeyer, S E Poynter, J M Kaplan, R S Chima, E Stalets, R K Basu, B M Varisco, F E Barr, Crit Care Med. 39Wong HR, Cvijanovich NZ, Allen GL, Thomas NJ, Freishtat RJ, Anas N, Meyer K, Checchia PA, Lin R, Shanley TP, Bigham MT, Wheeler DS, Doughty LA, Tegtmeyer K, Poynter SE, Kaplan JM, Chima RS, Stalets E, Basu RK, Varisco BM, Barr FE: Validation of a gene expression-based subclassifi cation strategy for pediatric septic shock. Crit Care Med 2011, 39:2511-2517.\\n\\nToward a clinically feasible gene expression-based subclassifi cation strategy for septic shock: proof of concept. H R Wong, D S Wheeler, K Tegtmeyer, S E Poynter, J M Kaplan, R S Chima, E Stalets, R K Basu, L A Doughty, Crit Care Med. 38Wong HR, Wheeler DS, Tegtmeyer K, Poynter SE, Kaplan JM, Chima RS, Stalets E, Basu RK, Doughty LA: Toward a clinically feasible gene expression-based subclassifi cation strategy for septic shock: proof of concept. Crit Care Med 2010, 38:1955-1961.\\n\\nInfl amm and Host Response to Injury Large Scale Collab. Res. Program: A network-based analysis of systemic infl ammation in humans. S E Calvano, W Xiao, D R Richards, R M Felciano, H V Baker, R J Cho, R O Chen, B H Brownstein, J P Cobb, S K Tschoeke, C Miller-Graziano, L L Moldawer, M N Mindrinos, R W Davis, R G Tompkins, S F Lowry, Nature. 437Calvano SE, Xiao W, Richards DR, Felciano RM, Baker HV, Cho RJ, Chen RO, Brownstein BH, Cobb JP, Tschoeke SK, Miller-Graziano C, Moldawer LL, Mindrinos MN, Davis RW, Tompkins RG, Lowry SF; Infl amm and Host Response to Injury Large Scale Collab. Res. Program: A network-based analysis of systemic infl ammation in humans. Nature 2005, 437:1032-1037.\\n\\nGenome-level longitudinal expression of signaling pathways and gene networks in pediatric septic shock. T P Shanley, N Cvijanovich, R Lin, G L Allen, N J Thomas, A Doctor, M Kalyanaraman, Penfi Tofi L Nm, S Monaco, M Odoms, K Barnes, M Sakthivel, B Aronow, B J Wong, H R , Mol Med. 13Shanley TP, Cvijanovich N, Lin R, Allen GL, Thomas NJ, Doctor A, Kalyanaraman M, Tofi l NM, Penfi l S, Monaco M, Odoms K, Barnes M, Sakthivel B, Aronow BJ, Wong HR: Genome-level longitudinal expression of signaling pathways and gene networks in pediatric septic shock. Mol Med 2007, 13:495-508.\\n\\nSuff redini AF: Compartmentalization of the acute cytokine response in humans after intravenous endotoxin administration. A J Boujoukos, G D Martich, E Supinski, J Appl Physiol. 74Boujoukos AJ, Martich GD, Supinski E, Suff redini AF: Compartmentalization of the acute cytokine response in humans after intravenous endotoxin administration. J Appl Physiol 1993, 74:3027-3033.\\n\\nActivation of the kallikrein-kinin system after endotoxin administration to normal human volunteers. Dela Cadena, R A Suff Redini, A F Page, J D Pixley, R A Kaufman, N Parrillo, J E Colman, R W , Blood. 81DeLa Cadena RA, Suff redini AF, Page JD, Pixley RA, Kaufman N, Parrillo JE, Colman RW: Activation of the kallikrein-kinin system after endotoxin administration to normal human volunteers. Blood 1993, 81:3313-3317.\\n\\nThe cardiovascular response of normal humans to the administration of endotoxin. A F Suff Redini, R E Fromm, M M Parker, M Brenner, J A Kovacs, R A Wesley, J E Parrillo, N Engl J Med. 321Suff redini AF, Fromm RE, Parker MM, Brenner M, Kovacs JA, Wesley RA, Parrillo JE: The cardiovascular response of normal humans to the administration of endotoxin. N Engl J Med 1989, 321:280-287.\\n\\nPromotion and subsequent inhibition of plasminogen activation after administration of intravenous endotoxin to normal subjects. A F Suff Redini, P C Harpel, J E Parrillo, N Engl J Med. 320Suff redini AF, Harpel PC, Parrillo JE: Promotion and subsequent inhibition of plasminogen activation after administration of intravenous endotoxin to normal subjects. N Engl J Med 1989, 320:1165-1172.\\n\\nSuff redini AF: Gene expression profi les of peripheral blood leukocytes after endotoxin challenge in humans. S Talwar, P J Munson, J Barb, C Fiuza, A P Cintron, C Logun, M Tropea, S Khan, D Reda, J H Shelhamer, R L Danner, Physiol Genomics. 25Talwar S, Munson PJ, Barb J, Fiuza C, Cintron AP, Logun C, Tropea M, Khan S, Reda D, Shelhamer JH, Danner RL, Suff redini AF: Gene expression profi les of peripheral blood leukocytes after endotoxin challenge in humans. Physiol Genomics 2006, 25:203-215.\\n\\nCorrelation of protein and gene expression profi les of infl ammatory proteins after endotoxin challenge in human subjects. U Prabhakar, T M Conway, P Murdock, J L Mooney, S Clark, P Hedge, B C Bond, E C Jazwinska, M R Barnes, F Tobin, V Damian-Iordachi, L Greller, M Hurle, A P Stubbs, Z Li, E I Valoret, C Erickson-Miller, L Cass, B Levitt, H M Davis, D K Jorkasky, W V Williams, DNA Cell Biol. 24Prabhakar U, Conway TM, Murdock P, Mooney JL, Clark S, Hedge P, Bond BC, Jazwinska EC, Barnes MR, Tobin F, Damian-Iordachi V, Greller L, Hurle M, Stubbs AP, Li Z, Valoret EI, Erickson-Miller C, Cass L, Levitt B, Davis HM, Jorkasky DK, Williams WV: Correlation of protein and gene expression profi les of infl ammatory proteins after endotoxin challenge in human subjects. DNA Cell Biol 2005, 24:410-431.\\n\\nThe use of gene-expression profi ling to identify candidate genes in human sepsis. B M Tang, A S Mclean, I W Dawes, S J Huang, R C Lin, Am J Respir Crit Care Med. 176Tang BM, McLean AS, Dawes IW, Huang SJ, Lin RC: The use of gene-expression profi ling to identify candidate genes in human sepsis. Am J Respir Crit Care Med 2007, 176:676-684.\\n\\nHuman endotoxemia: a model for mechanistic insight and therapeutic targeting. S F Lowry, Shock. 241SupplLowry SF: Human endotoxemia: a model for mechanistic insight and therapeutic targeting. Shock 2005, 24 Suppl 1:94-100.\\n\\nGene expression profi les diff erentiate between sterile SIRS and early sepsis. S B Johnson, M Lissauer, G V Bochicchio, R Moore, A S Cross, T M Scalea, Ann Surg. 245Johnson SB, Lissauer M, Bochicchio GV, Moore R, Cross AS, Scalea TM: Gene expression profi les diff erentiate between sterile SIRS and early sepsis. Ann Surg 2007, 245:611-621.\\n\\nSystemic transcriptional analysis in survivor and non-survivor septic shock patients: a preliminary study. A Pachot, A Lepape, S Vey, J Bienvenu, B Mougin, G Monneret, Immunol Lett. 106Pachot A, Lepape A, Vey S, Bienvenu J, Mougin B, Monneret G: Systemic transcriptional analysis in survivor and non-survivor septic shock patients: a preliminary study. Immunol Lett 2006, 106:63-71.\\n\\nGene profi ling in human blood leucocytes during recovery from septic shock. D Payen, A C Lukaszewicz, I Belikova, V Faivre, C Gelin, S Russwurm, J M Launay, N Sevenet, Intensive Care Med. 34Payen D, Lukaszewicz AC, Belikova I, Faivre V, Gelin C, Russwurm S, Launay JM, Sevenet N: Gene profi ling in human blood leucocytes during recovery from septic shock. Intensive Care Med 2008, 34:1371-1376.\\n\\nExpression profi ling: toward an application in sepsis diagnostics. M Prucha, A Ruryk, H Boriss, E Moller, R Zazula, Herold I Claus, R A Reinhart, K A Deigner, P Russwurm, S , Shock. 22Prucha M, Ruryk A, Boriss H, Moller E, Zazula R, Herold I, Claus RA, Reinhart KA, Deigner P, Russwurm S: Expression profi ling: toward an application in sepsis diagnostics. Shock 2004, 22:29-33.\\n\\nGeneexpression profi ling of gram-positive and gram-negative sepsis in critically ill patients. B M Tang, A S Mclean, I W Dawes, S J Huang, M J Cowley, R C Lin, Crit Care Med. 36Tang BM, McLean AS, Dawes IW, Huang SJ, Cowley MJ, Lin RC: Gene- expression profi ling of gram-positive and gram-negative sepsis in critically ill patients. Crit Care Med 2008, 36:1125-1128.\\n\\nGene-expression profi ling of peripheral blood mononuclear cells in sepsis. B M Tang, A S Mclean, I W Dawes, S J Huang, R C Lin, Crit Care Med. 37Tang BM, McLean AS, Dawes IW, Huang SJ, Lin RC: Gene-expression profi ling of peripheral blood mononuclear cells in sepsis. Crit Care Med 2009, 37:882-888.\\n\\nGenomics of Pediatric SIRS/ Septic Shock Investigators: Genomic expression profi ling across the pediatric systemic infl ammatory response syndrome, sepsis, and septic shock spectrum. H R Wong, N Cvijanovich, G L Allen, R Lin, N Anas, K Meyer, R J Freishtat, M Monaco, K Odoms, B Sakthivel, T P Shanley, Crit Care Med. 37Wong HR, Cvijanovich N, Allen GL, Lin R, Anas N, Meyer K, Freishtat RJ, Monaco M, Odoms K, Sakthivel B, Shanley TP; Genomics of Pediatric SIRS/ Septic Shock Investigators: Genomic expression profi ling across the pediatric systemic infl ammatory response syndrome, sepsis, and septic shock spectrum. Crit Care Med 2009, 37:1558-1566.\\n\\nLeukocyte subsetderived genomewide expression profi les in pediatric septic shock. H R Wong, R J Freishtat, M Monaco, K Odoms, T P Shanley, Pediatr Crit Care Med. 11Wong HR, Freishtat RJ, Monaco M, Odoms K, Shanley TP: Leukocyte subset- derived genomewide expression profi les in pediatric septic shock. Pediatr Crit Care Med 2010, 11:349-355.\\n\\nGenomics of Pediatric SIRS/ Septic Shock Investigators: Genome-level expression profi les in pediatric septic shock indicate a role for altered zinc homeostasis in poor outcome. H R Wong, T P Shanley, B Sakthivel, N Cvijanovich, R Lin, G L Allen, N J Thomas, A Doctor, M Kalyanaraman, Penfi Tofi L Nm, S Monaco, M Tagavilla, M A Odoms, K Dunsmore, K Barnes, M Aronow, B J , Physiol Genomics. 30Wong HR, Shanley TP, Sakthivel B, Cvijanovich N, Lin R, Allen GL, Thomas NJ, Doctor A, Kalyanaraman M, Tofi l NM, Penfi l S, Monaco M, Tagavilla MA, Odoms K, Dunsmore K, Barnes M, Aronow BJ; Genomics of Pediatric SIRS/ Septic Shock Investigators: Genome-level expression profi les in pediatric septic shock indicate a role for altered zinc homeostasis in poor outcome. Physiol Genomics 2007, 30:146-155.\\n\\nSepsis: time to reconsider the concept. J Carlet, J Cohen, Calandra T Opal, S M Masur, H , Crit Care Med. 36Carlet J, Cohen J, Calandra T, Opal SM, Masur H: Sepsis: time to reconsider the concept. Crit Care Med 2008, 36:964-966.\\n\\nSuch stuff as dreams are made on: mediator-directed therapy in sepsis. J C Marshall, Nat Rev Drug Discov. 2Marshall JC: Such stuff as dreams are made on: mediator-directed therapy in sepsis. Nat Rev Drug Discov 2003, 2:391-405.\\n\\nMeasures, markers, and mediators: toward a staging system for clinical sepsis. A report of the Fifth Toronto Sepsis Roundtable. J C Marshall, J L Vincent, M P Fink, D J Cook, G Rubenfeld, D Foster, C J Fisher, Jr, E Faist, K Reinhart, Crit Care Med. 31Marshall JC, Vincent JL, Fink MP, Cook DJ, Rubenfeld G, Foster D, Fisher CJ, Jr, Faist E, Reinhart K: Measures, markers, and mediators: toward a staging system for clinical sepsis. A report of the Fifth Toronto Sepsis Roundtable, Toronto, Ontario, Canada, October 25-26, 2000. Crit Care Med 2003, 31:1560-1567.\\n\\nGene expression in human sepsis: what have we learned?. J A Russell, Crit Care. 15121Russell JA: Gene expression in human sepsis: what have we learned? Crit Care 2011, 15:121.\\n\\nPro-versus anti-infl ammatory cytokine profi le in patients with severe sepsis: a marker for prognosis and future therapeutic options. C A Gogos, E Drosou, H P Bassaris, A Skoutelis, J Infect Dis. 181Gogos CA, Drosou E, Bassaris HP, Skoutelis A: Pro-versus anti-infl ammatory cytokine profi le in patients with severe sepsis: a marker for prognosis and future therapeutic options. J Infect Dis 2000, 181:176-180.\\n\\nCirculating cytokine/ inhibitor profi les reshape the understanding of the SIRS/CARS continuum in sepsis and predict mortality. M F Osuchowski, K Welch, J Siddiqui, D G Remick, J Immunol. 177Osuchowski MF, Welch K, Siddiqui J, Remick DG: Circulating cytokine/ inhibitor profi les reshape the understanding of the SIRS/CARS continuum in sepsis and predict mortality. J Immunol 2006, 177:1967-1974.\\n\\nChronic sepsis mortality characterized by an individualized infl ammatory response. M F Osuchowski, K Welch, H Yang, J Siddiqui, D G Remick, J Immunol. 179Osuchowski MF, Welch K, Yang H, Siddiqui J, Remick DG: Chronic sepsis mortality characterized by an individualized infl ammatory response. J Immunol 2007, 179:623-630.\\n\\nThe pathophysiology and treatment of sepsis. R S Hotchkiss, I E Karl, N Engl J Med. 348Hotchkiss RS, Karl IE: The pathophysiology and treatment of sepsis. N Engl J Med 2003, 348:138-150.\\n\\nImmunotherapy for sepsis -a new approach against an ancient foe. R S Hotchkiss, S Opal, N Engl J Med. 363Hotchkiss RS, Opal S: Immunotherapy for sepsis -a new approach against an ancient foe. N Engl J Med 2010, 363:87-89.\\n\\nIL-7 promotes T cell viability, traffi cking, and functionality and improves survival in sepsis. J Unsinger, M Mcglynn, K R Kasten, A S Hoekzema, E Watanabe, J T Muenzer, J S Mcdonough, J Tschoep, T A Ferguson, J E Mcdunn, M Morre, D A Hildeman, C C Caldwell, R S Hotchkiss, J Immunol. 184Unsinger J, McGlynn M, Kasten KR, Hoekzema AS, Watanabe E, Muenzer JT, McDonough JS, Tschoep J, Ferguson TA, McDunn JE, Morre M, Hildeman DA, Caldwell CC, Hotchkiss RS: IL-7 promotes T cell viability, traffi cking, and functionality and improves survival in sepsis. J Immunol 2010, 184:3768-3779.\\n\\nInterleukin-7 (IL-7) treatment accelerates neutrophil recruitment through gamma delta T-cell IL-17 production in a murine model of sepsis. K R Kasten, P S Prakash, J Unsinger, H S Goetzman, L G England, C M Cave, A P Seitz, C N Mazuski, T T Zhou, M Morre, R S Hotchkiss, D A Hildeman, C C Caldwell, Infect Immun. 78Kasten KR, Prakash PS, Unsinger J, Goetzman HS, England LG, Cave CM, Seitz AP, Mazuski CN, Zhou TT, Morre M, Hotchkiss RS, Hildeman DA, Caldwell CC: Interleukin-7 (IL-7) treatment accelerates neutrophil recruitment through gamma delta T-cell IL-17 production in a murine model of sepsis. Infect Immun 2010, 78:4714-4722.\\n\\n. J L Wynn, N Z Cvijanovich, G L Allen, N J Thomas, R J Freishtat, N Anas, K Meyer, Wong, Critical Care. 16204Wynn JL, Cvijanovich NZ, Allen GL, Thomas NJ, Freishtat RJ, Anas N, Meyer K, Wong Critical Care 2012, 16:204\\n\\nThe infl uence of developmental age on the early transcriptomic response of children with septic shock. P A Checchia, R Lin, T P Shanley, M T Bigham, S Banschbach, E Beckman, H R Wong, Mol Med. Epub ahead of printChecchia PA, Lin R, Shanley TP, Bigham MT, Banschbach S, Beckman E, Wong HR: The infl uence of developmental age on the early transcriptomic response of children with septic shock. Mol Med 2011 [Epub ahead of print].\\n\\nTREM-1 amplifi es infl ammation and is a crucial mediator of septic shock. A Bouchon, F Facchetti, M A Weigand, M Colonna, Nature. 410Bouchon A, Facchetti F, Weigand MA, Colonna M: TREM-1 amplifi es infl ammation and is a crucial mediator of septic shock. Nature 2001, 410:1103-1107.\\n\\nDiff erential expression of toll-like receptor genes: sepsis compared with sterile infl ammation 1 day before sepsis diagnosis. M E Lissauer, S B Johnson, G V Bochicchio, C J Feild, A S Cross, J D Hasday, C C Whiteford, W A Nussbaumer, M Towns, T M Scalea, Shock. 31Lissauer ME, Johnson SB, Bochicchio GV, Feild CJ, Cross AS, Hasday JD, Whiteford CC, Nussbaumer WA, Towns M, Scalea TM: Diff erential expression of toll-like receptor genes: sepsis compared with sterile infl ammation 1 day before sepsis diagnosis. Shock 2009, 31:238-244.\\n\\nValidation of the riboleukogram to detect ventilator-associated pneumonia after severe injury. J P Cobb, E E Moore, D L Hayden, J P Minei, J Cuschieri, J Yang, Q Li, N Lin, B H Brownstein, L Hennessy, P H Mason, W S Schierding, D J Dixon, R G Tompkins, H S Warren, D A Schoenfeld, R V Maier, Ann Surg. 250Cobb JP, Moore EE, Hayden DL, Minei JP, Cuschieri J, Yang J, Li Q, Lin N, Brownstein BH, Hennessy L, Mason PH, Schierding WS, Dixon DJ, Tompkins RG, Warren HS, Schoenfeld DA, Maier RV: Validation of the riboleukogram to detect ventilator-associated pneumonia after severe injury. Ann Surg 2009, 250:531-539.\\n\\nPlasticity of the systemic infl ammatory response to acute infection during critical illness: development of the riboleukogram. J E Mcdunn, K D Husain, A D Polpitiya, A Burykin, J Ruan, Q Li, W Schierding, N Lin, D Dixon, W Zhang, C M Coopersmith, W M Dunne, M Colonna, B K Ghosh, J P Cobb, PLoS One. 31564McDunn JE, Husain KD, Polpitiya AD, Burykin A, Ruan J, Li Q, Schierding W, Lin N, Dixon D, Zhang W, Coopersmith CM, Dunne WM, Colonna M, Ghosh BK, Cobb JP: Plasticity of the systemic infl ammatory response to acute infection during critical illness: development of the riboleukogram. PLoS One 2008, 3:e1564.\\n\\nChaussabel D: Gene expression patterns in blood leukocytes discriminate patients with acute infections. O Ramilo, W Allman, W Chung, A Mejias, M Ardura, C Glaser, K M Wittkowski, B Piqueras, J Banchereau, A K Palucka, Blood. 109Ramilo O, Allman W, Chung W, Mejias A, Ardura M, Glaser C, Wittkowski KM, Piqueras B, Banchereau J, Palucka AK, Chaussabel D: Gene expression patterns in blood leukocytes discriminate patients with acute infections. Blood 2007, 109:2066-2077.\\n\\nDecreased expression of the fractalkine receptor CX3CR1 on circulating monocytes as new feature of sepsis-induced immunosuppression. A Pachot, M A Cazalis, F Venet, F Turrel, C Faudot, N Voirin, J Diasparra, N Bourgoin, F Poitevin, B Mougin, A Lepape, G Monneret, J Immunol. 180Pachot A, Cazalis MA, Venet F, Turrel F, Faudot C, Voirin N, Diasparra J, Bourgoin N, Poitevin F, Mougin B, Lepape A, Monneret G: Decreased expression of the fractalkine receptor CX3CR1 on circulating monocytes as new feature of sepsis-induced immunosuppression. J Immunol 2008, 180:6421-6429.\\n\\nAdmiss ion chemokine (C-C motif) ligand 4 levels predict survival in pediatric septic shock. J E Nowak, D S Wheeler, K K Harmon, H R Wong, Pediatr Crit Care Med. 11Nowak JE, Wheeler DS, Harmon KK, Wong HR: Admiss ion chemokine (C-C motif) ligand 4 levels predict survival in pediatric septic shock. Pediatr Crit Care Med 2010, 11:213-216.\\n\\nInterleukin-8 as a stratifi cation tool for interventional trials involving pediatric septic shock. H R Wong, N Cvijanovich, D S Wheeler, M T Bigham, M Odoms, K Macias, W L Williams, M D , Am J Respir Crit Care Med. 178Wong HR, Cvijanovich N, Wheeler DS, Bigham MT, M onaco M, Odoms K, Macias WL, Williams MD: Interleukin-8 as a stratifi cation tool for interventional trials involving pediatric septic shock. Am J Respir Crit Care Med 2008, 178:276-282.\\n\\nPlasma interleukin-8 is not an eff ective risk stratifi cation tool for adults with vasopressor-dependent septic shock. C S Calfee, B T Thompson, P E Parsons, L B Ware, M A Mat Thay, H R Wong, Crit Care Med. 38Calfee CS, Thompson BT, Parsons PE, Ware LB, Mat thay MA, Wong HR: Plasma interleukin-8 is not an eff ective risk stratifi cation tool for adults with vasopressor-dependent septic shock. Crit Care Med 2010, 38:1436-1441.\\n\\nBiomarker discovery and deve lopment in pediatric critical care medicine. J M Kaplan, H R Wong, Pediatr Crit Care Med. 12Kaplan JM, Wong HR: Biomarker discovery and deve lopment in pediatric critical care medicine. Pediatr Crit Care Med 2011, 12:165-173.\\n\\nBiomarkers for pediatric s epsis and septic shock. S W Standage, H R Wong, Expert Rev Anti Infect Ther. 9Standage SW, Wong HR: Biomarkers for pediatric s epsis and septic shock. Expert Rev Anti Infect Ther 2011, 9:71-79.\\n\\nToner M; Infl ammation and the Host Response to Injury Collaborative Research Program: Clinical microfl uidics for neutrophil genomics and proteomics. K T Kotz, W Xiao, C Miller-Graziano, W J Qian, A Rus Som, E A Warner, L L Moldawer, A De, P E Bankey, B O Petritis, Camp Dg 2nd, A E Rosenbach, J Goverman, S P Fagan, B H Brownstein, D Irimia, W Xu, J Wilhelmy, M N Mindrinos, R D Smith, R W Davis, R G Tompkins, Nat Med. 16Kotz KT, Xiao W, Miller-Graziano C, Qian WJ, Rus som A, Warner EA, Moldawer LL, De A, Bankey PE, Petritis BO, Camp DG 2nd, Rosenbach AE, Goverman J, Fagan SP, Brownstein BH, Irimia D, Xu W, Wilhelmy J, Mindrinos MN, Smith RD, Davis RW, Tompkins RG, Toner M; Infl ammation and the Host Response to Injury Collaborative Research Program: Clinical microfl uidics for neutrophil genomics and proteomics. Nat Med 2010, 16:1042-1047.\\n\\nDirect multiplexed measurement of gene expression with color-coded probe pairs. G K Geiss, R E Bumgarner, B Birditt, T Dahl, N Dowid Ar, D L Dunaway, H P Fell, S Ferree, R D George, T Grogan, J J James, M Maysuria, J D Mitton, P Oliveri, J L Osborn, T Peng, A L Ratcliff E, P J Webster, E H Davidson, L Hood, K Dimitrov, Nat Biotechnol. 26Geiss GK, Bumgarner RE, Birditt B, Dahl T, Dowid ar N, Dunaway DL, Fell HP, Ferree S, George RD, Grogan T, James JJ, Maysuria M, Mitton JD, Oliveri P, Osborn JL, Peng T, Ratcliff e AL, Webster PJ, Davidson EH, Hood L, Dimitrov K: Direct multiplexed measurement of gene expression with color-coded probe pairs. Nat Biotechnol 2008, 26:317-325.\\n\\nRole of interleukin 6 in myocardial dysfunction of meningococcal septic shock. N Pathan, C A Hemingway, A A Alizadeh, A C Stephens, J C Boldrick, E E Oragui, C Mccabe, S B Welch, A Whitney, P O&amp;apos;gara, S Nadel, D A Relman, S E Harding, M Levin, Lancet. 363Pathan N, Hemingway CA, Alizadeh AA, Stephens AC , Boldrick JC, Oragui EE, McCabe C, Welch SB, Whitney A, O'Gara P, Nadel S, Relman DA, Harding SE, Levin M: Role of interleukin 6 in myocardial dysfunction of meningococcal septic shock. Lancet 2004, 363:203-209.\\n\\nZinc homeostasis and immunity. L Rink, H Haase, Trends Immunol. 28Rink L, Haase H: Zinc homeostasis and immunity. Trends Immunol 2007, 28:1-4.\\n\\nZinc homeostasis in pediatric critical illness. N Z Cvijanovich, J C King, H R Flori, G Gildengorin, H R Wong, Pediatr Crit Care Med. 10Cvijanovich NZ, King JC, Flori HR, Gildengorin G , Wong HR: Zinc homeostasis in pediatric critical illness. Pediatr Crit Care Med 2009, 10:29-34.\\n\\nZin c supplementation in critically ill patients: a key pharmaconutrient?. D K Heyland, N Jones, N Z Cvijanovich, H Wong, J Parenter Enteral Nutr. 32Heyland DK, Jones N, Cvijanovich NZ, Wong H: Zin c supplementation in critically ill patients: a key pharmaconutrient? J Parenter Enteral Nutr 2008, 32:509-519.\\n\\nEff ects of pharmaconutrients on cellular dysfunction and the microcirculation in critical illness. L R Weitzel, W J Mayles, P A Sandoval, P Wischmeyer, Curr Opin Anaesthesiol. 22Weitzel LR, Mayles WJ, Sandoval PA, Wischmeyer P E: Eff ects of pharmaconutrients on cellular dysfunction and the microcirculation in critical illness. Curr Opin Anaesthesiol 2009, 22:177-183.\\n\\nZinc modulates the innate immune response in vivo to polymicrobial sepsis through regulation of NF-kappaB. S Bao, M J Liu, B Lee, B Besecker, J P Lai, D C Guttri Dge, D L Knoell, Am J Physiol Lung Cell Mol Physiol. 298Bao S, Liu MJ, Lee B, Besecker B, Lai JP, Guttri dge DC, Knoell DL: Zinc modulates the innate immune response in vivo to polymicrobial sepsis through regulation of NF-kappaB. Am J Physiol Lung Cell Mol Physiol 2010, 298:L744-754.\\n\\nZinc defi ciency increases organ damage and mortality in a murine model of polymicrobial sepsis. D L Knoell, M W Julian, S Bao, B Besecker, J E Macre, G D Leikauf, R A Disilvestro, E D Crouser, Crit Care Med. 37Knoell DL, Julian MW, Bao S, Besecker B, Macre J E, Leikauf GD, DiSilvestro RA, Crouser ED: Zinc defi ciency increases organ damage and mortality in a murine model of polymicrobial sepsis. Crit Care Med 2009, 37:1380-1388.\\n\\nA comparison of zinc metabolism, infl ammation, and disease severity in critically ill infected and noninfected adults early after intensive care unit admission. B Y Besecker, M C Exline, J Hollyfi Eld, G Phillips, R A Disilvestro, M D Wewers, D L Knoell, Am J Clin Nutr. 93Besecker BY, Exline MC, Hollyfi eld J, Phillips G , Disilvestro RA, Wewers MD, Knoell DL: A comparison of zinc metabolism, infl ammation, and disease severity in critically ill infected and noninfected adults early after intensive care unit admission. Am J Clin Nutr 2011, 93:1356-1364.\\n\\nImpact of zinc metabolism on innate immune function in the setting of sepsis. D L Knoell, M J Liu, Int J Vitam Nutr Res. 80Knoell DL, Liu MJ: Impact of zinc metabolism on innate immune function in the setting of sepsis. Int J Vitam Nutr Res 2010, 80:271-277.\\n\\nA novel role for matrix metalloproteinase-8 in sepsis. P D Solan, K E Dunsmore, A G Denenberg, K Odom, B Zin Garelli, H R Wong, Crit Care Med. Epub ahead of printSolan PD, Dunsmore KE, Denenberg AG, Odom K, Zin garelli B, Wong HR: A novel role for matrix metalloproteinase-8 in sepsis. Crit Care Med 2011 [Epub ahead of print].\\n\\nMatrix metalloproteinase-8 : cleavage can be decisive. P Van Lint, C Libert, Cytokine Growth Factor Rev. 17Van Lint P, Libert C: Matrix metalloproteinase-8 : cleavage can be decisive. Cytokine Growth Factor Rev 2006, 17:217-223.\\n\\nMatrix metalloproteinases as drug targets in infections caused by gram-negative bacteria and in septic shock. I Vanlaere, C Libert, 10.1186/cc10537Clin Microbiol Rev. 22Vanlaere I, Libert C: Matrix metalloproteinases as drug targets in infections caused by gram-negative bacteria and in septic shock. Clin Microbiol Rev 2009, 22:224-239. doi:10.1186/cc10537\\n\\nCite this article as: Wong HR: Clinical review: Sepsis and septic shock -the potential of gene arrays. Critical Care. 16204Cite this article as: Wong HR: Clinical review: Sepsis and septic shock -the potential of gene arrays. Critical Care 2012, 16:204.\\n</td>\n      <td>4251</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>5694</td>\n      <td>36733880</td>\n      <td>\\n\\n2014\\n\\n\\nKremenchugskaya str, 3-4-198 Moscow121352russia\\n\\nInternational Journal of Nanomedicine\\n201410.2147/IJN.S67938\\n\\n\\nDear editor\\n\\nWith great interest we have read an article by Wu et al 1 recently published in the International Journal of Nanomedicine aimed to estimate cytotoxicity, toxicity, and histopathological changes, as well as the postsurgical antiadhesion potential of biodegradable and thermosensitive micelles by combining in vitro and in vivo models.\\n\\nOur congratulations to Wu et al 1 for their new, precisely designed, and promising study of nanoparticles used in the prevention of postsurgical adhesions. Even more so when we know that in most clinical studies the adhesion prevention adjuvants have failed. Taking into account medical and financial problems associated with postsurgical adhesions worldwide in the health care system, new developments in this area are welcomed and call for further investigation.\\n\\nWe do not question the definite study question and design: the experimental models are precisely formed and trigger correctly described results, a knowledgeable discussion and subsequent reasonable conclusions in line with their achievements.\\n\\nHowever, in our opinion, for these biodegradable and thermosensitive micelles to be recommended for further clinical applications, experimental studies should be designed examining different doses of nanoparticles in order to establish the optimal dose of micelles, an evaluation of the impact of these materials in different time points after their application, ie, 14, 21, 28 days, as well as their long-term impact after 6 and 12 months.\\n\\nThe impact of these particles on embryos in in vivo models should also be evaluated: do these particles last longer or do they have an impact on uterine wall tissue?\\n\\nIn the future a personalized adhesion prevention strategy could be developed, with application of state of the art technologies, taking into account genetic and constitutional predisposing factors of the patients undergoing these surgical procedures, with targeted predisposing genetic and constitutional conditions. 2 Subsequently, these biodegradable and self-assembling micelles could be designed to contain medication such as recombinant tissue plasminogen activator (rtPA), 3 tPA genes, inhibitors of plasminogen activator inhibitor-1 (PAI-1) genes, or other cytokines activating the expression of individually targeted genes at certain times after surgery in order to lyse and remove temporary fibrinogenous adhesions.\\n\\nIn conclusion, biodegradable micelles containing certain medications should be a basis for future personalized adhesion prevention strategies.\\n DovepressInternational Journal of Nanomedicine 2014:9 submit your manuscript | www.dovepress.comDovepressDovepress3376Mynbaev et alDisclosureThe authors report no conflicts of interest or financial ties to disclose in regards to this communication.\\nBiodegradable and thermosensitive micelles inhibit ischemia-induced postoperative peritoneal adhesion. Q Wu, L Li, N Wang, Int J Nanomedicine. 9Wu Q, Li L, Wang N, et al. Biodegradable and thermosensitive micelles inhibit ischemia-induced postoperative peritoneal adhesion. Int J Nanomedicine. 2014;9:727-734.\\n\\nComparison of lovastatin and hyaluronic acid/carboxymethyl cellulose on experimental created peritoneal adhesion model in rats. E Arslan, T Talih, B Oz, B Halaclar, K Caglayan, M Sipahi, Int J Surg. 122Arslan E, Talih T, Oz B, Halaclar B, Caglayan K, Sipahi M. Comparison of lovastatin and hyaluronic acid/carboxymethyl cellulose on experimen- tal created peritoneal adhesion model in rats. Int J Surg. 2014;12(2): 120-124.\\n\\nTowards gene therapy of postoperative adhesions: fiber and transcriptional modifications enhance adenovirus targeting towards human adhesion cells. S Nair, G M Saed, H M Atta, Gynecol Obstet Invest. 762Nair S, Saed GM, Atta HM, et al. Towards gene therapy of postoperative adhesions: fiber and transcriptional modifications enhance adenovirus targeting towards human adhesion cells. Gynecol Obstet Invest. 2013; 76(2):119-124.\\n</td>\n      <td>4281</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>5774</td>\n      <td>149444454</td>\n      <td>\\nSecond-generation molecular subgrouping of medulloblastoma: an international meta-analysis of Group 3 and Group 4 subtypes\\n\\n\\nT Sharma \\nE C Schwalbe \\nSecond-generation molecular subgrouping of medulloblastoma: an international meta-analysis of Group 3 and Group 4 subtypes\\n\\nAppendixSupplementary MethodsWe undertook a retrospective, integrative analysis of Grp3/4 substructure from three published component studies [1][2][3], whose respective major findings are summarized in figure 1. Quality control checks revealed no biases attributable to sample type (i.e. fresh frozen vs formalin-fixed, paraffin embedded derivatives), array type (450k vs 850k), or source study/institution (supplementaryFigure 3). 69 previously published samples were removed due to low methylation classifier prediction scores [4] and 80 samples were additionally removed due to shared genotypes (i.e. duplicate samples/patients). 28 samples were removed due to a 'non-MB' prediction by the methylation classifier (supplementary Tables 1 and 2; supplementaryFigure 2. The cohorts and their demographics are summarized inTable 1.MB classifierBefore training the classifier, a correction for the type of material tissue (FFPE or frozen) and array (450k or EPIC) was performed by fitting a linear model to the log2-transformed intensity values (removeBatchEffect function, limma package v.3.36 [5] including the material tissue and array type as factor variables. The methylated and unmethylated signals were corrected individually. Estimated batch effects were also used to adjust diagnostic samples or test samples within the cross-validation. Beta values were calculated from the retransformed intensities using an offset of 100 (as recommended by Illumina).To train the RF classifier, the randomForest R package [6] was used. For feature selection, the 50,000 CpG probes with highest SD were selected. In a second step, an RF was trained to calculate a RF permutation-based variable importance measure for the selected CpG probes. Finally, the 10,000 CpG probes with highest variable importance were used to train the final RF model. For both RF models 5,000 trees were fitted and imbalanced class sizes were accounted for by downsampling to the minority class; this ensures an identical number of samples per class and tree [7]. The performance of the RF model was validated by a threefold cross-validation, validating in each fold all steps of the classifier development, starting with the batch adjustment.The classification scores generated by the RF model (that is, the proportion of trees voting for a class) perform already well when used to predict the correct Grp3/4 subtype, but they do not reflect well calibrated class probabilities that may guide clinical decision making [6]. Furthermore, the distribution of the RF scores may vary between classes, which make inter-class comparisons difficult. To obtain scores that are comparable between classes and that are improved estimates of the certainty of individual predictions, we performed a classification score recalibration by mapping the original scores to more accurate class probabilities [8]. To find such a mapping, an L2-penalized, multinomial, logistic regression model was fitted, which takes the methylation class as response variable and the RF scores as explanatory variables. The R package glmnet [9] was used to fit this model. In addition, the model was fitted by incorporating a small ridge-penalty (L2) on the likelihood to prevent overfitting, as well as to stabilize estimation in situations in which classes are perfectly separable. The amount of this regularization, that is, the penalization parameter, is determined by running a tenfold cross-validation and choosing penalization parameter that minimizes the mean squared error (MSE). Independent RF scores are needed to fit this model, that is, the scores need to be generated by a RF classifier that was not trained using the same samples, otherwise the RF scores will be systematically biased and not comparable to scores of unseen cases. As such, RF scores generated by the threefold cross-validation are used. To validate the class predictions generated by using the recalibrated scores of the calibration model, an additional, nested threefold cross-validation loop is incorporated into the main threefold cross-validation. Within each\\n\\nSupplementary Methods\\n\\nWe undertook a retrospective, integrative analysis of Grp3/4 substructure from three published component studies [1][2][3], whose respective major findings are summarized in figure 1. Quality control checks revealed no biases attributable to sample type (i.e. fresh frozen vs formalin-fixed, paraffin embedded derivatives), array type (450k vs 850k), or source study/institution (supplementary Figure 3). 69 previously published samples were removed due to low methylation classifier prediction scores [4] and 80 samples were additionally removed due to shared genotypes (i.e. duplicate samples/patients). 28 samples were removed due to a 'non-MB' prediction by the methylation classifier (supplementary Tables 1 and 2; supplementary Figure 2. The cohorts and their demographics are summarized in Table 1.\\n\\n\\nMB classifier\\n\\nBefore training the classifier, a correction for the type of material tissue (FFPE or frozen) and array (450k or EPIC) was performed by fitting a linear model to the log2-transformed intensity values (removeBatchEffect function, limma package v.3.36 [5] including the material tissue and array type as factor variables. The methylated and unmethylated signals were corrected individually. Estimated batch effects were also used to adjust diagnostic samples or test samples within the cross-validation. Beta values were calculated from the retransformed intensities using an offset of 100 (as recommended by Illumina).\\n\\nTo train the RF classifier, the randomForest R package [6] was used. For feature selection, the 50,000 CpG probes with highest SD were selected. In a second step, an RF was trained to calculate a RF permutation-based variable importance measure for the selected CpG probes. Finally, the 10,000 CpG probes with highest variable importance were used to train the final RF model. For both RF models 5,000 trees were fitted and imbalanced class sizes were accounted for by downsampling to the minority class; this ensures an identical number of samples per class and tree [7]. The performance of the RF model was validated by a threefold cross-validation, validating in each fold all steps of the classifier development, starting with the batch adjustment.\\n\\nThe classification scores generated by the RF model (that is, the proportion of trees voting for a class) perform already well when used to predict the correct Grp3/4 subtype, but they do not reflect well calibrated class probabilities that may guide clinical decision making [6]. Furthermore, the distribution of the RF scores may vary between classes, which make inter-class comparisons difficult. To obtain scores that are comparable between classes and that are improved estimates of the certainty of individual predictions, we performed a classification score recalibration by mapping the original scores to more accurate class probabilities [8]. To find such a mapping, an L2-penalized, multinomial, logistic regression model was fitted, which takes the methylation class as response variable and the RF scores as explanatory variables. The R package glmnet [9] was used to fit this model. In addition, the model was fitted by incorporating a small ridge-penalty (L2) on the likelihood to prevent overfitting, as well as to stabilize estimation in situations in which classes are perfectly separable. The amount of this regularization, that is, the penalization parameter, is determined by running a tenfold cross-validation and choosing penalization parameter that minimizes the mean squared error (MSE). Independent RF scores are needed to fit this model, that is, the scores need to be generated by a RF classifier that was not trained using the same samples, otherwise the RF scores will be systematically biased and not comparable to scores of unseen cases. As such, RF scores generated by the threefold cross-validation are used. To validate the class predictions generated by using the recalibrated scores of the calibration model, an additional, nested threefold cross-validation loop is incorporated into the main threefold cross-validation. Within each cross-validation run this nested threefold cross-validation is applied to generate independent RF scores, which are then used to train a calibration model. The predicted RF scores resulting from predicting the one-third test data of the outer cross-validation loop are then recalibrated by applying the calibration model that was fitted on the RF scores generated on the other two-third of the data by the nested cross-validation. Performances of the resulting classifier predictions and scores generated by the cross-validation were assessed by the misclassification error, multiclass area under receiver operating characteristic (ROC) curve (AUC) and the multiclass Brier score. The misclassification error measures the frequency of falsely assigned class labels when using the maximum of the RF scores or re-calibrated scores as a cutoff to determine the predicted class, that is, the majority vote. To measure the AUC for our multiclass RF the generalization of the AUC for multiclass classification problems[10] was used. To measure how well the resulting RF scores and recalibrated scores perform when used as class probabilities, the multiclass Brier score was used [11,12].    Figure 4. Selection of optimal subtype number through application of consensus clustering using t-SNE, NMF and SNF. a. Heatmap shows sample reproducibility for differing combinations of dimensions and resultant clusters using t-SNE. Maximal reproducibility was observed at 2 and 5 clusters. There was no support beyond a maximum of 8 clusters. b. Sankey plot shows relationship between Grp3/4, the 5 clusters and 8 clusters identified through consensus t-SNE clustering. c. Heatmap shows sample reproducibility for differing combinations of NMF-defined metagenes, denoted by K, and clusters, denoted by C. Maximal sample reproducibility was observed at 5 clusters, defined by 6 metagenes. Subsequently, 2nd-order NMF was performed individually on each of the five robust clusters identified to characterize any further subtype splits. d. Sankey plot shows relataionships between Grp3 and Grp4, the five clusters identified with 1st-order NMF and the additional subtypes identified with 2nd-order NMF. e. Heatmap shows sample reproducibility for SNF, with differing combinations of the k parameter and clusters, denoted by C. Maximal sample reproducibility was observed at 2 clusters, with a further local maximum at 6 clusters. Support for 8 stable clusters was also observed. f. Sankey plot shows relationships between Grp3 and Grp4, and for 6 and 8 clusters defined using SNF.         \\n\\nSupplementary Figure 2 .Supplementary Figure 3 .\\n23Classification of study-specific samples designated as non-MB by methylation classifier. The detailed description of each of the classes can be found at : https://www.molecularneuropathology.org/mnp/classifierQuality control investigations reveal no bias. a. t-SNE plot of entire Grp3/Grp4 cohort coloured by Grp3/4 classification (Grp3, yellow; Grp4, green) by Heidelberg MNP2.0 classifier. b. t-SNE plot of entire Grp3/4 cohort coloured by derivative type (fresh-frozen (KRYO), blue; formalin-fixed, paraffin embedded (FFPE), red\\n\\nFigure 6 .\\n6SNF recapitulates study specific subtypes for Grp3/4 subsets of Schwalbe et al. dataset. a. The similarity matrix for SNF performed on RNA-seq data only identifies 4 subtypes for Schwalbe et al. dataset. b. The similarity matrix for SNF performed on DNA methylation data identifies 4/5 clusters. c. Stratification of the Schwalbe et al. dataset into 4 clusters using SNF matrix derived from combination of both DNA methylation and gene expression retrieves subtypes identified in the original study. d. Similarity matrix for SNF performed on gene expression data only on Northcott et al. subset fails to clearly stratify the data. e. Similarity matrix for SNF performed on DNA methylation data only on Northcott et al. subset identifies 5 clusters. f. Similarity matrix for SNF performed on paired DNA methylation and gene expression on Northcott et al.\\n\\nSupplementary Figure 9 .\\n9NMF incorrectly partitions consensus subtypes III/IV. a,b.\\n\\nFigure 10 :\\n10Survival analysis identifies cohort-specific survival differences. a. Kaplan-Meier plot shows progression free survival for Schwalbe et al. and Northcott et al. cohorts. b. Kapan-Meier plot shows overall survival for Schwalbe et al., Cavalli et al. and Northcott et al. cohorts. At risk tables are shown in two-year increments.\\n\\n\\nSupplementary table 2.List of samples excluded from study due to non-medulloblastoma prediction using Heidelberg molecular neuropathology brain tumor classifier. For each sample, its IDAT identifier, study ID, study, provisional classification and prediction score are given.Supplementary Table 3. Inter-technique mapping between subtype assignments.Idat \\n\\nStudy_ID \\nDerivative \\nMaterial \\n\\nArray \\nStudy \\nPrediction \\nScore \\n\\n200788260028_R06C01 \\n93622 \\nFFPE \\nEPIC \\nNorthcott \\nMB, G4 \\n0.179 \\n\\n200397860036_R06C01 \\nP036_90802_diag \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.085 \\n\\n200784530098_R06C01 \\nP036_93530_diag \\nFFPE \\nEPIC \\nNorthcott \\nMB, G3 \\n0.449 \\n\\n7970376085_R02C01 \\nTO_MB10 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.829 \\n\\n7970376048_R01C01 \\nTO_MB1 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.829 \\n\\n7810920080_R04C01 \\nTO_MB33 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.368 \\n\\n7970376149_R04C01 \\nTO_MB5 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.705 \\n\\n7973219019_R04C02 \\nMB-0580 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G4 \\n0.787 \\n\\n8942326039_R01C02 \\nMB-2631 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.633 \\n\\n7970376151_R01C01 \\nMB-MT1179 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.877 \\n\\n7796806110_R01C01 \\nMB-MT314 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.695 \\n\\n7810920068_R02C01 \\nID_7810920068_R02C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.688 \\n\\n7810920071_R01C01 \\nID_7810920071_R01C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.898 \\n\\n7810920078_R05C02 \\nID_7810920078_R05C02 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.661 \\n\\n7810920079_R05C01 \\nID_7810920079_R05C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.818 \\n\\n7970368003_R01C01 \\nID_7970368003_R01C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.818 \\n\\n7970368007_R02C01 \\nID_7970368007_R02C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.877 \\n\\n7970368045_R06C02 \\nID_7970368045_R06C02 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.897 \\n\\n7970368069_R04C01 \\nID_7970368069_R04C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.769 \\n\\n7970368130_R04C02 \\nID_7970368130_R04C02 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.776 \\n\\n7970376002_R04C01 \\nID_7970376002_R04C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.871 \\n\\n7970376148_R02C01 \\nID_7970376148_R02C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.737 \\n\\n7973201069_R04C02 \\nID_7973201069_R04C02 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.670 \\n\\n7973201149_R04C01 \\nID_7973201149_R04C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.440 \\n\\n7973201166_R02C01 \\nID_7973201166_R02C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.383 \\n\\n7973219003_R01C01 \\nID_7973219003_R01C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.493 \\n\\n7973219008_R04C01 \\nID_7973219008_R04C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.899 \\n\\n7973219018_R01C01 \\nID_7973219018_R01C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.860 \\n\\n7973219040_R04C02 \\nID_7973219040_R04C02 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.875 \\n\\n8942326054_R03C02 \\nID_8942326054_R03C02 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.481 \\n\\n9326189004_R03C02 \\nID_9326189004_R03C02 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.758 \\n\\n9326189013newLazer_R01C01 ID_9326189013newLazer_R01C01 Fresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.882 \\n\\n9326189013newLazer_R03C01 ID_9326189013newLazer_R03C01 Fresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.529 \\n\\n9326189023_R06C01 \\nID_9326189023_R06C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.798 \\n\\n9326189024newLazer_R03C01 ID_9326189024newLazer_R03C01 Fresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.746 \\n\\n9326189024newLazer_R04C02 ID_9326189024newLazer_R04C02 Fresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.491 \\n\\n9326189025newLazer_R04C02 ID_9326189025newLazer_R04C02 Fresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.854 \\n\\n9341679041_R06C01 \\nID_9341679041_R06C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.610 \\n\\n9344737005_R01C01 \\nID_9344737005_R01C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.656 \\n\\n9344737005_R04C02 \\nID_9344737005_R04C02 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.583 \\n\\n9344737044_R06C01 \\nID_9344737044_R06C01 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G4 \\n0.878 \\n\\n9344737073_R01C02 \\nID_9344737073_R01C02 \\nFresh-frozen \\n450k \\nCavalli \\nMB, G3 \\n0.332 \\n\\n9421912024_R01C01 \\n\\nNMB119 \\nFresh-frozen \\n450k \\nSchwalbe \\nMB, G4 \\n0.731 \\n\\n9421912022_R01C02 \\nNMB269 \\nFresh-frozen \\n450k \\nSchwalbe \\nMB, G3 \\n0.721 \\n\\n9647455134_R01C01 \\nNMB325 \\nFFPE \\n450k \\nSchwalbe \\nMB, G4 \\n0.488 \\n\\n7970368088_R05C01 \\nNMB440 \\nFresh-frozen \\n450k \\nSchwalbe \\nMB, G3 \\n0.721 \\n\\n7973201123_R06C02 \\nNMB445 \\nFresh-frozen \\n450k \\nSchwalbe \\nMB, G3 \\n0.533 \\n\\n9647455136_R04C01 \\nNMB476 \\nFresh-frozen \\n450k \\nSchwalbe \\nMB, G3 \\n0.733 \\n\\n9421912041_R01C01 \\nNMB49 \\nFresh-frozen \\n450k \\nSchwalbe \\nMB, G4 \\n0.889 \\n\\n6229017107_R01C02 \\nNMB583 \\nFresh-frozen \\n450k \\nSchwalbe \\nMB, G4 \\n0.519 \\n\\n7973201167_R01C02 \\nNMB717 \\nFresh-frozen \\n450k \\nSchwalbe \\nMB, G4 \\n0.522 \\n\\n9403904011_R01C02 \\nNMB795 \\nFresh-frozen \\n450k \\nSchwalbe \\nMB, G3 \\n0.697 \\n\\n7970368034_R06C02 \\nPNET30032 \\nFFPE \\n450k \\nSchwalbe \\nMB, G4 \\n0.617 \\n\\n9647455150_R05C01 \\nPNET350120 \\nFFPE \\n450k \\nSchwalbe \\nMB, G3 \\n0.239 \\n\\n9647455134_R02C02 \\nPNET350241 \\nFFPE \\n450k \\nSchwalbe \\nMB, G4 \\n0.801 \\n\\n8622007029_R01C02 \\n5M3 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G4 \\n0.686 \\n\\n9374341006_R05C02 \\nICGC_MB223 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.884 \\n\\n9406921070_R01C02 \\nP036_64906_diag \\nFFPE \\n450k \\nNorthcott \\nMB, G4 \\n0.180 \\n\\n9553932002_R05C01 \\nAK_2652_12 \\nFFPE \\n450k \\nNorthcott \\nMB, G3 \\n0.791 \\n\\n9610361029_R06C02 \\nAK_4480_08 \\nFFPE \\n450k \\nNorthcott \\nMB, G3 \\n0.809 \\n\\n9610361022_R06C02 \\nAK_4981_10 \\nFFPE \\n450k \\nNorthcott \\nMB, G3 \\n0.516 \\n\\n9761749088_R02C02 \\nICGC_MB289 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.861 \\n\\n9968646163_R05C02 \\nP036_69190_ref \\nFFPE \\n450k \\nNorthcott \\nMB, G3 \\n0.462 \\n\\n9969477090_R04C01 \\nICGC_MB291 \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.787 \\n\\n3999543042_R02C02 \\nP036_75374_ref \\nFFPE \\n450k \\nNorthcott \\nMB, G4 \\n0.384 \\n\\n3998920116_R05C02 \\nT51_MB \\nFresh-frozen \\n450k \\nNorthcott \\nMB, G3 \\n0.804 \\n\\n9741950157_R05C02 \\nSJMB03-UPN10669654 \\nFFPE \\n450k \\nSJMB03 \\nMB, G3 \\n0.707 \\n\\n9741950172_R01C01 \\nSJMB03-UPN13648866 \\nFFPE \\n450k \\nSJMB03 \\nMB, G3 \\n0.849 \\n\\n9934987066_R04C01 \\nSJMB03-UPN18600500 \\nFFPE \\n450k \\nSJMB03 \\nMB, G4 \\n0.211 \\n\\nSupplementary table 1. List of samples excluded from study due to low classifier prediction score \\nusing Heidelberg molecular neuropathology brain tumour classifier. For each sample, its IDAT \\nidentifier, study ID, study, provisional classification and prediction score are given. \\nIdat \\n\\nStudy ID \\nDerivative \\nMaterial \\n\\nArray \\nStudy \\nPrediction \\nScore \\n\\n200498360016_R03C01 \\n91426 \\nFFPE \\nEPIC \\nNorthcott \\nCONTR, CEBM \\n0.497 \\n\\n200790260040_R07C01 \\nP036_93488_diag \\nFresh-frozen \\nEPIC \\nNorthcott \\nGBM, RTK I \\n0.318 \\n\\n7970368034_R04C01 \\nNMB625 \\nFFPE \\n450k \\nSchwalbe \\nATRT, SHH \\n0.999 \\n\\n9647455136_R02C01 \\nNMB128 \\nFresh-frozen \\n450k \\nSchwalbe \\nCONTR, CEBM \\n0.999 \\n\\n6222421011_R05C02 \\nNMB168 \\nFresh-frozen \\n450k \\nSchwalbe \\nCONTR, CEBM \\n0.998 \\n\\n7970368035_R02C02 \\nNMB437 \\nFresh-frozen \\n450k \\nSchwalbe \\nGBM, MYCN \\n0.416 \\n\\n7970368032_R04C01 \\nNMB662 \\nFFPE \\n450k \\nSchwalbe \\nETMR \\n0.998 \\n\\n7970368052_R06C02 \\nNMB724 \\nFresh-frozen \\n450k \\nSchwalbe \\nETMR \\n0.996 \\n\\n9422491011_R03C02 \\nNMB869 \\nFresh-frozen \\n450k \\nSchwalbe \\nGBM, MID \\n0.511 \\n\\n9647450105_R06C02 \\nPNET30035 \\nFFPE \\n450k \\nSchwalbe \\nEPN, PF A \\n0.942 \\n\\n9403904116_R02C02 \\nNMB63 \\nFresh-frozen \\n450k \\nSchwalbe \\nCONTR, CEBM \\n0.957 \\n\\n7973201123_R01C02 \\nNMB441 \\nFresh-frozen \\n450k \\nSchwalbe \\nPLEX, PED B \\n0.758 \\n\\n9647455137_R03C02 \\nNMB823 \\nFFPE \\n450k \\nSchwalbe \\nGBM, MID \\n0.822 \\n\\n7970368138_R02C02 \\nNMB345 \\nFFPE \\n450k \\nSchwalbe \\nATRT, SHH \\n0.383 \\n\\n9403904132_R05C02 \\nNMB400 \\nFresh-frozen \\n450k \\nSchwalbe \\nCONTR, CEBM \\n0.945 \\n\\n7973201008_R03C02 \\nNMB759 \\nFresh-frozen \\n450k \\nSchwalbe \\nETMR \\n0.787 \\n\\n9647455134_R06C01 \\nPNET350049 \\nFFPE \\n450k \\nSchwalbe \\nPLEX, PED B \\n0.487 \\n\\n9647455169_R03C01 \\nNMB485 \\nFresh-frozen \\n450k \\nSchwalbe \\nPLEX, PED B \\n0.192 \\n\\n7970368032_R03C02 \\nNMB617 \\nFFPE \\n450k \\nSchwalbe \\nO IDH \\n0.319 \\n\\n9305651213newLazer_R06C01 \\nMB-0174 \\nFresh-frozen \\n450k \\nNorthcott \\nCONTR, HEMI \\n0.999 \\n\\n7970376150_R06C01 \\nMB-1351 \\nFresh-frozen \\n450k \\nNorthcott \\nRETB \\n0.589 \\n\\n9326189011newLazer_R04C02 \\nMB-2897 \\nFresh-frozen \\n450k \\nNorthcott \\nSUBEPN, SPINE \\n0.340 \\n\\n7810920078_R06C01 \\nID_7810920078_R06C01 \\nFresh-frozen \\n450k \\nCavalli \\nDMG, K27 \\n0.420 \\n\\n7970368124_R03C01 \\nID_7970368124_R03C01 \\nFresh-frozen \\n450k \\nCavalli \\nCONTR, CEBM \\n0.665 \\n\\n8959312036_R02C02 \\nID_8959312036_R02C02 \\nFresh-frozen \\n450k \\nCavalli \\nCONTR, CEBM \\n0.983 \\n\\n9326189009newLazer_R06C01 \\nID_9326189009newLazer \\n_R06C01 \\n\\nFresh-frozen \\n450k \\nCavalli \\nCONTR, CEBM \\n0.298 \\n\\n9344737044_R06C02 \\nID_9344737044_R06C02 \\nFresh-frozen \\n450k \\nCavalli \\nCONTR, CEBM \\n0.437 \\n\\n7973201069_R01C02 \\nID_7973201069_R01C02 \\nFresh-frozen \\n450k \\nCavalli \\nPLEX, PED B \\n0.301 \\n\\n6 \\n\\ntSNE \\nNMF \\nSNF \\nAgreement \\n\\nI \\nGrp3.HR.2 \\n-\\n2/3 \\n\\nII \\nGrp3.HR.1 \\n5 \\n3/3 \\n\\nIII \\nGrp3.LR.2/Grp3.LR.1 \\n2 \\n2/3 \\n\\nIV \\nGrp3.LR.2/Grp3.LR.1 \\n6 \\n2/3 \\n\\nV \\nGrp4.LR2.2 \\n-\\n2/3 \\n\\nVI \\nGrp4.LR2.1 \\n-\\n2/3 \\n\\nVII \\nGrp4.LR1.1 \\n\\nGrp4.LR1.2 \\n\\n7 \\n\\n4 \\n\\n2/3 \\n\\n2/3 \\n\\nVIII \\nGrp4.HR \\n-\\n2/3 \\n\\nNorthcott et al \\nn= 1256 \\n\\nSchwalbe et al \\nn= 428 \\n\\nCavalli et al \\nn= 763 \\n\\nFinal cohort= 1501 samples \\n\\nAdditional samples = 153 \\nSamples with paired DNA &amp; gene expression = 852 \\n\\nShared genotype= 80 \\nLow prediction scores=69 \\n\\nQC and normalization \\nnon MB prediction= 28 \\n\\nMB Grp3/4 classifier \\n\\nIdentify common and \\nrobust subtypes \\nEstablish consensus \\nsubtypes \\n\\nCheck clinico-pathological \\nsignificance \\n\\nCross compare sample calls \\nwith all the three methods \\n\\ntSNE/DBSCAN \\nConsensus NMF \\nSNF + spectral \\nclustering \\n\\nVerify consistency with published results \\n\\nAdjust parameters to maximize robustness \\n\\nCommon resampling and robustness measurements \\n\\n\\n\\n\\nt-SNE plot of subtypes III and IV labelled (a) by NMF-derived subtypes, and (b) by t-SNE-derived subtypes. c, d. PCA plot of subtypes III and IV labelled by (c) NMF-derived subtypes, and (d) t-SNE-derived subtypes. e. Distinct copy-number differences are observed for III/IV subtypes that were not identified in the NMFderived subtypes. f, g. Progression-free survival plots for (f), NMF-derived subtypes, and (g), t-SNEderived subtypes support the adoption of III/IV subtypes in favor of NMF-derived subtypes.Time since diagnosis (years)a \\n\\n0 \\n2 \\n4 \\n6 \\n8 \\n10 \\n\\n0.0 \\n\\n0.2 \\n\\n0.4 \\n\\n0.6 \\n\\n0.8 \\n\\n1.0 \\n\\n239 \\n148 \\n103 \\n62 \\n36 \\n19 Schwalbe et al. \\n264 \\n205 \\n158 \\n103 \\n57 \\n42 Northcott et al. \\nProgression free survival(%) \\n\\nTime since diagnosis (years) \\n\\nOverall survival(%) \\n\\nNumber at risk \\n\\n0 \\n2 \\n4 \\n6 \\n8 \\n10 \\n\\n0.0 \\n\\n0.2 \\n\\n0.4 \\n\\n0.6 \\n\\n0.8 \\n\\n1.0 \\n\\n249 \\n192 \\n131 \\n84 \\n59 \\n40 Cavalli et al. \\n244 \\n167 \\n117 \\n71 \\n43 \\n20 Schwalbe et al. \\n264 \\n212 \\n152 \\n87 \\n39 \\n30 Northcott et al. \\n\\nNumber at risk \\np=0.01 \\n\\np=0.00077 \\n\\n\\nFigure 11: Subtype-specific differentially methylated loci are not enriched in corresponding significantly altered chromosomes. For each subtype I-VIII, the chromosomal distribution of differentially methylated loci (mean change in beta value &gt;0.3, qval &lt; 0.0001) is shown, relative to the overall number of testable CpG loci per chromosome. Significant, subtype-specific chromosomal alterations (i.e. gain/loss) are shown in the corresponding subtype colour. Chromosomes without significant copy number alterations in a subtype are shown grey.\\nUnmethylated loci are shown blue, methylated are shown red. e. Copy number differences between subtype VII variants. f. Gene expression differences between SNF-derived subtype VII variants using Cavalli et al. transcriptome dataset. g, h. Progression-free and overall survival plots for subtype VII variants SNF4 and SNF7. Supplementary Figure 8. SNF identifies two variants of subtype VII. a. t-SNE plot of entire Grp3/ Grp4 cohort, showing two subtype VII variants. light blue; SNF4, dark blue; other samples, black). b. PCA plot of subtype VII SNF-variants only, showing separation of variants. c. Age distributions of subtype VII, SNF4 and SNF7. d. Heatmap shows most variably methylated lociSupplementary Figure 8. SNF identifies two variants of subtype VII. a. t-SNE plot of entire Grp3/ Grp4 cohort, showing two subtype VII variants (SNF 7, light blue; SNF4, dark blue; other samples, black). b. PCA plot of subtype VII SNF-variants only, showing separation of variants. c. Age distributions of subtype VII, SNF4 and SNF7. d. Heatmap shows most variably methylated loci. Unmethylated loci are shown blue, methylated are shown red. e. Copy number differences between subtype VII variants. f. Gene expression differences between SNF-derived subtype VII variants using Cavalli et al. transcriptome dataset. g, h. Progression-free and overall survival plots for subtype VII variants SNF4 and SNF7.\\n</td>\n      <td>4360</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>5787</td>\n      <td>28215090</td>\n      <td>\\nFamily Binomial name Common name Accession number LNA-Mit63 LNA-Mit1492\\n\\n\\nFamily Binomial name Common name Accession number LNA-Mit63 LNA-Mit1492\\n\\n\\n\\nTable S1\\nS1List of the SSU rRNA genes from plant mitochondria used for design of LNA oligonucleotides that competed with modified 63f and 1492r. Symbol ○ indicates the genes that were identical with designed LNA oligonicleotides, while symbol ▲ indicates the genes that showed mismatches in the concerned regions with designed LNA oligonucleotide. The -indicates that the corresponding sequence were not registered in DNA database. The N.D. indicates not determined.Convolvulaceae \\n\\nIpomoea batatas \\nSweet potato \\nN.D. \\n\\n-\\n-\\n\\nCrossosomataceae Crossosoma bigelovii \\nRagged rockflower \\nAF194001 \\n\\n○ \\n○ \\n\\nCucurbitaceae \\nCitrullus lanatus \\nWatermelon \\nNC_014043 \\n\\n○ \\n○ \\n\\nCucumis africanus \\nWild cucumber \\nAY357205 \\n\\n○ \\n\\n-\\n\\nCucumis melo \\nMelon \\nJF412792 \\n\\n○ \\n○ \\n\\nCucumis sativa \\nCucumber \\nNC_016005 \\n▲ \\n\\n○ \\n\\nCucumis zeyherii \\nSouth African spiny cucumber AY357207 \\n\\n○ \\n\\n-\\n\\nCucurbita moschata \\nWinter squash \\nAY357208 \\n\\n-\\n\\n○ \\n\\nCucurbita pepo \\nSummer squash \\nAY357209 \\n\\n-\\n\\n○ \\n\\nCucurbita moschata \\nPumpkin \\nN.D. \\n\\n-\\n-\\n\\nLuffa cylindrical \\nLoofah \\nN.D. \\n\\n-\\n-\\n\\nCupressaceae \\nCallitropsis nootkatensis \\nNootka cypress \\nAB029366 \\n▲ \\n\\n○ \\n\\nCryptomeria japonica \\nJapanese cedar \\nAB029367 \\n▲ \\n\\n○ \\n\\nJuniperus chinensis \\nChinese juniper \\nAB029368 \\n▲ \\n\\n○ \\n\\nCycadaceae \\nCycas taitungensis \\nPrince sago \\nAP009381 \\n▲ \\n\\n○ \\n\\nCyclanthaceae \\nCarludovica palmata \\nPanama hat plant \\nDQ008694 \\n\\n○ \\n○ \\n\\nCytinaceae \\nBdallophyton americanum \\nN.D. \\nU82653 \\n\\n○ \\n\\n-\\n\\nCytinus ruber \\nN.D. \\nU82639 \\n\\n○ \\n\\n-\\n\\nDegeneriaceae \\nDegeneria vitiensis \\nN.D. \\nDQ008695 \\n\\n○ \\n○ \\n\\nEquisetaceae \\nEquisetum arvense \\nField horsetail \\nAF058663 \\n\\n○ \\n\\n-\\n\\nEuphorbiaceae \\nJatropha curcas \\nBarbados nut \\nGQ386927 \\n\\n○ \\n\\n-\\n\\nRicinus communis \\nCastor bean \\nHQ874649 \\n\\n○ \\n○ \\n\\nManihot utilissima \\nCassava \\nN.D. \\n\\n-\\n-\\n\\nEupomatiaceae \\nEupomatia bennettii \\nSmall bolwarra \\nDQ008698 \\n\\n○ \\n○ \\n\\nEupteleaceae \\nEuptelea polyandra \\nN.D. \\nAF193997 \\n\\n○ \\n○ \\n\\nFabaceae \\nCajanus cajan \\nGandule bean \\nEU307415 \\n\\n○ \\n\\n-\\n\\nFabaceae \\nGlycine max \\nSoybean \\nM16859 \\n\\n○ \\n○ \\n\\nLotus japonicus \\nN.D. \\nJN872551 \\n\\n○ \\n\\n▲ \\nLupinus luteus \\nEuropean yellow lupine \\nZ11512 \\n\\n○ \\n○ \\n\\nMillettia pinnata \\nIndian beech \\nJN872550 \\n\\n○ \\n○ \\n\\nPhaseolus coccineus \\nRunner bean \\nEU307418 \\n\\n○ \\n\\n-\\n\\nPhaseolus vulgaris \\nCommon bean \\nEU307370 \\n\\n○ \\n\\n-\\n\\nPisum sativum \\nPea \\nX98799 \\n\\n○ \\n\\n-\\n\\nVicia faba \\nBroad bean \\nKC189947 \\n\\n○ \\n○ \\n\\nVigna angularis \\nAzuki bean \\nAP012599 \\n\\n○ \\n○ \\n\\nVigna radiata \\nMung bean \\nHM367685 \\n\\n○ \\n○ \\n\\nVigna unguiculata \\nCowpea \\nEU307376 \\n\\n○ \\n\\n-\\n\\nMedicago sativa \\nAlfalfa \\nN.D. \\n\\n-\\n-\\n\\nGeraniaceae \\nGeranium himalayense \\nLilac cranesbill \\nAF194003 \\n\\n○ \\n○ \\n\\nGeranium macrorrhizum \\nBigroot geranium \\nDQ317028 \\n\\n○ \\n○ \\n\\nHypseocharis pimpinellifolia \\nN.D. \\nAF194002 \\n\\n○ \\n○ \\n\\nMonsonia emarginata \\nMountain geranium \\nDQ317029 \\n\\n○ \\n\\n▲ \\nErodium pelargoniflorum \\nN.D. \\nDQ317006 \\n\\n○ \\n○ \\n\\nErodium trifolium \\nN.D. \\nDQ317007 \\n\\n○ \\n○ \\n\\nSarcocaulon vanderietiae \\nN.D. \\nDQ317004 \\n\\n○ \\n○ \\n\\nGesneriaceae \\nBoea hygrometrica \\nN.D. \\nJN107812 \\n▲ \\n\\n○ \\n\\nHernandiaceae \\nGyrocarpus americanus \\nHelicopter tree \\nDQ008702 \\n▲ \\n\\n○ \\n\\nHimantandraceae \\nGalbulimima belgraveana \\nWhite magnolia \\nDQ008704 \\n\\n○ \\n○ \\n\\nHydnoraceae \\nHydnora africana \\nJackal's kost \\nU82637 \\n\\n○ \\n\\n-\\n\\nLardizabalaceae \\nAkebia quinata \\nChocolate vine \\nDQ008709 \\n\\n○ \\n○ \\n\\nLardizabala biternata \\nLardizabala \\nDQ008710 \\n\\n○ \\n○ \\n\\nSargentodoxa cuneata \\nN.D. \\nDQ008729 \\n\\n-\\n\\n○ \\n\\nLauraceae \\nCinnamomum camphora \\nCamphor Laurel \\nDQ008711 \\n\\n○ \\n○ \\n\\nLindera benzoin \\nWild allspice \\nU82646 \\n\\n○ \\n○ \\n\\nLemnoideae \\nSpirodela polyrhiza \\nGreat duckweed \\nJQ804980 \\n\\n○ \\n○ \\n\\nLiliaceae \\nPleea tenuifolia \\nN.D. \\nDQ008735 \\n\\n○ \\n○ \\n\\nAllium cepa \\nOnion \\nN.D. \\n\\n-\\n-\\n\\nAllium fistulosum \\nWelsh onion \\nN.D. \\n\\n-\\n-\\n\\nAllium sativum \\nGarlic \\nN.D. \\n\\n-\\n-\\n\\nAsparagus officinalis \\nAsparagus \\nDQ008678 \\n\\n-\\n-\\n\\nMagnoliaceae \\nLiriodendron tulipifera \\nTulip tree \\nKC821969 \\n\\n○ \\n○ \\n\\nMagnolia grandiflora \\nSouthern magnolia \\nAF161089 \\n\\n○ \\n○ \\n\\nMagnolia × soulangeana \\nSaucer magnolia \\nU82648 \\n\\n○ \\n\\n-\\n\\nMalvaceae \\nGossypium hirsutum \\nCotton \\nN.D. \\n\\n-\\n-\\n\\nHibiscus esculentus \\nGumbo \\nN.D. \\n\\n-\\n-\\n\\nMenispermaceae \\nCocculus trilobus \\nN.D. \\nDQ008714 \\n\\n○ \\n○ \\n\\nMonimiaceae \\n\\nHedycarya arborea \\nPigeonwood \\nDQ008716 \\n\\n-\\n\\n○ \\n\\nHortonia floribunda \\nN.D. \\nDQ008717 \\n\\n-\\n\\n○ \\n\\nMyristicaceae \\nMyristica fragrans \\nNutmeg \\nDQ008718 \\n▲ \\n\\n○ \\n\\nNelumbonaceae \\nNelumbo nucifera \\nLotus \\nAF193983 \\n\\n○ \\n\\n▲ \\nNymphaeaceae \\nNuphar sp. \\nPond lily \\nAF193981 \\n\\n○ \\n\\n▲ \\nNymphaea sp. \\nWater lilies \\nAF161091 \\n\\n○ \\n○ \\n\\nEuryale sp. \\nN.D. \\nAF193980 \\n\\n○ \\n\\n▲ \\nVictoria cf. amazonica \\nN.D. \\nAF193979 \\n\\n○ \\n○ \\n\\nOnagraceae \\nOenothera berteriana \\nBertero's evening primrose \\nX61277 \\n\\n○ \\n○ \\n\\nOrchidaceae \\nDendrobium officinale \\nN.D. \\nKC465231 \\n▲ \\n\\n○ \\n\\nNeottia nidus-avis \\nBird's nest Orchid \\nU82644 \\n\\n○ \\n\\n-\\n\\nPapaveraceae \\nSanguinaria canadensis \\nBloodwort \\nDQ008720 \\n\\n○ \\n○ \\n\\nDicentra sp. \\nN.D. \\nDQ008700 \\n\\n○ \\n○ \\n\\nPedaliaceae \\nSesamum indicum \\nSesame \\nN.D. \\n\\n-\\n-\\n\\nPinaceae \\nAbies homolepis \\nNikko fir \\nAB029360 \\n\\n-\\n\\n○ \\n\\nPinus strobus \\nEastern white pine \\nAF058659 \\n\\n-\\n\\n○ \\n\\nPiperaceae \\nPeperomia obtusifolia \\nBaby rubber plant \\nDQ008722 \\n▲ \\n\\n○ \\n\\nPiper betle \\nBetel \\nAF161088 \\n▲ \\n\\n○ \\n\\nPeperomia obtusifolia \\nBaby rubber plant \\nDQ008722 \\n▲ \\n\\n○ \\n\\nPiper betle \\nBetel \\nAF161088 \\n▲ \\n\\n○ \\n\\nPeperomia fosterii \\nN.D. \\nAF193986 \\n▲ \\n\\n○ \\n\\nPiper nigrum \\nPepper \\nN.D. \\n\\n-\\n-\\n\\nPlantaginaceae \\nPlantago atrata \\nN.D. \\nAJ389618 \\n\\n-\\n\\n○ \\n\\nPlantago australis \\nN.D. \\nAY818949 \\n\\n-\\n\\n○ \\n\\nPlantago coronopus \\nMinutina \\nAJ389617 \\n\\n-\\n\\n○ \\n\\nPlantago cyanops \\nN.D. \\nAJ389620 \\n\\n-\\n\\n○ \\n\\nPlantago lanceolata \\nRibwort plantain \\nAJ389619 \\n\\n-\\n\\n○ \\n\\nPlantago media \\nHoary plantain \\nAJ389614 \\n\\n-\\n\\n○ \\n\\nPlantago rigida \\nN.D. \\nAJ389616 \\n\\n-\\n\\n○ \\n\\nPlantago sericea \\nN.D. \\nAJ389621 \\n\\n-\\n\\n○ \\n\\nPlatanaceae \\nPlatanus occidentalis \\nAmerican sycamore \\nU82650 \\n\\n○ \\n\\n-\\n\\nPoaceae \\nAegilops ventricosa \\nGoatgrass \\nDQ307262 \\n\\n○ \\n\\n-\\n\\nBambusa oldhamii \\nBamboo \\nEU365401 \\n\\n○ \\n○ \\n\\nEleusine coracana \\nFinger millet \\nKC119749 \\n\\n-\\n\\n○ \\n\\nEremopyrum bonaepartis \\nN.D. \\nDQ307271 \\n\\n○ \\n\\n-\\n\\nLolium perenne \\nPerennial ryegrass \\nJX999996 \\n\\n○ \\n○ \\n\\nOryza rufipogon \\nBrownbeard rice \\nAP011076 \\n\\n○ \\n○ \\n\\nOryza sativa Indica \\nIndica rice \\nNC_007866 \\n\\n○ \\n○ \\n\\nOryza sativa Japonica \\nJaponica rice \\nNC_011033 \\n\\n○ \\n○ \\n\\nPsathyrostachys juncea \\nRussian wildrye \\nDQ307265 \\n\\n○ \\n\\n-\\n\\nSecale cereale \\nRye \\nZ14059 \\n\\n○ \\n○ \\n\\nSecale strictum \\nMountain rye \\nDQ307259 \\n\\n○ \\n\\n-\\n\\nSecale sylvestre \\nN.D. \\nDQ307272 \\n\\n○ \\n\\n-\\n\\nSorghum bicolor \\nSorghum \\nDQ984518 \\n\\n○ \\n○ \\n\\nThinopyrum bessarabicum \\nWheatgrass \\nDQ307269 \\n\\n○ \\n\\n-\\n\\nThinopyrum intermedium \\nIntermediate wheatgrass \\nDQ677514 \\n\\n○ \\n\\n-\\n\\nTripsacum dactyloides \\nGamagrass \\nDQ984517 \\n\\n○ \\n○ \\n\\nTriticum aestivum \\nWheat \\nGU985444 \\n\\n○ \\n○ \\n\\nZea luxurians \\nGuatemalan teosinte \\nDQ645537 \\n\\n○ \\n○ \\n\\nZea mays \\nMaize \\nAY506529 \\n\\n○ \\n○ \\n\\nZea perennis \\nPerennial teosinte \\nDQ645538 \\n\\n○ \\n○ \\n\\nCrithopsis delileana \\nN.D. \\nDQ307273 \\n\\n○ \\n\\n-\\n\\nDasypyrum breviaristatum \\nN.D. \\nGU168590 \\n\\n○ \\n\\n-\\n\\nFerrocalamus rimosivaginus \\nN.D. \\nJQ235168 \\n\\n○ \\n○ \\n\\nHaynaldia villosa \\nN.D. \\nDQ307266 \\n\\n○ \\n\\n-\\n\\nLolium temulentum \\nDarnel \\nN.D. \\n\\n-\\n-\\n\\nHordeum vulgare \\nBarley \\nN.D. \\n\\n-\\n-\\n\\nRagmites australis \\nReed \\nN.D. \\n\\n-\\n-\\nRafflesiaceae \\n\\nRafflesia pricei \\nRafflesia \\nU96694 \\n\\n○ \\n○ \\n\\nRanunculaceae \\nRanunculus sp. \\nButtercup \\nAF161093 \\n\\n○ \\n\\n-\\n\\nRanunculoideae \\nClematis sp. \\ntraveller's joy \\nAF193994 \\n▲ \\n\\n○ \\n\\nRosaceae \\nMalus domestica \\nApple \\nFR714868 \\n\\n○ \\n○ \\n\\nFragaria x ananassa \\nStrawberry \\nN.D. \\n\\n-\\n-\\n\\nRutaceae \\nCitrus maxima \\nPomelo \\nFJ356261 \\n\\n-\\n\\n○ \\n\\nSabiaceae \\nMeliosma squamulata \\nN.D. \\nDQ008728 \\n\\n○ \\n\\n-\\n\\nSabia sp. \\nN.D. \\nDQ008727 \\n\\n○ \\n○ \\n\\nSantalaceae \\nLepidoceras chilense \\nN.D. \\nU82641 \\n▲ \\n\\n○ \\n\\nSchisandraceae \\nKadsura japonica \\nKadsura \\nAF193985 \\n\\n○ \\n○ \\n\\nSchisandra sphenanthera \\nfive flavor berry \\nAF193984 \\n▲ \\n\\n○ \\n\\nSciadopityaceae \\nSciadopitys verticillata \\nJapanese umbrella pine \\nAF161085 \\n▲ \\n\\n○ \\n\\nScrophulariaceae \\nMimulus guttatus \\nMonkey flower \\nJN098455 \\n▲ \\n\\n○ \\n\\nSiparunaceae \\nSiparuna decipiens \\nN.D. \\nDQ008733 \\n\\n○ \\n\\n▲ \\nSmilacaceae \\nSmilax hispida \\nN.D. \\nU82645 \\n\\n○ \\n\\n-\\n\\nSolanaceae \\nNicotiana tabacum \\nTobacco \\nBA000042 \\n\\n○ \\n○ \\n\\nSolanum tuberosum \\nPotato \\nAJ252732 \\n\\n○ \\n\\n-\\n\\nSolanum melongena \\nEggplant \\nN.D. \\n\\n-\\n-\\n\\nLycopersicum esculentum \\nTomato \\nN.D. \\n\\n-\\n-\\n\\nStemonaceae \\nCroomia pauciflora \\nN.D. \\nDQ008734 \\n\\n○ \\n○ \\n\\nTaxaceae \\nTaxus brevifolia \\nPacific yew \\nAB029370 \\n▲ \\n\\n○ \\n\\nThelypteridaceae \\nPhegopteris hexagonoptera \\nBroad beech fern \\nAF058666 \\n\\n○ \\n\\n▲ \\nTiliaceae \\nCorchorus capsularis \\nJute \\nN.D. \\n\\n-\\n-\\n\\nTofieldiaceae \\nTofieldia calyculata \\nN.D. \\nDQ008736 \\n\\n○ \\n○ \\n\\nTrochodendraceae Tetracentron sinense \\nSpur leaf \\nAF193998 \\n\\n○ \\n○ \\n\\nTrochodendron aralioides \\nWheel tree \\nAF161092 \\n\\n○ \\n○ \\n\\nVitaceae \\nVitis vinifera \\nGrape \\nFM179380 \\n\\n○ \\n○ \\n\\nWinteraceae \\nDrimys winteri \\nWinter's bark \\nAF197162 \\n▲ \\n\\n○ \\n\\nTakhtajania perrieri \\nN.D. \\nDQ008740 \\n\\n○ \\n○ \\n\\nTasmannia insipida \\nBrush pepperbush \\nDQ008739 \\n▲ \\n\\n○ \\n\\nZamiaceae \\nDioon edule \\nPalma de la Virgen \\nAF058657 \\n▲ \\n\\n○ \\n\\nZingiberaceae \\nZingiber officinale \\nGinger \\nN.D. \\n\\n-\\n-\\n\\nFamily \\n\\nBinomial name \\nCommon name \\nAccession number LNA-Pla63 LNA-Pla1492 \\n\\nAcoraceae \\nAcorus calamus \\nSweetflag \\nAJ879453 \\n\\n• \\n• \\n\\nAmaranthaceae Beta vulgaris \\nSugar beet \\nEF534108 \\n\\n• \\n• \\n\\nSpinacia oleracea \\nSpinach \\nNC_002202 \\n\\n• \\n• \\n\\nAmborellaceae \\nAmborella trichopoda \\nN.D. \\nAJ506156 \\n\\n• \\n• \\n\\nApiaceae \\nDaucus carota \\nCarrot \\nNC_008325 \\n\\n• \\n• \\n\\nAraceae \\nLemna minor \\nCommon duckweed \\nDQ400350 \\n\\n• \\n• \\n\\nAsparagaceae \\nAsparagus officinalis \\nGarden asparagus \\nHQ183492 \\n\\n• \\n• \\n\\nAsteraceae \\nHelianthus annuus \\nCommon sunflower \\nNC_007977 \\n\\n• \\n• \\n\\nLactuca sativa \\nGarden lettuce \\nNC_007578 \\n\\n• \\n• \\n\\nBerberidaceae \\nNandina domestica \\nHeavenly bamboo \\nDQ923117 \\n\\n• \\n• \\n\\nBrassicaceae \\nAethionema cordifolium \\nLebanon stonecress \\nAP009366 \\n\\n• \\n• \\n\\nArabidopsis thaliana \\nMouse-ear cress \\nAP000423 \\n\\n• \\n• \\n\\nArabis alpina \\nAlpine rock-cress \\nHF934132 \\n\\n○ \\n\\n• \\n\\nArabis hirsuta \\nHairy rock-cress \\nAP009369 \\n\\n○ \\n\\n• \\n\\nBarbarea verna \\nLand cress \\nAP009370 \\n\\n• \\n• \\n\\nBrassica rapa \\nChinese cabbage \\nNC_015139 \\n\\n• \\n• \\n\\nDraba nemorosa \\nWoodland whitlowgrass \\nAP009373 \\n\\n○ \\n\\n• \\n\\nLepidium virginicum \\nVirginia pepperweed \\nAP009374 \\n\\n• \\n• \\n\\nLobularia maritima \\nSweet alyssum \\nAP009375 \\n\\n• \\n• \\n\\nBuxaceae \\nBuxus microphylla \\nLittleleaf boxwood \\nEF380351 \\n\\n• \\n• \\n\\nCalycanthaceae Calycanthus floridus \\nEastern sweetshrub \\nAJ428413 \\n\\n• \\n• \\n\\nCampanulaceae Trachelium caeruleum \\nBlue throatwort \\nEU090187 \\n\\n• \\n• \\n\\nCaricaceae \\nCarica papaya \\nPapaya \\nEU431223 \\n\\n• \\n• \\n\\nCeratophyllaceae Ceratophyllum demersum \\nRigid hornwort \\nEF614270 \\n\\n• \\n• \\n\\nChloranthaceae Chloranthus spicatus \\nChulantree \\nEF380352 \\n\\n• \\n• \\n\\nConvolvulaceae Cuscuta campestris \\nField dodder \\nAJ401353 \\n\\n○ \\n\\n• \\n\\nCuscuta exaltata \\nTall dodder \\nEU189132 \\n▲ \\n\\n• \\n\\nCuscuta gronovii \\nCommon dodder \\nAM711639 \\n\\n○ \\n\\n• \\n\\nCuscuta obtusiflora \\nPeruvian dodder \\nEU189133 \\n\\n○ \\n\\n• \\n\\nCuscuta odorata \\nN.D. \\nAJ278624 \\n\\n○ \\n\\n• \\n\\nCuscuta platyloba \\nN.D. \\nAJ401355 \\n\\n○ \\n\\n• \\n\\nCuscuta reflexa \\nSouthern Asian dodder \\nAM711640 \\n▲ \\n\\n• \\n\\nCuscuta subinclusa \\nN.D. \\nAJ278625 \\n\\n○ \\n\\n• \\n\\nIpomoea purpurea \\nCommon morning glory \\nEU118126 \\n\\n• \\n• \\n\\nCucurbitaceae \\nCucumis melo \\nMuskmelon \\nNC_015983 \\n\\n• \\n• \\n\\nCucumis sativus \\nCucumber \\nNC_007144 \\n\\n• \\n• \\n\\nCupressaceae \\nCalocedrus formosana \\nN.D. \\nAB831010 \\n\\n○ \\n\\n• \\n\\nCalocedrus macrolepis \\nN.D. \\nEU161517 \\n\\n○ \\n\\n▲ \\nChamaecyparis pisifera \\nSawara false-cypress \\nEU161522 \\n\\n○ \\n\\n▲ \\nCryptomeria fortunei \\nN.D. \\nDQ478788 \\n\\n○ \\n\\n• \\n\\nCryptomeria japonica \\nJapanese cedar \\nAP009377 \\n\\n○ \\n\\n• \\n\\nCupressus gigantea \\nN.D. \\nEU161525 \\n\\n○ \\n\\n▲ \\nFokienia hodginsii \\nN.D. \\nEU161531 \\n\\n○ \\n\\n▲ \\nGlyptostrobus pensilis \\nN.D. \\nEU161532 \\n\\n○ \\n\\n▲ \\nJuniperus virginiana \\nEastern redcedar \\nU24586 \\n\\n○ \\n\\n• \\n\\nLibocedrus sp. \\nN.D. \\nEU161538 \\n\\n○ \\n\\n• \\n\\nMicrobiota decussata \\nN.D. \\nEU161541 \\n\\n○ \\n\\n• \\n\\nPlatycladus orientalis \\nN.D. \\nEU161549 \\n\\n○ \\n\\n• \\n\\nTaxodium mucronatum \\nN.D. \\nEU161559 \\n\\n○ \\n\\n▲ \\nTetraclinis articulata \\nN.D. \\nEU161560 \\n\\n○ \\n\\n• \\n\\nThujopsis dolabrata \\nHiba arborvitae \\nEU161562 \\n\\n○ \\n\\n• \\n\\nCycadaceae \\nCycas taitungensis \\nPrince sago \\nAP009339 \\n\\n• \\n• \\n\\nDalbergieae \\nArachis hypogaea \\nPeanut \\nEU982406 \\n\\n○ \\n\\n-\\n\\nDioscoreaceae \\nDioscorea elephantipes \\nElephant's foot \\nEF380353 \\n\\n• \\n• \\n\\nEcdeiocoleaceae Ecdeiocolea monostachya \\nN.D. \\nHQ183501 \\n\\n○ \\n\\n▲ \\nGeorgeantha hexandra \\nN.D. \\nHQ183505 \\n\\n• \\n\\n○ \\n\\n\\n\\nTable S2\\nS2List of the SSU rRNA genes from plant plastid used for design of LNA oligonucleotides that competed with modified 63f and 1492r. Symbol ○ indicates the genes that were identical with designed LNA oligonicleotides, while symbol • indicates the genes that were identical with the other plant groups. Symbol ▲ indicates the genes that showed mismatches in the concerned regions with both plant groups. The -indicates that the corresponding sequence were not registered in DNA database. The N.D. indicates not determined.Ephedraceae \\n\\nEphedra distachya \\nJoint-fir \\nEU161530 \\n\\n○ \\n\\n▲ \\nEphedra equisetina \\nMa huang \\nAP010819 \\n\\n○ \\n\\n▲ \\nEphedra trifurca \\nMexican-tea \\nU24584 \\n\\n○ \\n\\n▲ \\nEuphorbiaceae \\nManihot utilissima \\nCassava \\nNC_010433 \\n\\n• \\n• \\n\\nFabaceae \\nGlycine max \\nSoybean \\nNC_007942 \\n\\n• \\n• \\n\\nLotus japonicus \\nN.D. \\nAP002983 \\n\\n• \\n• \\n\\nMedicago sativa \\nAlfalfa \\nJX185400 \\n\\n○ \\n\\n• \\n\\nMedicago truncatula \\nBarrel medic \\nAC093544 \\n\\n○ \\n\\n• \\n\\nPhaseolus vulgaris \\nKidney bean \\nDQ886273 \\n\\n• \\n• \\n\\nGeraniaceae \\nPelargonium × hortorum \\nCommon geranium \\nDQ897681 \\n\\n• \\n\\n▲ \\nGnetaceae \\nGnetum leyboldii \\nN.D. \\nAF244555 \\n\\n○ \\n\\n▲ \\nGnetum montanum \\nN.D. \\nKC427271 \\n\\n○ \\n\\n▲ \\nGnetum parvifolium \\nSmall-leaved jointfir \\nAP009569 \\n\\n○ \\n\\n▲ \\nJuncaceae \\nJuncus effusus \\nSoft rush \\nHQ183510 \\n\\n○ \\n\\n▲ \\nLamiaceae \\nGlechoma hederacea \\nGround-ivy \\nDQ417652 \\n\\n• \\n\\n○ \\n\\nMagnoliaceae \\nLiriodendron tulipifera \\nTuliptree \\nDQ899947 \\n\\n• \\n• \\n\\nMalvaceae \\nGossypium barbadense \\nSea-island cotton \\nAP009123 \\n\\n• \\n• \\n\\nGossypium hirsutum \\nUpland cotton \\nNC_007944 \\n\\n• \\n• \\n\\nMoraceae \\nMorus indica \\nMulberry \\nDQ226511 \\n\\n• \\n• \\n\\nMyristicaceae \\nMyristica yunnanensis \\nN.D. \\nDQ629588 \\n\\n○ \\n\\n• \\n\\nNelumbonaceae Nelumbo nucifera \\nSacred lotus \\nNC_015610 \\n\\n• \\n• \\n\\nNymphaeaceae \\nNymphaea alba \\nCastalia alba \\nAJ627251 \\n\\n• \\n• \\n\\nOleaceae \\nJasminum nudiflorum \\nWinter jasmine \\nDQ673255 \\n\\n• \\n• \\n\\nOnagraceae \\nOenothera argillicola \\nAppalachian evening primrose EU262887 \\n\\n• \\n\\n▲ \\nOrobanchaceae Epifagus virginiana \\nBeechdrops \\nM81884 \\n\\n• \\n• \\n\\nPassifloraceaePassiflora bifloraMit63Mit1492 Pla63 Pla1492Acidobacteria69299Phylogenetic groups in division levelsNumber of matched sequences</td>\n      <td>4373</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"execution_count":null,"metadata":{},"output_type":"execute_result"}]},{"cell_type":"code","source":"table_name = \"abstract\"\n\nsql = text(f'''\n    SELECT * FROM abstracts LIMIT 5; \n''')\n\nwith engine.connect() as conn:\n    a_query = conn.execute(sql)\n\nabstract_df = pd.DataFrame(a_query.fetchall())\nabstract_df.head()","metadata":{},"execution_count":null,"outputs":[{"data":{"application/vnd.hex.export+parquet":{"success":true,"exportKey":"604dd43f-9bf2-4f53-b2ad-a57886ed9dc6/1e7b90bc-1996-4336-8653-50313b6d2f75/exports/0fec2510-899f-4591-8941-a08d118fd266"},"text/html":"<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>id</th>\n      <th>paperId</th>\n      <th>corpusId</th>\n      <th>abstract</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>10187</td>\n      <td>b262c08e1b7ba06e43b949590cddd82235ed651a</td>\n      <td>2548148</td>\n      <td>Background— Calcific aortic valve disease is common in the elderly, is correlated with common cardiovascular risk factors, and is associated with increased cardiovascular event risk; however, whether metabolic syndrome is associated with an increased prevalence of aortic valve calcium (AVC) is not known. Methods and Results— The prevalence of AVC, as assessed by computed tomography, was compared in 6780 Multi-Ethnic Study of Atherosclerosis (MESA) participants with metabolic syndrome (n=1550; National Cholesterol Education Program’s Adult Treatment Panel III [ATP III] criteria), diabetes mellitus (n=1016), or neither condition (n=4024). The prevalence of AVC for those with neither condition, metabolic syndrome, or diabetes mellitus was, respectively, 8%, 12%, and 17% in women (P&lt;0.001) and 14%, 22%, and 24% in men (P&lt;0.001). Compared with those with neither condition, the adjusted relative risks for the presence of AVC were 1.45 (95% CI 1.11 to 1.90) for metabolic syndrome and 2.12 (95% CI 1.54 to 2.92) for diabetes mellitus in women and 1.70 (95% CI 1.32 to 2.19) for metabolic syndrome and 1.73 (95% CI 1.33 to 2.25) for diabetes mellitus in men. There was a graded, linear relationship between AVC prevalence and the number of metabolic syndrome components in both women and men (both P&lt;0.001). Similar results were obtained when the International Diabetes Federation metabolic syndrome definition was used. Conclusions— In the MESA cohort, the metabolic syndrome and diabetes mellitus are associated with increased risk of AVC, and AVC prevalence is increased with increasing number of metabolic syndrome components.</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>10188</td>\n      <td>8ffce7230bfba33d1c2d53cd5f3dd890a4dc091e</td>\n      <td>205791884</td>\n      <td>Immune enhancement is desirable in situations where decreased immunity results in increased morbidity. We investigated whether blocking the surface inhibitory receptor PD‐1 and/or p38 MAP kinase could enhance the proliferation of the effector memory CD8+ T‐cell subset that re‐expresses CD45RA (EMRA) and exhibits characteristics of senescence, which include decreased proliferation and telomerase activity but increased expression of the DNA damage response related protein γH2AX. Blocking of both PD‐1 and p38 MAPK signaling in these cells enhanced proliferation and the increase was additive when both pathways were inhibited simultaneously in both young and old human subjects. In contrast, telomerase activity in EMRA CD8+ T cells was only enhanced by blocking the p38 but not the PD‐1 signaling pathway, further indicating that nonoverlapping signaling pathways were involved. Although blocking p38 MAPK inhibits TNF‐α secretion in the EMRA population, this decrease was counteracted by the simultaneous inhibition of PD‐1 signaling in these cells. Therefore, end‐stage characteristics of EMRA CD8+ T cells are stringently controlled by distinct and reversible cell signaling events. In addition, the inhibition of PD‐1 and p38 signaling pathways together may enable the enhancement of proliferation of EMRA CD8+ T cells without compromising their capacity for cytokine secretion.</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>10190</td>\n      <td>5444be002d34ef9b227113b16ab69bc53f530190</td>\n      <td>13001400</td>\n      <td>Background—Based on the current understanding of cardiac conduction system development and the observation that arrhythmogenic foci can originate in areas near the atrioventricular annuli, we hypothesized that focal annular tachycardias, whether atrial or ventricular, share a common mechanism. We, therefore, prospectively evaluated this hypothesis in patients with sustained atrial and ventricular tachycardia originating from the peri-tricuspid and mitral annuli. Methods and Results—Forty-nine consecutive patients with sustained, focal annular tachycardia comprised the study group. All underwent electrophysiological evaluation and the mode of tachycardia initiation, termination, sensitivity to catecholamine infusion, and response to adenosine/verapamil were evaluated. Electroanatomical activation maps identified the sites of arrhythmia origin. Tachycardias could be initiated or terminated or both with programmed stimulation in 46 of 46 patients and most (70%) were catecholamine facilitated. Of the 9 patients with sustained annular ventricular tachycardia, 3 were localized to the tricuspid annulus, and 6 to the mitral annulus. All the 9 ventricular tachycardias (100%) terminated with adenosine, 2 of 2 terminated with verapamil, and 2 of 2 terminated with Valsalva. Of the 40 patients with annular atrial tachycardia, 4 tachycardias were localized to the mitral annulus and 37 to the tricuspid annulus (including 9 para-Hisian), and all were adenosine sensitive. Conclusions—Peri-annular atrial and ventricular tissue correspond to a region enriched with arrhythmogenic foci, which may reflect a common developmental origin. Furthermore, the sensitivity of these tachycardias to adenosine provides evidence for a shared arrhythmia mechanism, consistent with intracellular calcium overload and triggered activity.</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>10191</td>\n      <td>33c7ee5ca368c25fb87521107a4e4f477ba63ba6</td>\n      <td>14166630</td>\n      <td>Background and Objectives: Hepatitis C Virus genotyping appears to be vital for predicting the response to antiviral therapy. The present study aimed to analyze the HCV genotypes in relation to persistence or clearance of the virus in residents of Hamadan Province, West-Iran. Material and Methods: A total of 1159 recorded questionnaires of HCV infected people were evaluated in this prospective study. Several parameters including HCV genotypes, anti-HCV antibodies, viral load, drug treatment, response to therapy and amount of ALT and AST were analyzed. Results: HCV genotyping in 637 samples revealed a predominance of type 1a (52.1%) followed by 3a (42.4%), type 1b (2.7%) and type 2 (0.2%) respectively. Mixed genotypes (3a-1a) were detected in 0.9%, and 1.7% had untypable genotype. High frequency of genotypes 1a and 3a were observed in drug-resistant (group-a) and drug-sensitive (group-b) patients respectively (P&lt;0.0001). Additionally, duration of drug treatment was significantly higher in group-a than group-b (P&lt;0.0001). During follow-up period, 92 cases showed spontaneous clearance of HCV infection and more importantly 86 of 92 cases were positive for anti-HCV antibodies compared with 59 of 455 antibody positive cases with treatment-induced clearance of HCV infection (P&lt;0.0001). Conclusion: HCV genotyping and also antibody screening could be useful for proper therapeutic intervention in HCV infected subjects.</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>10218</td>\n      <td>fc3ffada1b9cdd9437939cb4023cb3411c1153c8</td>\n      <td>30451760</td>\n      <td>Controversy surrounds the question of whether there is a specific pattern of psychopathology or personality style observed in patients with temporal lobe epilepsy (TLE) or whether the symptoms of psychological distress reflect a common disorder such as depression. Measurement equivalence was examined to test the hypothesis that the latent variable model underlying scores on the Beck Depression Inventory (BDI) was equivalent across samples of patients with TLE (n = 187) and patients with heterogeneous neurological disorders (n = 150). A well-replicated model of depression or psychological distress comprising three related variables, negative attitude, performance difficulty, and somatic elements, displayed a pattern of strict metric invariance. This result suggests that the same set of latent variables is measured with the same metric relationship between item scores and latent variables in patients with TLE and in patients with heterogeneous neurological disorders.</td>\n    </tr>\n  </tbody>\n</table>\n</div>"},"execution_count":null,"metadata":{},"output_type":"execute_result"}]},{"cell_type":"code","source":"input_1 = _hex_json.loads(\"\\\"\\\"\")","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"# birdtokenizer = AutoTokenizer.from_pretrained(\"google/bigbird-pegasus-large-arxiv\")\nlatestbirdtokenizer = AutoTokenizer.from_pretrained(\"pszemraj/bigbird-pegasus-large-K-booksum\")\nlatestbirdmodel =  AutoModelForSeq2SeqLM.from_pretrained(\"pszemraj/bigbird-pegasus-large-K-booksum\")\n# birdmodel = BigBirdPegasusForConditionalGeneration.from_pretrained(\"google/bigbird-pegasus-large-arxiv\")\n# birdpipeline = pipeline(\"summarization\", model=birdmodel, tokenizer=birdtokenizer)\nlatestbirdpipeline = pipeline(\"summarization\", model=latestbirdmodel, tokenizer=latestbirdtokenizer)","metadata":{},"execution_count":null,"outputs":[]},{"cell_type":"code","source":"sample_abstract = abstract_df['abstract'][0]\nlatestbirdpipeline(sample_abstract, num_beams=2, no_repeat_ngram_size=2, top_p=80, top_k=7, min_length=100)","metadata":{},"execution_count":null,"outputs":[{"data":{"text/plain":"Your max_length is set to 512, but your input_length is only 328. Since this is a summarization task, where outputs shorter than the input are typically wanted, you might consider decreasing max_length manually, e.g. summarizer('...', max_length=164)\nAttention type 'block_sparse' is not possible if sequence_length: 328 <= num global tokens: 2 * config.block_size + min. num sliding tokens: 3 * config.block_size + config.num_random_blocks * config.block_size + additional buffer: config.num_random_blocks * config.block_size = 704 with config.block_size = 64, config.num_random_blocks = 3. Changing attention type to 'original_full'...\n"},"execution_count":null,"metadata":{},"output_type":"execute_result"},{"data":{"text/plain":"[{'summary_text': \"The present invention relates to a method for assessing the presence of aortic valve calcium in a subject. The method involves measuring the amount of calcium present in the patient's blood and comparing the measured amount to known amounts for a patient with the same type of heart disease. It is found that the method can be used to assess the risk of developing aortic stenosis, a disease that is associated with an increased number of strokes and heart attacks in elderly subjects. In addition, it is believed that this discovery will have a significant impact on the treatment of other diseases related to blood flow, such as stroke.\"}]"},"execution_count":null,"metadata":{},"output_type":"execute_result"}]},{"cell_type":"code","source":"tokenized_abstract = birdtokenizer(sample_abstract, return_tensors='pt')\ngenerated_summary = birdmodel.generate(tokenized_abstract['input_ids'],\n                                        num_beams=3,\n                                        no_repeat_ngram_size=2,\n                                        top_p=80,\n                                        top_k=7,\n                                        min_length=100)\nprint(birdtokenizer.decode(generated_summary[0], skip_special_tokens=True, clean_up_tokenization_spaces=True))","metadata":{},"execution_count":null,"outputs":[{"data":{"text/plain":"increasing prevalence of metabolic syndrome was determined for participants in the age range 40 to 70 years.<n> the case study was a group of participants with high calcium content ( , ng ) who were selected on the basis of their association with calcium - rich sources ; the calcium was obtained from the study of the distribution of calcium in a region near to the kidneys of patients with diabetes at the national institute of standards and technology of uzbekistan ( nizt - ust)[qcd]$ ] ( see also the complete list of citations : http://arxiv.org/abs/1304.2126[ar xiv:1302.4126 ]) ]. in this study<n> we report on a systematic analysis of data for more than 2000 people with type ia diabetes ( diad - type ii - insulin - dependent diabetes or dia - dm - ii diabetes for those with at least two copies of insulin _ in their system_. the data were collected from clinical trials for the years 2000 to 2010 ( fig.1a.4 )(fig.2b.3 ).<n> results were obtained by means of a direct measurement of prevalence and risk for each participant in both age and diabetes cohort \n"},"execution_count":null,"metadata":{},"output_type":"execute_result"}]},{"cell_type":"code","source":"","metadata":{},"execution_count":null,"outputs":[]}],"metadata":{"orig_nbformat":4,"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"hex_info":{"author":"Nic Brathwaite","project_id":"1e7b90bc-1996-4336-8653-50313b6d2f75","version":"draft","exported_date":"Mon Aug 07 2023 03:50:54 GMT+0000 (Coordinated Universal Time)"}},"nbformat":4,"nbformat_minor":4}