{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: transformers in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (4.29.2)\n",
      "Requirement already satisfied: filelock in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers) (3.12.2)\n",
      "Requirement already satisfied: huggingface-hub<1.0,>=0.14.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers) (0.15.1)\n",
      "Requirement already satisfied: numpy>=1.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers) (1.23.5)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers) (23.1)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers) (6.0)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers) (2022.7.9)\n",
      "Requirement already satisfied: requests in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers) (2.29.0)\n",
      "Requirement already satisfied: tokenizers!=0.11.3,<0.14,>=0.11.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers) (0.13.2)\n",
      "Requirement already satisfied: tqdm>=4.27 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers) (4.65.0)\n",
      "Requirement already satisfied: fsspec in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface-hub<1.0,>=0.14.1->transformers) (2023.4.0)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface-hub<1.0,>=0.14.1->transformers) (4.6.3)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->transformers) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->transformers) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->transformers) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->transformers) (2023.5.7)\n",
      "Requirement already satisfied: langchain in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.0.219)\n",
      "Requirement already satisfied: PyYAML>=5.4.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (6.0)\n",
      "Requirement already satisfied: SQLAlchemy<3,>=1.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.4.39)\n",
      "Requirement already satisfied: aiohttp<4.0.0,>=3.8.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (3.8.3)\n",
      "Requirement already satisfied: async-timeout<5.0.0,>=4.0.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (4.0.2)\n",
      "Requirement already satisfied: dataclasses-json<0.6.0,>=0.5.7 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (0.5.8)\n",
      "Requirement already satisfied: langchainplus-sdk>=0.0.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (0.0.17)\n",
      "Requirement already satisfied: numexpr<3.0.0,>=2.8.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (2.8.4)\n",
      "Requirement already satisfied: numpy<2,>=1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.23.5)\n",
      "Requirement already satisfied: openapi-schema-pydantic<2.0,>=1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.2.4)\n",
      "Requirement already satisfied: pydantic<2,>=1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.10.9)\n",
      "Requirement already satisfied: requests<3,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (2.29.0)\n",
      "Requirement already satisfied: tenacity<9.0.0,>=8.1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (8.2.2)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (22.1.0)\n",
      "Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (2.0.4)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (6.0.2)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.8.1)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.3.3)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.2.0)\n",
      "Requirement already satisfied: marshmallow<4.0.0,>=3.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (3.19.0)\n",
      "Requirement already satisfied: marshmallow-enum<2.0.0,>=1.5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (1.5.1)\n",
      "Requirement already satisfied: typing-inspect>=0.4.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (0.9.0)\n",
      "Requirement already satisfied: typing-extensions>=4.2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pydantic<2,>=1->langchain) (4.6.3)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (2023.5.7)\n",
      "Requirement already satisfied: greenlet!=0.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from SQLAlchemy<3,>=1.4->langchain) (2.0.1)\n",
      "Requirement already satisfied: packaging>=17.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from marshmallow<4.0.0,>=3.3.0->dataclasses-json<0.6.0,>=0.5.7->langchain) (23.1)\n",
      "Requirement already satisfied: mypy-extensions>=0.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from typing-inspect>=0.4.0->dataclasses-json<0.6.0,>=0.5.7->langchain) (1.0.0)\n",
      "Requirement already satisfied: openai in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.27.4)\n",
      "Requirement already satisfied: requests>=2.20 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from openai) (2.29.0)\n",
      "Requirement already satisfied: tqdm in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from openai) (4.65.0)\n",
      "Requirement already satisfied: aiohttp in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from openai) (3.8.3)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.20->openai) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.20->openai) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.20->openai) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.20->openai) (2023.5.7)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (22.1.0)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (6.0.2)\n",
      "Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (4.0.2)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (1.8.1)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (1.3.3)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp->openai) (1.2.0)\n",
      "Requirement already satisfied: langchain in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.0.219)\n",
      "Requirement already satisfied: boto3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (1.26.163)\n",
      "Requirement already satisfied: PyYAML>=5.4.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (6.0)\n",
      "Requirement already satisfied: SQLAlchemy<3,>=1.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.4.39)\n",
      "Requirement already satisfied: aiohttp<4.0.0,>=3.8.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (3.8.3)\n",
      "Requirement already satisfied: async-timeout<5.0.0,>=4.0.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (4.0.2)\n",
      "Requirement already satisfied: dataclasses-json<0.6.0,>=0.5.7 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (0.5.8)\n",
      "Requirement already satisfied: langchainplus-sdk>=0.0.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (0.0.17)\n",
      "Requirement already satisfied: numexpr<3.0.0,>=2.8.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (2.8.4)\n",
      "Requirement already satisfied: numpy<2,>=1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.23.5)\n",
      "Requirement already satisfied: openapi-schema-pydantic<2.0,>=1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.2.4)\n",
      "Requirement already satisfied: pydantic<2,>=1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (1.10.9)\n",
      "Requirement already satisfied: requests<3,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (2.29.0)\n",
      "Requirement already satisfied: tenacity<9.0.0,>=8.1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from langchain) (8.2.2)\n",
      "Requirement already satisfied: botocore<1.30.0,>=1.29.163 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from boto3) (1.29.163)\n",
      "Requirement already satisfied: jmespath<2.0.0,>=0.7.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from boto3) (1.0.1)\n",
      "Requirement already satisfied: s3transfer<0.7.0,>=0.6.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from boto3) (0.6.1)\n",
      "Requirement already satisfied: attrs>=17.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (22.1.0)\n",
      "Requirement already satisfied: charset-normalizer<3.0,>=2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (2.0.4)\n",
      "Requirement already satisfied: multidict<7.0,>=4.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (6.0.2)\n",
      "Requirement already satisfied: yarl<2.0,>=1.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.8.1)\n",
      "Requirement already satisfied: frozenlist>=1.1.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.3.3)\n",
      "Requirement already satisfied: aiosignal>=1.1.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from aiohttp<4.0.0,>=3.8.3->langchain) (1.2.0)\n",
      "Requirement already satisfied: python-dateutil<3.0.0,>=2.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from botocore<1.30.0,>=1.29.163->boto3) (2.8.2)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.25.4 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from botocore<1.30.0,>=1.29.163->boto3) (1.26.16)\n",
      "Requirement already satisfied: marshmallow<4.0.0,>=3.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (3.19.0)\n",
      "Requirement already satisfied: marshmallow-enum<2.0.0,>=1.5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (1.5.1)\n",
      "Requirement already satisfied: typing-inspect>=0.4.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dataclasses-json<0.6.0,>=0.5.7->langchain) (0.9.0)\n",
      "Requirement already satisfied: typing-extensions>=4.2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pydantic<2,>=1->langchain) (4.6.3)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (3.4)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests<3,>=2->langchain) (2023.5.7)\n",
      "Requirement already satisfied: greenlet!=0.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from SQLAlchemy<3,>=1.4->langchain) (2.0.1)\n",
      "Requirement already satisfied: packaging>=17.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from marshmallow<4.0.0,>=3.3.0->dataclasses-json<0.6.0,>=0.5.7->langchain) (23.1)\n",
      "Requirement already satisfied: six>=1.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from python-dateutil<3.0.0,>=2.1->botocore<1.30.0,>=1.29.163->boto3) (1.16.0)\n",
      "Requirement already satisfied: mypy-extensions>=0.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from typing-inspect>=0.4.0->dataclasses-json<0.6.0,>=0.5.7->langchain) (1.0.0)\n",
      "Requirement already satisfied: pypdf in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (3.11.1)\n",
      "Requirement already satisfied: semanticscholar in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.4.1)\n",
      "Requirement already satisfied: requests in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from semanticscholar) (2.29.0)\n",
      "Requirement already satisfied: tenacity in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from semanticscholar) (8.2.2)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->semanticscholar) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->semanticscholar) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->semanticscholar) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->semanticscholar) (2023.5.7)\n",
      "Requirement already satisfied: metapub in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.5.5)\n",
      "Requirement already satisfied: setuptools in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (67.8.0)\n",
      "Requirement already satisfied: lxml in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (4.9.2)\n",
      "Requirement already satisfied: requests in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (2.29.0)\n",
      "Requirement already satisfied: eutils in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (0.6.0)\n",
      "Requirement already satisfied: habanero in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (1.2.3)\n",
      "Requirement already satisfied: tabulate in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (0.9.0)\n",
      "Requirement already satisfied: cssselect in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (1.2.0)\n",
      "Requirement already satisfied: unidecode in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (1.3.6)\n",
      "Requirement already satisfied: docopt in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (0.6.2)\n",
      "Requirement already satisfied: six in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (1.16.0)\n",
      "Requirement already satisfied: tox in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (4.6.3)\n",
      "Requirement already satisfied: pytest in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (7.4.0)\n",
      "Requirement already satisfied: coloredlogs in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (15.0.1)\n",
      "Requirement already satisfied: python-Levenshtein in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from metapub) (0.21.1)\n",
      "Requirement already satisfied: humanfriendly>=9.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from coloredlogs->metapub) (10.0)\n",
      "Requirement already satisfied: pytz in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from eutils->metapub) (2022.7)\n",
      "Requirement already satisfied: tqdm in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from habanero->metapub) (4.65.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->metapub) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->metapub) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->metapub) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->metapub) (2023.5.7)\n",
      "Requirement already satisfied: iniconfig in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (2.0.0)\n",
      "Requirement already satisfied: packaging in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (23.1)\n",
      "Requirement already satisfied: pluggy<2.0,>=0.12 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (1.2.0)\n",
      "Requirement already satisfied: exceptiongroup>=1.0.0rc8 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (1.1.1)\n",
      "Requirement already satisfied: tomli>=1.0.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from pytest->metapub) (2.0.1)\n",
      "Requirement already satisfied: Levenshtein==0.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from python-Levenshtein->metapub) (0.21.1)\n",
      "Requirement already satisfied: rapidfuzz<4.0.0,>=2.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from Levenshtein==0.21.1->python-Levenshtein->metapub) (3.1.1)\n",
      "Requirement already satisfied: cachetools>=5.3.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (5.3.1)\n",
      "Requirement already satisfied: chardet>=5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (5.1.0)\n",
      "Requirement already satisfied: colorama>=0.4.6 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (0.4.6)\n",
      "Requirement already satisfied: filelock>=3.12.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (3.12.2)\n",
      "Requirement already satisfied: platformdirs>=3.5.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (3.8.0)\n",
      "Requirement already satisfied: pyproject-api>=1.5.2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (1.5.2)\n",
      "Requirement already satisfied: virtualenv>=20.23.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tox->metapub) (20.23.1)\n",
      "Requirement already satisfied: distlib<1,>=0.3.6 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from virtualenv>=20.23.1->tox->metapub) (0.3.6)\n",
      "Requirement already satisfied: tiktoken in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.4.0)\n",
      "Requirement already satisfied: regex>=2022.1.18 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tiktoken) (2022.7.9)\n",
      "Requirement already satisfied: requests>=2.26.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from tiktoken) (2.29.0)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests>=2.26.0->tiktoken) (2023.5.7)\n",
      "Requirement already satisfied: sentence_transformers in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (2.2.2)\n",
      "Requirement already satisfied: transformers<5.0.0,>=4.6.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (4.29.2)\n",
      "Requirement already satisfied: tqdm in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (4.65.0)\n",
      "Requirement already satisfied: torch>=1.6.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (2.0.1)\n",
      "Requirement already satisfied: torchvision in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (0.15.2)\n",
      "Requirement already satisfied: numpy in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (1.23.5)\n",
      "Requirement already satisfied: scikit-learn in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (1.2.2)\n",
      "Requirement already satisfied: scipy in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (1.10.1)\n",
      "Requirement already satisfied: nltk in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (3.8.1)\n",
      "Requirement already satisfied: sentencepiece in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (0.1.99)\n",
      "Requirement already satisfied: huggingface-hub>=0.4.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from sentence_transformers) (0.15.1)\n",
      "Requirement already satisfied: filelock in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence_transformers) (3.12.2)\n",
      "Requirement already satisfied: fsspec in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence_transformers) (2023.4.0)\n",
      "Requirement already satisfied: requests in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence_transformers) (2.29.0)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence_transformers) (6.0)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence_transformers) (4.6.3)\n",
      "Requirement already satisfied: packaging>=20.9 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface-hub>=0.4.0->sentence_transformers) (23.1)\n",
      "Requirement already satisfied: sympy in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from torch>=1.6.0->sentence_transformers) (1.11.1)\n",
      "Requirement already satisfied: networkx in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from torch>=1.6.0->sentence_transformers) (2.8.4)\n",
      "Requirement already satisfied: jinja2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from torch>=1.6.0->sentence_transformers) (3.1.2)\n",
      "Requirement already satisfied: regex!=2019.12.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers<5.0.0,>=4.6.0->sentence_transformers) (2022.7.9)\n",
      "Requirement already satisfied: tokenizers!=0.11.3,<0.14,>=0.11.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from transformers<5.0.0,>=4.6.0->sentence_transformers) (0.13.2)\n",
      "Requirement already satisfied: click in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from nltk->sentence_transformers) (8.0.4)\n",
      "Requirement already satisfied: joblib in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from nltk->sentence_transformers) (1.2.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from scikit-learn->sentence_transformers) (2.2.0)\n",
      "Requirement already satisfied: pillow!=8.3.*,>=5.3.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from torchvision->sentence_transformers) (9.4.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from jinja2->torch>=1.6.0->sentence_transformers) (2.1.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->huggingface-hub>=0.4.0->sentence_transformers) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->huggingface-hub>=0.4.0->sentence_transformers) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->huggingface-hub>=0.4.0->sentence_transformers) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->huggingface-hub>=0.4.0->sentence_transformers) (2023.5.7)\n",
      "Requirement already satisfied: mpmath>=0.19 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages/mpmath-1.2.1-py3.10.egg (from sympy->torch>=1.6.0->sentence_transformers) (1.2.1)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-10 18:31:45.985837: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2023-07-10 18:31:47.901435: I tensorflow/tsl/cuda/cudart_stub.cc:28] Could not find cuda drivers on your machine, GPU will not be used.\n",
      "2023-07-10 18:31:47.903911: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.\n",
      "To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.\n",
      "2023-07-10 18:31:50.964560: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT\n",
      "2023-07-10 18:31:57 ip-172-31-90-198 metapub.config[1527] WARNING NCBI_API_KEY was not set.\n"
     ]
    }
   ],
   "source": [
    "!pip install transformers\n",
    "!pip install langchain\n",
    "!pip install openai\n",
    "!pip install langchain boto3\n",
    "!pip install pypdf\n",
    "!pip install semanticscholar\n",
    "!pip install metapub\n",
    "!pip install tiktoken\n",
    "!pip install sentence_transformers\n",
    "from langchain.llms import OpenAI\n",
    "from langchain import PromptTemplate, SagemakerEndpoint\n",
    "from langchain.llms.sagemaker_endpoint import LLMContentHandler\n",
    "from langchain.chains.question_answering import load_qa_chain\n",
    "from langchain.chains.summarize import load_summarize_chain\n",
    "import json\n",
    "from langchain.document_loaders import TextLoader\n",
    "from langchain.document_loaders.csv_loader import CSVLoader\n",
    "from langchain.chains import RetrievalQA\n",
    "from langchain.embeddings.openai import OpenAIEmbeddings\n",
    "from langchain.vectorstores import Chroma\n",
    "from langchain.document_loaders import JSONLoader\n",
    "from langchain.document_loaders import PyPDFLoader\n",
    "from langchain.docstore.document import Document\n",
    "from langchain.text_splitter import CharacterTextSplitter\n",
    "from pathlib import Path\n",
    "from pprint import pprint\n",
    "from langchain import OpenAI, PromptTemplate, LLMChain\n",
    "from langchain.chains.mapreduce import MapReduceChain\n",
    "from langchain.prompts import PromptTemplate\n",
    "from transformers import AutoTokenizer, pipeline, BioGptTokenizer, BioGptForCausalLM\n",
    "from transformers import AutoModelForSeq2SeqLM\n",
    "from transformers import Seq2SeqTrainer, Seq2SeqTrainingArguments\n",
    "import os\n",
    "import glob\n",
    "import tensorflow as tf\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import ast\n",
    "import torch\n",
    "from transformers import T5Tokenizer, TFT5Model, TFT5ForConditionalGeneration, AutoTokenizer, TFBertModel, AutoModel, BertTokenizer, TFBertForQuestionAnswering\n",
    "import sentencepiece\n",
    "from metapub import PubMedFetcher\n",
    "from semanticscholar import SemanticScholar\n",
    "from metapub import FindIt\n",
    "import requests\n",
    "import urllib\n",
    "import json\n",
    "# from transformers import BioGptModel, BioGptConfig, BioGptTokenizer, BioGptForCausalLM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: huggingface_hub in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (0.15.1)\n",
      "Requirement already satisfied: filelock in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface_hub) (3.12.2)\n",
      "Requirement already satisfied: fsspec in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface_hub) (2023.4.0)\n",
      "Requirement already satisfied: requests in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface_hub) (2.29.0)\n",
      "Requirement already satisfied: tqdm>=4.42.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface_hub) (4.65.0)\n",
      "Requirement already satisfied: pyyaml>=5.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface_hub) (6.0)\n",
      "Requirement already satisfied: typing-extensions>=3.7.4.3 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface_hub) (4.6.3)\n",
      "Requirement already satisfied: packaging>=20.9 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from huggingface_hub) (23.1)\n",
      "Requirement already satisfied: charset-normalizer<4,>=2 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->huggingface_hub) (2.0.4)\n",
      "Requirement already satisfied: idna<4,>=2.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->huggingface_hub) (3.4)\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->huggingface_hub) (1.26.16)\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from requests->huggingface_hub) (2023.5.7)\n",
      "Collecting dask\n",
      "  Downloading dask-2023.7.0-py3-none-any.whl (1.2 MB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m1.2/1.2 MB\u001b[0m \u001b[31m27.0 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m00:01\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: click>=8.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dask) (8.0.4)\n",
      "Collecting cloudpickle>=1.5.0 (from dask)\n",
      "  Downloading cloudpickle-2.2.1-py3-none-any.whl (25 kB)\n",
      "Requirement already satisfied: fsspec>=2021.09.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dask) (2023.4.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dask) (23.1)\n",
      "Collecting partd>=1.2.0 (from dask)\n",
      "  Downloading partd-1.4.0-py3-none-any.whl (18 kB)\n",
      "Requirement already satisfied: pyyaml>=5.3.1 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dask) (6.0)\n",
      "Collecting toolz>=0.10.0 (from dask)\n",
      "  Downloading toolz-0.12.0-py3-none-any.whl (55 kB)\n",
      "\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m55.8/55.8 kB\u001b[0m \u001b[31m6.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hRequirement already satisfied: importlib-metadata>=4.13.0 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from dask) (6.7.0)\n",
      "Requirement already satisfied: zipp>=0.5 in /home/ubuntu/anaconda3/envs/ta/lib/python3.10/site-packages (from importlib-metadata>=4.13.0->dask) (3.15.0)\n",
      "Collecting locket (from partd>=1.2.0->dask)\n",
      "  Downloading locket-1.0.0-py2.py3-none-any.whl (4.4 kB)\n",
      "Installing collected packages: toolz, locket, cloudpickle, partd, dask\n",
      "Successfully installed cloudpickle-2.2.1 dask-2023.7.0 locket-1.0.0 partd-1.4.0 toolz-0.12.0\n"
     ]
    }
   ],
   "source": [
    "!pip install huggingface_hub\n",
    "!pip install dask"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "import yaml\n",
    "from sqlalchemy import create_engine\n",
    "engine = create_engine(f'postgresql://postgres:qRd71PqwOsbv62WAvboR@database-1.cuaho2dof33c.us-east-1.rds.amazonaws.com:5432/postgres')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>empty</th>\n",
       "      <th>corpusid</th>\n",
       "      <th>text</th>\n",
       "      <th>id</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1410</td>\n",
       "      <td>250929149</td>\n",
       "      <td>\\nOPEN ACCESS EDITED BY\\n\\n\\nAlexander Nikolae...</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1411</td>\n",
       "      <td>215793097</td>\n",
       "      <td>\\n\\n\\n\\nCi Song \\nState Key Laboratory of Repr...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1412</td>\n",
       "      <td>246020601</td>\n",
       "      <td>\\nCOVID-19 and the Otolaryngology Residency Ma...</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1413</td>\n",
       "      <td>18163935</td>\n",
       "      <td>\\nAnalysis of biophysical and functional conse...</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1414</td>\n",
       "      <td>237935836</td>\n",
       "      <td>\\nImpact of Plant-Based Meat Alternatives on t...</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  empty   corpusid                                               text  id\n",
       "0  1410  250929149  \\nOPEN ACCESS EDITED BY\\n\\n\\nAlexander Nikolae...   1\n",
       "1  1411  215793097  \\n\\n\\n\\nCi Song \\nState Key Laboratory of Repr...   2\n",
       "2  1412  246020601  \\nCOVID-19 and the Otolaryngology Residency Ma...   3\n",
       "3  1413   18163935  \\nAnalysis of biophysical and functional conse...   4\n",
       "4  1414  237935836  \\nImpact of Plant-Based Meat Alternatives on t...   5"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Retreive Full Text from Table\n",
    "from sqlalchemy import text\n",
    "\n",
    "table_name = 'fulltext'\n",
    "\n",
    "sql = text(f''' \n",
    "    SELECT empty, corpusid, text, id FROM {table_name} LIMIT 10;\n",
    "''')\n",
    "\n",
    "with engine.connect() as conn: \n",
    "    query = conn.execute(sql)\n",
    "\n",
    "full_text = pd.DataFrame(query.fetchall())\n",
    "full_text.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>paperId</th>\n",
       "      <th>corpusId</th>\n",
       "      <th>abstract</th>\n",
       "      <th>input_ids</th>\n",
       "      <th>attention_mask</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>859c91de1ab22aeff85558dcf676ee5ffc4981a5</td>\n",
       "      <td>33235381</td>\n",
       "      <td>Summary This work aims at applying concepts of...</td>\n",
       "      <td>[20698, 100, 161, 3, 8345, 44, 6247, 6085, 13,...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2460a37a3305b3bb072770e5bb57ed95496ecf80</td>\n",
       "      <td>72970529</td>\n",
       "      <td>Summary Objectives: To diagnose the hospital i...</td>\n",
       "      <td>[20698, 27919, 7, 10, 304, 18730, 8, 2833, 251...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>e92f481f3be6f0956e0cd6c160a2a384c4eacd76</td>\n",
       "      <td>26375252</td>\n",
       "      <td>Holdoff et al.1 described a retrospective, mon...</td>\n",
       "      <td>[8470, 1647, 3, 15, 17, 491, 5, 536, 3028, 3, ...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>d51f04cfdc8fe907e4a66e948028bf09f3a6af7a</td>\n",
       "      <td>11026954</td>\n",
       "      <td>1. In forty‐one out of forty‐seven dogs under ...</td>\n",
       "      <td>[1300, 86, 19662, 2, 782, 91, 13, 19662, 2, 7,...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>111d6c65ad374861a3c9c70b210996b76b0a7080</td>\n",
       "      <td>95308098</td>\n",
       "      <td>Calculated and observed excited singlet state ...</td>\n",
       "      <td>[18555, 920, 11, 6970, 2787, 712, 17, 538, 703...</td>\n",
       "      <td>[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    paperId  corpusId  \\\n",
       "0  859c91de1ab22aeff85558dcf676ee5ffc4981a5  33235381   \n",
       "1  2460a37a3305b3bb072770e5bb57ed95496ecf80  72970529   \n",
       "2  e92f481f3be6f0956e0cd6c160a2a384c4eacd76  26375252   \n",
       "3  d51f04cfdc8fe907e4a66e948028bf09f3a6af7a  11026954   \n",
       "4  111d6c65ad374861a3c9c70b210996b76b0a7080  95308098   \n",
       "\n",
       "                                            abstract  \\\n",
       "0  Summary This work aims at applying concepts of...   \n",
       "1  Summary Objectives: To diagnose the hospital i...   \n",
       "2  Holdoff et al.1 described a retrospective, mon...   \n",
       "3  1. In forty‐one out of forty‐seven dogs under ...   \n",
       "4  Calculated and observed excited singlet state ...   \n",
       "\n",
       "                                           input_ids  \\\n",
       "0  [20698, 100, 161, 3, 8345, 44, 6247, 6085, 13,...   \n",
       "1  [20698, 27919, 7, 10, 304, 18730, 8, 2833, 251...   \n",
       "2  [8470, 1647, 3, 15, 17, 491, 5, 536, 3028, 3, ...   \n",
       "3  [1300, 86, 19662, 2, 782, 91, 13, 19662, 2, 7,...   \n",
       "4  [18555, 920, 11, 6970, 2787, 712, 17, 538, 703...   \n",
       "\n",
       "                                      attention_mask  \n",
       "0  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  \n",
       "1  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  \n",
       "2  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  \n",
       "3  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  \n",
       "4  [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, ...  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### Retreive Abstracts from Table\n",
    "from sqlalchemy import text\n",
    "\n",
    "table_name = 'abstracts_encodings'\n",
    "\n",
    "sql = text(f''' \n",
    "    SELECT * FROM {table_name} LIMIT 10;\n",
    "''')\n",
    "\n",
    "with engine.connect() as conn: \n",
    "    query = conn.execute(sql)\n",
    "\n",
    "abstract_text = pd.DataFrame(query.fetchall())\n",
    "abstract_text.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\"\\nHeat stress induces apoptosis through transcription-independent p53-mediated mitochondrial pathways in human umbilical vein endothelial cell\\nPublished 26 March 2014\\n\\nZ T Gu \\nState Administration of Traditional Chinese Medicine\\nSchool of Traditional Chinese Medicine\\nThe Key Laboratory of Molecular Biology\\nSouthern Medical University\\nGuangzhouP R China\\n\\nDepartment of intensive care unit\\nKey Laboratory of Tropical Zone Trauma Care and Tissue Repair of PLA\\nGeneral Hospital of Guangzhou Military Command\\nGuangzhouPR China\\n\\nH Wang \\nDepartment of Oncology\\nCancer Research Institute\\nSouthern Medical University\\nGuangzhouPR China\\n\\nL Li \\nState Administration of Traditional Chinese Medicine\\nSchool of Traditional Chinese Medicine\\nThe Key Laboratory of Molecular Biology\\nSouthern Medical University\\nGuangzhouP R China\\n\\nDepartment of intensive care unit\\nKey Laboratory of Tropical Zone Trauma Care and Tissue Repair of PLA\\nGeneral Hospital of Guangzhou Military Command\\nGuangzhouPR China\\n\\nY S Liu \\nDepartment of intensive care unit\\nKey Laboratory of Tropical Zone Trauma Care and Tissue Repair of PLA\\nGeneral Hospital of Guangzhou Military Command\\nGuangzhouPR China\\n\\nX B Deng \\nDepartment of Oncology\\nCancer Research Institute\\nSouthern Medical University\\nGuangzhouPR China\\n\\nS F Huo \\nDepartment of Oncology\\nCancer Research Institute\\nSouthern Medical University\\nGuangzhouPR China\\n\\nF F Yuan \\nState Administration of Traditional Chinese Medicine\\nSchool of Traditional Chinese Medicine\\nThe Key Laboratory of Molecular Biology\\nSouthern Medical University\\nGuangzhouP R China\\n\\nDepartment of intensive care unit\\nKey Laboratory of Tropical Zone Trauma Care and Tissue Repair of PLA\\nGeneral Hospital of Guangzhou Military Command\\nGuangzhouPR China\\n\\nZ F Liu \\nDepartment of intensive care unit\\nKey Laboratory of Tropical Zone Trauma Care and Tissue Repair of PLA\\nGeneral Hospital of Guangzhou Military Command\\nGuangzhouPR China\\n\\nH S Tong \\nDepartment of intensive care unit\\nKey Laboratory of Tropical Zone Trauma Care and Tissue Repair of PLA\\nGeneral Hospital of Guangzhou Military Command\\nGuangzhouPR China\\n\\n&amp; L Su \\nDepartment of intensive care unit\\nKey Laboratory of Tropical Zone Trauma Care and Tissue Repair of PLA\\nGeneral Hospital of Guangzhou Military Command\\nGuangzhouPR China\\n\\nHeat stress induces apoptosis through transcription-independent p53-mediated mitochondrial pathways in human umbilical vein endothelial cell\\nPublished 26 March 201410.1038/srep04469Received 11 October 2013 Accepted 5 March 2014OPEN SUBJECT AREAS: TRAUMA APOPTOSIS Correspondence and requests for materials should be addressed to L.S. (sulei_ICU@163. com) * These authors contributed equally to this work. SCIENTIFIC REPORTS | 4 : 4469 |\\nCells apoptosis induced by intense heat stress is the prominent feature of heat-related illness. However, little is known about the biological effects of heat stress on cells apoptosis. Herein, we presented evidence that intense heat stress could induce early apoptosis of HUVEC cells through activating mitochondrial pathway with changes in mitochondrial membrane potential(DYm), release of cytochrome c, and activation of caspase-9 and -3. We further revealed that p53 played a crucial role in heat stress-induced early apoptosis, with p53 protein rapidly translocated into mitochondria. Using pifithrin-a(PFT), a p53's mitochondrial translocation inhibitor, we found that pretreated with PFT, heat stress induced mitochondrial p53 translocation was significantly suppressed, accompanied by a significant alleviation in the loss of DYm, cytochrome c release and caspase-9 activation. Furthermore, we also found that generation of reactive oxygen species (ROS) was a critical mediator in heat stress-induced apoptosis. In addition, the antioxidant MnTMPyP significantly decreased the heat stress-induced p53's mitochondrial translocation, followed by the loss of DYm, cytochrome c release, caspase-9 activation and heat stress-mediated apoptosis. Conclusively, these findings indicate the contribution of the transcription-independent mitochondrial p53 pathway to early apoptosis in HUVEC cells induced by oxidative stress in response to intense heat stress.\\n\\nH eat is likely to be gained by way of infrared radiation from the sun, or direct contact with hot objects, and absorbed when the ambient temperature rises above body temperature, as well as generated as a byproduct of cellular metabolism and the mechanical work of skeletal muscle 1 . Thus, heat stress is widely considered as a major extracellular stimulus 2 . Moreover, an increase in cellular temperature incurs protein denaturation and interrupts critical cellular processes, thus resulting in apoptosis and cell death 3 . It is well established that intense heat stress wrecks damages to the organism mainly by direct toxicity to cell 4 and radiation 5 .Temperaturesgreater than 41.6uC to 42uC,the critical human body temperature point, can lead to cell death from apoptosis even in a few hours 6 ,whereas extreme high temperatures of 49uC to 50uC can give rise to cellular death from necrosis in less than 5 minutes 7 . Therefore, excessive gain of heat from the elevated ambient temperature has been proposed to be a risk factor for the severely life-threatening disorder known as heat stroke characterized by an increased core body temperature that rapidly rises to above 40uC and engenders central nervous system dysfunction 8 . Besides, the mortality thereof is found very high even after adequate hypothermia or other intensive care therapy 9 . For the heat wave that affected Europe in the summer of 2003, 45 000 deaths have been reported of which one-third was attributed to heatstroke 10,11 . With the prospect of increased global temperatures in near future, the mortality and morbidity may continue to rise 11 . However, the pathogenesis of tissue injury and cell death in heat-stroke attributed to heat stress is not well understood. Studies in cell lines and animal models have suggested that endothelial-cell is an early target of heat stress injury and also is a prominent feature of severe heat-stroke 12,13 . Recent studies suggest that endothelial-cell can induce significant apoptosis during acute-phase response to heat stress 14 .Thus,endothelial-cell apoptosis appears to be a mechanism of heat-stroke,but little is known about the biological effects of heat stress on endothelial-cell apoptosis.\\n\\nThe tumor suppressor gene p53 is a key regulator of apoptosis, which has proapoptotic activity through transcription-dependent or independent pathway 15 . Under stressed conditions, p53 protein works as a transcription factor, activating a series of genes such as Puma, Noxa, Bax, and Bid 16 . On the other hand, cytoplasmic p53 directly activates mitochondrial death pathway through a transcription-independent manner 17 . It recently becomes clear that early in the course of p53-dependent apoptosis, caused by death signal stimuli, a fraction of p53 proteins rapidly undergo translocation to mitochondria(detectable at 30 min-1 h), and trigger an early first wave of marked caspase 3 activation, followed by an early wave of apoptosis 18 . This rapid first wave of apoptosis is transcription-independent, and precedes a second slower wave that is transcription-dependent 18 . Besides, such death signal inducers, which induce mitochondrial p53 accumulation and rapid apoptosis, have been confirmed including exposure to oxidative stress, DNA-damaging drugs, c-irradiation or hypoxia 19 . Oxidative stress have also been indentified in numerous studies correlated with p53-directed cell apoptosis 20,21 . In addition, preliminary data have indicated that ROS plays a signaling role in mitochondrial migration of p53 22 and oxidative stress is thought to play a pivotal role in heat stress induced apoptosis 23 . However, the field about how ROS regulates p53-mediated cell apoptosis and heat stress induced apoptosis remains rarely trodden.\\n\\nTherefore, in this study, we investigated the molecular mechanism of apoptosis induced by intense heat stress. We reported for the first time that intense heat stress can induce early wave apoptosis in HUVEC cells which occurs through the mechanism that intense heat stress induces p53 translocation to mitochondria, and triggers mitochondrial apoptotic pathways. We also found that ROS produced by heat stress might be involved in transcription-independent p53mediated mitochondrial pathways.\\n\\n\\nResults\\n\\nHeat shock inhibits cell proliferation in HUVEC cells. To investigate the effect of heat stress on inhibition of HUVEC cells proliferation, we initially examined the effect of heat stress on the proliferation of HUVEC cells. HUVEC cells were maintained in culture media for 48 h and the culture dishes were sealed with parafilm and immersed for 2 h in a circulating water bath thermoregulated at 37uC, 39uC, 41uC, 43uC, or 45uC for the heat stress treatments 2 . Culture media were replaced with fresh media and the cells were further incubated at 37uC for 6 h. As shown in (Figure 1), exposure of HUVEC cells to heat stress for 2 h and further incubation for 6 h registered an inhibition of the growth of HUVEC cells in a temperature-dependent manner.\\n\\nIntense heat stress induces early wave of apoptosis by triggering the mitochondrial pathway in HUVEC cells. A quantitative evaluation was made, with flow cytometry adopted to assess the apoptotic cells. When cells were exposed to intense heat stress, the number of early apoptotic cells increased markedly under the treatment of HUVEC cells at 43uC in 2 h and apoptosis induced amounted to 27.7% by 6 h (Figure 2). In order to determine the apoptotic pathways involved in intense heat stress-induced early wave of apoptosis,we examined the changes in the activities of initiator caspase (caspase-8,-9, and-4) and effector caspase (caspase-3). As shown in Figure 3a,d caspase-9 and -3 activation were detected at 3 h after heat stress treatment. The levels of caspase-9 and -3 activities were increased by more than 2 fold and 3 fold, respectively. Caspase-8 and -4 were not activated. We also analyzed the changes in mitochondrial membrane potential(DYm) and release of cytochrome c from mitochondria in heat stress-treated HUVEC cells. As shown in Figure 3e the exposure of HUVEC cells to HUVEC cells viability after heat stress was tested by WST-1. Results are expressed as the percentage of cell proliferation relative to the proliferation of control treatment. The data shown are the mean from three independent experiments. Each value is the mean 6 SD of three determinations. The asterisk indicates a significant difference between control and test groups, *p , 0.05. intense heat stress resulted in the reduction of mitochondrial membrane potential in a time-dependent manner followed by significant increases in the release of cytochrome c from mitochondria( Figure 3f). These results indicated that intense heat stress might induce the mitochondrial pathway leading to caspase-9/ caspase-3-mediated early apoptosis in HUVEC cells. To further explore whether heat stress induces apoptosis through the mitochondrial pathway, we infected HUVEC cells stably over-expressBcl-2 or the vector alone (control cells) with several strains. As shown in Figure 4a,b and c, the overexpressing Bcl-2 of cells significantly decreased the heat stress-induced apoptosis, the loss of DYm and cytochrome c release. These data further supports the hypothesis that the mitochondrial pathway may mediate the intense heat stress-induced early wave of apoptosis of HUVEC ells.\\n\\nP53 activation of the mitochondrial apoptosis pathway. Previous study has reported heat stress induces apoptotic death in cells through mitochondrial pathway 24 . Our study has also found that intense heat stress results in caspase-9 activation. However, the upstream of mitochondrial apoptotic pathway induced by intense heat stress has not been fully illustrated. Nevertheless, It is well known that p53 protein plays a central role in the regulation of apoptosis. To investigate the involvement of p53 in heat stressinduced apoptosis, the effects of the transfection of the P53 siRNA on the cellular features of apoptosis were studied. As shown in Figure 5a, the transfected cells were assayed for P53 protein expression by western blot assay. Of the several tested siRNA target sequences, one siRNA was effective in decreasing protein levels. P53 mRNA level measured by RT-PCR was significantly decreased in P53 siRNA-transfected cells compared to control cells (scrambled siRNA) ( Figure 5b). The P53 siRNA-transfected cells expressed reduced level of p53 protein, and were less susceptible to intense heat stress-induced early apoptosis compared with control cells (Figure 5c). The caspase-9 activation, release of cytochrome c and mitochondrial membrane potential(DYm) in response to intense heat stress were also studied in P53 siRNA transfected HUVEC cells. When compared with control cells, P53 siRNA transfected HUVEC cells showed a drastic reduction in caspase-9 activity ( Figure 5d) and the release of cytochrome c (Figure 5f). The loss of DYm was alleviated in P53 siRNA transfected HUVEC cells compared with control cells (Figure 5e). These results indicate that signaling of heat stress-induced apoptosis is in part dependent on the expression of p53.\\n\\nTranslocation of p53 to Mitochondria in heatstress-treated HUVEC cells. Based on evidence for the crucial role of transcriptionindependent p53-mediated cells apoptosis in this process of early wave apoptosis 25 , we analyzed further the subcellular distribution of p53 in HUVEC cells during early apoptosis induced by intense heat stress. Highly enriched mitochondria fractions e purified from untreated or heat-treated HUVEC cells were prepared by sucrose gradient centrifugation 26 and analyzed by Western blot. Exposure of HUVEC cells to heat stress for 2 h, and further incubation for different times exhibited a significant amount of p53 accumulation at the mitochondria in a time -dependent manner, which occurs at 0 h after cells were treated with 43uC and peaked at 6 h ( Figure 6a). To directly localize p53 in mitochondria of early apoptosing cells, we performed indirect immunofluorescence. P53 mitochondrial translocation was detected by immunofluorescence staining (Figure 6b). These findings thus suggest that intense heat stress induces rapid mitochondrial translocation of p53, which may activate the apoptotic program.\\n\\nMitochondrial localization of p53 promotes mitochondrial apoptosis pathways. To verify further that mitochondrial p53 accumulation is associated with heat stress induced cells apoptotic death, we administered pifithrin-a(PFT), a highly selective and potent synthetic p53 inhibitor to show that PFT can block p53's mitochondrial translocation 27 . HUVEC cells were pretreated with PFT for 1.5 h, then treated with intense heat stress (43uC) for 2 h. As shown in Figure 7a, after pretreatment with PFT the heat stress induced mitochondrial p53 translocation displayed a significant reduction. To investigate the downstream of the mitochondrial p53 pathway,we examined cytochrome c release to the cytosol, caspase-9 activation and mitochondrial membrane potential(DYm)in PFT treated HUVEC cells. A significant decrease in cytosolic cytochrome c (Figure 7b), and caspase-9 activation (Figure 7c) were observed. The loss of DYm was also alleviated in pretreated with PFT HUVEC cells compared with control cells (figure 7d). These results indicate that p53 molecules accumulated in response to heat stress translocate into the mitochondria and activate mitochondrial apoptosis pathway.\\n\\nIntense heat stress increases the generation of ROS in HUVEC cells. Based on the evidence that ROS generation plays an important role in heat stress 28 , we further validated the possibility that heat stress-induced apoptosis prompts ROS accumulation. We utilized the fluorescent dye DCFH-DA, which produces enhanced The role of ROS on heat stress-induced apoptosis. To investigate whether ROS generation via heat stress also involves p53-mediated mitochondrial pathways, we assessed mitochondrial apoptosis related proteins in HUVEC cells which had been pretreated with antioxidant MnTMPyP (10 umol/l) for 1 h, then treated with intense heat stress (43uC). The depletion of ROS by the cell-   permeable ROS scavenger MnTMPyP significantly decreased the heatstress-mediated translocation of p53 proteins to mitochondria, cytochrome c release from mitochondria, caspase-9 activity, and apoptosis ( Figure 9). These data confirm that heat stress treatment of HUVEC cells activates the mitochondria via release of ROS.\\n\\n\\nDiscussion\\n\\nExposure to high ambient temperature and an elevation of body temperature has been proposed to be a risk factor for heat-related illness representing a continuum of disorders from minor syndromes such as heat cramps, heat syncope, and heat exhaustion to the severely life-threatening disorder known as heat stroke 29 . Although lots of researches were performed on this during the past decade, the mechanism of heat-related illnesses is not clear 30 . Studies in animal models found that there was widespread apoptosis in spleen, lymph node, thymus, small bowel, and cells of lymphoid origin in the lamina propria 12 . Using in vitro approach, numerous studies have established that apoptosis can be promoted by application of heat stress [31][32][33] . These data imply that apoptosis may play an important role in the physiology/pathophysiology of heat-related illness attributed to heat stress. In our study, we focused on endothelial cells because (1) endothelial cells are exposed to diverse stressors,which serve not only as a barrier between the bloodstream, its elements and toxins, and tissue parenchyma but also as generalized transducing cells that modulate the parenchymal responses to those stimuli 34 ; (2) Endothelial-cell injury in acute-phase response to heat stress are the prominent features of heat stroke 8 ; and (3) Endothelial-cell apoptosis has been shown to increase the extent of injury in cell lines and animal models attributable to heat stress 12,35 . In this work, we investigated the biological effects of intense heat stress on cells apoptosis, and showed that intense heat stress could induce early apoptosis through the transcription-independent mitochondrial p53 pathway, which might involve reactive oxygen species(ROS) attributed to intense heat stress.\\n\\nMitochondria are central integrators and transducers for proapoptotic signals, forming the nexus between the non-specific inducer phase and the final execution phase of apoptosis 36 . Among studies about cell lines, Hsu YL et al found that heat stress triggered the mitochondrial apoptotic pathway resulting in caspase-9 activity 24 . A previous study by Milleron RS et al. also found that under intense heat stress, cells underwent early phase of apoptosis (within 4 h of heat stress),which triggered the mitochondrial apoptotic pathway by activating the apical protease that induced mitochondrial outer membrane permeabilization(MOMP), a loss in DYm and caspase-3 activation 37 . Consistently, our results showed that intense heat stress results in a significant increase of release of cytochrome c from mitochondria and decrease of DYm. In addition, our results also revealed the degree of caspase-9 and caspase-3 activity after HUVEC cells were treated with intense heat stress. Caspase-8 and -4 was not activated in this study. Therefore, our results suggested that intense heat stress might initiate mitochondrial other than death receptors or the endoplasmic reticulum (ER) to promote the early wave of apoptosis of HUVEC cells.\\n\\nP53 is a transcription factor which modulates expression of numerous target genes that control apoptosis,cell cycle arrest, senescence, DNA repair and genetic stability in response to death signal stressor 38 . In addition to its transcriptional route, P53 also promotes apoptosis through transcription-independent mechanisms, primarily signals through the mitochondrial pathway 39 . Mitochondrial p53 targeting occurs in a wide spectrum of cell types and after a variety of stress signals including oxidative stress, DNA damage, irradiation and hypoxic stress 40 . Previous studies demonstrated that early in the course of p53-dependent apoptosis, caused by such stressor, a fraction of wild-type p53 protein rapidly migrates to the mitochondria 41 . Once at the mitochondrion, p53 leads to mitochondrial dysfunction that can directly trigger the release of pro-apoptotic factors from the mitochondrial intermembrane space 42 . Tian XJ et al. 43 recently demonstrated that at the population level, apoptosis appears in two waves, i.e the fast wave mediated by the mitochondrial p53 pathway and the slow wave by the nuclear p53 pathway. We have shown in this study that mitochondrial p53 translocation started within 1 h and was significant at 3 h after intense heat-treatment, followed by early cells apoptosis (peak at27.7% at 6 h after43uC for 2 h).To identify whether mitochondrial p53 accumulation is associated with heat stress-induced early cells apoptosis, we found that knockdown of p53 by P53 siRNA, reduced DYm loss, release of cytochrome c and caspase-9 and -3 activation, as well as early wave of apoptosis ( Figure 5).This shows that P53 plays an important role in intense heat stress-mediated HUVEC cellular early apoptosis. We also further evaluated the effect of Pifithrin-a(PFT) on apoptosis cells induced by intense heat stress. We found that PFT can significantly block p53's mitochondrial translocation, followed by decreased cytochrome c release (Figure 7), caspase-9 activation, as well as alleviated DYm decrease in heat-treated cells. Therefore, it is plausible that the p53 in heattreated HUVEC cells translocates to mitochondria where it triggers mitochondrial apoptotic pathways. These results are consistent with the notion that mitochondrial p53 can trigger fast apoptosis at higher damage levels 43 . Our data also explains the intriguing observation by Milleron RS et al 37 .who showed that intense heat stressinduced cells an early apoptosis through an unknown upstream of mitochondria.\\n\\nOxidative damage to cellular compartments has great implication for animal production as well as human health 44 . Under physiological conditions, the redox status of all aerobic cells is balanced by enzyme and non-enzyme systems. However, under certain stressful conditions, increases in generation of ROS and/or decreases in anti-oxidant capacity enhanced oxidative stress 45 . Pucciariello C et al. 46 discovered that acute heat stress could disturb the balance between the ROS production and the antioxidant systems. Several reports link oxidative stress with heat stress and suggest synergistic augmentation of cell death, and report increased ROS generation in heat exposed cells [47][48][49] . What's more, recent studies have revealed that ROS acts as both an up-stream signal that triggers p53 activation and as a downstream factor that mediates apoptosis 21 . In other previous studies, ROS induced p53 (,2%) moves into mitochondria and initiates apoptosis 50 . Our study has shown that intense heat stressmediated oxidative stress works primarily by increasing the production of ROS. We further observed that heat stress generates ROS, which activates caspase-9 rather than caspase-8 and -4. Furthermore, using MnTMPyP, an agent which is a cell-permeable ROS scavenger 51 , able to block caspase-9 activity, P53 moves to mitochondria and engenders apoptosis. All in all, ROS, as the production of acute heat stress, might act as an upstream signal that triggers transcription-independent p53-mediated mitochondrial pathways to promote the early wave of apoptosis of intense heat stress on HUVEC cells.\\n\\nIn conclusion, our data indicate that HUVEC cells undergo apoptosis soon after intense heat stress. Intense heat stress-induced early wave of apoptosis is associated with the mitochondrial apoptosis pathway involving translocation of p53 to mitochondria, which are in turn mediated by ROS generation. Our study provides for the first time an evidence of ROS involved transcriptional-independent p53mediated mitochondrial pathways in heat stress-mediated apoptosis of HUVEC cells.\\n\\n\\nMethods\\n\\nCell culture, treatments and cell proliferation assay. Human umbilical vein endothelial cells (HUVECs) were purchased from Shanghai Institute of Cell Biology, Chinese Academy of Sciences. Cells passage 3, were grown in media recommended by the manufacturer. HUVEC cells maintained in culture media Culture dishes were sealed with parafilm and immersed for 2 h in a circulating water bath thermoregulated at 37uC 6 0.5uC for the control treatment or at 39uC,41uC, 43uC,or45uC 6 0.5uC for the heat stress treatments. Culture media were replaced with fresh media and the cells further incubated for the indicated times. Cell proliferation was assessed by Premixed WST-1 Cell Proliferation Reagent (Clontech Laboratories Inc., Mountain View, CA, USA) according to the manufacturer's instructions.\\n\\nFlow cytometric analysis of cell apoptosis using Annexin V-FITC/PI staining. Cell cycle phase were either kept untreated or exposed to 2 h 43uC before being analyzed by flow cytometry. The detection was performed according to the manual of AnnexinV-FITC apoptosis detection kit (invitrogen). About 1 3 10 6 cells were collected, washed with ice-cold PBS, and resuspended in the binding buffer containing suitable amount of Annexin V-FITC. After 10 min incubation in the dark at room temperature, the buffer was removed by centrifugation. The cells were resuspended in reaction buffer containing propidium iodide (PI). Then flow cytometric analysis was immediately performed to detect apoptosis.\\n\\nMeasurements of ROS. Levels of intracellular ROS were assessed by use of a reactive oxygen species assay kit (Beyotime, Nanjing, China). dichlorofluoresceindiacetate(DCFH-DA, Molecular Probes,Beyotime) was used. This reagent enters the cells and reacts with ROS, producing the fluorophoredichlorofluorescein (DCF). Briefly, cell cycle phase was either kept untreated or exposed to 2 h 43uC, and further incubated for different times, then about 3 3 10 5 were harvested, washed with serum free DMEM culture medium and the cells were stained with10 mM dichlorofluoresceindiacetate (DCFH-DA, Molecular Probes,Beyotime) for 30 min at 37uC in the darkness. Then the cells were harvested, washed and resuspended in serum-free DMEM culture medium for three times. The cells' fluorescence intensity was determined using flow cytometer 52 . 200 mM H 2 O 2 -treated HUVEC cells were as a positive control.\\n\\nAssay for caspase activity. After treatment with 43uC for 2 h, cells were harvested, and cell lysates were prepared at 280uC for 30 min and then incubated with the following appropriate caspase substrates at 37uC using a Quadruple MonochromatorMicroplate Reader (Infinite M1000, Tecan US, NC, USA). Caspase activities were measured by cleavage of the following fluorogenic peptide substrates 53,54 : Ac-LEHD-AFC for caspase-9, Ac-DEVD-AMC for caspase-3, Ac-ATAD-AFC for caspase-8 and Ac-LEVD-AFC for caspase-4. Activities of caspases are represented as relative cumulative fluorescence of the kinetic reaction and compared to untreated controls.\\n\\nImmunofluorescence analysis. HUVEC cells were either left without or with heat stress for 2 h and further incubated for 3 h, then removed of the culture medium. All cells were labeled by adding pre-warmed (37uC) staining solution containing 200 nM MitoTrackerH probe and incubated for 20 minutes under 37uC. The cells were washed with PBS twice, subsequently fixed in 3.5% paraformaldehyde, permeabilized in KB solution and 0.2% Triton X-100 at room temperature. After the blocking solution was washed out, cells were incubated with a primary antibody (rabbit anti-human p53, 15500) (diluted in KB) for 30-45 min at 37uC and subsequently f washed with KB twice. After incubated for 30-45 min at 37uC with secondary antibody (diluted in KB), cells were washed with KB again before mounted onto slide with mounting solution containing 0.2 mg/ml DAPI and sealed with nail polisher. The slide were stored at 4uC in a dark box and observed under a fluorescent microscope.\\n\\nMitochondrial membrane potential assay. The mitochondrial membrane potential was detected using5,59,6,69-tetrachloro-1, 19,3,39 tetraethylbenzimidazolcarbocyanine iodide (JC-1; Molecular Probes, Eugene, OR, USA). Following treatment with heat stress, cells were rinsed once with complete culture medium and incubated with JC-1 for 30 min at 37uC, centrifuged, washed twice with cold PBS, transferred to a 96-well plate (10 5 cells/well), and assayed in a fluorescence plate reader with the following settings:excitation at 490 nm, emission at 540 nm and 590 nm, Changes in the ratio between the measurement at test wave lengths of 590 nm (red) and 540 nm (green). Fluorescence intensities are indicative of changes in the mitochondria membrane potential 55 .\\n\\nPlasmid construction and stable transfection. Eukaryotic expression vectors of pcDNA3.1-Bcl-2 were constructed by Shanghai Genechem (Genechem Incorporation, Shanghai, China). HUVEC cells were transfected with empty vector(pcDNA3.1), Bcl-2 expression vector (pcDNA3.1-Bcl-2)by Lipofectamine 2000. Culture medium containing G418 was used to select stable transfectants.\\n\\nsiRNA transfection. Small-interfering RNA (siRNA) for P53 was designed and synthesized by Guangzhou RiboBio (RiboBioInc, China). The sequence of each gene and their scrambled are shown in follows. For p53,sense and antisense siRNAs are 59-GAAAUUUGCGUGUGGAGUA (dTdT)-39 and 59 -UACUCCACACGCAAAU-UUC(dTdT)-39, respectively. For scrambled control, sense and antisense siRNAs are 59-UGGUUUACAUGUCGACUAA(dTdT)-39 and 59-UUAGUCGACAUGUA-AACCA(dTdT)-39, respectively. Twenty-four hours prior to transfection, HUVEC cells were plated onto a 6-well plate (Nest, Biotech,China) at 30-50% confluence, which were then transfected into cells using TurboFect TM siRNA Transfection Reagent (Fermentas, Vilnius, Lithuania) according to the manufacturer's protocol. Cells were collected after 48-72 h for the further experiments. The mRNA and protein levels of P53 were estimated by RT-PCR and Western blotting. After 2 days of incubation at 37uC in a humidified atmosphere of 5% CO2, cells were exposed to heat stress.\\n\\nWestern blot analysis. HUVEC cells were pretreated with or without 43uC for 2 h, and further incubated for different times as indicated. Western blot was carried out as described 56,57 with monoclonal Anti-p53 antibodies(151000; Cell signaling technology,Danvers, USA) and Polyclonal Anti-Cytochrome C antibodies(151000; Cell signaling technology,Danvers, USA). An HRP-conjugated anti-rabbit IgG antibody was used as the secondary antibody (ZhongshanInc, China). Signals were detected using enhanced chemiluminescence reagents (Pierce, Rockford, IL, USA).\\n\\nRNA isolation, reverse transcription, and qRT-PCR. Total RNA was extracted from the p53siRNA and ScrsiRNA HUVEC cells using Trizol(Takara, Shiga, Japan). For P53, RNA was transcribed into cDNA and amplified with Specific sense: 5-ACACGCTTCCCTGGATTG-3; antisense primer: 5-CAAGAAGCCCAGAC-GGAAAC-3. Reverse transcription and QPCR were performed in accordance with manufacturer's instructions (Takara, Shiga, Japan). The PCR reaction for each gene was repeated three times. GAPDH was used as inner control for normalizing p53 expression. Differential expression of p53 was calculated using the 2-DDCt method.\\n\\nStatistical analysis. All data were analyzed for statistical significance using SPSS 13.0 software (SPSS, Chicago, IL, USA). Data were expressed as means 6 SD from at least 3 independent experiments performed in duplicate. Statistical comparisons of the results were made using analysis of One-way ANOVA. p , 0.05 was considered statistically significant.\\n\\nFigure 1 |\\n1The effects of heat stress on cell viability in HUVECs cells.\\n\\nFigure 2 |\\n2Intense heat stress-induced apoptosis. HUVEC cells were treated with 43uC for 2 h, culture media were replaced with fresh media and the cells were further incubated for different times as indicated, and apoptosis was analyzed by flow cytometry using Annexin V-FITC/PI staining. Data are presented as means 6 SD of three separate experiments. The asterisk indicates a significant difference between control(37uC) and test groups(43uC), *p , 0.05.www.nature.com/scientificreports SCIENTIFIC REPORTS | 4 : 4469 | DOI: 10.1038/srep04469\\n\\nFigure 3 |\\n3Heat stress-induced apoptosis through the initiation of the mitochondrial pathway. HUVEC cells were treated with 43uC for 2 h, culture media were replaced with fresh media and the cells were further incubated for different times as indicated. (a-d) Enzymatic activity of caspase-9,-3,-8 and -4 was measured in cell lysates using fluorogenic substrate Ac-LEHD-AFC,Ac-DEVD-AMC, Ac-ATAD-AFC and Ac-IETD-pNA, respectively, and was expressed relative to the control at 37uC (100%). (e) The loss of DYm was measured by JC-1 and flow cytometry. (f) The expression levels of cytochrome C were determined by Western Blots(cropped).COX IV, mitochondrial loading control. Each value is the mean 6 SD of three separate determinations, *P , 0.05 compared with control group(37uC).www.nature.com/scientificreports SCIENTIFIC REPORTS | 4 : 4469 | DOI: 10.1038/srep04469\\n\\nFigure 4 |Figure 5 |\\n45Heat stress-induced apoptosis in HUVEC cells over-expressing protein Bcl-2. HUVEC cells stably over-expressing the anti-apoptotic protein Bcl-2 or vector alone were treated with 43uC for 2 h, the cells further incubated for 6 h, and(a) Western Blots of bcl-2 protein expressed(cropped) in HUVEC transfectant cells. b-actin was run as an internal control.(b) apoptosis was analyzed by flow cytometry using Annexin V-FITC/PI staining. (c) The loss of DYm was measured by JC-1 and flow cytometry. (d)Intracellular location of cytochrome C was determined by Western Blots(cropped). COX IV, mitochondrial loading control. Each value is the mean 6 SD of three separate determinations. *P , 0.05 versus vector alone cells exposed to heat stress. Heat stress-induced apoptosis in P53 siRNA transfectant HUVEC cells. (a) HUVEC cells were transfected with scrambled siRNA (Scr) or P53 siRNA (P53). Western Blots of P53 protein expressed(cropped) in HUVEC transfectant cells.b-actin was run as an internal control. (b) RT-PCR analysis of P53 in HUVEC transfectant cells. Total RNA was isolated from HUVEC transfectant cells and P53 mRNA levels were determined by RT-PCR and analyzed on agarose gel electrophoresis. b-actin from the same samples was amplified as control. (c-f) HUVEC transfectant cells were treated with 43uC for 2 h with the cells further incubated for 6 h, and (c) apoptosis was analyzed by flow cytometry using Annexin V-FITC/PI staining. (d) Enzymatic activity of caspase-9 was measured in cell lysates using fluorogenic substrate Ac-LEHD-AFC and was expressed relative to the control at 37uC (100%).(e) The loss of DYm was measured by JC-1 and flow cytometry. (f) Intracellular location of cytochrome C was determined by Western Blots(cropped). COX IV, mitochondrial loading control. Each value is the mean 6 SD of three separate determinations. *P , 0.05 versus scrambled siRNA-transfected cells exposed to heat stress. www.nature.com/scientificreports SCIENTIFIC REPORTS | 4 : 4469 | DOI: 10.1038/srep04469 fluorescence when cells generate ROS. H 2 O 2 -treated HUVEC cells were used as a positive control. DCFH-DA-based detection revealed that intracellular ROS production in HUVEC cells increased due to the treatment with heat stress. As shown in Figure 8, we observed an increase in a time-dependent manner in intracellular ROS levels in HUVEC cells after the treatment with heat stress which peaked at 2 h.\\n\\nFigure 6 |\\n6Intense heat stress induces translocation of p53 from nucleus to mitochondria. (a) To determine localization of p53, HUVEC cells were treated with 43uC for 2 h, further incubated for different time as indicated, and the levels of p53 (mitochondrial and cytosolic fraction) proteins were assessed by Western blot using corresponding antibodies(cropped). COX IV, mitochondrial loading control. (b) immunofluorescence localization of p53 to mitochondria in heat-treated cells. HUVEC cells were treated with 43uC for 2 h, and stained with anti-p53 antibody (green), MitoTracker (red) and nuclear(blue). Merged images are shown in the right panels.\\n\\nFigure 7 |\\n7Mitochondrial p53 activates mitochondrial apoptosis pathway in PFT treated HUVEC cells. HUVEC cells were pretreated with or without pifithrin-a for 1.5 h followed by exposure to 43uC for 2 h, and the cells further incubated for 6 h. (a-b) Western blot(cropped) analysis of the effect of PFT on heat stress-induced P53 mitochondrial translocation and release of cytochrome c in HUVEC cells. COX IV, mitochondrial loading control. (c) The effect of PFT on enzymatic activity of caspase-9 was measured in cell lysate using fluorogenic substrate Ac-LEHD-AFC and was expressed relative to the control at 37uC (100%). (d) Flow cytometry and JC-1 measure the effet of PFT on DYm. Data are presented as mean 6 SD for three independent experiments, *P , 0.05 compared with HS group.\\n\\nFigure 8 |\\n8The effect of heat stress on the production of ROS in HUVEC cells. HUVEC cells were treated for 2 h in a water bath at 37uC or 43uC.Culture media were replaced with fresh media and the cells were further incubated for the indicated time. The amounts of ROS were assayed by DCFH staining. H 2 O 2 was used as a positive control. Data are presented as mean 6 SD for three independent experiments, *P , 0.05 compared with control group(37uC).\\n\\nFigure 9 |\\n9ROS involved in P53-mediated mitochondrial apoptotic pathway. HUVEC cells were pretreated with or without MnTMPyP (10 umol/l) for 1 h before exposure to heat stress. (a) the effect of MnTMPyP on the production of ROS in HUVEC cells were assayed by DCFH staining.(b)Flow cytometry analysis of the effect of MnTMPyP on heat stress-induced cells apoptosis. H 2 O 2 was used as a positive control. (c)The effect of MnTMPyP on enzymatic activity of caspase-9 was measured in cell lysates using fluorogenic substrate Ac-LEHD-AFC and was expressed relative to the control at 37uC (100%). (d) Flow cytometry and JC-1 to measure the effect of MnTMPyP on DYm. (e-f): Westen blots analysis(cropped) on the effects of MnTMPyP on P53 mitochondrial translocation, and cytochrome c release from mitochondria induced by intense heat stress. COX IV, mitochondrial loading control. Data are presented as mean 6 SD for three independent experiments, *P , 0.05 compared with HS group.\\nSCIENTIFIC REPORTS | 4 : 4469 | DOI: 10.1038/srep04469\\nAcknowledgmentsThis study was supported by National Natural Science Foundation of China (NO.81101467, 81101406), the project team of Natural Science Foundation of Guangdong Province(s2013030013217), the Project of Medical Research of PLABWS12J108 and Natural Science Foundation of Guangdong Province(NO.10151001002000001).Additional informationCompeting financial interests: The authors declare no competing financial interests.\\nHeat-related illness. N M Lugo-Amador, T Rothenhaus, P Moyer, Emerg Med Clin N Am. 22Lugo-Amador, N. M., Rothenhaus, T. & Moyer, P. Heat-related illness. Emerg Med Clin N Am. 22, 315-327 (2004).\\n\\nHeat shock-induced matrix metalloproteinase (MMP)-1 and MMP-3 are mediated through ERK and JNK activation and via an autocrine interleukin-6 loop. C H Park, J Invest Dermatol. 123Park, C. H. et al. Heat shock-induced matrix metalloproteinase (MMP)-1 and MMP-3 are mediated through ERK and JNK activation and via an autocrine interleukin-6 loop. J Invest Dermatol. 123, 1012-1019 (2004).\\n\\nSuppression of cytochrome c release and apoptosis in testeswith heat stress by minocycline. S Matsuki, BBRC. 312Matsuki, S. et al. Suppression of cytochrome c release and apoptosis in testeswith heat stress by minocycline. BBRC. 312, 843-849 (2003).\\n\\nHeat stroke in small animal medicine: a clinical practice review. S I Johnson, M Mcmichael, G White, J Vet Emerg Crit Care. 16Johnson, S. I., McMichael, M. & White, G. Heat stroke in small animal medicine: a clinical practice review. J Vet Emerg Crit Care. 16, 112-119 (2006).\\n\\nAcidic environment modifies heat or radiation-induced apoptosis in human maxillary cancer cells. T Ohtsubo, Int J Radiation Oncology BiolPhys. 49Ohtsubo, T. et al. Acidic environment modifies heat or radiation-induced apoptosis in human maxillary cancer cells. Int J Radiation Oncology BiolPhys. 49, 1391-1399 (2001).\\n\\nApoptosis in tumors andnormal tissues induced by whole body hyperthermia in rats. Y Sakaguchi, Cancer Res. 55Sakaguchi, Y. et al. Apoptosis in tumors andnormal tissues induced by whole body hyperthermia in rats. Cancer Res. 55, 5459-5464 (1995).\\n\\nA light and electron microscopic study of thermally in-jured cultured cells. I K Buckley, Lab Invest. 26Buckley, I. K. A light and electron microscopic study of thermally in-jured cultured cells. Lab Invest. 26, 201-209 (1972).\\n\\nHeat stroke. A Bouchama, J P Knochel, N Engl J Med. 346Bouchama, A. & Knochel, J. P. Heat stroke. N Engl J Med. 346, 1978-1988 (2002). www.nature.com/scientificreports\\n\\n. 10.1038/srep04469SCIENTIFIC REPORTS |. 4SCIENTIFIC REPORTS | 4 : 4469 | DOI: 10.1038/srep04469\\n\\nCooling and hemodynamic management in heatstroke: practical recommendations. A Bouchama, M Dehbi, E Chaves-Carballo, Crit Care. 1154Bouchama, A., Dehbi, M. & Chaves-Carballo, E. Cooling and hemodynamic management in heatstroke: practical recommendations. Crit Care. 11, R54 (2007).\\n\\nThe heat wave in France in. D Hemon, E Jougla, Rev Epidemiol Sante. 52Hemon, D. & Jougla, E. The heat wave in France in August 2003. Rev Epidemiol Sante. 52, 3-5 (2004).\\n\\nImpact of regional climate change on human health. J A Patz, Nature. 438Patz, J. A. et al. Impact of regional climate change on human health. Nature. 438, 310-317 (2005).\\n\\nMicrovascular Injury, Thrombosis, Inflammation, andApoptosis in the Pathogenesis of Heatstroke A Study in Baboon Model. G T Roberts, Roberts, G. T. et al. Microvascular Injury, Thrombosis, Inflammation, andApoptosis in the Pathogenesis of Heatstroke A Study in Baboon Model.\\n\\n. Arterioscl Throm Vas, 28Arterioscl Throm Vas. 28, 1130-1136 (2008).\\n\\nEvidence for endothelial cell activation/injury in heatstroke. A Bouchama, Crit Care Med. 24Bouchama, A. et al. Evidence for endothelial cell activation/injury in heatstroke. Crit Care Med. 24, 1173-1178 (1996).\\n\\nThermal sensitivity of endothelial cells on synthetic vascular graft material. M R Brinton, Int J Hyperther. 28Brinton, M. R. Thermal sensitivity of endothelial cells on synthetic vascular graft material. Int J Hyperther. 28, 163-174 (2012).\\n\\np53's mitochondrial translocation and MOMP action isindependent of Puma and Bax and severely disrupts mitochondrial membrane integrity. S Wolff, Cell Res. 18Wolff, S. et al. p53's mitochondrial translocation and MOMP action isindependent of Puma and Bax and severely disrupts mitochondrial membrane integrity. Cell Res. 18, 733-744 (2008).\\n\\nStabilization and translocation of p53 to mitochondria is linked to Bax translocation to mitochondria in simvastatin-induced apoptosis. S K Lee, Biochem Bioph Res Co. 391Lee, S. K. et al. Stabilization and translocation of p53 to mitochondria is linked to Bax translocation to mitochondria in simvastatin-induced apoptosis. Biochem Bioph Res Co. 391, 1592-1597 (2010).\\n\\nDissecting p53-dependent apoptosis. J E Chipuk, D R Green, Cell Death Differ. 13Chipuk, J. E. & Green, D. R. Dissecting p53-dependent apoptosis. Cell Death Differ. 13, 994-1002 (2006).\\n\\nTranscription-independent pro-apoptotic functions of p53. U M Moll, Curr Opin Cell Biol. 17Moll, U. M. et al. Transcription-independent pro-apoptotic functions of p53. Curr Opin Cell Biol. 17, 631-636 (2005).\\n\\nThe mitochondrial p53 pathway. A V Vaseva, U M Moll, BBA-Biomembranes. 1787Vaseva, A. V. & Moll, U. M. The mitochondrial p53 pathway. BBA- Biomembranes. 1787, 414-420 (2009).\\n\\nC-myc down-regulation increases susceptibility to cisplatin through reactive oxygen species-mediated apoptosis in m14 human melanoma cells. A Biroccio, Mol Pharmacol. 60Biroccio, A. et al. C-myc down-regulation increases susceptibility to cisplatin through reactive oxygen species-mediated apoptosis in m14 human melanoma cells. Mol Pharmacol. 60, 174-182 (2001).\\n\\n. B Liu, Y Chen, D K Clair, Free Radic Biol Med. 44Liu, B., Chen, Y. & Clair, D. K. ROS and p53: versatile partnership. Free Radic Biol Med. 44, 1529-1535 (2008).\\n\\nMitochondrial and nuclear p53 localization in cardiomyocytes: Redox modulation by doxorubicin (adriamycin)? Antioxid Redox Sign. R Nithipongvanitch, 9Nithipongvanitch, R. et al. Mitochondrial and nuclear p53 localization in cardiomyocytes: Redox modulation by doxorubicin (adriamycin)? Antioxid Redox Sign. 9, 1001-1008 (2007).\\n\\nRegulation of heat shock-induced apoptosis by sensitive to apoptosis gene protein. S J Lee, Free Radic Biol Med. 45Lee, S. J. et al. Regulation of heat shock-induced apoptosis by sensitive to apoptosis gene protein. Free Radic Biol Med. 45, 167-176 (2008).\\n\\nHeat shock induces apoptosis through reactive oxygen species involving mitochondrial and death receptor pathways in corneal cells. Y L Hsu, Exp Eye Res. 93Hsu, Y. L. et al. Heat shock induces apoptosis through reactive oxygen species involving mitochondrial and death receptor pathways in corneal cells. Exp Eye Res. 93, 405-421 (2011).\\n\\nThe prolyl-isomerase Pin1 activates the mitochondrial death program of p53. G Sorrentino, Cell Death Differ. 20Sorrentino, G. et al. The prolyl-isomerase Pin1 activates the mitochondrial death program of p53. Cell Death Differ. 20, 198-208 (2013).\\n\\nThe Number of Mitochondrial Deoxyribonucleic Acid Genomes in Mouse L and Human HeLa Cells. D Bogenhagen, D A Clayton, J Biol Chem. 249Bogenhagen, D. & Clayton, D. A. The Number of Mitochondrial Deoxyribonucleic Acid Genomes in Mouse L and Human HeLa Cells. J Biol Chem. 249, 7991-7995 (1974).\\n\\nMediates the Apoptotic Response to GTP Depletion after Renal Ischemia-Reperfusion: Protective Role of a p53 Inhibitor. K J Kelly, Z Plotkin, S L Vulgamott, P C Dagher, P53, J Am Soc Nephrol. 14Kelly, K. J., Plotkin, Z., Vulgamott, S. L. & Dagher, P. C. P53 Mediates the Apoptotic Response to GTP Depletion after Renal Ischemia-Reperfusion: Protective Role of a p53 Inhibitor. J Am Soc Nephrol. 14, 128-138 (2003).\\n\\nMild thermotolerance induced at 40uC protects HeLa cells against activation of death receptor-mediated apoptosis by hydrogen peroxide. Free Radical Bio Med. P Pallepati, D A Averill-Bates, 50Pallepati, P. & Averill-Bates, D. A. Mild thermotolerance induced at 40uC protects HeLa cells against activation of death receptor-mediated apoptosis by hydrogen peroxide. Free Radical Bio Med. 50, 667-679 (2011).\\n\\nMechanisms of circulatory and intestinal barrier dysfunction during whole body hyperthermia. D M Hall, Am J Physiol-Heart C. 280Hall, D. M. et al. Mechanisms of circulatory and intestinal barrier dysfunction during whole body hyperthermia. Am J Physiol-Heart C. 280, H509-H521 (2001).\\n\\nExertional heatstroke: clinical characteristics, diagnostic and therapeutic considerations. L Zeller, Eur J Intern Med. 22Zeller, L. et al. Exertional heatstroke: clinical characteristics, diagnostic and therapeutic considerations. Eur J Intern Med. 22, 296-299 (2011).\\n\\nBcl-Xl Overexpression Restricts Heat-Induced Apoptosis and Influences hsp70, bcl-2, and Bax Protein Levels in FL5. J D Robertson, K Datta, J P Kehrer, 12Robertson, J. D., Datta, K. & Kehrer, J. P. Bcl-Xl Overexpression Restricts Heat- Induced Apoptosis and Influences hsp70, bcl-2, and Bax Protein Levels in FL5.12\\n\\n. Cells, Biochem Bioph Res Co. 241Cells. Biochem Bioph Res Co. 241, 164-168 (1997).\\n\\nLong lasting heat shock stimulation of TRAIL-induced apoptosis in transformed T lymphocytes. M Moulin, A P Arrigo, Exp Cell Res. 312Moulin, M. & Arrigo, A. P. Long lasting heat shock stimulation of TRAIL-induced apoptosis in transformed T lymphocytes. Exp Cell Res. 312, 1765-1784 (2006).\\n\\nEvidence of apoptosis in primary neuronal cultures after heat shock. P Vogel, E Dux, C Wiessner, Brain Res. 764Vogel, P., Dux, E. & Wiessner, C. Evidence of apoptosis in primary neuronal cultures after heat shock. Brain Res. 764, 205-213 (1997).\\n\\n. C Michiels, Endothelial Cell Functions. J Cell Physiol. 196Michiels, C. Endothelial Cell Functions. J Cell Physiol. 196, 430-443 (2003).\\n\\nEndothelial cell apoptosis is accelerated by inorganicironandheatvia an oxygen radical dependent mechanism. A K Jacob, Sugery. 122Jacob, A. K. et al. Endothelial cell apoptosis is accelerated by inorganicironandheatvia an oxygen radical dependent mechanism. Sugery. 122, 243-254 (1997).\\n\\nNuclear and mitochondrial apoptotic pathways of p53. U M Moll, A Zaika, FEBS Lett. 493Moll, U. M. & Zaika, A. Nuclear and mitochondrial apoptotic pathways of p53. FEBS Lett. 493, 65-69 (2001).\\n\\nHeat Shock Induces Apoptosis Independently of Any Known Initiator Caspase-activating Complex. R S Milleron, S B Bratton, JBC. 281Milleron, R. S. & Bratton, S. B. Heat Shock Induces Apoptosis Independently of Any Known Initiator Caspase-activating Complex. JBC. 281, 16991-17000 (2006).\\n\\nMechanisms of p53-mediated mitochondrial membrane permeabilization. E Morselli, Cell Res. 18Morselli, E. et al. Mechanisms of p53-mediated mitochondrial membrane permeabilization. Cell Res. 18, 708-710 (2008).\\n\\nCytoplasmic functions of the tumour suppressor p53. D R Green, G Kroemer, Nature. 458Green, D. R. & Kroemer, G. Cytoplasmic functions of the tumour suppressor p53. Nature. 458, 1127-1130 (2009).\\n\\nDeath Signal-induced Localization of p53 Protein to Mitochondria. N D Marchenko, A Zaika, U M Moll, JBC. 275Marchenko, N. D., Zaika, A. & Moll, U. M. Death Signal-induced Localization of p53 Protein to Mitochondria. JBC. 275, 16202-16212 (2000).\\n\\np53 Translocation to Mitochondria Precedes Its Nuclear Translocation and Targets Mitochondrial Oxidative Defense Protein-Manganese Superoxide Dismutase. Y F Zhao, Cancer Res. 65Zhao, Y. F. et al. p53 Translocation to Mitochondria Precedes Its Nuclear Translocation and Targets Mitochondrial Oxidative Defense Protein-Manganese Superoxide Dismutase. Cancer Res. 65, 3745-3750 (2005).\\n\\nIn Vivo Mitochondrial p53 Translocation Triggers a Rapid First Wave of Cell Death in Response to DNA Damage That Can Precede p53 Target Gene Activation. S Erster, Mol Cell Biol. 24Erster, S. et al. In Vivo Mitochondrial p53 Translocation Triggers a Rapid First Wave of Cell Death in Response to DNA Damage That Can Precede p53 Target Gene Activation. Mol Cell Biol. 24, p6728-6741 (2004).\\n\\nA Two-Step Mechanism for Cell Fate Decision byCoordination of Nuclear and Mitochondrial p53 Activities. X J Tian, PLoS ONE. 738164Tian, X. J. et al. A Two-Step Mechanism for Cell Fate Decision byCoordination of Nuclear and Mitochondrial p53 Activities. PLoS ONE. 7, e38164 (2012).\\n\\nDequalinium induces cell death in human leukemia cells by early mitochondrial alterations which enhance ROS production. P Sancho, Leukemia Res. 31Sancho, P. et al. Dequalinium induces cell death in human leukemia cells by early mitochondrial alterations which enhance ROS production. Leukemia Res. 31, 969-978 (2007).\\n\\nReactive oxygen species, cellular redox systems, and apoptosis. Free Radical Bio Med. M L Circu, T Y Aw, 48Circu, M. L. & Aw, T. Y. Reactive oxygen species, cellular redox systems, and apoptosis. Free Radical Bio Med. 48, 749-762 (2010).\\n\\nROS signaling as common element in low oxygen and heat stresses. C Pucciariello, V Banti, P Perata, Plant Physiol Bioch. 59Pucciariello, C., Banti, V. & Perata, P. ROS signaling as common element in low oxygen and heat stresses. Plant Physiol Bioch. 59, 3-10 (2012).\\n\\nOxidative stress and heat shockprotein induction in human cells. R H Burdon, V M Gill, C Rice-Evans, Free Radic Res Commun. 3Burdon, R. H., Gill, V. M. & Rice-Evans, C. Oxidative stress and heat shockprotein induction in human cells. Free Radic Res Commun. 3, 129-139 (1987).\\n\\nOxidative stress as a precursor to the irreversible hepatocellular injury caused byhyperthermia. J L Skibba, Int J yperther. 7Skibba, J. L. et al. Oxidative stress as a precursor to the irreversible hepatocellular injury caused byhyperthermia. Int J yperther. 7, 749-761 (1991).\\n\\nHyperthermia increases exercise-induced oxidative stress. S R Mcanulty, Int J Sports Med. 26McAnulty, S. R. et al. Hyperthermia increases exercise-induced oxidative stress. Int J Sports Med. 26, 188-192 (2005).\\n\\nDeath signal-induced localization of p53 protein to mitochondria. A potential role in apoptotic signaling. N D Marchenko, A Zaika, U M Moll, J Biol Chem. 275Marchenko, N. D., Zaika, A. & Moll, U. M. Death signal-induced localization of p53 protein to mitochondria. A potential role in apoptotic signaling. J Biol Chem. 275, 16202-16212 (2000).\\n\\nROS leads to MnSODupregulation through ERK2 translocation and p53 activation in selenite-induced apoptosis of NB4 cells. Z Li, FEBS. 584Li, Z. et al. ROS leads to MnSODupregulation through ERK2 translocation and p53 activation in selenite-induced apoptosis of NB4 cells. FEBS. 584, 2291-2297 (2010).\\n\\nNG, a novel PABA/NO-based oleanolic acid derivative, induces Human hepatoma cell apoptosis via a ROS/MAPK-dependent mitochondrial pathway. L Liu, Eur J Pharmacol. 691Liu, L. et al. NG, a novel PABA/NO-based oleanolic acid derivative, induces Human hepatoma cell apoptosis via a ROS/MAPK-dependent mitochondrial pathway. Eur J Pharmacol. 691, 61-68 (2012).\\n\\nCorosolic acid induces apoptosis through mitochondrial pathway and caspases activation in human cervix adenocarcinoma HeLa cells. Y F Xu, Cancer Lett. 284Xu, Y. F. et al. Corosolic acid induces apoptosis through mitochondrial pathway and caspases activation in human cervix adenocarcinoma HeLa cells. Cancer Lett. 284, 229-237 (2009).\\n\\nMild thermotolerance induced at 40uC increases antioxidants and protectsHeLa cells against mitochondrial apoptosis induced by hydrogen peroxide:Role of p53. P Pallepati, D A Averill-Bates, Arch Biochem Biophys. 495Pallepati, P. & Averill-Bates, D. A. Mild thermotolerance induced at 40uC increases antioxidants and protectsHeLa cells against mitochondrial apoptosis induced by hydrogen peroxide:Role of p53. Arch Biochem Biophys. 495, 97-111 (2010).\\n\\nThermotolerance induced at a fever temperature of 40 C protects cells against hyperthermia-induced apoptosis mediated by death receptor signalling. A Bettaieb, D A Averill-Bates, Cell Biol. 86Bettaieb, A. & Averill-Bates, D. A. Thermotolerance induced at a fever temperature of 40 C protects cells against hyperthermia-induced apoptosis mediated by death receptor signalling. Cell Biol. 86, 521-538 (2008).\\n\\np53 Acetylation Is Crucial for Its Transcription-independent ProapoptoticFunctions. H Yamaguchi, JBC. 284Yamaguchi, H. et al. p53 Acetylation Is Crucial for Its Transcription-independent ProapoptoticFunctions. JBC. 284, 11171-11183 (2009).\\n\\nHyperglycemia induces apoptosis and p53 mobilization to mitochondria in RINm5F cells. C Ortega-Camarillo, Mol Cell Biochem. 281Ortega-Camarillo, C. et al. Hyperglycemia induces apoptosis and p53 mobilization to mitochondria in RINm5F cells. Mol Cell Biochem. 281, 163-171 (2006).\\n\""
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "full_doc_example = full_text['text'][9]\n",
    "abstract_doc_example = abstract_text['abstract'][9]\n",
    "full_doc_example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "from getpass import getpass\n",
    "from langchain import HuggingFaceHub\n",
    "from langchain.embeddings import HuggingFaceEmbeddings\n",
    "# \"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\"\n",
    "HUGGINGFACEHUB_API_TOKEN = getpass()\n",
    "os.environ[\"HUGGINGFACEHUB_API_TOKEN\"] = \"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\"\n",
    "\n",
    "\n",
    "repo_id = \"allenai/led-large-16384\"  \n",
    "repo_id2 = \"google/pegasus-large\"\n",
    "\n",
    "llm = HuggingFaceHub(repo_id=repo_id2, model_kwargs={\"temperature\": 0.5, \"max_length\": 3000})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-05 20:50:27 ip-172-31-90-198 sentence_transformers.SentenceTransformer[1972] INFO Load pretrained SentenceTransformer: google/pegasus-large\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f034f37ce7b44b5ca8640f31e3b0d200",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)b6b77/.gitattributes:   0%|          | 0.00/391 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b3830086b5e1410b87c3c875c3bf4e18",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)8ed57b6b77/README.md:   0%|          | 0.00/3.33k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "112d867b9e0044d79449a86accae1df0",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)d57b6b77/config.json:   0%|          | 0.00/3.09k [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "36dec2c57337415ca8328cfe0f40d981",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)neration_config.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ce10c576a5214434b2ad15c570cfc904",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)arization_aeslc.json:   0%|          | 0.00/259 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7dad753a3ac6486ab50e475aa176a93a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)arization_arxiv.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "1b6a8f13725e48458a206c69c9023f0e",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)tion_big_patent.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0b728c85caec4eec8472e3521525bd25",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)ization_billsum.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8ff77b1259e340d099300ea806bc36c6",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)n_cnn_dailymail.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a442d25d2804424b993f35d68c923813",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)zation_gigaword.json:   0%|          | 0.00/259 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "fef51051fa2d4af9921a147dfe44e2da",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)arization_large.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8bdeda9095ce445887993faad784a86b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)tion_multi_news.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dd9ba81c58724cdcaf3fd0578a00b806",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)zation_newsroom.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3fc4e0650a9d466f8add68a9702fa944",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)rization_pubmed.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "a89b3047017d40f7ae8b81244666efd7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)ion_reddit_tifu.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "7b758fc957534d479488654ff562b2a1",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)ization_wikihow.json:   0%|          | 0.00/260 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "67368f5458c443eb83d94f88e954c4eb",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)marization_xsum.json:   0%|          | 0.00/259 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ec48058a00824463a48b85dd86ea0bb5",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading pytorch_model.bin:   0%|          | 0.00/2.28G [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e166d7f7ec7f485eb8a995976aa771a7",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)cial_tokens_map.json:   0%|          | 0.00/65.0 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "b5015d83aabe45e298ae6c4ab6f56a8d",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)57b6b77/spiece.model:   0%|          | 0.00/1.91M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "979d2917a2ad4311a80e3cbf46823399",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)okenizer_config.json:   0%|          | 0.00/88.0 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "2023-07-05 20:51:30 ip-172-31-90-198 sentence_transformers.SentenceTransformer[1972] WARNING No sentence-transformers model found with name /home/ubuntu/.cache/torch/sentence_transformers/google_pegasus-large. Creating a new one with MEAN pooling.\n",
      "Some weights of the model checkpoint at /home/ubuntu/.cache/torch/sentence_transformers/google_pegasus-large were not used when initializing PegasusModel: ['final_logits_bias']\n",
      "- This IS expected if you are initializing PegasusModel from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).\n",
      "- This IS NOT expected if you are initializing PegasusModel from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).\n",
      "Some weights of PegasusModel were not initialized from the model checkpoint at /home/ubuntu/.cache/torch/sentence_transformers/google_pegasus-large and are newly initialized: ['model.decoder.embed_positions.weight', 'model.encoder.embed_positions.weight']\n",
      "You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "Descriptors cannot not be created directly.\nIf this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.\nIf you cannot immediately regenerate your protos, some other possible workarounds are:\n 1. Downgrade the protobuf package to 3.20.x or lower.\n 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).\n\nMore information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m embeddings \u001b[39m=\u001b[39m HuggingFaceEmbeddings(model_name\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mgoogle/pegasus-large\u001b[39;49m\u001b[39m\"\u001b[39;49m )\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/langchain/embeddings/huggingface.py:59\u001b[0m, in \u001b[0;36mHuggingFaceEmbeddings.__init__\u001b[0;34m(self, **kwargs)\u001b[0m\n\u001b[1;32m     53\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mImportError\u001b[39;00m \u001b[39mas\u001b[39;00m exc:\n\u001b[1;32m     54\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mImportError\u001b[39;00m(\n\u001b[1;32m     55\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mCould not import sentence_transformers python package. \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     56\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mPlease install it with `pip install sentence_transformers`.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     57\u001b[0m     ) \u001b[39mfrom\u001b[39;00m \u001b[39mexc\u001b[39;00m\n\u001b[0;32m---> 59\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mclient \u001b[39m=\u001b[39m sentence_transformers\u001b[39m.\u001b[39;49mSentenceTransformer(\n\u001b[1;32m     60\u001b[0m     \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmodel_name, cache_folder\u001b[39m=\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mcache_folder, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49m\u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49mmodel_kwargs\n\u001b[1;32m     61\u001b[0m )\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sentence_transformers/SentenceTransformer.py:97\u001b[0m, in \u001b[0;36mSentenceTransformer.__init__\u001b[0;34m(self, model_name_or_path, modules, device, cache_folder, use_auth_token)\u001b[0m\n\u001b[1;32m     95\u001b[0m         modules \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_load_sbert_model(model_path)\n\u001b[1;32m     96\u001b[0m     \u001b[39melse\u001b[39;00m:   \u001b[39m#Load with AutoModel\u001b[39;00m\n\u001b[0;32m---> 97\u001b[0m         modules \u001b[39m=\u001b[39m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_load_auto_model(model_path)\n\u001b[1;32m     99\u001b[0m \u001b[39mif\u001b[39;00m modules \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39mand\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(modules, OrderedDict):\n\u001b[1;32m    100\u001b[0m     modules \u001b[39m=\u001b[39m OrderedDict([(\u001b[39mstr\u001b[39m(idx), module) \u001b[39mfor\u001b[39;00m idx, module \u001b[39min\u001b[39;00m \u001b[39menumerate\u001b[39m(modules)])\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sentence_transformers/SentenceTransformer.py:806\u001b[0m, in \u001b[0;36mSentenceTransformer._load_auto_model\u001b[0;34m(self, model_name_or_path)\u001b[0m\n\u001b[1;32m    802\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    803\u001b[0m \u001b[39mCreates a simple Transformer + Mean Pooling model and returns the modules\u001b[39;00m\n\u001b[1;32m    804\u001b[0m \u001b[39m\"\"\"\u001b[39;00m\n\u001b[1;32m    805\u001b[0m logger\u001b[39m.\u001b[39mwarning(\u001b[39m\"\u001b[39m\u001b[39mNo sentence-transformers model found with name \u001b[39m\u001b[39m{}\u001b[39;00m\u001b[39m. Creating a new one with MEAN pooling.\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m.\u001b[39mformat(model_name_or_path))\n\u001b[0;32m--> 806\u001b[0m transformer_model \u001b[39m=\u001b[39m Transformer(model_name_or_path)\n\u001b[1;32m    807\u001b[0m pooling_model \u001b[39m=\u001b[39m Pooling(transformer_model\u001b[39m.\u001b[39mget_word_embedding_dimension(), \u001b[39m'\u001b[39m\u001b[39mmean\u001b[39m\u001b[39m'\u001b[39m)\n\u001b[1;32m    808\u001b[0m \u001b[39mreturn\u001b[39;00m [transformer_model, pooling_model]\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/sentence_transformers/models/Transformer.py:31\u001b[0m, in \u001b[0;36mTransformer.__init__\u001b[0;34m(self, model_name_or_path, max_seq_length, model_args, cache_dir, tokenizer_args, do_lower_case, tokenizer_name_or_path)\u001b[0m\n\u001b[1;32m     28\u001b[0m config \u001b[39m=\u001b[39m AutoConfig\u001b[39m.\u001b[39mfrom_pretrained(model_name_or_path, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mmodel_args, cache_dir\u001b[39m=\u001b[39mcache_dir)\n\u001b[1;32m     29\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_load_model(model_name_or_path, config, cache_dir)\n\u001b[0;32m---> 31\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mtokenizer \u001b[39m=\u001b[39m AutoTokenizer\u001b[39m.\u001b[39;49mfrom_pretrained(tokenizer_name_or_path \u001b[39mif\u001b[39;49;00m tokenizer_name_or_path \u001b[39mis\u001b[39;49;00m \u001b[39mnot\u001b[39;49;00m \u001b[39mNone\u001b[39;49;00m \u001b[39melse\u001b[39;49;00m model_name_or_path, cache_dir\u001b[39m=\u001b[39;49mcache_dir, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mtokenizer_args)\n\u001b[1;32m     33\u001b[0m \u001b[39m#No max_seq_length set. Try to infer from model\u001b[39;00m\n\u001b[1;32m     34\u001b[0m \u001b[39mif\u001b[39;00m max_seq_length \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/transformers/models/auto/tokenization_auto.py:711\u001b[0m, in \u001b[0;36mAutoTokenizer.from_pretrained\u001b[0;34m(cls, pretrained_model_name_or_path, *inputs, **kwargs)\u001b[0m\n\u001b[1;32m    709\u001b[0m tokenizer_class_py, tokenizer_class_fast \u001b[39m=\u001b[39m TOKENIZER_MAPPING[\u001b[39mtype\u001b[39m(config)]\n\u001b[1;32m    710\u001b[0m \u001b[39mif\u001b[39;00m tokenizer_class_fast \u001b[39mand\u001b[39;00m (use_fast \u001b[39mor\u001b[39;00m tokenizer_class_py \u001b[39mis\u001b[39;00m \u001b[39mNone\u001b[39;00m):\n\u001b[0;32m--> 711\u001b[0m     \u001b[39mreturn\u001b[39;00m tokenizer_class_fast\u001b[39m.\u001b[39;49mfrom_pretrained(pretrained_model_name_or_path, \u001b[39m*\u001b[39;49minputs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs)\n\u001b[1;32m    712\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[1;32m    713\u001b[0m     \u001b[39mif\u001b[39;00m tokenizer_class_py \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/transformers/tokenization_utils_base.py:1812\u001b[0m, in \u001b[0;36mPreTrainedTokenizerBase.from_pretrained\u001b[0;34m(cls, pretrained_model_name_or_path, *init_inputs, **kwargs)\u001b[0m\n\u001b[1;32m   1809\u001b[0m     \u001b[39melse\u001b[39;00m:\n\u001b[1;32m   1810\u001b[0m         logger\u001b[39m.\u001b[39minfo(\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mloading file \u001b[39m\u001b[39m{\u001b[39;00mfile_path\u001b[39m}\u001b[39;00m\u001b[39m from cache at \u001b[39m\u001b[39m{\u001b[39;00mresolved_vocab_files[file_id]\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m)\n\u001b[0;32m-> 1812\u001b[0m \u001b[39mreturn\u001b[39;00m \u001b[39mcls\u001b[39;49m\u001b[39m.\u001b[39;49m_from_pretrained(\n\u001b[1;32m   1813\u001b[0m     resolved_vocab_files,\n\u001b[1;32m   1814\u001b[0m     pretrained_model_name_or_path,\n\u001b[1;32m   1815\u001b[0m     init_configuration,\n\u001b[1;32m   1816\u001b[0m     \u001b[39m*\u001b[39;49minit_inputs,\n\u001b[1;32m   1817\u001b[0m     use_auth_token\u001b[39m=\u001b[39;49muse_auth_token,\n\u001b[1;32m   1818\u001b[0m     cache_dir\u001b[39m=\u001b[39;49mcache_dir,\n\u001b[1;32m   1819\u001b[0m     local_files_only\u001b[39m=\u001b[39;49mlocal_files_only,\n\u001b[1;32m   1820\u001b[0m     _commit_hash\u001b[39m=\u001b[39;49mcommit_hash,\n\u001b[1;32m   1821\u001b[0m     _is_local\u001b[39m=\u001b[39;49mis_local,\n\u001b[1;32m   1822\u001b[0m     \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs,\n\u001b[1;32m   1823\u001b[0m )\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/transformers/tokenization_utils_base.py:1975\u001b[0m, in \u001b[0;36mPreTrainedTokenizerBase._from_pretrained\u001b[0;34m(cls, resolved_vocab_files, pretrained_model_name_or_path, init_configuration, use_auth_token, cache_dir, local_files_only, _commit_hash, _is_local, *init_inputs, **kwargs)\u001b[0m\n\u001b[1;32m   1973\u001b[0m \u001b[39m# Instantiate tokenizer.\u001b[39;00m\n\u001b[1;32m   1974\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m-> 1975\u001b[0m     tokenizer \u001b[39m=\u001b[39m \u001b[39mcls\u001b[39;49m(\u001b[39m*\u001b[39;49minit_inputs, \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49minit_kwargs)\n\u001b[1;32m   1976\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mOSError\u001b[39;00m:\n\u001b[1;32m   1977\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mOSError\u001b[39;00m(\n\u001b[1;32m   1978\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mUnable to load vocabulary from file. \u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1979\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39mPlease check that the provided vocabulary is accessible and not corrupted.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1980\u001b[0m     )\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/transformers/models/pegasus/tokenization_pegasus_fast.py:142\u001b[0m, in \u001b[0;36mPegasusTokenizerFast.__init__\u001b[0;34m(self, vocab_file, tokenizer_file, pad_token, eos_token, unk_token, mask_token, mask_token_sent, additional_special_tokens, offset, **kwargs)\u001b[0m\n\u001b[1;32m    139\u001b[0m     additional_special_tokens \u001b[39m=\u001b[39m [mask_token_sent] \u001b[39mif\u001b[39;00m mask_token_sent \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m \u001b[39melse\u001b[39;00m []\n\u001b[1;32m    140\u001b[0m     additional_special_tokens \u001b[39m+\u001b[39m\u001b[39m=\u001b[39m [\u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m<unk_\u001b[39m\u001b[39m{\u001b[39;00mi\u001b[39m}\u001b[39;00m\u001b[39m>\u001b[39m\u001b[39m\"\u001b[39m \u001b[39mfor\u001b[39;00m i \u001b[39min\u001b[39;00m \u001b[39mrange\u001b[39m(\u001b[39m2\u001b[39m, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39moffset)]\n\u001b[0;32m--> 142\u001b[0m \u001b[39msuper\u001b[39;49m()\u001b[39m.\u001b[39;49m\u001b[39m__init__\u001b[39;49m(\n\u001b[1;32m    143\u001b[0m     vocab_file,\n\u001b[1;32m    144\u001b[0m     tokenizer_file\u001b[39m=\u001b[39;49mtokenizer_file,\n\u001b[1;32m    145\u001b[0m     pad_token\u001b[39m=\u001b[39;49mpad_token,\n\u001b[1;32m    146\u001b[0m     eos_token\u001b[39m=\u001b[39;49meos_token,\n\u001b[1;32m    147\u001b[0m     unk_token\u001b[39m=\u001b[39;49munk_token,\n\u001b[1;32m    148\u001b[0m     mask_token\u001b[39m=\u001b[39;49mmask_token,\n\u001b[1;32m    149\u001b[0m     mask_token_sent\u001b[39m=\u001b[39;49mmask_token_sent,\n\u001b[1;32m    150\u001b[0m     offset\u001b[39m=\u001b[39;49moffset,\n\u001b[1;32m    151\u001b[0m     additional_special_tokens\u001b[39m=\u001b[39;49madditional_special_tokens,\n\u001b[1;32m    152\u001b[0m     \u001b[39m*\u001b[39;49m\u001b[39m*\u001b[39;49mkwargs,\n\u001b[1;32m    153\u001b[0m )\n\u001b[1;32m    154\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mvocab_file \u001b[39m=\u001b[39m vocab_file\n\u001b[1;32m    155\u001b[0m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mcan_save_slow_tokenizer \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mvocab_file \u001b[39melse\u001b[39;00m \u001b[39mTrue\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/transformers/tokenization_utils_fast.py:114\u001b[0m, in \u001b[0;36mPreTrainedTokenizerFast.__init__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m    111\u001b[0m     fast_tokenizer \u001b[39m=\u001b[39m TokenizerFast\u001b[39m.\u001b[39mfrom_file(fast_tokenizer_file)\n\u001b[1;32m    112\u001b[0m \u001b[39melif\u001b[39;00m slow_tokenizer \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    113\u001b[0m     \u001b[39m# We need to convert a slow tokenizer to build the backend\u001b[39;00m\n\u001b[0;32m--> 114\u001b[0m     fast_tokenizer \u001b[39m=\u001b[39m convert_slow_tokenizer(slow_tokenizer)\n\u001b[1;32m    115\u001b[0m \u001b[39melif\u001b[39;00m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mslow_tokenizer_class \u001b[39mis\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39mNone\u001b[39;00m:\n\u001b[1;32m    116\u001b[0m     \u001b[39m# We need to create and convert a slow tokenizer to build the backend\u001b[39;00m\n\u001b[1;32m    117\u001b[0m     slow_tokenizer \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39mslow_tokenizer_class(\u001b[39m*\u001b[39margs, \u001b[39m*\u001b[39m\u001b[39m*\u001b[39mkwargs)\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/transformers/convert_slow_tokenizer.py:1303\u001b[0m, in \u001b[0;36mconvert_slow_tokenizer\u001b[0;34m(transformer_tokenizer)\u001b[0m\n\u001b[1;32m   1295\u001b[0m     \u001b[39mraise\u001b[39;00m \u001b[39mValueError\u001b[39;00m(\n\u001b[1;32m   1296\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mAn instance of tokenizer class \u001b[39m\u001b[39m{\u001b[39;00mtokenizer_class_name\u001b[39m}\u001b[39;00m\u001b[39m cannot be converted in a Fast tokenizer instance.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1297\u001b[0m         \u001b[39m\"\u001b[39m\u001b[39m No converter was found. Currently available slow->fast convertors:\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1298\u001b[0m         \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m \u001b[39m\u001b[39m{\u001b[39;00m\u001b[39mlist\u001b[39m(SLOW_TO_FAST_CONVERTERS\u001b[39m.\u001b[39mkeys())\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m   1299\u001b[0m     )\n\u001b[1;32m   1301\u001b[0m converter_class \u001b[39m=\u001b[39m SLOW_TO_FAST_CONVERTERS[tokenizer_class_name]\n\u001b[0;32m-> 1303\u001b[0m \u001b[39mreturn\u001b[39;00m converter_class(transformer_tokenizer)\u001b[39m.\u001b[39mconverted()\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/transformers/convert_slow_tokenizer.py:445\u001b[0m, in \u001b[0;36mSpmConverter.__init__\u001b[0;34m(self, *args)\u001b[0m\n\u001b[1;32m    441\u001b[0m requires_backends(\u001b[39mself\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mprotobuf\u001b[39m\u001b[39m\"\u001b[39m)\n\u001b[1;32m    443\u001b[0m \u001b[39msuper\u001b[39m()\u001b[39m.\u001b[39m\u001b[39m__init__\u001b[39m(\u001b[39m*\u001b[39margs)\n\u001b[0;32m--> 445\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m \u001b[39mimport\u001b[39;00m sentencepiece_model_pb2 \u001b[39mas\u001b[39;00m model_pb2\n\u001b[1;32m    447\u001b[0m m \u001b[39m=\u001b[39m model_pb2\u001b[39m.\u001b[39mModelProto()\n\u001b[1;32m    448\u001b[0m \u001b[39mwith\u001b[39;00m \u001b[39mopen\u001b[39m(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39moriginal_tokenizer\u001b[39m.\u001b[39mvocab_file, \u001b[39m\"\u001b[39m\u001b[39mrb\u001b[39m\u001b[39m\"\u001b[39m) \u001b[39mas\u001b[39;00m f:\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/transformers/utils/sentencepiece_model_pb2.py:91\u001b[0m\n\u001b[1;32m     25\u001b[0m _sym_db \u001b[39m=\u001b[39m _symbol_database\u001b[39m.\u001b[39mDefault()\n\u001b[1;32m     28\u001b[0m DESCRIPTOR \u001b[39m=\u001b[39m _descriptor\u001b[39m.\u001b[39mFileDescriptor(\n\u001b[1;32m     29\u001b[0m     name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39msentencepiece_model.proto\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     30\u001b[0m     package\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39msentencepiece\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     80\u001b[0m     ),\n\u001b[1;32m     81\u001b[0m )\n\u001b[1;32m     84\u001b[0m _TRAINERSPEC_MODELTYPE \u001b[39m=\u001b[39m _descriptor\u001b[39m.\u001b[39mEnumDescriptor(\n\u001b[1;32m     85\u001b[0m     name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mModelType\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     86\u001b[0m     full_name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39msentencepiece.TrainerSpec.ModelType\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     87\u001b[0m     filename\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m     88\u001b[0m     file\u001b[39m=\u001b[39mDESCRIPTOR,\n\u001b[1;32m     89\u001b[0m     create_key\u001b[39m=\u001b[39m_descriptor\u001b[39m.\u001b[39m_internal_create_key,\n\u001b[1;32m     90\u001b[0m     values\u001b[39m=\u001b[39m[\n\u001b[0;32m---> 91\u001b[0m         _descriptor\u001b[39m.\u001b[39;49mEnumValueDescriptor(\n\u001b[1;32m     92\u001b[0m             name\u001b[39m=\u001b[39;49m\u001b[39m\"\u001b[39;49m\u001b[39mUNIGRAM\u001b[39;49m\u001b[39m\"\u001b[39;49m,\n\u001b[1;32m     93\u001b[0m             index\u001b[39m=\u001b[39;49m\u001b[39m0\u001b[39;49m,\n\u001b[1;32m     94\u001b[0m             number\u001b[39m=\u001b[39;49m\u001b[39m1\u001b[39;49m,\n\u001b[1;32m     95\u001b[0m             serialized_options\u001b[39m=\u001b[39;49m\u001b[39mNone\u001b[39;49;00m,\n\u001b[1;32m     96\u001b[0m             \u001b[39mtype\u001b[39;49m\u001b[39m=\u001b[39;49m\u001b[39mNone\u001b[39;49;00m,\n\u001b[1;32m     97\u001b[0m             create_key\u001b[39m=\u001b[39;49m_descriptor\u001b[39m.\u001b[39;49m_internal_create_key,\n\u001b[1;32m     98\u001b[0m         ),\n\u001b[1;32m     99\u001b[0m         _descriptor\u001b[39m.\u001b[39mEnumValueDescriptor(\n\u001b[1;32m    100\u001b[0m             name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mBPE\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    101\u001b[0m             index\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m,\n\u001b[1;32m    102\u001b[0m             number\u001b[39m=\u001b[39m\u001b[39m2\u001b[39m,\n\u001b[1;32m    103\u001b[0m             serialized_options\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    104\u001b[0m             \u001b[39mtype\u001b[39m\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    105\u001b[0m             create_key\u001b[39m=\u001b[39m_descriptor\u001b[39m.\u001b[39m_internal_create_key,\n\u001b[1;32m    106\u001b[0m         ),\n\u001b[1;32m    107\u001b[0m         _descriptor\u001b[39m.\u001b[39mEnumValueDescriptor(\n\u001b[1;32m    108\u001b[0m             name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mWORD\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    109\u001b[0m             index\u001b[39m=\u001b[39m\u001b[39m2\u001b[39m,\n\u001b[1;32m    110\u001b[0m             number\u001b[39m=\u001b[39m\u001b[39m3\u001b[39m,\n\u001b[1;32m    111\u001b[0m             serialized_options\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    112\u001b[0m             \u001b[39mtype\u001b[39m\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    113\u001b[0m             create_key\u001b[39m=\u001b[39m_descriptor\u001b[39m.\u001b[39m_internal_create_key,\n\u001b[1;32m    114\u001b[0m         ),\n\u001b[1;32m    115\u001b[0m         _descriptor\u001b[39m.\u001b[39mEnumValueDescriptor(\n\u001b[1;32m    116\u001b[0m             name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mCHAR\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    117\u001b[0m             index\u001b[39m=\u001b[39m\u001b[39m3\u001b[39m,\n\u001b[1;32m    118\u001b[0m             number\u001b[39m=\u001b[39m\u001b[39m4\u001b[39m,\n\u001b[1;32m    119\u001b[0m             serialized_options\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    120\u001b[0m             \u001b[39mtype\u001b[39m\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    121\u001b[0m             create_key\u001b[39m=\u001b[39m_descriptor\u001b[39m.\u001b[39m_internal_create_key,\n\u001b[1;32m    122\u001b[0m         ),\n\u001b[1;32m    123\u001b[0m     ],\n\u001b[1;32m    124\u001b[0m     containing_type\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    125\u001b[0m     serialized_options\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,\n\u001b[1;32m    126\u001b[0m     serialized_start\u001b[39m=\u001b[39m\u001b[39m1294\u001b[39m,\n\u001b[1;32m    127\u001b[0m     serialized_end\u001b[39m=\u001b[39m\u001b[39m1347\u001b[39m,\n\u001b[1;32m    128\u001b[0m )\n\u001b[1;32m    129\u001b[0m _sym_db\u001b[39m.\u001b[39mRegisterEnumDescriptor(_TRAINERSPEC_MODELTYPE)\n\u001b[1;32m    131\u001b[0m _MODELPROTO_SENTENCEPIECE_TYPE \u001b[39m=\u001b[39m _descriptor\u001b[39m.\u001b[39mEnumDescriptor(\n\u001b[1;32m    132\u001b[0m     name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39mType\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m    133\u001b[0m     full_name\u001b[39m=\u001b[39m\u001b[39m\"\u001b[39m\u001b[39msentencepiece.ModelProto.SentencePiece.Type\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    190\u001b[0m     serialized_end\u001b[39m=\u001b[39m\u001b[39m2184\u001b[39m,\n\u001b[1;32m    191\u001b[0m )\n",
      "File \u001b[0;32m~/anaconda3/envs/ta/lib/python3.10/site-packages/google/protobuf/descriptor.py:796\u001b[0m, in \u001b[0;36mEnumValueDescriptor.__new__\u001b[0;34m(cls, name, index, number, type, options, serialized_options, create_key)\u001b[0m\n\u001b[1;32m    793\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__new__\u001b[39m(\u001b[39mcls\u001b[39m, name, index, number,\n\u001b[1;32m    794\u001b[0m             \u001b[39mtype\u001b[39m\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m,  \u001b[39m# pylint: disable=redefined-builtin\u001b[39;00m\n\u001b[1;32m    795\u001b[0m             options\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, serialized_options\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m, create_key\u001b[39m=\u001b[39m\u001b[39mNone\u001b[39;00m):\n\u001b[0;32m--> 796\u001b[0m   _message\u001b[39m.\u001b[39;49mMessage\u001b[39m.\u001b[39;49m_CheckCalledFromGeneratedFile()\n\u001b[1;32m    797\u001b[0m   \u001b[39m# There is no way we can build a complete EnumValueDescriptor with the\u001b[39;00m\n\u001b[1;32m    798\u001b[0m   \u001b[39m# given parameters (the name of the Enum is not known, for example).\u001b[39;00m\n\u001b[1;32m    799\u001b[0m   \u001b[39m# Fortunately generated files just pass it to the EnumDescriptor()\u001b[39;00m\n\u001b[1;32m    800\u001b[0m   \u001b[39m# constructor, which will ignore it, so returning None is good enough.\u001b[39;00m\n\u001b[1;32m    801\u001b[0m   \u001b[39mreturn\u001b[39;00m \u001b[39mNone\u001b[39;00m\n",
      "\u001b[0;31mTypeError\u001b[0m: Descriptors cannot not be created directly.\nIf this call came from a _pb2.py file, your generated code is out of date and must be regenerated with protoc >= 3.19.0.\nIf you cannot immediately regenerate your protos, some other possible workarounds are:\n 1. Downgrade the protobuf package to 3.20.x or lower.\n 2. Set PROTOCOL_BUFFERS_PYTHON_IMPLEMENTATION=python (but this will use pure-Python parsing and will be much slower).\n\nMore information: https://developers.google.com/protocol-buffers/docs/news/2022-05-06#python-updates"
     ]
    }
   ],
   "source": [
    "embeddings = HuggingFaceEmbeddings(model_name=\"google/pegasus-large\" )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "text_splitter = CharacterTextSplitter()\n",
    "text = text_splitter.split_text(full_doc_example)\n",
    "docs = [Document(page_content=t) for t in text[:3]]\n",
    "\n",
    "chain = load_summarize_chain(llm, chain_type=\"map_reduce\")\n",
    "# chain.run(docs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "4d82c2d3f8194c179cf63e8972d58f1a",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Batches:   0%|          | 0/1 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCannot execute code, session has been disposed. Please try restarting the Kernel."
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "query_documents = embeddings.embed_documents([full_doc_example])\n",
    "print(query_documents)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ta",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.11"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
