{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "21658a25",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "import requests\n",
    "from pandas import json_normalize\n",
    "import numpy as np\n",
    "from semanticscholar import SemanticScholar\n",
    "sch = SemanticScholar()\n",
    "import csv\n",
    "import time\n",
    "import heapq\n",
    "import numpy as np\n",
    "from sklearn.metrics.pairwise import cosine_similarity\n",
    "from scipy.spatial.distance import cdist\n",
    "import time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a9d841a",
   "metadata": {},
   "source": [
    "### Import Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "df2b4e74",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('final.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b8255d36",
   "metadata": {},
   "source": [
    "### Pull Spector 1 Embeddings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ae57b189",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "490\n",
      "980\n",
      "1470\n",
      "1960\n",
      "2450\n",
      "2940\n",
      "3430\n",
      "3920\n",
      "4410\n",
      "4900\n",
      "5390\n",
      "5880\n",
      "6370\n",
      "6860\n",
      "7350\n",
      "7840\n",
      "8330\n",
      "8820\n",
      "9310\n",
      "9800\n",
      "10290\n",
      "10780\n",
      "11270\n",
      "11760\n",
      "12250\n",
      "12740\n",
      "13230\n",
      "13720\n",
      "14210\n",
      "14700\n",
      "15190\n",
      "15680\n",
      "16170\n",
      "16660\n",
      "17150\n",
      "17640\n",
      "18130\n",
      "18620\n",
      "19110\n",
      "19600\n",
      "20090\n",
      "20580\n",
      "21070\n",
      "21560\n",
      "22050\n",
      "22540\n",
      "23030\n",
      "23520\n",
      "24010\n",
      "24500\n",
      "24990\n",
      "25480\n",
      "25970\n",
      "26460\n",
      "26950\n",
      "27440\n",
      "27930\n",
      "28420\n",
      "28910\n",
      "29400\n",
      "29890\n",
      "30380\n",
      "30870\n",
      "31360\n",
      "31850\n",
      "32340\n",
      "32830\n",
      "33320\n",
      "33810\n",
      "34300\n",
      "34790\n",
      "35280\n",
      "35770\n",
      "36260\n",
      "36750\n",
      "37240\n",
      "37730\n",
      "38220\n",
      "38710\n",
      "39200\n",
      "39690\n",
      "40180\n",
      "40670\n",
      "41160\n",
      "41650\n",
      "42140\n",
      "42630\n",
      "43120\n",
      "43610\n",
      "44100\n",
      "44590\n",
      "45080\n",
      "45570\n",
      "46060\n",
      "46550\n",
      "47040\n",
      "47530\n",
      "48020\n",
      "48510\n",
      "49000\n",
      "49490\n",
      "49980\n",
      "50470\n",
      "50960\n",
      "51450\n",
      "51940\n",
      "52430\n",
      "52920\n",
      "53410\n",
      "53900\n",
      "54390\n",
      "54880\n",
      "55370\n",
      "55860\n",
      "56350\n",
      "56840\n",
      "57330\n",
      "57820\n",
      "58310\n",
      "58800\n",
      "59290\n",
      "59780\n",
      "60270\n",
      "60760\n",
      "61250\n",
      "61740\n",
      "62230\n",
      "62720\n",
      "63210\n",
      "63700\n",
      "64190\n",
      "64680\n",
      "65170\n",
      "65660\n",
      "66150\n",
      "66640\n",
      "67130\n",
      "67620\n",
      "68110\n",
      "68600\n",
      "69090\n",
      "69580\n",
      "70070\n",
      "70560\n",
      "71050\n",
      "71540\n",
      "72030\n",
      "72520\n",
      "73010\n",
      "73500\n",
      "73990\n",
      "74480\n",
      "74970\n",
      "75460\n",
      "75950\n",
      "76440\n",
      "76930\n",
      "77420\n",
      "77910\n",
      "78400\n",
      "78890\n",
      "79380\n",
      "79870\n",
      "80360\n",
      "80850\n",
      "81340\n",
      "81830\n",
      "82320\n",
      "82810\n",
      "83300\n",
      "83790\n",
      "84280\n",
      "84770\n",
      "85260\n",
      "85750\n",
      "86240\n",
      "86730\n",
      "87220\n",
      "87710\n",
      "88200\n",
      "88690\n",
      "89180\n",
      "89670\n",
      "90160\n",
      "90650\n",
      "91140\n",
      "91630\n",
      "92120\n",
      "92610\n",
      "93100\n",
      "93590\n",
      "94080\n",
      "94570\n",
      "95060\n",
      "95550\n",
      "96040\n",
      "96530\n",
      "97020\n",
      "97510\n",
      "98000\n",
      "98490\n",
      "98980\n",
      "99470\n",
      "99960\n",
      "100450\n",
      "100940\n",
      "101430\n",
      "101920\n",
      "102410\n",
      "102900\n",
      "103390\n",
      "103880\n",
      "104370\n",
      "104860\n",
      "105350\n",
      "105840\n",
      "106330\n",
      "106820\n",
      "107310\n",
      "107800\n",
      "108290\n",
      "108780\n",
      "109270\n",
      "109760\n",
      "110250\n",
      "110740\n",
      "111230\n",
      "111720\n",
      "112210\n",
      "112700\n",
      "113190\n",
      "113680\n",
      "114170\n",
      "114660\n",
      "115150\n",
      "115640\n",
      "116130\n",
      "116620\n",
      "117110\n",
      "117600\n",
      "118090\n",
      "118580\n",
      "119070\n",
      "119560\n",
      "120050\n",
      "120540\n",
      "121030\n",
      "121520\n",
      "122010\n",
      "122500\n",
      "122990\n",
      "123480\n",
      "123970\n",
      "124460\n",
      "124950\n",
      "125440\n",
      "125930\n",
      "126420\n",
      "126910\n",
      "127400\n",
      "127890\n",
      "128380\n",
      "128870\n",
      "129360\n",
      "129850\n",
      "130340\n",
      "130830\n",
      "131320\n",
      "131810\n",
      "132300\n",
      "132790\n",
      "133280\n",
      "133770\n",
      "134260\n",
      "134750\n",
      "135240\n",
      "135730\n",
      "136220\n",
      "136710\n",
      "137200\n",
      "137690\n",
      "138180\n",
      "138670\n",
      "139160\n",
      "139650\n",
      "140140\n",
      "140630\n",
      "141120\n",
      "141610\n",
      "142100\n",
      "142590\n",
      "143080\n",
      "143570\n",
      "144060\n",
      "144550\n",
      "145040\n",
      "145530\n",
      "146020\n",
      "146510\n",
      "147000\n",
      "147490\n",
      "147980\n",
      "148470\n",
      "148960\n",
      "149450\n",
      "149940\n",
      "150430\n",
      "150920\n",
      "151410\n",
      "151900\n",
      "152390\n",
      "152880\n",
      "153370\n",
      "153860\n",
      "154350\n",
      "154840\n",
      "155330\n",
      "155820\n",
      "156310\n",
      "156800\n",
      "157290\n",
      "157780\n",
      "158270\n",
      "158760\n",
      "159250\n",
      "159740\n",
      "160230\n",
      "160720\n",
      "161210\n",
      "161700\n",
      "162190\n",
      "162680\n",
      "163170\n",
      "163660\n",
      "164150\n",
      "164640\n",
      "165130\n",
      "165620\n",
      "166110\n",
      "166600\n",
      "167090\n",
      "167580\n",
      "168070\n",
      "168560\n",
      "169050\n",
      "169540\n",
      "170030\n",
      "170520\n",
      "171010\n",
      "171500\n",
      "171990\n",
      "172480\n",
      "172970\n",
      "173460\n",
      "173950\n",
      "174440\n",
      "174930\n",
      "175420\n",
      "175910\n",
      "176400\n",
      "176890\n",
      "177380\n",
      "177870\n",
      "178360\n",
      "178850\n",
      "179340\n",
      "179830\n",
      "180320\n",
      "180810\n",
      "181300\n",
      "181790\n",
      "182280\n",
      "182770\n",
      "183260\n",
      "183750\n",
      "184240\n",
      "184730\n",
      "185220\n",
      "185710\n",
      "186200\n",
      "186690\n",
      "187180\n",
      "187670\n",
      "188160\n",
      "188650\n",
      "189140\n",
      "189630\n",
      "190120\n",
      "190610\n",
      "191100\n",
      "191590\n",
      "192080\n",
      "192570\n",
      "193060\n",
      "193550\n",
      "194040\n",
      "194530\n",
      "195020\n",
      "195510\n",
      "196000\n",
      "196490\n",
      "196980\n",
      "197470\n",
      "197960\n",
      "198450\n",
      "198940\n",
      "199430\n",
      "199920\n",
      "200410\n",
      "200900\n",
      "201390\n",
      "201880\n",
      "202370\n",
      "202860\n",
      "203350\n",
      "203840\n",
      "204330\n",
      "204820\n",
      "205310\n",
      "205800\n",
      "206290\n",
      "206780\n",
      "207270\n",
      "207760\n",
      "208250\n",
      "208740\n",
      "209230\n",
      "209720\n",
      "210210\n",
      "210700\n",
      "211190\n",
      "211680\n",
      "212170\n",
      "212660\n",
      "213150\n",
      "213640\n",
      "214130\n",
      "214620\n",
      "215110\n",
      "215600\n",
      "216090\n",
      "216580\n",
      "217070\n",
      "217560\n",
      "218050\n",
      "218540\n",
      "219030\n",
      "219520\n",
      "220010\n",
      "220500\n",
      "220990\n",
      "221480\n",
      "221970\n",
      "222460\n",
      "222950\n",
      "223440\n",
      "223930\n",
      "224420\n",
      "224910\n",
      "225400\n",
      "225890\n",
      "226380\n",
      "226870\n",
      "227360\n",
      "227850\n",
      "228340\n",
      "228830\n",
      "229320\n",
      "229810\n",
      "230300\n",
      "230790\n",
      "231280\n",
      "231770\n",
      "232260\n",
      "232750\n",
      "233240\n",
      "233730\n",
      "234220\n",
      "234710\n",
      "235200\n",
      "235690\n",
      "236180\n",
      "236670\n",
      "237160\n",
      "237650\n",
      "238140\n",
      "238630\n",
      "239120\n",
      "239610\n",
      "240100\n",
      "240590\n",
      "241080\n",
      "241570\n",
      "242060\n",
      "242550\n",
      "243040\n",
      "243530\n",
      "244020\n",
      "244510\n",
      "245000\n",
      "245490\n",
      "245980\n",
      "246470\n",
      "246960\n",
      "247450\n",
      "247940\n",
      "248430\n",
      "248920\n",
      "249410\n",
      "249900\n",
      "250390\n",
      "250880\n",
      "251370\n",
      "251860\n",
      "252350\n",
      "252840\n",
      "253330\n",
      "253820\n",
      "254310\n",
      "254800\n",
      "255290\n",
      "255780\n",
      "256270\n",
      "256760\n",
      "257250\n",
      "257740\n",
      "258230\n",
      "258720\n",
      "259210\n",
      "259700\n",
      "260190\n",
      "260680\n",
      "261170\n",
      "261660\n",
      "262150\n",
      "262640\n",
      "263130\n",
      "263620\n",
      "264110\n",
      "264600\n",
      "265090\n",
      "265580\n",
      "266070\n",
      "266560\n",
      "267050\n",
      "267540\n",
      "268030\n",
      "268520\n",
      "269010\n",
      "269500\n",
      "269990\n",
      "270480\n",
      "270970\n",
      "271460\n",
      "271950\n",
      "272440\n",
      "272930\n",
      "273420\n",
      "273910\n",
      "274400\n",
      "274890\n",
      "275380\n",
      "275870\n",
      "276360\n",
      "276850\n",
      "277340\n",
      "277830\n",
      "278320\n",
      "278810\n",
      "279300\n",
      "279790\n",
      "280280\n",
      "280770\n",
      "281260\n",
      "281750\n",
      "282240\n",
      "282730\n",
      "283220\n",
      "283710\n",
      "284200\n",
      "284690\n",
      "285180\n",
      "285670\n",
      "286160\n",
      "286650\n",
      "287140\n",
      "287630\n",
      "288120\n",
      "288610\n",
      "289100\n",
      "289590\n",
      "290080\n",
      "290570\n",
      "291060\n",
      "291550\n",
      "292040\n",
      "292530\n",
      "293020\n",
      "293510\n",
      "294000\n",
      "294490\n",
      "294980\n",
      "295470\n",
      "295960\n",
      "296450\n",
      "296940\n",
      "297430\n",
      "297920\n",
      "298410\n",
      "298900\n",
      "299390\n",
      "299880\n",
      "300370\n",
      "300860\n",
      "301350\n",
      "301840\n",
      "302330\n",
      "302820\n",
      "303310\n",
      "303800\n",
      "304290\n",
      "304780\n",
      "305270\n",
      "305760\n",
      "306250\n",
      "306740\n",
      "307230\n",
      "307720\n",
      "308210\n",
      "308700\n",
      "309190\n",
      "309680\n",
      "310170\n",
      "310660\n",
      "311150\n",
      "311640\n",
      "312130\n",
      "312620\n",
      "313110\n",
      "313600\n",
      "314090\n",
      "314580\n",
      "315070\n",
      "315560\n",
      "316050\n",
      "316540\n",
      "317030\n",
      "317520\n",
      "318010\n",
      "318500\n",
      "318990\n",
      "319480\n",
      "319970\n",
      "320460\n",
      "320950\n",
      "321440\n",
      "321930\n",
      "322420\n",
      "322910\n",
      "323400\n",
      "323890\n",
      "324380\n",
      "324870\n",
      "325360\n",
      "325850\n",
      "326340\n",
      "326830\n",
      "327320\n",
      "327810\n",
      "328300\n",
      "328790\n",
      "329280\n",
      "329770\n",
      "330260\n",
      "330750\n",
      "331240\n",
      "331730\n",
      "332220\n",
      "332710\n",
      "333200\n",
      "333690\n",
      "334180\n",
      "334670\n",
      "335160\n",
      "335650\n",
      "336140\n",
      "336630\n",
      "337120\n",
      "337610\n",
      "338100\n",
      "338590\n",
      "339080\n",
      "339570\n",
      "340060\n",
      "340550\n",
      "341040\n",
      "341530\n",
      "342020\n",
      "342510\n",
      "343000\n",
      "343490\n",
      "343980\n",
      "344470\n",
      "344960\n",
      "345450\n",
      "345940\n",
      "346430\n",
      "346920\n",
      "347410\n",
      "347900\n",
      "348390\n",
      "348880\n",
      "349370\n",
      "349860\n",
      "350350\n",
      "350840\n",
      "351330\n",
      "351820\n",
      "352310\n",
      "352800\n",
      "353290\n",
      "353780\n",
      "354270\n",
      "354760\n",
      "355250\n",
      "355740\n",
      "356230\n",
      "356720\n",
      "357210\n",
      "357700\n",
      "358190\n",
      "358680\n",
      "359170\n",
      "359660\n",
      "360150\n",
      "360640\n",
      "361130\n",
      "361620\n",
      "362110\n",
      "362600\n",
      "363090\n",
      "363580\n",
      "364070\n",
      "364560\n",
      "365050\n",
      "365540\n",
      "366030\n",
      "366520\n",
      "367010\n",
      "367500\n",
      "367990\n",
      "368480\n",
      "368970\n",
      "369460\n",
      "369950\n",
      "370440\n",
      "370930\n",
      "371420\n",
      "371910\n",
      "372400\n",
      "372890\n",
      "373380\n",
      "373870\n",
      "374360\n",
      "374850\n",
      "375340\n",
      "375830\n",
      "376320\n",
      "376810\n",
      "377300\n",
      "377790\n",
      "378280\n",
      "378770\n",
      "379260\n",
      "379750\n",
      "380240\n",
      "380730\n",
      "381220\n",
      "381710\n",
      "382200\n",
      "382690\n",
      "383180\n",
      "383670\n",
      "384160\n",
      "384650\n",
      "385140\n",
      "385630\n",
      "386120\n",
      "386610\n",
      "387100\n",
      "387590\n",
      "388080\n",
      "388570\n",
      "389060\n",
      "389550\n",
      "390040\n",
      "390530\n",
      "391020\n",
      "391510\n",
      "392000\n",
      "392490\n",
      "392980\n",
      "393470\n",
      "393960\n",
      "394450\n",
      "394940\n",
      "395430\n",
      "395920\n",
      "396410\n",
      "396900\n",
      "397390\n",
      "397880\n",
      "398370\n",
      "398860\n",
      "399350\n",
      "399840\n",
      "400330\n",
      "400820\n",
      "401310\n",
      "401800\n",
      "402290\n",
      "402780\n",
      "403270\n",
      "403760\n",
      "404250\n",
      "404740\n",
      "405230\n",
      "405720\n",
      "406210\n",
      "406700\n",
      "407190\n",
      "407680\n",
      "408170\n",
      "408660\n",
      "409150\n",
      "409640\n",
      "410130\n",
      "410620\n",
      "411110\n",
      "411600\n",
      "412090\n",
      "412580\n",
      "413070\n",
      "413560\n",
      "414050\n",
      "414540\n",
      "415030\n",
      "415520\n",
      "416010\n",
      "416500\n",
      "416990\n",
      "417480\n",
      "417970\n",
      "418460\n",
      "418950\n",
      "419440\n",
      "419930\n",
      "420420\n",
      "420910\n",
      "421400\n",
      "421890\n",
      "422380\n",
      "422870\n",
      "423360\n",
      "423850\n",
      "424340\n",
      "424830\n",
      "425320\n",
      "425810\n",
      "426300\n",
      "426790\n",
      "427280\n",
      "427770\n",
      "428260\n",
      "428750\n",
      "429240\n",
      "429730\n",
      "430220\n",
      "430710\n",
      "431200\n",
      "431690\n",
      "432180\n",
      "432670\n",
      "433160\n",
      "433650\n",
      "434140\n",
      "434630\n",
      "435120\n",
      "435610\n",
      "436100\n",
      "436590\n",
      "437080\n",
      "437570\n",
      "438060\n",
      "438550\n",
      "439040\n",
      "439530\n",
      "440020\n",
      "440510\n",
      "441000\n",
      "441490\n",
      "441980\n",
      "442470\n",
      "442960\n",
      "443450\n",
      "443940\n",
      "444430\n",
      "444920\n",
      "445410\n",
      "445900\n",
      "446390\n",
      "446880\n",
      "447370\n",
      "447860\n",
      "448350\n",
      "448840\n",
      "449330\n",
      "449820\n",
      "450310\n",
      "450800\n",
      "451290\n",
      "451780\n",
      "452270\n",
      "452760\n",
      "453250\n",
      "453740\n",
      "454230\n",
      "454720\n",
      "455210\n",
      "455700\n",
      "456190\n",
      "456680\n",
      "457170\n",
      "457660\n",
      "458150\n",
      "458640\n",
      "459130\n",
      "459620\n",
      "460110\n",
      "460600\n",
      "461090\n",
      "461580\n",
      "462070\n",
      "462560\n",
      "463050\n",
      "463540\n",
      "464030\n",
      "464520\n",
      "465010\n",
      "465500\n",
      "465990\n",
      "466480\n",
      "466970\n",
      "467460\n",
      "467950\n",
      "468440\n",
      "468930\n",
      "469420\n",
      "469910\n",
      "470400\n",
      "470890\n",
      "471380\n",
      "471870\n",
      "472360\n",
      "472850\n",
      "473340\n",
      "473830\n",
      "474320\n",
      "474810\n",
      "475300\n",
      "475790\n",
      "476280\n",
      "476770\n",
      "477260\n",
      "477750\n",
      "478240\n",
      "478730\n",
      "479220\n",
      "479710\n",
      "480200\n",
      "480690\n",
      "481180\n",
      "481670\n",
      "482160\n",
      "482650\n",
      "483140\n",
      "483630\n",
      "484120\n",
      "484610\n",
      "485100\n",
      "485590\n",
      "486080\n",
      "486570\n",
      "487060\n",
      "487550\n",
      "488040\n",
      "488530\n",
      "489020\n",
      "489510\n",
      "490000\n",
      "490490\n",
      "490980\n",
      "491470\n",
      "491960\n",
      "492450\n",
      "492940\n",
      "493430\n",
      "493920\n",
      "494410\n",
      "494900\n",
      "495390\n",
      "495880\n",
      "496370\n",
      "496860\n",
      "497350\n",
      "497840\n",
      "498330\n",
      "498820\n",
      "499310\n",
      "499800\n",
      "500290\n",
      "500780\n",
      "501270\n",
      "501760\n",
      "502250\n",
      "502740\n",
      "503230\n",
      "503720\n",
      "504210\n",
      "504700\n",
      "505190\n",
      "505680\n",
      "506170\n",
      "506660\n",
      "507150\n",
      "507640\n",
      "508130\n",
      "508620\n",
      "509110\n",
      "509600\n",
      "510090\n",
      "510580\n",
      "511070\n",
      "511560\n",
      "512050\n",
      "512540\n",
      "513030\n",
      "513520\n",
      "514010\n",
      "514500\n",
      "514990\n",
      "515480\n",
      "515970\n",
      "516460\n",
      "516950\n",
      "517440\n",
      "517930\n",
      "518420\n",
      "518910\n",
      "519400\n",
      "519890\n",
      "520380\n",
      "520870\n",
      "521360\n",
      "521850\n",
      "522340\n",
      "522830\n",
      "523320\n",
      "523810\n",
      "524300\n",
      "524790\n",
      "525280\n",
      "525770\n",
      "526260\n",
      "526750\n",
      "527240\n",
      "527730\n",
      "528220\n",
      "528710\n",
      "529200\n",
      "529690\n",
      "530180\n",
      "530670\n",
      "531160\n",
      "531650\n",
      "532140\n",
      "532630\n",
      "533120\n",
      "533610\n",
      "534100\n",
      "534590\n",
      "535080\n",
      "535570\n",
      "536060\n",
      "536550\n",
      "537040\n",
      "537530\n",
      "538020\n",
      "538510\n",
      "539000\n",
      "539490\n",
      "539980\n",
      "540470\n",
      "540960\n",
      "541450\n",
      "541940\n",
      "542430\n",
      "542920\n",
      "543410\n",
      "543900\n",
      "544390\n",
      "544880\n",
      "545370\n",
      "545860\n",
      "546350\n",
      "546840\n",
      "547330\n",
      "547820\n",
      "548310\n",
      "548800\n",
      "549290\n",
      "549780\n",
      "550270\n",
      "550760\n",
      "551250\n",
      "551740\n",
      "552230\n",
      "552720\n",
      "553210\n",
      "553700\n",
      "554190\n",
      "554680\n",
      "555170\n",
      "555660\n",
      "556150\n",
      "556640\n",
      "557130\n",
      "557620\n",
      "558110\n",
      "558600\n",
      "559090\n",
      "559580\n",
      "560070\n",
      "560560\n",
      "561050\n",
      "561540\n",
      "562030\n",
      "562520\n",
      "563010\n",
      "563500\n",
      "563990\n",
      "564480\n",
      "564970\n",
      "565460\n",
      "565950\n",
      "566440\n",
      "566930\n",
      "567420\n",
      "567910\n",
      "568400\n",
      "568890\n",
      "569380\n",
      "569870\n",
      "570360\n",
      "570850\n",
      "571340\n",
      "571830\n",
      "572320\n",
      "572810\n",
      "573300\n",
      "573790\n",
      "574280\n",
      "574770\n",
      "575260\n",
      "575750\n",
      "576240\n",
      "576730\n",
      "577220\n",
      "577710\n",
      "578200\n",
      "578690\n",
      "579180\n",
      "579670\n",
      "580160\n",
      "580650\n",
      "581140\n",
      "581630\n",
      "582120\n",
      "582610\n",
      "583100\n",
      "583590\n",
      "584080\n",
      "584570\n",
      "585060\n",
      "585550\n",
      "586040\n",
      "586530\n",
      "587020\n",
      "587510\n",
      "588000\n",
      "588490\n",
      "588980\n",
      "589470\n",
      "589960\n",
      "590450\n",
      "590940\n",
      "591430\n",
      "591920\n",
      "592410\n",
      "592900\n",
      "593390\n",
      "593880\n",
      "594370\n",
      "594860\n",
      "595350\n",
      "595840\n",
      "596330\n",
      "596820\n",
      "597310\n",
      "597800\n",
      "598290\n",
      "598780\n",
      "599270\n",
      "599760\n",
      "600250\n",
      "600740\n",
      "601230\n",
      "601720\n",
      "602210\n",
      "602700\n",
      "603190\n",
      "603680\n",
      "604170\n",
      "604660\n",
      "605150\n",
      "605640\n",
      "606130\n",
      "606620\n",
      "607110\n",
      "607600\n",
      "608090\n",
      "608580\n",
      "609070\n",
      "609560\n",
      "610050\n",
      "610540\n",
      "611030\n",
      "611520\n",
      "612010\n",
      "612500\n",
      "612990\n",
      "613480\n",
      "613970\n",
      "614460\n",
      "614950\n",
      "615440\n",
      "615930\n",
      "616420\n",
      "616910\n",
      "617400\n",
      "617890\n",
      "618380\n",
      "618870\n",
      "619360\n",
      "619850\n",
      "620340\n",
      "620830\n",
      "621320\n",
      "621810\n",
      "622300\n",
      "622790\n",
      "623280\n",
      "623770\n",
      "624260\n",
      "624750\n",
      "625240\n",
      "625730\n",
      "626220\n",
      "626710\n",
      "627200\n",
      "627690\n",
      "628180\n",
      "628670\n",
      "629160\n",
      "629650\n",
      "630140\n",
      "630630\n",
      "631120\n",
      "631610\n",
      "632100\n",
      "632590\n",
      "633080\n",
      "633570\n",
      "634060\n",
      "634550\n",
      "635040\n",
      "635530\n",
      "636020\n",
      "636510\n",
      "637000\n",
      "637490\n",
      "637980\n",
      "638470\n",
      "638960\n",
      "639450\n",
      "639940\n",
      "640430\n",
      "640920\n",
      "641410\n",
      "641900\n",
      "642390\n",
      "642880\n",
      "643370\n",
      "643860\n",
      "644350\n",
      "644840\n",
      "645330\n",
      "645820\n",
      "646310\n",
      "646800\n",
      "647290\n",
      "647780\n",
      "648270\n",
      "648760\n",
      "649250\n",
      "649740\n",
      "650230\n",
      "650720\n",
      "651210\n",
      "651700\n",
      "652190\n",
      "652680\n",
      "653170\n",
      "653660\n",
      "654150\n",
      "654640\n",
      "655130\n",
      "655620\n",
      "656110\n",
      "656600\n",
      "657090\n",
      "657580\n",
      "658070\n",
      "658560\n",
      "659050\n",
      "659540\n",
      "660030\n",
      "660520\n",
      "661010\n",
      "661500\n",
      "661990\n",
      "662480\n",
      "662970\n",
      "663460\n",
      "663950\n",
      "664440\n",
      "664930\n",
      "665420\n",
      "665910\n",
      "666400\n",
      "666890\n",
      "667380\n",
      "667870\n",
      "668360\n",
      "668850\n",
      "669340\n",
      "669830\n",
      "670320\n",
      "670810\n",
      "671300\n",
      "671790\n",
      "672280\n",
      "672770\n",
      "673260\n",
      "673750\n",
      "674240\n",
      "674730\n",
      "675220\n",
      "675710\n",
      "676200\n",
      "676690\n",
      "677180\n",
      "677670\n",
      "678160\n",
      "678650\n",
      "679140\n",
      "679630\n",
      "680120\n",
      "680610\n",
      "681100\n",
      "681590\n",
      "682080\n",
      "682570\n",
      "683060\n",
      "683550\n",
      "684040\n",
      "684530\n",
      "685020\n",
      "685510\n",
      "686000\n",
      "686490\n",
      "686980\n",
      "687470\n",
      "687960\n",
      "688450\n",
      "688940\n",
      "689430\n",
      "689920\n",
      "690410\n",
      "690900\n",
      "691390\n",
      "691880\n",
      "692370\n",
      "692860\n",
      "693350\n",
      "693840\n",
      "694330\n",
      "694820\n",
      "695310\n",
      "695800\n",
      "696290\n",
      "696780\n",
      "697270\n",
      "697760\n",
      "698250\n",
      "698740\n",
      "699230\n",
      "699720\n",
      "700210\n",
      "700700\n",
      "701190\n",
      "701680\n",
      "702170\n",
      "702660\n",
      "703150\n",
      "703640\n",
      "704130\n",
      "704620\n",
      "705110\n",
      "705600\n",
      "706090\n",
      "706580\n",
      "707070\n",
      "707560\n",
      "708050\n",
      "708540\n",
      "709030\n",
      "709520\n",
      "710010\n",
      "710500\n",
      "710990\n",
      "711480\n",
      "711970\n",
      "712460\n",
      "712950\n",
      "713440\n",
      "713930\n",
      "714420\n",
      "714910\n",
      "715400\n",
      "715890\n",
      "716380\n",
      "716870\n",
      "717360\n",
      "717850\n",
      "718340\n",
      "718830\n",
      "719320\n",
      "719810\n",
      "720300\n",
      "720790\n",
      "721280\n",
      "721770\n",
      "722260\n",
      "722750\n",
      "723240\n",
      "723730\n",
      "724220\n",
      "724710\n",
      "725200\n",
      "725690\n",
      "726180\n",
      "726670\n",
      "727160\n",
      "727650\n",
      "728140\n",
      "728630\n",
      "729120\n",
      "729610\n",
      "730100\n",
      "730590\n",
      "731080\n",
      "731570\n",
      "732060\n",
      "732550\n",
      "733040\n",
      "733530\n",
      "734020\n",
      "734510\n",
      "735000\n",
      "735490\n",
      "735980\n",
      "736470\n",
      "736960\n",
      "737450\n",
      "737940\n",
      "738430\n",
      "738920\n",
      "739410\n",
      "739900\n",
      "740390\n",
      "740880\n",
      "741370\n",
      "741860\n",
      "742350\n",
      "742840\n",
      "743330\n",
      "743820\n",
      "744310\n",
      "744800\n",
      "745290\n",
      "745780\n",
      "746270\n",
      "746760\n",
      "747250\n",
      "747740\n",
      "748230\n",
      "748720\n",
      "749210\n",
      "749700\n",
      "750190\n",
      "750680\n",
      "751170\n",
      "751660\n",
      "752150\n",
      "752640\n",
      "753130\n",
      "753620\n",
      "754110\n",
      "754600\n",
      "755090\n",
      "755580\n",
      "756070\n",
      "756560\n",
      "757050\n",
      "757540\n",
      "758030\n",
      "758520\n",
      "759010\n",
      "759500\n",
      "759990\n",
      "760480\n",
      "760970\n",
      "761460\n",
      "761950\n",
      "762440\n",
      "762930\n",
      "763420\n",
      "763910\n",
      "764400\n",
      "764890\n",
      "765380\n",
      "765870\n",
      "766360\n",
      "766850\n",
      "767340\n",
      "767830\n",
      "768320\n",
      "768810\n",
      "769300\n",
      "769790\n",
      "770280\n",
      "770770\n",
      "771260\n",
      "771750\n",
      "772240\n",
      "772730\n",
      "773220\n",
      "773710\n",
      "774200\n",
      "774690\n",
      "775180\n",
      "775670\n",
      "776160\n",
      "776650\n",
      "777140\n",
      "777630\n",
      "778120\n",
      "778610\n",
      "779100\n",
      "779590\n",
      "780080\n",
      "780570\n",
      "781060\n",
      "781550\n",
      "782040\n",
      "782530\n",
      "783020\n",
      "783510\n",
      "784000\n",
      "784490\n",
      "784980\n",
      "785470\n",
      "785960\n",
      "786450\n",
      "786940\n",
      "787430\n",
      "787920\n",
      "788410\n",
      "788900\n",
      "789390\n",
      "789880\n",
      "790370\n",
      "790860\n",
      "791350\n",
      "791840\n",
      "792330\n",
      "792820\n",
      "793310\n",
      "793800\n",
      "794290\n",
      "794780\n",
      "795270\n",
      "795760\n",
      "796250\n",
      "796740\n",
      "797230\n",
      "797720\n",
      "798210\n",
      "798700\n",
      "799190\n",
      "799680\n",
      "800170\n",
      "800660\n",
      "801150\n",
      "801640\n",
      "802130\n",
      "802620\n",
      "803110\n",
      "803600\n",
      "804090\n",
      "804580\n",
      "805070\n",
      "805560\n",
      "806050\n",
      "806540\n",
      "807030\n",
      "807520\n",
      "808010\n",
      "808500\n",
      "808990\n",
      "809480\n",
      "809970\n",
      "810460\n",
      "810950\n",
      "811440\n",
      "811930\n",
      "812420\n",
      "812910\n",
      "813400\n",
      "813890\n",
      "814380\n",
      "814870\n",
      "815360\n",
      "815850\n",
      "816340\n",
      "816830\n",
      "817320\n",
      "817810\n",
      "818300\n",
      "818790\n",
      "819280\n",
      "819770\n",
      "820260\n",
      "820750\n",
      "821240\n",
      "821730\n",
      "822220\n",
      "822710\n",
      "823200\n",
      "823690\n",
      "824180\n",
      "824670\n",
      "825160\n",
      "825650\n",
      "826140\n",
      "826630\n",
      "827120\n",
      "827610\n",
      "828100\n",
      "828590\n",
      "829080\n",
      "829570\n",
      "830060\n",
      "830550\n",
      "831040\n",
      "831530\n",
      "832020\n",
      "832510\n",
      "833000\n",
      "833490\n",
      "833980\n",
      "834470\n",
      "834960\n",
      "835450\n",
      "835940\n",
      "836430\n",
      "836920\n",
      "837410\n",
      "837900\n",
      "838390\n",
      "838880\n",
      "839370\n",
      "839860\n",
      "840350\n",
      "840840\n",
      "841330\n",
      "841820\n",
      "842310\n",
      "842800\n",
      "843290\n",
      "843780\n",
      "844270\n",
      "844760\n",
      "845250\n",
      "845740\n",
      "846230\n",
      "846720\n",
      "847210\n",
      "847700\n",
      "848190\n",
      "848680\n",
      "849170\n",
      "849660\n",
      "850150\n",
      "850640\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/f3/1bf0zv1s7991qswb6sl4v9k00000gn/T/ipykernel_25378/3172597103.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     18\u001b[0m     \u001b[0mtest_df\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'papers_abstracts.csv'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m \u001b[0;34m=\u001b[0m\u001b[0;34m'a'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mheader\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m     \u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msleep\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     21\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m     \u001b[0mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "df['DOI'] = df.apply(lambda x: json.loads(x['externalids'])['DOI'], axis = 1)\n",
    "id_list = df['DOI'].to_list()\n",
    "\n",
    "i = 0\n",
    "\n",
    "while i < 1000000:\n",
    "    r = requests.post(\n",
    "            'https://api.semanticscholar.org/graph/v1/paper/batch',\n",
    "            params={'fields': 'abstract,embedding'},\n",
    "            json={\"ids\": id_list[i:i + 490]}\n",
    "        )\n",
    "\n",
    "    i = i + 490\n",
    "\n",
    "    df_json = json.dumps(r.json(), indent=2)\n",
    "    test_df = json_normalize(json.loads(df_json))\n",
    "    test_df = test_df[~test_df['abstract'].isnull()]\n",
    "    test_df.to_csv('papers_abstracts.csv', mode ='a', header=False)\n",
    "\n",
    "    time.sleep(5)\n",
    "        \n",
    "    print (i)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ade035cb",
   "metadata": {},
   "source": [
    "### Format Embeddings Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "caf1a718",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   level_0  index   1                                      id_1  \\\n",
      "0        0      0   2  c856627242a754d2d756b32843523e6d7a089148   \n",
      "3        3      3   7  9b9659a4e9a4a48d7c532c76dd14ee9ccd696025   \n",
      "4        4      4   8  314f8d8aba804f3ba77ac167161e73950f406332   \n",
      "5        5      5  15  76796fdb8de6a4625a865fca751ee1ceb4989502   \n",
      "6        6      6  17  78c025475dd2a940b0ebdc3be885ad1a01fde2ca   \n",
      "\n",
      "                                            abstract           model  \\\n",
      "0  Summary: The current work characterizes young ...  specter@v0.1.1   \n",
      "3  Background The Brain Injury Guidelines provide...  specter@v0.1.1   \n",
      "4  A 66-year-old woman presented with pustular le...  specter@v0.1.1   \n",
      "5  Abstract Background Danon disease is an X-link...  specter@v0.1.1   \n",
      "6  BACKGROUND Biologics are often required for tr...  specter@v0.1.1   \n",
      "\n",
      "                                           embedding  \n",
      "0  [-3.783419132232666, -2.987468957901001, -3.34...  \n",
      "3  [-5.099650859832764, -5.582947731018066, 0.164...  \n",
      "4  [-1.8056416511535645, -2.358576774597168, 3.89...  \n",
      "5  [0.689982533454895, -3.432250499725342, 1.7640...  \n",
      "6  [-7.241069316864014, -4.761958122253418, -2.08...  \n",
      "240151\n"
     ]
    }
   ],
   "source": [
    "mapping = {df_embed.columns[1]: 'id_1', df_embed.columns[2]: 'abstract', df_embed.columns[3]: 'model', df_embed.columns[4]: 'embedding'}\n",
    "df_embed = df_embed.rename(columns=mapping)\n",
    "df_embed = df_embed.drop_duplicates(subset = 'abstract', keep = 'last')\n",
    "df_embed = df_embed.reset_index()\n",
    "df_embed.to_csv('abstracts_embedding')\n",
    "print (df_embed.head(5))\n",
    "print (len(df_embed))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "dd87acce",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_embed = pd.read_csv('abstracts_embedding')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c78a7168",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'str'>\n",
      "<class 'numpy.ndarray'>\n",
      "   Unnamed: 0  level_0  index   1                                      id_1  \\\n",
      "0           0        0      0   2  c856627242a754d2d756b32843523e6d7a089148   \n",
      "1           3        3      3   7  9b9659a4e9a4a48d7c532c76dd14ee9ccd696025   \n",
      "2           4        4      4   8  314f8d8aba804f3ba77ac167161e73950f406332   \n",
      "3           5        5      5  15  76796fdb8de6a4625a865fca751ee1ceb4989502   \n",
      "4           6        6      6  17  78c025475dd2a940b0ebdc3be885ad1a01fde2ca   \n",
      "\n",
      "                                            abstract           model  \\\n",
      "0  Summary: The current work characterizes young ...  specter@v0.1.1   \n",
      "1  Background The Brain Injury Guidelines provide...  specter@v0.1.1   \n",
      "2  A 66-year-old woman presented with pustular le...  specter@v0.1.1   \n",
      "3  Abstract Background Danon disease is an X-link...  specter@v0.1.1   \n",
      "4  BACKGROUND Biologics are often required for tr...  specter@v0.1.1   \n",
      "\n",
      "                                           embedding  \\\n",
      "0  [-3.783419132232666, -2.987468957901001, -3.34...   \n",
      "1  [-5.099650859832764, -5.582947731018066, 0.164...   \n",
      "2  [-1.8056416511535645, -2.358576774597168, 3.89...   \n",
      "3  [0.689982533454895, -3.432250499725342, 1.7640...   \n",
      "4  [-7.241069316864014, -4.761958122253418, -2.08...   \n",
      "\n",
      "                                     embedding array  \n",
      "0  [[-3.783419132232666, -2.987468957901001, -3.3...  \n",
      "1  [[-5.099650859832764, -5.582947731018066, 0.16...  \n",
      "2  [[-1.8056416511535645, -2.358576774597168, 3.8...  \n",
      "3  [[0.689982533454895, -3.432250499725342, 1.764...  \n",
      "4  [[-7.241069316864014, -4.761958122253418, -2.0...  \n"
     ]
    }
   ],
   "source": [
    "df_embed['embedding array'] = df_embed.apply(lambda row : np.array(row['embedding'].strip('][').split(', '), dtype = float).reshape(1,-1), axis = 1)\n",
    "df_embed = df_embed.reset_index(drop = True)\n",
    "print (type(df_embed['embedding'][0]))\n",
    "print (type(df_embed['embedding array'][0]))\n",
    "print (df_embed.head(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e84821e2",
   "metadata": {},
   "source": [
    "\n",
    "### Calc Similarity of Embeddings Example 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "af1fc35c",
   "metadata": {},
   "outputs": [],
   "source": [
    "example = df_embed['embedding array'][80]\n",
    "cosin = []\n",
    "\n",
    "for i in range(0,len(df_embed)):\n",
    "    try: \n",
    "        res = df_embed['embedding array'][i]\n",
    "        val_out = 1 - cdist(res, example, 'cosine')\n",
    "        cosin.append(val_out)\n",
    "    except:\n",
    "        print (i)\n",
    "        cosin.append(0)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "794e9f42",
   "metadata": {},
   "source": [
    "### Confirm calc by checking max similarity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "301c3d16",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "240151\n"
     ]
    }
   ],
   "source": [
    "max_value = max(cosin)\n",
    "max_index = cosin.index(max_value)\n",
    "\n",
    "print (df_embed['abstract'][80] == df_embed.loc[max_index]['abstract'])\n",
    "print (len(cosin))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ca38204a",
   "metadata": {},
   "source": [
    "### Display Top 5 similarity ranking"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ac3034a2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1.]]\n",
      "[[0.81100269]]\n",
      "[[0.80840463]]\n",
      "[[0.8025388]]\n",
      "[[0.8020896]]\n",
      "[[0.80160081]]\n",
      "[[0.80029626]]\n",
      "[[0.80012671]]\n",
      "[[0.79925581]]\n",
      "[[0.79634739]]\n",
      "[[0.79417126]]\n",
      "[[0.79398333]]\n",
      "[[0.79386986]]\n",
      "[[0.79366036]]\n",
      "[[0.78924773]]\n",
      "[[0.78813912]]\n",
      "[[0.78767597]]\n",
      "[[0.78288072]]\n",
      "[[0.78204333]]\n",
      "[[0.78175017]]\n",
      "[[0.78004856]]\n",
      "[[0.77966376]]\n",
      "[[0.77931865]]\n",
      "[[0.77758856]]\n",
      "[[0.77748765]]\n",
      "[[0.77682937]]\n",
      "[[0.77643427]]\n",
      "[[0.77349939]]\n",
      "[[0.77297556]]\n",
      "[[0.77267169]]\n",
      "[[0.77251627]]\n",
      "[[0.77238161]]\n",
      "[[0.77149863]]\n",
      "[[0.77116522]]\n",
      "[[0.7708249]]\n",
      "[[0.77055611]]\n",
      "[[0.77042392]]\n",
      "[[0.77038357]]\n",
      "[[0.76987788]]\n",
      "[[0.7691085]]\n",
      "[[0.76897641]]\n",
      "[[0.76868324]]\n",
      "[[0.76850764]]\n",
      "[[0.76715418]]\n",
      "[[0.76685403]]\n",
      "[[0.76679032]]\n",
      "[[0.76664837]]\n",
      "[[0.7665276]]\n",
      "[[0.7660499]]\n",
      "[[0.76588579]]\n",
      "[[0.76573136]]\n",
      "[[0.76403629]]\n",
      "[[0.76383959]]\n",
      "[[0.76285659]]\n",
      "[[0.76191269]]\n",
      "[[0.76066767]]\n",
      "[[0.76047025]]\n",
      "[[0.76042579]]\n",
      "[[0.76029325]]\n",
      "[[0.76010876]]\n",
      "[[0.75870135]]\n",
      "[[0.7575599]]\n",
      "[[0.75731395]]\n",
      "[[0.7572666]]\n",
      "[[0.75676671]]\n",
      "[[0.75660028]]\n",
      "[[0.75606981]]\n",
      "[[0.7560321]]\n",
      "[[0.75573641]]\n",
      "[[0.75569514]]\n",
      "[[0.75425296]]\n",
      "[[0.75387653]]\n",
      "[[0.75373915]]\n",
      "[[0.75345378]]\n",
      "[[0.75343963]]\n",
      "[[0.7534396]]\n",
      "[[0.7533492]]\n",
      "[[0.75315001]]\n",
      "[[0.7530581]]\n",
      "[[0.75292111]]\n",
      "[[0.75262234]]\n",
      "[[0.75250156]]\n",
      "[[0.75239285]]\n",
      "[[0.75212067]]\n",
      "[[0.75210332]]\n",
      "[[0.7515225]]\n",
      "[[0.75090528]]\n",
      "[[0.75077956]]\n",
      "[[0.7507725]]\n",
      "[[0.74983755]]\n",
      "[[0.74975636]]\n",
      "[[0.74949446]]\n",
      "[[0.74942712]]\n",
      "[[0.74928802]]\n",
      "[[0.74844945]]\n",
      "[[0.7475424]]\n",
      "[[0.74753444]]\n",
      "[[0.74750474]]\n",
      "[[0.74740371]]\n",
      "[[0.74701337]]\n"
     ]
    }
   ],
   "source": [
    "top_5 = heapq.nlargest(100, range(len(cosin)), key=cosin.__getitem__)\n",
    "\n",
    "max_index = cosin.index(max_value)\n",
    "\n",
    "abstract = []\n",
    "for i in top_5:\n",
    "    print (cosin[i])\n",
    "    abstract.append(df_embed.loc[i]['abstract'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0ca980c3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>This study aims to investigate the influence o...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Due to the advantages of charged particles com...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>The successful treatment of cancer, including ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>The efficacy of conventional radiation therapy...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Purpose This study was designed to evaluate th...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>95</th>\n",
       "      <td>We previously reported frequent loss of microR...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>Background/Aim: Anti-apoptotic proteins, inclu...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>97</th>\n",
       "      <td>Changes in p53 expression, apoptosis and cell ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>Radiotherapy (RT), the major anti-cancer modal...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>Abstract— Photofrin (25 mg/kg) was administere...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>100 rows × 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    0\n",
       "0   This study aims to investigate the influence o...\n",
       "1   Due to the advantages of charged particles com...\n",
       "2   The successful treatment of cancer, including ...\n",
       "3   The efficacy of conventional radiation therapy...\n",
       "4   Purpose This study was designed to evaluate th...\n",
       "..                                                ...\n",
       "95  We previously reported frequent loss of microR...\n",
       "96  Background/Aim: Anti-apoptotic proteins, inclu...\n",
       "97  Changes in p53 expression, apoptosis and cell ...\n",
       "98  Radiotherapy (RT), the major anti-cancer modal...\n",
       "99  Abstract— Photofrin (25 mg/kg) was administere...\n",
       "\n",
       "[100 rows x 1 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64b62c2a",
   "metadata": {},
   "source": [
    "### Calc Similarity of Embedding Example 2, Check Calcs, and Time Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "c06d67a6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "240151\n",
      "[[1.]]\n",
      "Amplicon sequencing generates chimeric reads which can cause spurious inferences of biological variation. I describe UCHIME2, an update of the popular UCHIME chimera detection algorithm with new modes optimized for high-resolution biological sequence reconstruction (“denoising”) and other applications. I show that chimera frequency correlates inversely with divergence, that error-free chimera prediction from sequence is impossible in principle, and that UCHIME2 achieves higher detection accuracy than previous methods.\n",
      "[[0.73703842]]\n",
      "Motivation Advances in the sequencing of uncultured environmental samples, dubbed metagenomics, raise a growing need for accurate taxonomic assignment. Accurate identification of organisms present within a community is essential to understanding even the most elementary ecosystems. However, current high-throughput sequencing technologies generate short reads which partially cover full-length marker genes and this poses difficult bioinformatic challenges for taxonomy identification at high resolution Results We designed MATAM, a software dedicated to the fast and accurate targeted assembly of short reads sequenced from a genomic marker of interest. The method implements a stepwise process based on construction and analysis of a read overlap graph. It is applied to the assembly of 16S rRNA markers and is validated on simulated, synthetic and genuine metagenomes. We show that MATAM outperforms other available methods in terms of low error rates and recovered genome fractions and is suitable to provide improved assemblies for precise taxonomic assignments. Availability https://github.com/bonsai-team/matam Contact pierre.pericard@gmail.com, helene.touzet@univ-lille1.fr\n",
      "[[0.7153147]]\n",
      "Mitochondrial resources are of known utility to many fields of phylogenetic, population and molecular biology. Their combination of faster and slower‐evolving regions and high copy number enables them to be used in many situations where other loci are unsuitable, with degraded samples and after recent speciation events.The advent of next‐generation sequencing technologies (and notably the Illumina platform) has led to an explosion in the number of samples that can be studied at transcriptomic level, at relatively low cost. Here we describe a robust pipeline for the recovery of mitochondrial genomes from these RNA‐sequencing resources. This pipeline can be used on sequencing of a variety of depths, and reliably recovers the protein coding and ribosomal gene complements of mitochondria from almost any transcriptomic sequencing experiment. The complete sequence of the mitochondrial genome can also be recovered when sequencing is performed in sufficient depth. We show the efficacy of our pipeline using data from eight nonmodel invertebrates of six disparate phyla. Interestingly, among our poriferan data, where microbiological symbionts are known empirically to make mitochondrial assembly difficult, this pipeline proved especially useful. Our pipeline will allow the recovery of mitochondrial data from a variety of previously sequenced samples, and add an additional angle of enquiry to future RNA‐sequencing efforts, simplifying the process of mitochondrial genome assembly for even the most recalcitrant clades and adding these data to the scientific record for a range of future uses.\n",
      "[[0.71410368]]\n",
      "Modern sequencing technologies should make the assembly of the relatively small mitochondrial genomes an easy undertaking. However, few tools exist that address mitochondrial assembly directly. As part of the Vertebrate Genomes Project (VGP) we have developed mitoVGP, a fully automated pipeline for similarity-based identification of mitochondrial reads and de novo assembly of mitochondrial genomes that incorporates both long (>10 kbp, PacBio or Nanopore) and short (100-300 bp, Illumina) reads. Our pipeline led to successful complete mitogenome assemblies of 100 vertebrate species of the VGP. We have observed that tissue type and library size selection have considerable impact on mitogenome sequencing and assembly. Comparing our assemblies to purportedly complete reference mitogenomes based on short-read sequencing, we have identified errors, missing sequences, and incomplete genes in those references, particularly in repeat regions. Our assemblies have also identified novel gene region duplications, shedding new light on mitochondrial genome evolution and organization.\n",
      "[[0.71091685]]\n",
      "The explosion of bioinformatics technologies in the form of next generation sequencing (NGS) has facilitated a massive influx of genomics data in the form of short reads. Short read mapping is therefore a fundamental component of next generation sequencing pipelines which routinely match these short reads against reference genomes for contig assembly. However, such techniques have seldom been applied to microbial marker gene sequencing studies, which have mostly relied on novel heuristic approaches. We propose NINJA Is Not Just Another OTU-Picking Solution (NINJA-OPS, or NINJA for short), a fast and highly accurate novel method enabling reference-based marker gene matching (picking Operational Taxonomic Units, or OTUs). NINJA takes advantage of the Burrows-Wheeler (BW) alignment using an artificial reference chromosome composed of concatenated reference sequences, the “concatesome,” as the BW input. Other features include automatic support for paired-end reads with arbitrary insert sizes. NINJA is also free and open source and implements several pre-filtering methods that elicit substantial speedup when coupled with existing tools. We applied NINJA to several published microbiome studies, obtaining accuracy similar to or better than previous reference-based OTU-picking methods while achieving an order of magnitude or more speedup and using a fraction of the memory footprint. NINJA is a complete pipeline that takes a FASTA-formatted input file and outputs a QIIME-formatted taxonomy-annotated BIOM file for an entire MiSeq run of human gut microbiome 16S genes in under 10 minutes on a dual-core laptop.\n",
      "[[0.70849587]]\n",
      "The capability to generate densely sampled single nucleotide polymorphism (SNP) data is essential in diverse subdisciplines of biology, including crop breeding, pathology, forensics, forestry, ecology, evolution, and conservation. However, the wet-lab expertise and bioinformatics training required to conduct genome-scale variant discovery remain limiting factors for investigators with limited resources. Here we present ISSRseq, a PCR-based method for reduced representation of genomic variation using simple sequence repeats as priming sites to sequence inter simple sequence repeat (ISSR) regions. Briefly, ISSR regions are amplified with single primers, pooled, used to construct sequencing libraries with a commercially-available kit, and sequenced on the Illumina platform. We also present a flexible bioinformatic pipeline that assembles ISSR loci, calls and hard filters variants, outputs data matrices in common formats, and conducts population analyses using R. Using three angiosperm species as case studies, we demonstrate that ISSRseq is highly repeatable, necessitates only simple wet-lab skills and commonplace instrumentation, is flexible in terms of the number of single primers used, and can generate genomic-scale variant discovery on par with existing RRS methods which require more complex wet lab procedures. ISSRseq represents a straightforward approach to SNP genotyping in any organism, and we predict that this method will be particularly useful for those studying population genomics and phylogeography of non-model organisms. Furthermore, the ease of ISSRseq relative to other RRS methods should prove useful to those lacking advanced expertise in wet lab methods or bioinformatics.\n",
      "4.698361873626709\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "example = df_embed['embedding array'][9000]\n",
    "cosin = []\n",
    "\n",
    "for i in range(0,len(df_embed)):\n",
    "    try: \n",
    "        res = df_embed['embedding array'][i]\n",
    "        val_out = 1 - cdist(res, example, 'cosine')\n",
    "        cosin.append(val_out)\n",
    "    except:\n",
    "        print (i)\n",
    "        cosin.append(0)\n",
    "        \n",
    "max_value = max(cosin)\n",
    "max_index = cosin.index(max_value)\n",
    "\n",
    "\n",
    "print (df_embed['abstract'][9000] == df_embed.loc[max_index]['abstract'])\n",
    "print (len(cosin))\n",
    "\n",
    "top_5 = heapq.nlargest(6, range(len(cosin)), key=cosin.__getitem__)\n",
    "\n",
    "max_index = cosin.index(max_value)\n",
    "\n",
    "for i in top_5:\n",
    "    print (cosin[i])\n",
    "    print (df_embed.loc[i]['abstract'])\n",
    "    \n",
    "print (time.time() - start)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fae00d38",
   "metadata": {},
   "source": [
    "### Calc Similarity of Embedding Example 3, Check Calcs, Time function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "6829ba0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "240151\n",
      "[[1.]]\n",
      "Studies were carried out to compare the early morphologic changes in the cecal mucosa of mice either infected with Serpulina hyodysenteriae or exposed to the β-hemolysin of S. hyodysenteriae. Sixty-five 12–24-week-old C3H/HeOuJ mice were infected with S. hyodysenteriae by gastric intubation. Two mice were necropsied every hour for 30 hours following infection. S. hyodysenteriae was isolated from the cecal contents of each mouse at all time points. Macroscopic lesions were first apparent at 14 hours postinfection (PI), and light microscopic lesions were first apparent at 10 hours PI, earlier than has been previously reported. Ultrastructural changes, first evident at 6 hours PI, included disarray and loss of microvilli and terminal web, with dilatation of intercellular spaces. Luminal bacteria were translocated through epithelial cells to the lamina propria, where capillaries exhibited changes indicative of increased permeability. In another experiment, solutions containing between 2,500 and 25,000 hemolytic units of purified S. hyodysenteriae hemolysin were placed within the lumen of surgically closed murine ceca (n = 10); ceca were collected for examination 3 hours following treatment. Ultrastructural changes consisted of loss of microvilli and terminal web and marked vacuolation and exfoliation of epithelial cells. Significant numbers of necrotic and apoptotic epithelial cells were present, and epithelial cells internalized moderate numbers of bacteria. The hemolysin of S. hyodysenteriae induces some of the same early ultrastructural changes in the cecal epithelium of mice as occur following infection with S. hyodysenteriae. Based on the observed bacterial translocation, luminal bacteria also appear to play a unique role in lesion development in this model.\n",
      "[[0.8525548]]\n",
      ". Ultrastructural studies were made on the inflammatory response in livers of two killifishes, Fundulus grandis and F. similis, experimentally infected with the coccidium Eimeria funduli. Coinciding with the formation and development of gamonts of E. funduli, changes took place in the liver and leucocytes appeared in that organ. Near areas of developing macrogamonts, many hepatocytes degenerated whereas others contained residual bodies which indicated sublethal damage. The damage was probably associated with merozoite formation and release. Filament-rich perisinusoidal cells hypertrophied during the development of E. funduli and contributed to walling off aggregates of oocysts from uninfected liver tissue. Leucocytes involved in the host response included lymphocytes, monocytes, heterophils, eosinophils and an unidentified third type of granulocyte. The heterophil was the first granulocyte to respond to the parasite and sometimes appeared to penetrate the parasitophorous vacuole of degenerating macrogamonts. Degranulating eosinophils were often observed near damaged tissues. Both the monocyte and eosinophil were phagocytically active. The monocyte was more phagocytically active and contained both host cells and parasite debris, whereas the eosinophil contained only debris from degenerating macrogamonts. The presence of immature granulocytes in the liver suggests that at least some granulocyte formation occurred in the liver.\n",
      "[[0.84139143]]\n",
      "Amoebiasis (a human intestinal infection affecting 50 million people every year) is caused by the protozoan parasite Entamoeba histolytica. To study the molecular mechanisms underlying human colon invasion by E. histolytica, we have set up an ex vivo human colon model to study the early steps in amoebiasis. Using scanning electron microscopy and histological analyses, we have established that E. histolytica caused the removal of the protective mucus coat during the first two hours of incubation, detached the enterocytes, and then penetrated into the lamina propria by following the crypts of Lieberkühn. Significant cell lysis (determined by the release of lactodehydrogenase) and inflammation (marked by the secretion of pro-inflammatory molecules such as interleukin 1 beta, interferon gamma, interleukin 6, interleukin 8 and tumour necrosis factor) were detected after four hours of incubation. Entamoeba dispar (a closely related non-pathogenic amoeba that also colonizes the human colon) was unable to invade colonic mucosa, lyse cells or induce an inflammatory response. We also examined the behaviour of trophozoites in which genes coding for known virulent factors (such as amoebapores, the Gal/GalNAc lectin and the cysteine protease 5 (CP-A5), which have major roles in cell death, adhesion (to target cells or mucus) and mucus degradation, respectively) were silenced, together with the corresponding tissue responses. Our data revealed that the signalling via the heavy chain Hgl2 or via the light chain Lgl1 of the Gal/GalNAc lectin is not essential to penetrate the human colonic mucosa. In addition, our study demonstrates that E. histolytica silenced for CP-A5 does not penetrate the colonic lamina propria and does not induce the host's pro-inflammatory cytokine secretion.\n",
      "[[0.83445396]]\n",
      "Rabbits are not usually susceptible to intestinal Shigella infection without extensive pretreatment, including starvation and administration of antimicrobial, antimotility, and toxic agents (carbon tetrachloride). Most animals infected this way die rapidly and do not always develop colonic lesions and signs of dysentery. We describe here a successful experimental infection in the adult rabbit which does not require preparatory treatment and which reproduced characteristic features of human shigellosis. Unstarved, untreated adult rabbits were infected by direct inoculation of virulent Shigella flexneri 2a (10 ml of 10(7) bacteria per ml) into the proximal colon after ligation of the distal cecum (cecal bypass). Within 24 h of infection, most inoculated animals consistently developed clinical dysentery, characterized by liquid stool mixed with mucus and blood, leukocytosis, anorexia, and weight loss. Histologically, there were edema, exudation, superficial ulceration, and polymorphonuclear infiltrations in the lamina propria; crypt abscess formation; focal hemorrhages; and the presence of immunohistochemically stained S. flexneri in the colonic mucosa. Successful bacterial colonization was indicated by the isolation of the challenge strain of S. flexneri 2a from the colonic contents. None of the control rabbits challenged with nonvirulent S. flexneri or without cecal bypass developed dysentery or colitis. We conclude that successful Shigella infection can be induced by direct colonic inoculation with virulent S. flexneri 2a in adult rabbits without starvation and pretreatment. The colitis is dependent on the virulence of the bacteria and on the cecal bypass. This model should be useful in studies of the immunopathogenesis of Shigella infection.\n",
      "[[0.82988651]]\n",
      "SUMMARY Mice inoculated with 2000 Ascaris suum eggs 7 days before an infection with 2 cysticercoids of Hymenolepis diminuta harboured significantly fewer and/or smaller tapeworms than control mice by day 7 post-infection. When the interval between the infections was increased, the effect on H. diminuta decreased and no effect was found 21 days after the A. suum infection or if the infections were given simultaneously in ńaive or in mice immune to A. suum. Two possible explanations for the rejection and/or stunting of H. diminuta in mice infected 7 days earlier with A. suum are suggested; either a host reaction in the small intestine stimulated by the returning larvae of A. suum after their hepato–pulmonary migration or an indirect effect of decreased food intake of the host caused by this migration. It is concluded that experiments on possible immunodepressive or immunostimulating effects of parasites ought to include studies on living agents and that they should not rely on measurements of immune responses only.\n",
      "[[0.82953416]]\n",
      "N. gonorrhoeae initiates infection by adhering to and invading columnar epithelial cells. Over time these activities often induce inflammation, with the influx of neutrophils and serum into the urethral lumen, cervical os, conjunctiva, and the like. At least some of these infected niches contain CMP-NANA (cytidine monophospho-N-acetyl neuraminic acid, also called CMP-sialic), contain sialylated gonococci, and are relatively or strictly anaerobic due to neutrophil and gonococcal metabolism and to the site of disease, that is, the peritoneal cavity. Gonococci thus encounter environmental conditions, reagents, and substrates in the human body that are not normally present in vitro. Knapp and Clark were the first to successfully grow gonococci anaerobically in an easily reproducible system, allowing researchers to begin to investigate in vitro the effects of anaerobiosis on gonococcal virulence traits. As a result of a series of elegant and in depth studies, Smith and Parsons and their colleagues showed that growth in CMP-NANA confers on the gonococcus a high degree of phenotypic (readily reversible) serum resistance and that CMP-NANA is available in vivo at sites of gonococcal infection and disease; gonococci become covalently coated with sialic acid and they become serum resistant (reviewed in refs. 8-10). Given that gonococci growing in the absence of oxygen or in the presence of CMP-NANA probably more closely resemble gonococci growing inside the human host, we studied several possible virulence traits of gonococci cultivated under these conditions. We first observed that anaerobic growth (in the absence of CMP-NANA) increases gonococcal resistance to killing by low (but not high) concentrations of normal human serum. We also asked whether anaerobic growth affected gonococcal association with host cells. Contrary to the effects on serum killing, anaerobic growth (in the absence of CMP-NANA) does not appear to affect the ability of gonococci (expressing certain adhesive outer membrane proteins called Opa proteins) to bind to and enter human epithelial cell lines or to bind to or resist killing by human neutrophils. The results from studies investigating the modulatory role of CMP-NANA were more striking. Growth in CMP-NANA dramatically inhibits the adherence of Opa+ gonococci to human neutrophils. It does not, however, appear to significantly decrease their sensitivity to phagocytic killing or to in vitro killing by lysosomal contents (aqueous extracts of human neutrophil granules).(ABSTRACT TRUNCATED AT 400 WORDS)\n",
      "4.688028812408447\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "example = df_embed['embedding array'][50000]\n",
    "cosin = []\n",
    "\n",
    "for i in range(0,len(df_embed)):\n",
    "    try: \n",
    "        res = df_embed['embedding array'][i]\n",
    "        val_out = 1 - cdist(res, example, 'cosine')\n",
    "        cosin.append(val_out)\n",
    "    except:\n",
    "        print (i)\n",
    "        cosin.append(0)\n",
    "        \n",
    "max_value = max(cosin)\n",
    "max_index = cosin.index(max_value)\n",
    "\n",
    "\n",
    "print (df_embed['abstract'][50000] == df_embed.loc[max_index]['abstract'])\n",
    "print (len(cosin))\n",
    "\n",
    "top_5 = heapq.nlargest(6, range(len(cosin)), key=cosin.__getitem__)\n",
    "\n",
    "max_index = cosin.index(max_value)\n",
    "\n",
    "for i in top_5:\n",
    "    print (cosin[i])\n",
    "    print (df_embed.loc[i]['abstract'])\n",
    "    \n",
    "print (time.time() - start)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49b65650",
   "metadata": {},
   "source": [
    "### Calc Similarity of Embedding Example 4, Check Calcs, Time function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "1a24dfe0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "240151\n",
      "[[1.]]\n",
      "Black radish (Raphanus sativus L. var. niger) root has been used in folk medicine since antiquity as a natural drug for the stimulation of bile function. According to in vitro studies the squeezed juice from black radish root exhibited significant antioxidant properties. In the present study, the beneficial effect of the black radish juice on some free radical reactions in rats fed with a diet rich in lipids (20% sunflower oil, 2% cholesterol, 0.5% cholic acid in normal chow) was examined. Thiobarbituric acid reactive substances and conjugated diene concentrations were significantly higher, while the antioxidant enzyme activities and the free radical scavenging capacity were lower in hyperlipidaemic rats compared with normal controls. Supplementation of the lipid‐rich diet with black radish juice resulted in a significant improvement of the parameters mentioned above. Although the exact mechanism of the biologically active compounds in black radish on the lipid metabolism and lipid peroxidation is not clear yet, a beneficial effect of the drug was evident in alimentary hyperlipidaemia. Copyright © 2005 John Wiley & Sons, Ltd.\n",
      "[[0.84510529]]\n",
      "A decoction of leaves and unripe fruits prepared from C. aronia syn. azarolus (L) (Rosaceae), the indigenous Israeli hawthorn, is used to treat diabetes in Arab traditional medicine. There is a growing body of evidence reporting the presence of systemic oxidative stress in insulin-dependent and non-insulin dependent diabetes. We have already reported that this decoction has antioxidant properties in vitro. In this report, we extended these in vitro observations to study its antioxidant action in the streptozotocin (STZ)-induced diabetic rat in which we measured its effect on plasma and tissue indices of oxidative stress, as well as blood glucose levels. In an experiment lasting 24 days, 30 diabetic and 30 control rats were each subdivided into three separate groups. Group 1 rats drank tap water, group 2 rats drank the decoction containing 50 mg grounded dried plant/mL and group 3 rats drank the decoction containing 350 mg grounded dried plant/mL. Upon completion of the treatment period, the rats were anesthetized and blood glucose was measured using a drop of blood obtained by tail vein puncture. The abdomen was opened and aortic blood was withdrawn for measurement of the plasma levels of lipid peroxides, glutathione, cholesterol and triglycerides and the livers were harvested for determination of the extent of lipid peroxidation. The results of these experiments show that a decoction prepared from the leaves and unripe fruits of C. aronia normalizes plasma lipid peroxide levels and lowers blood glucose levels in diabetic rats. These findings indicate that the therapeutic benefit of the decoction could be due to these two actions, thereby providing a basis for its use in treating diabetes in traditional Arab medicine.\n",
      "[[0.83984426]]\n",
      "Phytochemicals such as alkaloids, glycosides, phenols, tannins, carbohydrates and saponins have been known to possess antidiabetic activity. The crude aqueous ethanolic extract from roots of Barringtonia acutangula Linn (EBA) were prepared, subjected to preliminary photochemical analyses to know the presence of alkaloids, flavonoids, tannins and saponins, which are proved to act as potent antioxidants indicating the possibility of the antidiabetic nature. The crude aqueous extract at a dose of 2000 mg/kg bw/p.o were used for studying acute oral toxicity as per OECD 423 guideline. Neither lethality was observed nor any profound toxic reaction, indicating 2000 mg/kg /p.o of aqueous EBA to be safe for further studies. Thus the pharmacological screening was done with two doses, aqueous EBA 250 and 500 mg/kg b.w/p.o. In normal fasted rats, aqueous EBA treatment at both the dose levels were found to reduce the blood glucose levels, significantly. The aqueous EBA also showed a significant improvement in oral glucose tolerance test.\n",
      "[[0.83929412]]\n",
      "Abstract The aim of study was to investigate the antidiabetic and antioxidant properties of extracts obtained from dried Prosopis farcta fruit and seeds against streptozotocin-induced diabetes in rats. According to the results, glucose, haemoglobin A1c, α-glycosidase activity, liver and kidney damage biomarkers, and malondialdehyde contents of all of the diabetic groups were found to have increased significantly according to the control group. Furthermore, the insulin and C-peptide secretions increased, and liver malondialdehyde level decreased, which were determined as the result of fluctuations in the antioxidant enzyme activities with a dose of 400 mg/kg fruit extract, while seed extract dosages of 100 and 400 mg/kg caused an increase in hepatic demage biomarkers. It was concluded that fruit extract may have insulin secretion stimulating and lipid peroxidation inhibitory effects, whereas seed extract might have caused hepatocyte damage changes to the transport functions and membrane permeability of these cells, thus causing enzymes to leak.\n",
      "[[0.8363105]]\n",
      "Melissa Officinalis L. (MOL) domestic preparations appropriate for consumption were studied by monitoring content in Na, K, Ca, Li, phenolic bioactives (total phenols, hydroxycinnamic acid derivatives and flavonols), and antioxidant activity (1,1-diphenyl-2-picrylhydrazyl radical inhibition (DPPH) and ferric reducing ability (FRAP)). The effects of practice applied, material to solvent ratio, time of preparation, and solvent were studied. MOL decoctions and infusions, commonly prepared at home, were better or of equal nutritional value to preparations upon ultrasounds or maceration concerning the studied parameters. Aqueous MOL preparations were richer in total phenols (704–1949 mg per 250 mL) and the examined macroelements (1.1–2.9, 30.5–288.4 and 50.1–176.1 mg Na, K and Ca per 250 mL, respectively) and showed better antioxidant activity compared to ethanol counterparts. The 25% w/v hydroethanolic MOL preparations, suitable for consumption, presented a significant content in phenolic antioxidants and in the examined minerals, too. MOL infusions were significantly richer in total phenols with respective chamomile and olive leaf ones, comparatively examined. Overall acceptance scores of aqueous MOL preparations indicated that bitterness has to be masked for efficient reception by the consumers.\n",
      "[[0.83177208]]\n",
      "Cordyceps taii, an edible medicinal mushroom native to south China, is recognized as an unparalleled resource of healthy foods and drug discovery. In the present study, the antioxidant pharmacological properties of C. taii were systematically investigated. In vitro assays revealed the scavenging activities of the aqueous extract and polysaccharides of C. taii against various free radicals, that is, 1,1-diphenyl-2-picrylhydrazyl radical, hydroxyl radical, and superoxide anion radical. The EC50 values for superoxide anion-free radical ranged from 2.04 mg/mL to 2.49 mg/mL, which was at least 2.6-fold stronger than that of antioxidant thiourea. The polysaccharides also significantly enhanced the antioxidant enzyme activities (superoxide dismutase, catalase, and glutathione peroxidase) and markedly decreased the malondialdehyde production of lipid peroxidation in a D-galactose-induced aging mouse model. Interestingly, the immune function of the administration group was significantly boosted compared with the D-galactose-induced aging model group. Therefore, the C. taii polysaccharides possessed potent antioxidant activity closely associated with immune function enhancement and free radical scavenging. These findings suggest that the polysaccharides are a promising source of natural antioxidants and antiaging drugs. Consequently, a preliminary chemical investigation was performed using gas chromatography-mass spectroscopy and revealed that the polysaccharides studied were mainly composed of glucose, mannose, and galactose. Fourier-transform infrared spectra also showed characteristic polysaccharide absorption bands.\n",
      "4.87040376663208\n"
     ]
    }
   ],
   "source": [
    "start = time.time()\n",
    "\n",
    "example = df_embed['embedding array'][200000]\n",
    "cosin = []\n",
    "\n",
    "for i in range(0,len(df_embed)):\n",
    "    try: \n",
    "        res = df_embed['embedding array'][i]\n",
    "        val_out = 1 - cdist(res, example, 'cosine')\n",
    "        cosin.append(val_out)\n",
    "    except:\n",
    "        print (i)\n",
    "        cosin.append(0)\n",
    "        \n",
    "max_value = max(cosin)\n",
    "max_index = cosin.index(max_value)\n",
    "\n",
    "\n",
    "print (df_embed['abstract'][200000] == df_embed.loc[max_index]['abstract'])\n",
    "print (len(cosin))\n",
    "\n",
    "top_5 = heapq.nlargest(6, range(len(cosin)), key=cosin.__getitem__)\n",
    "\n",
    "max_index = cosin.index(max_value)\n",
    "\n",
    "for i in top_5:\n",
    "    print (cosin[i])\n",
    "    print (df_embed.loc[i]['abstract'])\n",
    "    \n",
    "print (time.time() - start)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a42967f",
   "metadata": {},
   "source": [
    "### Embed user input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "07fca21d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e56cb17e36ae4742a103cdee3c71ddf9",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)okenizer_config.json:   0%|          | 0.00/321 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "93ffd63d600448598ef60d384ef703ae",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)lve/main/config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "2d04fe44d7a541d6898bb0cbc009e0a2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)solve/main/vocab.txt: 0.00B [00:00, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "8a98b8423f1647b59e92091a7b0ba69c",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading (…)cial_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "21fbe442dfc24e3ba4a88dc5846ee956",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Downloading pytorch_model.bin:   0%|          | 0.00/440M [00:00<?, ?B/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from transformers import AutoTokenizer, AutoModel\n",
    "\n",
    "# load model and tokenizer\n",
    "tokenizer = AutoTokenizer.from_pretrained('allenai/specter')\n",
    "model = AutoModel.from_pretrained('allenai/specter')\n",
    "\n",
    "papers = [{'title': 'BERT', 'abstract': 'We introduce a new language representation model called BERT'},\n",
    "          {'title': 'Attention is all you need', 'abstract': ' The dominant sequence transduction models are based on complex recurrent or convolutional neural networks'}]\n",
    "\n",
    "# concatenate title and abstract\n",
    "title_abs = [d['title'] + tokenizer.sep_token + (d.get('abstract') or '') for d in papers]\n",
    "# preprocess the input\n",
    "inputs = tokenizer(title_abs, padding=True, truncation=True, return_tensors=\"pt\", max_length=512)\n",
    "result = model(**inputs)\n",
    "# take the first token in the batch as the embedding\n",
    "embeddings = result.last_hidden_state[:, 0, :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7bc4072",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adfcf079",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f76ada71",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
